{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "113f99ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nothing done.\n"
     ]
    }
   ],
   "source": [
    "%reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8b3787c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/bin/python\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.executable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dad2fed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matplotlib is building the font cache; this may take a moment.\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/muon/_core/preproc.py:31: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('scanpy')` instead\n",
      "  if Version(scanpy.__version__) < Version(\"1.10\"):\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import muon as mu\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "988c4c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/oak/stanford/groups/engreitz/Users/opushkar/cc_perturb/2.preprocessing_tonys_pipeline/2.normalized_data/normalized_data\"\n",
    "working_dir = \"/oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "567ea78c",
   "metadata": {},
   "source": [
    "### Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5f82567c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1598: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1461: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "d0_mdata = mu.read(f\"{data_dir}/d0_normalized_gex_and_guide.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75c7305b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 38609 × 18318\n",
       "  2 modalities\n",
       "    GEX:\t38609 x 16392\n",
       "      obs:\t&#x27;guides_per_cell&#x27;, &#x27;guide_umi_counts&#x27;, &#x27;pct_counts_ribo&#x27;, &#x27;n_genes_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;total_counts_mt&#x27;, &#x27;pct_counts_mt&#x27;, &#x27;pool&#x27;, &#x27;sample_id&#x27;, &#x27;round1&#x27;, &#x27;well&#x27;, &#x27;biological_sample&#x27;, &#x27;plate&#x27;, &#x27;bci&#x27;, &#x27;stype&#x27;, &#x27;day&#x27;, &#x27;cell_pool&#x27;, &#x27;log1p_n_genes_by_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;pct_counts_in_top_50_genes&#x27;, &#x27;pct_counts_in_top_100_genes&#x27;, &#x27;pct_counts_in_top_200_genes&#x27;, &#x27;pct_counts_in_top_500_genes&#x27;, &#x27;log1p_total_counts_mt&#x27;, &#x27;total_counts_ribo&#x27;, &#x27;log1p_total_counts_ribo&#x27;, &#x27;S_score&#x27;, &#x27;G2M_score&#x27;, &#x27;phase&#x27;, &#x27;outlier&#x27;, &#x27;size_factors&#x27;, &#x27;leiden_res_0.1&#x27;, &#x27;leiden_res_0.2&#x27;, &#x27;leiden_res_0.3&#x27;, &#x27;leiden_res_0.4&#x27;, &#x27;leiden_res_0.5&#x27;, &#x27;leiden_res_0.6&#x27;, &#x27;leiden_res_0.7&#x27;, &#x27;leiden_res_0.8&#x27;, &#x27;leiden_res_0.9&#x27;, &#x27;leiden_res_1.0&#x27;\n",
       "      var:\t&#x27;gene_id&#x27;, &#x27;symbol&#x27;, &#x27;mt&#x27;, &#x27;ribo&#x27;, &#x27;n_cells_by_counts&#x27;, &#x27;mean_counts&#x27;, &#x27;log1p_mean_counts&#x27;, &#x27;pct_dropout_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;n_cells&#x27;, &#x27;n_counts&#x27;, &#x27;highly_deviant&#x27;, &#x27;highly_deviant_2k&#x27;, &#x27;binomial_deviance&#x27;, &#x27;highly_variable&#x27;, &#x27;means&#x27;, &#x27;dispersions&#x27;, &#x27;dispersions_norm&#x27;, &#x27;orig_highly_variable&#x27;, &#x27;highly_variable_2k&#x27;\n",
       "      uns:\t&#x27;hvg&#x27;, &#x27;leiden_res_0.1&#x27;, &#x27;leiden_res_0.2&#x27;, &#x27;leiden_res_0.3&#x27;, &#x27;leiden_res_0.4&#x27;, &#x27;leiden_res_0.5&#x27;, &#x27;leiden_res_0.6&#x27;, &#x27;leiden_res_0.7&#x27;, &#x27;leiden_res_0.8&#x27;, &#x27;leiden_res_0.9&#x27;, &#x27;leiden_res_1.0&#x27;, &#x27;neighbors&#x27;, &#x27;pca&#x27;, &#x27;umap&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;, &#x27;guide_assignment&#x27;, &#x27;guide_counts&#x27;\n",
       "      varm:\t&#x27;PCs&#x27;\n",
       "      layers:\t&#x27;ambiguous&#x27;, &#x27;analytic_pearson_residuals&#x27;, &#x27;log1p_norm&#x27;, &#x27;mature&#x27;, &#x27;nascent&#x27;, &#x27;raw_counts&#x27;, &#x27;scran_normalization&#x27;\n",
       "      obsp:\t&#x27;connectivities&#x27;, &#x27;distances&#x27;\n",
       "    guide:\t38609 x 1926\n",
       "      obs:\t&#x27;bci&#x27;, &#x27;stype&#x27;, &#x27;well&#x27;, &#x27;day&#x27;, &#x27;cell_pool&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 38609 × 18318\n",
       "  2 modalities\n",
       "    GEX:\t38609 x 16392\n",
       "      obs:\t'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "      var:\t'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "      uns:\t'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "      obsm:\t'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "      varm:\t'PCs'\n",
       "      layers:\t'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "      obsp:\t'connectivities', 'distances'\n",
       "    guide:\t38609 x 1926\n",
       "      obs:\t'bci', 'stype', 'well', 'day', 'cell_pool'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "78510df9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 38609 × 16392\n",
       "    obs: 'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "    var: 'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "    uns: 'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "    varm: 'PCs'\n",
       "    layers: 'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['GEX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6fc81e19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 65423017 stored elements and shape (38609, 16392)>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['GEX'].layers[\"raw_counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8e287af7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000227232.5', 'ENSG00000238009.6', 'ENSG00000241860.7',\n",
       "       'ENSG00000279457.4', 'ENSG00000228463.10', 'ENSG00000290385.1',\n",
       "       'ENSG00000230021.10', 'ENSG00000225972.1', 'ENSG00000225630.1',\n",
       "       'ENSG00000237973.1',\n",
       "       ...\n",
       "       'ENSG00000210196.2', 'ENSG00000271254.7', 'ENSG00000276345.1',\n",
       "       'ENSG00000276700.1', 'ENSG00000275757.1', 'ENSG00000277196.4',\n",
       "       'ENSG00000276256.1', 'ENSG00000273748.1', 'ENSG00000274917.1',\n",
       "       'ENSG00000273730.1'],\n",
       "      dtype='object', name='gene_id', length=16392)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['GEX'].var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c538e7f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ACAA1_+_38178488.23-P1P2', 'ACAA1_+_38178517.23-P1P2',\n",
       "       'ACAA1_+_38178559.23-P1P2', 'ACAA1_+_38178570.23-P1P2',\n",
       "       'ACAA1_+_38178620.23-P1P2', 'ACAA1_-_38178575.23-P1P2',\n",
       "       'ACTL7B_+_111618040.23-P1P2', 'ACTL7B_+_111618198.23-P1P2',\n",
       "       'ACTL7B_+_111618251.23-P1P2', 'ACTL7B_-_111617995.23-P1P2',\n",
       "       ...\n",
       "       'SAFE_TARGETING_933', 'SAFE_TARGETING_986', 'WDR61_SGTN_13',\n",
       "       'WDR61_SGTN_14', 'SUZ12_SGTN_22', 'SUZ12_SGTN_23', 'MESDC1_SGTN_11',\n",
       "       'MESDC1_SGTN_12', 'TP53_SGTN_01', 'TP53_SGTN_02'],\n",
       "      dtype='object', name='feature_id', length=1926)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['guide'].var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3d6f059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 38609 × 1926\n",
       "    obs: 'bci', 'stype', 'well', 'day', 'cell_pool'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['guide']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "91061ca5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "barcode\n",
       "CCTAATCCAACAACCATCTATTAC_2    T\n",
       "GGTGCGAACCTCCTGAATCCGCGA_2    T\n",
       "AACTCACCATCATTCCCTGTCCCG_1    T\n",
       "TTCACGCACACTTCGAATTCATGG_2    T\n",
       "GTCTGTCAGAGCTGAACCGTTCTA_3    T\n",
       "                             ..\n",
       "CGAACTTAGAGTTAGCATTCATGG_3    T\n",
       "CCGTGAGAGAATCTGAATCGCATA_3    T\n",
       "TGAAGAGACCTCTATCTCTATTAC_3    T\n",
       "CCATCCTCGAGCTGAAAAACGATA_2    T\n",
       "TCCGTCTAAACCGAGACGTTCGAG_2    T\n",
       "Name: stype, Length: 38609, dtype: category\n",
       "Categories (1, object): ['T']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['guide'].obs[\"stype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f07a2cf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "ENSG00000227232.5              WASH7P\n",
       "ENSG00000238009.6     ENSG00000238009\n",
       "ENSG00000241860.7     ENSG00000241860\n",
       "ENSG00000279457.4              WASH9P\n",
       "ENSG00000228463.10    ENSG00000228463\n",
       "                           ...       \n",
       "ENSG00000277196.4               PRODH\n",
       "ENSG00000276256.1     ENSG00000276256\n",
       "ENSG00000273748.1     ENSG00000273748\n",
       "ENSG00000274917.1           RNA5-8SN5\n",
       "ENSG00000273730.1           5_8S_rRNA\n",
       "Name: symbol, Length: 16392, dtype: category\n",
       "Categories (16338, object): ['5_8S_rRNA', '7SK', 'A1BG', 'A1BG-AS1', ..., 'ZYG11B', 'ZYX', 'ZZEF1', 'ZZZ3']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d0_mdata['GEX'].var['symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75172427",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_meta_df = pd.read_csv(f\"{working_dir}/300genes_guide_metadata_v43.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e28c09a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guide_id</th>\n",
       "      <th>spacer</th>\n",
       "      <th>targeting</th>\n",
       "      <th>type</th>\n",
       "      <th>guide_chr</th>\n",
       "      <th>guide_start</th>\n",
       "      <th>guide_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>pam</th>\n",
       "      <th>genomic_element</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>intended_target_chr</th>\n",
       "      <th>intended_target_start</th>\n",
       "      <th>intended_target_end</th>\n",
       "      <th>putative_target_genes</th>\n",
       "      <th>reporter</th>\n",
       "      <th>imperfect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1_+_38178488.23-P1P2</td>\n",
       "      <td>TAGTGCTGGGCCACCTGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38137000.0</td>\n",
       "      <td>38137019.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ENSG00000060971</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38103129.0</td>\n",
       "      <td>38137242.0</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1_+_38178517.23-P1P2</td>\n",
       "      <td>GTTGACCTGCGCAATGCAG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38137029.0</td>\n",
       "      <td>38137048.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ENSG00000060971</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38103129.0</td>\n",
       "      <td>38137242.0</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1_+_38178559.23-P1P2</td>\n",
       "      <td>TAACTCCGCGGTCAGTTCC</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38137071.0</td>\n",
       "      <td>38137090.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ENSG00000060971</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38103129.0</td>\n",
       "      <td>38137242.0</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1_+_38178570.23-P1P2</td>\n",
       "      <td>CGGCTGTCTGTTAACTCCG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38137082.0</td>\n",
       "      <td>38137101.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ENSG00000060971</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38103129.0</td>\n",
       "      <td>38137242.0</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAA1_+_38178620.23-P1P2</td>\n",
       "      <td>GGCGGAGTTGAGGCCTTGG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38137132.0</td>\n",
       "      <td>38137151.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ENSG00000060971</td>\n",
       "      <td>chr3</td>\n",
       "      <td>38103129.0</td>\n",
       "      <td>38137242.0</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>SUZ12_SGTN_23</td>\n",
       "      <td>TTGCAGGCGCTTGCTCTCCG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31937180.0</td>\n",
       "      <td>31937203.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>SUZ12</td>\n",
       "      <td>ENSG00000178691</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31937007.0</td>\n",
       "      <td>32001038.0</td>\n",
       "      <td>SUZ12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1922</th>\n",
       "      <td>MESDC1_SGTN_11</td>\n",
       "      <td>AGAAACTGAAAGGAAGGGCG</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81001111.0</td>\n",
       "      <td>81001134.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>TLNRD1</td>\n",
       "      <td>ENSG00000140406</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81000923.0</td>\n",
       "      <td>81005788.0</td>\n",
       "      <td>TLNRD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1923</th>\n",
       "      <td>MESDC1_SGTN_12</td>\n",
       "      <td>CCGCGCGCGGAGAGGAAGC</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81001066.0</td>\n",
       "      <td>81001088.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>TLNRD1</td>\n",
       "      <td>ENSG00000140406</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81000923.0</td>\n",
       "      <td>81005788.0</td>\n",
       "      <td>TLNRD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>TP53_SGTN_01</td>\n",
       "      <td>GGACACTTTGCGTTCGGGC</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7687417.0</td>\n",
       "      <td>7687439.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7661779.0</td>\n",
       "      <td>7687538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1925</th>\n",
       "      <td>TP53_SGTN_02</td>\n",
       "      <td>AGCGCCAGTCTTGAGCACAT</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7687505.0</td>\n",
       "      <td>7687528.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NGG</td>\n",
       "      <td>promoter</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7661779.0</td>\n",
       "      <td>7687538.0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1926 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      guide_id                spacer  targeting       type  \\\n",
       "0     ACAA1_+_38178488.23-P1P2   TAGTGCTGGGCCACCTGAG       True  targeting   \n",
       "1     ACAA1_+_38178517.23-P1P2   GTTGACCTGCGCAATGCAG       True  targeting   \n",
       "2     ACAA1_+_38178559.23-P1P2   TAACTCCGCGGTCAGTTCC       True  targeting   \n",
       "3     ACAA1_+_38178570.23-P1P2   CGGCTGTCTGTTAACTCCG       True  targeting   \n",
       "4     ACAA1_+_38178620.23-P1P2   GGCGGAGTTGAGGCCTTGG       True  targeting   \n",
       "...                        ...                   ...        ...        ...   \n",
       "1921             SUZ12_SGTN_23  TTGCAGGCGCTTGCTCTCCG       True  targeting   \n",
       "1922            MESDC1_SGTN_11  AGAAACTGAAAGGAAGGGCG       True  targeting   \n",
       "1923            MESDC1_SGTN_12   CCGCGCGCGGAGAGGAAGC       True  targeting   \n",
       "1924              TP53_SGTN_01   GGACACTTTGCGTTCGGGC       True  targeting   \n",
       "1925              TP53_SGTN_02  AGCGCCAGTCTTGAGCACAT       True  targeting   \n",
       "\n",
       "     guide_chr  guide_start   guide_end strand  pam genomic_element  \\\n",
       "0         chr3   38137000.0  38137019.0      -  NGG        promoter   \n",
       "1         chr3   38137029.0  38137048.0      -  NGG        promoter   \n",
       "2         chr3   38137071.0  38137090.0      -  NGG        promoter   \n",
       "3         chr3   38137082.0  38137101.0      -  NGG        promoter   \n",
       "4         chr3   38137132.0  38137151.0      -  NGG        promoter   \n",
       "...        ...          ...         ...    ...  ...             ...   \n",
       "1921     chr17   31937180.0  31937203.0      +  NGG        promoter   \n",
       "1922     chr15   81001111.0  81001134.0      -  NGG        promoter   \n",
       "1923     chr15   81001066.0  81001088.0      +  NGG        promoter   \n",
       "1924     chr17    7687417.0   7687439.0      -  NGG        promoter   \n",
       "1925     chr17    7687505.0   7687528.0      +  NGG        promoter   \n",
       "\n",
       "     intended_target_name       ensembl_id intended_target_chr  \\\n",
       "0                   ACAA1  ENSG00000060971                chr3   \n",
       "1                   ACAA1  ENSG00000060971                chr3   \n",
       "2                   ACAA1  ENSG00000060971                chr3   \n",
       "3                   ACAA1  ENSG00000060971                chr3   \n",
       "4                   ACAA1  ENSG00000060971                chr3   \n",
       "...                   ...              ...                 ...   \n",
       "1921                SUZ12  ENSG00000178691               chr17   \n",
       "1922               TLNRD1  ENSG00000140406               chr15   \n",
       "1923               TLNRD1  ENSG00000140406               chr15   \n",
       "1924                 TP53  ENSG00000141510               chr17   \n",
       "1925                 TP53  ENSG00000141510               chr17   \n",
       "\n",
       "      intended_target_start  intended_target_end putative_target_genes  \\\n",
       "0                38103129.0           38137242.0                 ACAA1   \n",
       "1                38103129.0           38137242.0                 ACAA1   \n",
       "2                38103129.0           38137242.0                 ACAA1   \n",
       "3                38103129.0           38137242.0                 ACAA1   \n",
       "4                38103129.0           38137242.0                 ACAA1   \n",
       "...                     ...                  ...                   ...   \n",
       "1921             31937007.0           32001038.0                 SUZ12   \n",
       "1922             81000923.0           81005788.0                TLNRD1   \n",
       "1923             81000923.0           81005788.0                TLNRD1   \n",
       "1924              7661779.0            7687538.0                  TP53   \n",
       "1925              7661779.0            7687538.0                  TP53   \n",
       "\n",
       "      reporter  imperfect  \n",
       "0          NaN        NaN  \n",
       "1          NaN        NaN  \n",
       "2          NaN        NaN  \n",
       "3          NaN        NaN  \n",
       "4          NaN        NaN  \n",
       "...        ...        ...  \n",
       "1921       NaN        NaN  \n",
       "1922       NaN        NaN  \n",
       "1923       NaN        NaN  \n",
       "1924       NaN        NaN  \n",
       "1925       NaN        NaN  \n",
       "\n",
       "[1926 rows x 18 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9dfa99b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io as sio\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f78f22f",
   "metadata": {},
   "source": [
    "### Day 0 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8315f88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata = d0_mdata\n",
    "\n",
    "# Filter out cells that have no guides\n",
    "#mdata = mdata[mdata['GEX'].obs['guides_per_cell'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4500127f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing sparse matrices...\n"
     ]
    }
   ],
   "source": [
    "# Define a temporary folder for transfer\n",
    "temp_dir = working_dir + \"/d0_sceptre_transfer\"\n",
    "os.makedirs(temp_dir, exist_ok=True)\n",
    "\n",
    "# 1. EXTRACT MATRICES (Transpose them!)\n",
    "# Scanpy is (Cells × Genes), Sceptre needs (Genes × Cells)\n",
    "print(\"Writing sparse matrices...\")\n",
    "sio.mmwrite(f\"{temp_dir}/response_matrix.mtx\", mdata['GEX'].layers['raw_counts'].T)\n",
    "sio.mmwrite(f\"{temp_dir}/grna_matrix.mtx\", mdata['guide'].X.T)\n",
    "\n",
    "# 2. EXTRACT NAMES (Row/Col names are lost in .mtx)\n",
    "# Save Gene names and Guide names to assign them later in R\n",
    "pd.Series(mdata['GEX'].var_names).to_csv(f\"{temp_dir}/gene_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['guide'].var_names).to_csv(f\"{temp_dir}/guide_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['GEX'].obs_names).to_csv(f\"{temp_dir}/cell_names.csv\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "56a87e34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 38609 × 16392\n",
       "    obs: 'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "    var: 'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "    uns: 'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "    varm: 'PCs'\n",
       "    layers: 'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['GEX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7e2f3172",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>grna_n_nonzero</th>\n",
       "      <th>grna_n_umis</th>\n",
       "      <th>response_n_nonzero</th>\n",
       "      <th>response_n_umis</th>\n",
       "      <th>response_p_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CCTAATCCAACAACCATCTATTAC_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>5</td>\n",
       "      <td>195.0</td>\n",
       "      <td>2049</td>\n",
       "      <td>3969.242920</td>\n",
       "      <td>14.708926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTGCGAACCTCCTGAATCCGCGA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>5</td>\n",
       "      <td>176.0</td>\n",
       "      <td>2232</td>\n",
       "      <td>3808.424805</td>\n",
       "      <td>10.594932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTCACCATCATTCCCTGTCCCG_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>4</td>\n",
       "      <td>183.0</td>\n",
       "      <td>1656</td>\n",
       "      <td>3441.276367</td>\n",
       "      <td>9.923644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCACGCACACTTCGAATTCATGG_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>688</td>\n",
       "      <td>805.500061</td>\n",
       "      <td>1.738051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTCTGTCAGAGCTGAACCGTTCTA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>3</td>\n",
       "      <td>46.0</td>\n",
       "      <td>2623</td>\n",
       "      <td>5361.565430</td>\n",
       "      <td>12.086022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGAACTTAGAGTTAGCATTCATGG_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>12</td>\n",
       "      <td>233.0</td>\n",
       "      <td>1729</td>\n",
       "      <td>4035.552734</td>\n",
       "      <td>10.494226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGAGAGAATCTGAATCGCATA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>2</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1205</td>\n",
       "      <td>1955.549194</td>\n",
       "      <td>13.985841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGAAGAGACCTCTATCTCTATTAC_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>7</td>\n",
       "      <td>344.0</td>\n",
       "      <td>2099</td>\n",
       "      <td>3987.371338</td>\n",
       "      <td>9.166440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCATCCTCGAGCTGAAAAACGATA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>4</td>\n",
       "      <td>256.0</td>\n",
       "      <td>1923</td>\n",
       "      <td>3728.120605</td>\n",
       "      <td>14.229690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCCGTCTAAACCGAGACGTTCGAG_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>5</td>\n",
       "      <td>66.0</td>\n",
       "      <td>924</td>\n",
       "      <td>1596.008423</td>\n",
       "      <td>18.796894</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38609 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             batch  grna_n_nonzero  \\\n",
       "barcode                                                              \n",
       "CCTAATCCAACAACCATCTATTAC_2  300genes:eTN5_GEX_sub2               5   \n",
       "GGTGCGAACCTCCTGAATCCGCGA_2  300genes:eTN5_GEX_sub2               5   \n",
       "AACTCACCATCATTCCCTGTCCCG_1  300genes:eTN5_GEX_sub1               4   \n",
       "TTCACGCACACTTCGAATTCATGG_2  300genes:eTN5_GEX_sub2               0   \n",
       "GTCTGTCAGAGCTGAACCGTTCTA_3  300genes:eTN5_GEX_sub3               3   \n",
       "...                                            ...             ...   \n",
       "CGAACTTAGAGTTAGCATTCATGG_3  300genes:eTN5_GEX_sub3              12   \n",
       "CCGTGAGAGAATCTGAATCGCATA_3  300genes:eTN5_GEX_sub3               2   \n",
       "TGAAGAGACCTCTATCTCTATTAC_3  300genes:eTN5_GEX_sub3               7   \n",
       "CCATCCTCGAGCTGAAAAACGATA_2  300genes:eTN5_GEX_sub2               4   \n",
       "TCCGTCTAAACCGAGACGTTCGAG_2  300genes:eTN5_GEX_sub2               5   \n",
       "\n",
       "                            grna_n_umis  response_n_nonzero  response_n_umis  \\\n",
       "barcode                                                                        \n",
       "CCTAATCCAACAACCATCTATTAC_2        195.0                2049      3969.242920   \n",
       "GGTGCGAACCTCCTGAATCCGCGA_2        176.0                2232      3808.424805   \n",
       "AACTCACCATCATTCCCTGTCCCG_1        183.0                1656      3441.276367   \n",
       "TTCACGCACACTTCGAATTCATGG_2          3.0                 688       805.500061   \n",
       "GTCTGTCAGAGCTGAACCGTTCTA_3         46.0                2623      5361.565430   \n",
       "...                                 ...                 ...              ...   \n",
       "CGAACTTAGAGTTAGCATTCATGG_3        233.0                1729      4035.552734   \n",
       "CCGTGAGAGAATCTGAATCGCATA_3         57.0                1205      1955.549194   \n",
       "TGAAGAGACCTCTATCTCTATTAC_3        344.0                2099      3987.371338   \n",
       "CCATCCTCGAGCTGAAAAACGATA_2        256.0                1923      3728.120605   \n",
       "TCCGTCTAAACCGAGACGTTCGAG_2         66.0                 924      1596.008423   \n",
       "\n",
       "                            response_p_mito  \n",
       "barcode                                      \n",
       "CCTAATCCAACAACCATCTATTAC_2        14.708926  \n",
       "GGTGCGAACCTCCTGAATCCGCGA_2        10.594932  \n",
       "AACTCACCATCATTCCCTGTCCCG_1         9.923644  \n",
       "TTCACGCACACTTCGAATTCATGG_2         1.738051  \n",
       "GTCTGTCAGAGCTGAACCGTTCTA_3        12.086022  \n",
       "...                                     ...  \n",
       "CGAACTTAGAGTTAGCATTCATGG_3        10.494226  \n",
       "CCGTGAGAGAATCTGAATCGCATA_3        13.985841  \n",
       "TGAAGAGACCTCTATCTCTATTAC_3         9.166440  \n",
       "CCATCCTCGAGCTGAAAAACGATA_2        14.229690  \n",
       "TCCGTCTAAACCGAGACGTTCGAG_2        18.796894  \n",
       "\n",
       "[38609 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. EXTRACT COVARIATES \n",
    "rna = mdata['GEX']\n",
    "covariates_df = pd.DataFrame({\n",
    "    \"batch\": rna.obs['sample_id'],\n",
    "    \"grna_n_nonzero\": rna.obs['guides_per_cell'],\n",
    "    \"grna_n_umis\": rna.obs['guide_umi_counts'],\n",
    "    \"response_n_nonzero\": rna.obs['n_genes_by_counts'],\n",
    "    \"response_n_umis\": rna.obs['total_counts'],\n",
    "    \"response_p_mito\": rna.obs['pct_counts_mt'],\n",
    "})\n",
    "covariates_df.to_csv(f\"{temp_dir}/covariates.csv\")\n",
    "covariates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9bca97d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ All guides mapped successfully.\n",
      "\n",
      "First 5 rows of mapping (Must match mdata['guide'].var_names order):\n",
      "                    grna_id grna_target\n",
      "0  ACAA1_+_38178488.23-P1P2       ACAA1\n",
      "1  ACAA1_+_38178517.23-P1P2       ACAA1\n",
      "2  ACAA1_+_38178559.23-P1P2       ACAA1\n",
      "3  ACAA1_+_38178570.23-P1P2       ACAA1\n",
      "4  ACAA1_+_38178620.23-P1P2       ACAA1\n"
     ]
    }
   ],
   "source": [
    "# 4. PREPARE GUIDE MAPPING \n",
    "# Sceptre needs a table mapping 'grna_id' to 'grna_target'.\n",
    "\n",
    "# 1. Create a \"skeleton\" DataFrame from the actual matrix order\n",
    "# This ensures the order matches mdata['guide'].X perfectly\n",
    "# We name the index 'grna_id' to match Sceptre's expectation immediately\n",
    "guide_mapping = pd.DataFrame(index=mdata['guide'].var_names)\n",
    "guide_mapping.index.name = 'grna_id'\n",
    "\n",
    "# 2. Prepare metadata for merging\n",
    "# Ensure 'guide_id' is the index so we can map efficiently\n",
    "lookup_df = guide_meta_df.set_index('guide_id')\n",
    "\n",
    "# 3. Map the targets onto the skeleton\n",
    "# We use .loc[] to pull the specific target for each guide in the exact order required\n",
    "guide_mapping['grna_target'] = lookup_df.loc[guide_mapping.index, 'intended_target_name']\n",
    "\n",
    "# 4. Reset index so 'grna_id' becomes a column (Sceptre requires this)\n",
    "guide_mapping = guide_mapping.reset_index()\n",
    "\n",
    "# ==========================================\n",
    "# VALIDATION CHECKS \n",
    "# ==========================================\n",
    "\n",
    "# Check 1: Are there any NAs? (Guides in mdata that were missing from meta_df)\n",
    "na_count = guide_mapping['grna_target'].isna().sum()\n",
    "if na_count > 0:\n",
    "    print(f\"⚠️ WARNING: {na_count} guides have NaN targets! These will crash Sceptre.\")\n",
    "else:\n",
    "    print(\"✅ All guides mapped successfully.\")\n",
    "\n",
    "# Check 2: Verify the first few rows manually\n",
    "print(\"\\nFirst 5 rows of mapping (Must match mdata['guide'].var_names order):\")\n",
    "print(guide_mapping.head())\n",
    "\n",
    "# 5. Save for Sceptre\n",
    "guide_mapping.to_csv(f\"{temp_dir}/guide_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "aa5df091",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>grna_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1_+_38178488.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1_+_38178517.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1_+_38178559.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1_+_38178570.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAA1_+_38178620.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>SUZ12_SGTN_23</td>\n",
       "      <td>SUZ12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1922</th>\n",
       "      <td>MESDC1_SGTN_11</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1923</th>\n",
       "      <td>MESDC1_SGTN_12</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>TP53_SGTN_01</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1925</th>\n",
       "      <td>TP53_SGTN_02</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1926 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       grna_id grna_target\n",
       "0     ACAA1_+_38178488.23-P1P2       ACAA1\n",
       "1     ACAA1_+_38178517.23-P1P2       ACAA1\n",
       "2     ACAA1_+_38178559.23-P1P2       ACAA1\n",
       "3     ACAA1_+_38178570.23-P1P2       ACAA1\n",
       "4     ACAA1_+_38178620.23-P1P2       ACAA1\n",
       "...                        ...         ...\n",
       "1921             SUZ12_SGTN_23       SUZ12\n",
       "1922            MESDC1_SGTN_11      TLNRD1\n",
       "1923            MESDC1_SGTN_12      TLNRD1\n",
       "1924              TP53_SGTN_01        TP53\n",
       "1925              TP53_SGTN_02        TP53\n",
       "\n",
       "[1926 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "debf88ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Gene mapping saved.\n"
     ]
    }
   ],
   "source": [
    "# Create a mapping from the Gene Symbol (which matches your guides)\n",
    "# to the Gene Index (which matches the matrix rownames)\n",
    "\n",
    "# Check which column in mdata['GEX'].var holds the symbol\n",
    "symbol_col = 'symbol' \n",
    "\n",
    "gene_map = mdata['GEX'].var[[symbol_col]].reset_index()\n",
    "gene_map.columns = ['response_id', 'grna_target'] # Rename for easy merging in R\n",
    "\n",
    "# Save it\n",
    "gene_map.to_csv(f\"{temp_dir}/gene_symbol_map.csv\", index=False)\n",
    "print(\"✅ Gene mapping saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb2b01f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>WASH9P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16387</th>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>PRODH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16388</th>\n",
       "      <td>ENSG00000276256.1</td>\n",
       "      <td>ENSG00000276256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16389</th>\n",
       "      <td>ENSG00000273748.1</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16390</th>\n",
       "      <td>ENSG00000274917.1</td>\n",
       "      <td>RNA5-8SN5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16391</th>\n",
       "      <td>ENSG00000273730.1</td>\n",
       "      <td>5_8S_rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16392 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              response_id      grna_target\n",
       "0       ENSG00000227232.5           WASH7P\n",
       "1       ENSG00000238009.6  ENSG00000238009\n",
       "2       ENSG00000241860.7  ENSG00000241860\n",
       "3       ENSG00000279457.4           WASH9P\n",
       "4      ENSG00000228463.10  ENSG00000228463\n",
       "...                   ...              ...\n",
       "16387   ENSG00000277196.4            PRODH\n",
       "16388   ENSG00000276256.1  ENSG00000276256\n",
       "16389   ENSG00000273748.1  ENSG00000273748\n",
       "16390   ENSG00000274917.1        RNA5-8SN5\n",
       "16391   ENSG00000273730.1        5_8S_rRNA\n",
       "\n",
       "[16392 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a3b3acb",
   "metadata": {},
   "source": [
    "#### Import data to SCEPTRE (Rcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "498990ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e21a04eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading matrices into R...\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(sceptre)\n",
    "library(Matrix)\n",
    "\n",
    "# Define path\n",
    "data_dir <- \"/oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer\"\n",
    "\n",
    "# 1. LOAD MATRICES\n",
    "cat(\"Loading matrices into R...\\n\")\n",
    "response_mat <- readMM(file.path(data_dir, \"response_matrix.mtx\"))\n",
    "grna_mat <- readMM(file.path(data_dir, \"grna_matrix.mtx\"))\n",
    "\n",
    "# 2. ATTACH NAMES\n",
    "gene_names <- read.csv(file.path(data_dir, \"gene_names.csv\"), header=FALSE)$V1\n",
    "guide_names <- read.csv(file.path(data_dir, \"guide_names.csv\"), header=FALSE)$V1\n",
    "cell_names <- read.csv(file.path(data_dir, \"cell_names.csv\"), header=FALSE)$V1\n",
    "\n",
    "rownames(response_mat) <- gene_names\n",
    "colnames(response_mat) <- cell_names\n",
    "\n",
    "rownames(grna_mat) <- guide_names\n",
    "colnames(grna_mat) <- cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0bcc991d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renaming 'response_p_mito' to 'mt_percent'...\n",
      "[1] \"\\u2705 Sceptre object created!\"\n",
      "                           response_n_nonzero response_n_umis grna_n_nonzero\n",
      "CCTAATCCAACAACCATCTATTAC_2               1774        3969.243             13\n",
      "GGTGCGAACCTCCTGAATCCGCGA_2               1785        3808.425             23\n",
      "AACTCACCATCATTCCCTGTCCCG_1               1439        3441.277             38\n",
      "TTCACGCACACTTCGAATTCATGG_2                598         805.500              3\n",
      "GTCTGTCAGAGCTGAACCGTTCTA_3               2308        5361.566             10\n",
      "CTGGCATAAACCGAGATGTCTGAA_3               1068        2485.453            137\n",
      "                           grna_n_umis                  batch mt_percent\n",
      "CCTAATCCAACAACCATCTATTAC_2         195 300genes:eTN5_GEX_sub2  14.708926\n",
      "GGTGCGAACCTCCTGAATCCGCGA_2         176 300genes:eTN5_GEX_sub2  10.594932\n",
      "AACTCACCATCATTCCCTGTCCCG_1         183 300genes:eTN5_GEX_sub1   9.923644\n",
      "TTCACGCACACTTCGAATTCATGG_2           3 300genes:eTN5_GEX_sub2   1.738051\n",
      "GTCTGTCAGAGCTGAACCGTTCTA_3          46 300genes:eTN5_GEX_sub3  12.086022\n",
      "CTGGCATAAACCGAGATGTCTGAA_3        1699 300genes:eTN5_GEX_sub3  13.860650\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# 3. LOAD METADATA\n",
    "covariates <- read.csv(file.path(data_dir, \"covariates.csv\"), row.names=1)\n",
    "guide_mapping <- read.csv(file.path(data_dir, \"guide_mapping.csv\"))\n",
    "\n",
    "# ==============================================================================\n",
    "# FIX: RENAME 'response_p_mito' SO WE KEEP IT\n",
    "# Sceptre blocks the name 'response_p_mito' but doesn't calculate it.\n",
    "# We rename it to 'mt_percent' so we can use it as a covariate later.\n",
    "# ==============================================================================\n",
    "\n",
    "# 1. Rename response_p_mito if it exists\n",
    "if (\"response_p_mito\" %in% colnames(covariates)) {\n",
    "    cat(\"Renaming 'response_p_mito' to 'mt_percent'...\\n\")\n",
    "    names(covariates)[names(covariates) == \"response_p_mito\"] <- \"mt_percent\"\n",
    "}\n",
    "\n",
    "# 2. DROP OTHER RESERVED COLUMNS (which Sceptre DOES calculate)\n",
    "# We still drop n_umis and n_nonzero because Sceptre re-calculates those automatically.\n",
    "reserved_cols <- c(\"response_n_nonzero\", \"response_n_umis\", \"grna_n_nonzero\", \"grna_n_umis\")\n",
    "covariates <- covariates[, !(colnames(covariates) %in% reserved_cols), drop=FALSE]\n",
    "\n",
    "# 4. CREATE SCEPTRE OBJECT\n",
    "sceptre_object <- import_data(\n",
    "  response_matrix = response_mat,\n",
    "  grna_matrix = grna_mat,\n",
    "  grna_target_data_frame = guide_mapping,\n",
    "  extra_covariates = covariates,\n",
    "  moi = \"high\"\n",
    ")\n",
    "\n",
    "print(\"✅ Sceptre object created!\")\n",
    "print(head(sceptre_object@covariate_data_frame))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "04f62e98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Unique pairs to test: 281\"\n",
      "   grna_target        response_id\n",
      "1        ACAA1 ENSG00000060971.19\n",
      "7       ACVR1B ENSG00000135503.13\n",
      "13      ACVR2A ENSG00000121989.15\n",
      "19      ACVR2B ENSG00000114739.14\n",
      "25       ADCY2 ENSG00000078295.17\n",
      "31      AMOTL2 ENSG00000114019.15\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 1. Load the translation map\n",
    "gene_map <- read.csv(file.path(data_dir, \"gene_symbol_map.csv\"))\n",
    "\n",
    "# 2. Start with guide mapping (which has 'grna_target' as Symbol)\n",
    "discovery_pairs_highmoi <- guide_mapping\n",
    "\n",
    "# 3. Merge to attach the correct response_id (Gene ID)\n",
    "#    This matches 'grna_target' in guides to 'grna_target' in the gene map\n",
    "discovery_pairs_highmoi <- merge(discovery_pairs_highmoi, gene_map, by=\"grna_target\")\n",
    "\n",
    "# 4. Select ONLY the two columns Sceptre needs\n",
    "#    (We filter out non-targeting guides automatically via the merge)\n",
    "discovery_pairs_highmoi <- discovery_pairs_highmoi[, c(\"grna_target\", \"response_id\")]\n",
    "\n",
    "# 5. DEDUPLICATE\n",
    "#    Sceptre performs analysis at the *Target* level, not the *Guide* level.\n",
    "#    We only need one row per Target-Gene pair.\n",
    "discovery_pairs_highmoi <- unique(discovery_pairs_highmoi)\n",
    "\n",
    "print(paste(\"Unique pairs to test:\", nrow(discovery_pairs_highmoi)))\n",
    "print(head(discovery_pairs_highmoi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "96e78b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "sceptre_object_highmoi <- set_analysis_parameters(\n",
    "  sceptre_object = sceptre_object,\n",
    "  discovery_pairs = discovery_pairs_highmoi,\n",
    "  side = \"both\", control_group = \"complement\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "42f4e7da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/grna_count_distributions.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "png(file.path(data_dir, \"grna_count_distributions.png\"), width = 3000, height = 2000, res = 300)\n",
    "\n",
    "print(plot_grna_count_distributions(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  n_grnas_to_plot = 8\n",
    "))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"grna_count_distributions.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9ec0aa8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3deVxU9f4/8PcsMAMzMOziBoiACrhe1K+JZmnWNa/mlmaoZdnXupesjK+mqaC3xbUyy1yq283rbdHMPXcrlURcQATFhUU2WWZYZhhmPb8/jj+uV6AGBM5nZl7Ph39w3jNzzvvA4eWHzzlzRsRxHAEAAHvEQjcAAACNQ0ADADAKAQ0AwCgENAAAoySJiYlC99AeVq1adeTIkVOnTuXm5oaFhbm6ugrdUZNWrVpVWFgYFRXFL65ZsyYiIsLd3V3YrsBepKSkbNmy5eTJkxcvXpRKpZ07dxa6oybxrUZGRioUCiJKTU09c+ZMZGRk/RMOHz588+bN7t2784snT57MyMiIiIjgX9ipUydfX1/+oc2bNx87dmzo0KEpKSmnT5+u//Wxd84ygl61apVcLu/SpcuePXsGDRpkNBqF7qhJq1atmj17dkFBAb+4Zs2a8vJyYVsCO5KSknLgwIGwsDCO48aMGbNt2zahO2pSSkrKihUrVqxYwS+mpqbu2LHj3iccPnz48OHD9YsnT57ct28f/8J169atX7+er5eUlCxcuHDVqlX8Q99991077UDbkwrdQPsZO3ZsdHT0s88+6+3tfeXKlf79+xPRgQMHrly50r1796eeekosFhNRenr60aNHLRbLgAEDRo4cSUSJiYmzZ8/eu3evUqmcNm2aTCYjIrPZ/MMPP+Tn5w8aNGj48OH8JhITE19++eVdu3ZJJJK4uDg3N7dGV9jodu81YcKEpKSkLVu2NLojjW7Fxn4uXLiwZ8+e+lVNnz49IiKi0X4SExP/93//d+/evSEhIY8++mjDlQOzgoKC4uLiiMhgMHz//ff81wUFBfv37zcYDE8++SQ/Jq2rq9u5c2deXp6/v//48eMDAgJOnjxJRDU1NdevX3/00Uf79evHr/Dq1auHDx9WKBQTJkzw8fEhIv6ZZrM5LS1t+PDhAwcObHSFjW73Xk8++eS33377xhtvdOvWrVn7+MQTT+zfv3/dunWurq5ff/31xIkTf/zxx3uf0GgzdsdZRtD1LBYLx3H8QTZ9+vQjR4507dr12LFjkydPJqIrV65MmjTJz8+vS5cuycnJ/EuSkpIWL14cExNTUlIyZswYIuI47s9//vPBgwcDAgISEhLef//9+me+/fbbUVFR+fn506dPb2qFDbd7n4SEhH379l27dq3RXWi4Fdv7CQwMjI2NjY2NDQkJWbduncViaaqfpKSkBQsWdOvWrUuXLo2uHNhnNBr9/PyIKDU1dfLkyRKJRKFQjBs3LiUlhYief/755OTknj171tbW8gfbyZMnZ8yYoVarBw4c+MILLxw7doyIjh07NmbMGLlcfvv27QEDBlRUVPDPnDNnTn5+fv/+/ePi4i5evNjoChvd7r18fHxee+21pUuXNnfX3N3dR44cyY82tm3bNmPGjPue0LAZe+REI+ht27b5+/sfOnTogw8+CA4OTk1NTUlJefvtt+vq6gYOHPjmm28WFRXdvn07ICDg8ccf79Chw72vXbp0aXh4+ODBg3fs2HH27Fmz2VxYWHjkyBEiGjFiRGRk5P/93//xA8+kpKROnToNGTLEy8uL47iGK2x0u506dbp3c56engkJCUuWLGnqj7X7tnLmzBkb++nUqVOnTp20Wu2wYcO2bNnSq1ev3+knMTExNDT09OnTTa0c2HTjxo0PP/ywtLT07Nmz/BRHYmJibGysVColoiFDhmzdunXQoEE5OTnjxo0bP368RCKpf+3w4cNnzZpFRMuXL1+zZs3IkSPff//99957b+rUqUSUn5+/devWBQsWENEjjzwye/ZsInrmmWcOHTrUv3//hitsdLv3dfvaa69FRERkZGQ0dzeff/759957Lzg4uGPHjg2n2hvdO7vjXL9mVqtVrVbz/+tev37dw8Oj/qE1a9bI5fLHH3983LhxY8aM6dOnzyeffFL/aFBQEP9FcHBwXl5ebm5ujx497n2ofpqYjzb+iDQYDA1X2Oh2G7b617/+NSUl5fz5843uyH1bsb0fIrJYLNOmTZs8eTL/K/c7/fB/df7OyoFler3+3LlzN2/epP/+KcfGxk6bNo2ItmzZ8sMPP4SHh0+aNCkvL49/9L5DnYhyc3N79uzJFyMjI+ufWT+qUCqVlZWVja6w0e3ex93dffHixYsWLWr4kEwm4w9ansFguPeXZejQoTk5OStXrnzuuecavrbRvbM7TjSCjouLi46Ofumll3r16rV79+7Q0NDS0tJp06bdl48LFixYsGBBaWlpVFTU2LFjg4ODiSgrK4ufj8vMzAwODjabzVevXuU4TiQS5efnExH/h2RDIpHovhU2td37yGSyZcuWvfXWW7bsWkhIiI39ENFrr73m6+u7ePFifvF3+hGJRM1dObAgLCzstddeI6JOnTq9+uqr58+fDw0N9fHxuS/Ievfu/f3333Mcl5iYmJSU9MUXXxBRVlYW/yh/qBNRSEhIVlZW3759+WL9/9YNNVxho9tt6MUXX1y3bp2Xl9d99YiICL4rXnp6+pNPPnnvE2bOnLlu3bpt27bVn1T//b2zO04U0DyVSrV8+fJly5ZdvHhx+PDhY8aMmTFjhlarPXLkyJ49e3bv3p2amhoVFVVYWKhUKutPLKxYsWL8+PE///xzx44dBw8eTERdunSZNWvWI488snHjxrfffrupP/kbrjA4OLjhdht97cyZM9esWVNWVvaHOzV06FAb+/n6669/+OGHLVu2HD16lIj69es3ePDg3+/H9pUDa+bNm/fZZ5/t2rVr+fLlTz31lFqt7tatW1paWnR09PPPPz9nzpyYmBiVSvXLL7889dRT/EsyMzMXL17crVu3d999d+vWrUS0cOHCF198UaPRFBcXHz9+fPXq1U1truEKG91uwxe6uLgsX748Li5uypQp99anTp26du3a2bNnDxky5MKFCzdu3Jg5c+a9T+BHPzY2Y4+c5TpoIho6dCh/uWXfvn0rKyvDw8Ofe+65Tp065eTkuLi4zJo1KzAwMCAgoLq6Ojc3V6VSrV27lv8vPSkp6fjx45cvX+7evfv777/PTxdMnz7dYrGUlJQ899xz/Fly3ogRI+q/HjZsWGBgYMMVTp48+b7tNmxVLpeLxeIBAwZ07tx5xIgR910Hfd9WxGKxjf1UVFT4+/sXFhbm5+fn5+eHhYX5+vo21U/9a5taObCpc+fO/IXAEolkwIABGo3m8ccff/bZZ+/cuVNaWtq7d+9x48bJZLKAgIC8vDy1Wj116lT+HPLJkyd79+49fPjw27dvL1y4cMiQIUQUGhr6xBNPZGdn+/v7f/jhh/WXHoeEhISEhPBfd+3aNTQ0tOEKO3fu3HC797XKz59ER0fL5fLY2Nh7r4OWSqWzZ882m80lJSW9e/f+6KOP+F/he19Yz83NbejQofW737AZeyTC3ez+kEiE7xI4BX645jyDNvYhegAAGIXJRAAARiGgAQAYhYAGAGAUAhoAgFEIaAAARrXVG1U0Gs3SpUs//vhj219ScpCqr5DvQ+T7UAs3Wlpau3DhCZXKlYgMBsvAgV2efz66hetydJyVrq8jIur2Erl4Ct0N82w/nquzqGQ/yTtRUHMuvf3226snTuS4uUmJqKrKsGTJsG7dVC3u1vFUXaY7h8gtiLo+LXQr7autRtA6na65763UXKCivVST3fKNVlUZfvzxVkJCVEJCVGCg7MSJ/Javy+FxVLSXivaSRS90J/bA9uNZX0BFe6n81B8/Mztbs3DhL/y/NWvOKRRi/tA9ebKwtLT2QTt2LLX5VLSXKs4I3Ue7c9i3epeW6q9fr9y8OZ1fHDkyqHv3+9/pDyCgnJyqL79MnzIllIhyc6uGD7fLGxZDm7ob0MePHz979qxIJOLvak9EBw8e/PXXX4nI19d3/vz5RGSxWLZv315UVPTYY48NGDCg0Qo7bt2qKSio2rv3KhGlpJR5eIxEQANrOnVSLFrUj4h+/rlI6F6ARXcD2tXV9ZVXXjEajUuWLFm3bp27u3txcfEzzzzTu3fv+qfu2LHD09NzypQpixYt6tatm7e3d8OKQHvRuP/5n4CPPhpCRM8+e0LoXgAAmu3uHHRsbKxKpfL39/fy8tJqtXxx3759//jHP0pKSvjFs2fPjho1Si6XDxkyhL9PccMKAAC0lv+ag87KypJIJPw9NmNjY2Uy2Z07dxITE/kxdU1NDX8rKW9vb/5jbxpWiCg9Pf2zzz4zGAz33gYeAACa6z9XcRQXF2/atOn111/nFyMiIoKDgwcNGhQdHZ2dnU1EUqmUv7OSyWTib+7esEJEgYGBEydOHD169L0fhQAAAM11N6CrqqpWr179xhtvNPywDLVazd+MOCgoKDc3l4hu3brF3we2YYWIAgICRo0aNXToUKPR2C678MeuXq187bXj3btv4f/V1LDSGADA77g7xbF69Wpvb+/k5OTk5ORRo0b5+vquXbvW19e3rKxMKpWGh4cT0aRJkzZu3NirV6+ysrI+ffo0WmGT0WiNiwt76KEOVis3ffoJqxV3WAUAO3A3oGfOnFk/I+Hm5kZEc+fOLSgoUCqV9R+XGxERsXTp0vLy8qCgIP7T6hpWmNWjh9ewYYEWC6IZAOzG3YCOiIi47wGFQtHw0yGVSqVSqfz9CgAAtArcLAkAgFEO+1ZvADtlNFqvXlVLpWIikkrFffv6C90RCAYBDcCW0lL9woUnZTKJyWQ1GLjy8r8K3REIBgENwJytW4f37++bk1MzfvxRoXsBIWEOGgCAUQhoAABGIaABABiFgAYAYJRznSTk3+T93HM/ubiIiahHD+8VK2KFbgoAoHHOFdAcR0TUv7+nXC7Jyqo8ceK20B2B05k2bR8/UCgp0QndC7DOGac4pk4NjYsLGzGio9CNgDPaseNar16KmBiVQiF0K8A85xpBA7Bg8uQQX1+5Wm0oKSkQuhdgmjOOoAEA7AICGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABGIaABABiFgAYAYJQjvJNw5MjvcnOriMhksgrdCwBAq3GEgL59u+bVVyN79FBdvFixenW60O2A8L755pv09HSTyRQRETFnzhwiUqvVGzZs4DguPDx8+vTpNlYAhOUgUxxhYZ59+viEhnoI3QgwYcqUKe++++7q1atzcnLy8vKI6JtvvhkzZsyyZctu3LhhewVAWA4S0AD3kkgkRGSxWORyua+vLxFlZmb27duXiPr165eRkWFjBUBYjjDFAdDQtm3bfv311549e7q7uxORwWBwcXEhIoVCUV5ebmOFiE6ePPnmm28SkY+Pj0C7As4LI2hwTHFxcRs3bqyrq0tNTSUimUxmMpmISKfTKZVKGytEFBsbe/z48W3btmk0GqH2BZwWAhocEJ+zYrHYzc3N1dWViKKiotLS0ojo0qVL0dHRNlaISCqVenp6KpVKjv88HoB2hCkOcEAbN24sKyszm83du3fnp5WnTp26fv36PXv29OzZMzg42MYKgLAQ0OCAXn311fsqPj4+iYmJza0ACAtTHAAAjEJAAwAwCgENAMAo552Dtlg4s9mq0dTxi25uUrnceb8bAMAg5x1BX76sPnu22MdnA/9v7dpUoTsCAPgvzhvQRDRwoH9h4fTCwukzZ4YL3QsAwP2cOqABAFiGgAYAYBQCGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABG2euHPP3jHxkHDuTwXxcX64RtBgCgLdjrCPrSpVKNRhsTo4qJUen1JqHbAQBoffYa0ETUp49PXFxYXFyYWCwSuhcAgNZnr1McAA6vosJQU2N8+um9/OLw4V3+9rf+wrYE7cyOR9AAjk2vN0ulIn4ez2CoS0kpFrojaG8YQQOwSyaTxMWFEVFdneXGDb3Q7UB7wwgaAIBRGEETEZWW1lVVlWzenM4vTpwY7ufnJmxLAAAIaCKinJxqkUhksRiJ6NixooEDAxHQACA4BPRd48cHv/pqFBENHrxb6F4AAIgwBw0AwCwENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMwr04AOxAamrZoUOF+/bd4he3bBk9aVKEsC1BO8AIGsAOmM3co4923L9/9P79oyMiPI1Gq9AdQXvACBrAPiiVLsHBSiJyc5MI3Qu0EwQ0QJs7fbqwqEjLf81xwvYC9gQBDdDmVq5Myc6uCAiQE5HVioQGW90N6M2bN9+6dctisQwZMmTixIlEpFarN2zYwHFceHj49OnTbawAQKNefrnXlCndiKhz5+1C9wJ24+5JwnHjxr3//vsrV648ceKERqMhom+++WbMmDHLli27ceNGXl6ejRUAAGgtdwM6MDCQiMRisUgkEovFRJSZmdm3b18i6tevX0ZGho0VIjKbzdXV1VqtViQSCbBDAACO4r/moA8ePNizZ0+VSkVEBoPBxcWFiBQKRXl5uY0VIjp16tSbb75JRN7e3u26KwAAjuU/10GnpqaeOXNmzpw5/KJMJjOZTESk0+mUSqWNFSIaMWJEamrqjz/+qFar23dfAAAcyt2Azs7O/u6779566y1+RExEUVFRaWlpRHTp0qXo6GgbKwAA0FruTnGsX79eIpG89957RPTSSy917dp16tSp69ev37NnT8+ePYODg4nIlgoAALSWuwG9YcOG+x7w8fFJTExsbgUAAFoL7sUBAMAoBDQAAKMQ0AAAjEJAAwAwCgENAMAoBDQAAKMQ0AAAjEJAAwAwCgENAMAoBDQAAKMQ0AAAjEJAAwAwCh8aez+TicvKquA/2dPFRdynj7/QHQGAk0JA36+8XD9//nGZTGI0WolERUUvC90RADgpTHE04quvRpw5M+5f/xohdCMA4NQQ0AAAjEJAAwAwCgENAMAoBDQAAKMQ0AAAjEJAAwAwCgENAMAoBDQAAKMQ0AAAjEJAAwAwyp7uxVFQUHPgQA7/dUZGed++KmH7AQBoU/YU0FeuVCxYcDI2NpCIzp0rRkADgGOzp4AmopAQj02bYono8ccPCN0LAEDbwhw0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIyys9uNAoDRaC0trb11q4pfDAnxFItFwrYEbQQBDWBnrl6tXLTol2XLThFRVZWxsjJepZIJ3RS0CUxxANif1asHZ2ZOvnx5ktCNQNtCQAMAMAoBDQDAKAQ0AACjcJKwSbW15ro68/ffX+MXu3VTxcQECtsSADgVBHSTSkv1Op3p44/PEVFBge7RR0O2bkVAA0D7QUD/Hi8v12++eZSINm7MKi42C90OADgXzEEDADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKn6gCYK84jojoxx+vu7u7EJGfn9sjjwQJ3BO0KgQ0gL2yWjki2rTpglQqVqsNXl7up04hoB0KAhoc0PHjx8+ePSsSicaOHRsdHU1EFotl+/btRUVFjz322IABA2ys2IUvv3zYw8PlyJHCLVtuCN0LtDLMQYMDcnV1feWVV1544YUNGzbU1tYS0Y4dOzw9PefNm7dt2zaNRmNjBUBYCGhwQLGxsSqVyt/f38vLS6vVEtHZs2dHjRoll8uHDBly/vx5GysAwsIUBzisrKwsiUQSEBBARDU1NQqFgoi8vb0rKipsrBDR+fPnP/zwQ6PR6OnpKdSOgNPCCBocU3Fx8aZNm15//XV+USqVchxHRCaTSS6X21ghotDQ0Pj4+JkzZ+r1ekF2BJwZAhocUFVV1erVq9944w0/Pz++EhQUlJubS0S3bt0KCQmxsUJE3t7egwYN6tu3r8lkaue9AMAUBzig1atXe3t7JycnJycnjxo1ytfXd9KkSRs3buzVq1dZWVmfPn2IyJYKgLAQ0OCAZs6caTAY+K/d3NyIKCIiYunSpeXl5UFBQSKRyMYKgLD+E9AlJSVVVVU9evSoXywrKyMimUwWERHBF7VaLX/4isXipioAgqs/Yu+lVCqVSmVzKwACuhvQP/zww4ULF+7cubNlyxa+cuDAgerq6g4dOiiVSv5wz87O/vzzzyMjI3NycpYtWyYSiRpWBNsPAACHc3fYO3HixL///e/3JexDDz00YcKEv/zlL/zizp07586dO2vWLH9///T09EYrjuryZfXu3TdiYr7m/x0/ni90RwDg+Jqcl1AoFMeOHVu8ePEXX3zBV/Lz8/lT26GhofzJ7oYVIioqKvrhhx8OHTokk8natPX2pNWae/f2TkiISkiI0usNGk2d0B0BgONr8iTh1KlTiYjjuLfeeisvLy84ONhsNvNDbBcXl+rqaiJqWCGiysrKlJQUvV4vlTrUGcjAQLdhwwKJyMvLcf7jAQCW/UGGikSi4ODgioqK4OBgDw8PnU6nUCg0Go2vry8RNawQUWRk5Pvvv19QULB169ZWaXHatH38XbtKSnStskIAALvQ5BRHVlaW0WgsLCxMS0sLDQ0losGDBx85ckSv1ycnJ//pT39qtNIWduy41quXIiZGpVC00RYAAFh0N6DPnTu3ZMkSf3//JUuW3Lhxg4iuXLny7rvvbt++fd68eV5eXkQ0efLkmpqajz76KC4uztvbu9FKG5k8OSQuLqxvX9+22wQAAGvuTnEMHDhw4MCB9z4wefLkyZMn31uRSCQzZsz4/QoAALQWvLsEAIBRDnWhBQA7Ll4sPXw4l//6xo3KkSMDBG0H7BICGqBN/PZb0caNF/hLM3NyKoVuB+wSAhqgrfTt67ty5SAiOnSoQOhewC5hDhoAgFEIaAAARiGgAQAYhYAGAGAUAhoAgFEIaAAARiGgAQAYheugm+3q1cpZsw7OmXOYXywqmiuX49sIAK0PydJsFgu3eHG/ESM6GgyWRx7Zz3FCNwQADgoB3RJ+frLgYGVdnUXoRgDAkWEOGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABGIaABABiFgAYAYBQCGgCAUQhoAABG4a3eAI4gO7sqLa00JuZrfvGVV/rNnt1b2JbgwSGgARxBba25c2dFQkIUEX3++bXiYp3QHUErQEADOAhPT5dhwwKJ6MCB20L3Aq0Dc9AAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKHwmYcsZjVYiGjv2B7FYRER9+/qvWTNC4J4AwIEgoFvOauWIaMKELq6u4gsXyi9cKBW6IwBwKAjoBzVmTFe5XCISUXZ2vtC9AIBDwRw0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMAr3gwZwNKmpZd9/n7N2bSq/ePLk1D59/IVtCVoGI2gAR2OxcDNnhu3fP3r//tEymchi4YTuCFoII2gAB+TlJQsOVhKRVIpBmB3DDw8AgFEIaAAARiGgAQAYxegc9MqVKefP3+G/tlpxigMAnBGjI+jTpwtFIlNMjComRsUhnwHAKTEa0ET00EMd4uLC4uLChG4EAEAY7AY0AICTQ0ADADCK0ZOEdufq1crU1JKYmK/5xXnz/jRjRqSwLQGAvUNAt47aWnPXru4JCVFEtGnT1ZISndAdAYDdQ0C3Gi8v2bBhgUS0e3ee0L0AgCO4G9AZGRlHjhwxm80JCQl8xWKxbN++vaio6LHHHhswYICNFQAAaC13TxK6urqOHj36+vXr9Q/s2LHD09Nz3rx527Zt02g0NlYAAKC13A3oiIiIqKioex84e/bsqFGj5HL5kCFDzp8/b2MFAABaS5Nz0DU1NQqFgoi8vb0rKipsrBBRenr6Z599ZjAYPDw82mEHAAAcVZPXQUulUo7jiMhkMsnlchsrRBQYGDhx4sTRo0cbDIZ22AEAAEfVZEAHBQXl5uYS0a1bt0JCQmysEFFAQMCoUaOGDh1qNBrbtHUAAMd2d4rj5s2bqampOp3u22+/ffjhhwMDAydNmrRx48ZevXqVlZX16dOHiGypAABAa7kb0N7e3pGRkZGRkUTETytHREQsXbq0vLw8KChIJBLZWAEAgNZyN6B9fHx8fHzue0ypVCqVyuZWAACgVeBmSQAAjEJAAwAwCgENAMAo3CwJwJEZDNazZ4srKvREJJNJhg3rInRH0AwIaABHplbXrVuX4uYmqauz1NSYi4peFrojaAYENICD27Dhoeho72vXKp999mehe4HmwRw0AACjENAAAIxCQAMAMAoBDQDAKAQ0AACjcBVH67t9W1dcfIu/8pSI3ngjJiDAXdiWAMAeYQTd+oqKao1Go1ar1Wq1H398obxcL3RHAGCXMIJuEyNHdpo7txcRff99jtC9AIC9wggaAIBRCGgAAEYhoAEAGIWABgBgFAIaAIBRCGgAAEYhoAEAGIWABgBgFAIaAIBRCGgAAEYhoAEAGIWABgBgFAIaAIBRuJsdgFOoqjJqtaaVK1P4xT59/P/8527CtgR/CAEN4BQ0GoPZbM3IKCKiK1c01693QkCzDwEN4Cw8PKQrVw4ioo0bs4qLzUK3A38Mc9DggMrLy48ePbpnz557i5cuXdq7d295eXmzKgACQkCDA8rMzKyoqNi3b1995Zdffjl+/Hjnzp3feecdg8FgYwVAWAhocEDDhw+fOnXqvZVDhw7NnDlzwIAB/fv3T01NtZaWWiEAABnXSURBVLECICzMQYNTKCsr8/PzI6IOHTqUlJTYWCGi7OzsHTt26HQ6d3d8NDu0NwQ0wO+Ry+WdO3euqqqyWq1C9wJOB1Mc4BT8/f35U3937twJDAy0sUJEQUFBs2bNmjhxYl1d3R9upbBQ+8YbJ4jo4sU7S5eebrO9AWeBgAYHpNFoMjMzTSZTZmamTqcjotGjR3/11VepqakXL16MiYmxsdJcZrO1stJARN26eURHe7XqPoEzwhQHOKCKioq0tLTRo0enpaV5e3srFIqHH37Y09OzoKBg8eLFMpmMiGyptIBYLCIid3epSuXKca24T+CMENDggMLCwsLCwu4r9u/fv3///s2tAAgIUxwAAIxCQAMAMAoBDQDAKAQ0AACjENAAAIxCQAMAMIqhgL5xo5KIVq5M8fHZ8NNPOUK3AwAgMIYCmr/XwfTp3ffvH+3h4SJ0OwAAAmPujSpeXq6+wUqJRCR0I61DpzOvW5fq5+dGRN7e8gULBgndEQDYDYZG0A5JrzfX1tZqtdqcnPLPPksTuh0AsCfMjaAdz0sv9YyIUF2+rE5JSRa6FwCwJxhBAwAwCgENAMAoBDQAAKMQ0AAAjEJAAwAwCgENAMAoXGbXTkpK9GVltU8/vZdfHD065MUXewvbEgAwDiPodqLVmlxdxTExqpgYVXW17uLFO0J3BACswwi6/SiVLnFxYURUUWHQaoXuBgCYh4AGcDp37uivXq3ZvDmdX3zsseBu3VTCtgSNQkADOJ1bt2pu367au/cqEf32W6mv72gENJsQ0ADOaOjQDmvWDCaiKVOOCd0LNAknCQEAGIWABgBgFAIaAIBRCGgAAEYhoAEAGIWABgBgFC6zA3Bqd+7ot23LPHeuhF9csSLWxQXjNlbgJyGA5OTSTz+9JBKt4f/t23dT6I7AeZWX19XV1Wm12srK6pUrU8xmq9AdwX9gBC2MZ57pvmRJfyKaPv2E0L2As5s4MeTJJ4Pq6iybNl0Vuhf4LwhoYbi6ilUqVyKSSkVC9wIAjMIUBwAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMavIyuy+++CItLU2hUPj6+s6fP5+I1Gr1hg0bOI4LDw+fPn16oxUAAGgtv3cd9Isvvti7d+/6xW+++WbMmDExMTHLly/Py8sLDg5uWGn7hgEAnMXvTXF8+eWXb731VnJyMr+YmZnZt29fIurXr19GRkajFSIym83V1dVarVYkwlswAABarskR9JQpU2bPnl1dXf3WW2+Fh4f7+fkZDAYXFxciUigU5eXlRNSwQkSnTp168803icjb27s99gAAwEE1GdAeHh5E5OnpGR0dnZeX5+fnJ5PJTCaTi4uLTqdTKpVE1LBCRCNGjEhNTS0oKOjRo0f77AMAgENqcopDr9cTkcViuX79emBgIBFFRUWlpaUR0aVLl6KjoxutAABAa2lyBL1hwwa1Wm0ymR5++OHOnTsT0dSpU9evX79nz56ePXvy5wMbVgAAoLU0GdAJCQn3VXx8fBITE3+/AgB2ymCwEFGnTp/xZ/djYzvv2TNB4J6cHm43CgBERBxHRLRr10iZTHLyZPGRI3eE7ggQ0ABwj6AgpVwu8fOTCd0IEOGt3gAAzEJAAwAwClMcAisrq/vqqyunThXyi++9NxxvwAQAHkbQAquoqNPr9VqtVq2uXrkyhePP1AAAYATNgqefDh01qnN1tWnLFnymMgD8B0bQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCjci4MVer2ZiKZO3cffza5fP/9Fi/5H4J4AQFAIaFaYzRwRxcR4ikSiS5cqfv21UOiOAEBgCGi2PPtsmFgsUiik+/YVC90LOC+NxlBaWvv999f4xT59/Hv08BG2JeeEgAaA++Xk1BQUVH/88Tkiys6umj9/UEICAloACGgAaERUlPc33zxKRG++eVboXpwXAppFJSX6W7eqVq5M4RdHjQr+0586CNsSALQ/gQO6okK/aNGv/Ne9ciL6ifBnFBFRQYFWq63LyCgiotOn77i5SRHQAE5I4OugtVrTP/6R4ecn9vMTGwxmYZthSo8eXitXDlq5ctCAAX5C9wIAwhB+isPVVfLXv0YS0e4UExmF7gYAgBnCBzQAsEynMxcUaM+fv8MvRkX5yuXIjXaCbzQA/J6MDM2vv5bs3n2NiAoKdOnpz0VG+grdlLNAQAPAH/jb3yLnzu1FRP377xK6F+eCmyUBADAKAQ0AwCgENAAAoxDQAACMQkADADAKAQ0AwCgENOvOnSubN++4SLSG/3fjRqXQHQFAO0FA24FFi/plZk7OzJysUrkK3QsAtB+8UcUOyOUSPpr5jysEACeBETQAAKMQ0AAAjEJAAwAwCnPQAGArvd78+eeXAwLciUilks2d21fojhwcAhoAbKXTmbOzS0tLXfPytKdP36n/vLoXXui9evXDwvbmkBDQ9kSrNU+Y8KNMJiGioCDPH34YL3RH4HTmz+8dEaHatSv32rXKr74aQUT//vfN2lqT0H05JgS0PbFYrHPmhHfo4Hb9etXXX98Suh1wajKZtE8fHyI6caJYqxW6GweFk4R2JibGf9iwwH798EmyAI4PAQ0AwChMcdgltdqgVtetXJnCL8bEBI4cGSRsSwDQ6hDQdqmsrM5gMGdkFBFRWpq6pESHgAZwPAhoe+XvL1+5chARrVmTznFCdwMAbQBz0AAAjMIIGgAeSHq6+vz58rNni/nFDz54ZNiwLsK25DAQ0ADwQHQ6c79+Pi+80IOIli49X1VlELojx4GABoAH1amT+7BhgUTk6ekidC8OBXPQAACMwggaAFqNxcJptUaNpo5f9PaWC9uPvUNA273k5Du//Vb24Yfn+cUXXugdHu5NRGKxKCFhoKCtgdO5fr3qmWf21y9aLPPFYnxQW8thisMRzJgRxn+qrJubNCenIiOjKDX1dv2tIAHa01dfPVxYOD0ra4rQjTgCjKAdgVwurf9U2Rdf7PHwwx1zcmp27sx97LHv+ScMGdJp+fKhgvYIAM2GgHZMJpNVLKYpU7oS0Zkzd9LSSoXuCJwL/+5WjaaOn+JwcZEolbjAo9kQ0A5LJBKNHRtERLW15hMnyoRux5ElJxfpdCYiunOnVuheWKHVmojIz+8TfvGJJ7odPDhJ0I7sEgIa4EHNmXO4rs6oUEh1OlMvChG6HYbcvv2MWCzatSt3375ioXuxSwhogFawevWgwYMDUlPLtsSXCN0LOA5cxQEAwCgENAC0rdTUsp9+yhGJ1vD/1q+/IHRHdgNTHADQ5mJjO2zePIyInn762JdfZpw6VcjX169/NDBQIWhrTMMI2rnodCaNpo7/J3Qv4ERcXCQqlatK5VpdbQoKcouJUcXEqPbvv8lf7AFNQUA7l9mzf/Lx2eDjsyEg4FOhewEnFRPjFxcXFhcX5uqK/PkDmOJwfDduVKem3nn66b1E9NtvRe+/P2jGjDDOQtfjhO4MnBvHUVWVgf9jTiIRe3q6Ct0RcxDQjk+tNnh5SWNiVER05EjOfY++8spRrUTftavH2rUjBGgOnFh1tTEm5mv+64gI72vXXhC2HwYhoJ1Cly6KuLgwIvroo4z7HurTx+NaieHw4VwB2gKnd/r0uJAQ5YULFfPnp/CVU6cKT58urH/C00/36NZNJVB3wkNAO7sJE0K276/Lyrrm47OBr7z66oDExIeE7QqcSm6uNj+/mh9NFxVpVSppTIw/ER08WNC/fwACGpyaxcIFByu2bXuEiJYsSf3pp5y6OjMRpaaWhIV5e3nJ9Hrz3r03u3f34p8/bFiXpUuHCNkxOBaDweLtLUtIiCKipKQLsbGBiYkDiOjyZQ3/BP7WS2azVaMxkjPdegkBDURErq6S4GAlEVVWmqxWq1arJaKTJws8PERSqaK0VF9YWLNwYW8i+vXXksuXceslaGVubhL+Uw1VqkaSV6s1Ern++mvBQJ/viCgy0nfmzCj+occfD+nXL6A9W21PCGi435AhHRYt6kdEW7dei4sLHzIkIC2tYt++2/y98SorjefOVQrdIziFmzerJ03a4+IiHund669dRg4eHJD5zeSlS8//9ltpRkYREf30U8Hx4/n9+98N6IULB3t5yQRtuZUhoKF5rl6t/OWXgvqPApg/P+aJJ7oJ2xI4KquVe++9mIED/e8clNHPJJGIVCpXFxdxZKTXypWDiOj48SKj0cD/wbdxY9bp04VyuZSIQkI8t2x5XODuWwMCGpqnqsrUpYsb/1EAGzdmFRVp+frWrZetVo7/2s/PbeLEcMFaBAfi7y8PDlYavJr8YMMnnug6Y0YYEX36aeaTT3by8ZH9/HPJl19e2bnzOv8Euz7p/aABrdVqy8vLg4KCxGK8KchZdOzozk93rFt3eePGtJ9+yiWinTuzR47sJJNJCgp0xcV1KSl37//76KNBo0eHENHt2zUmk5UvisUUEsLiqXkcz3Zt5MhOnTsrSkvr6k96v/TSr0lJZ5KSzvBPOHFi6ogRXQXtsXkeKKCzs7M///zzyMjInJycZcuWiUT4+F7nUllp6N3bh38LzPffc6tWDQoIcPv008zNm6/yf3WePn3HxUXMB/To0d8XFNRIJCKLhbNYuGXL7g5qIiN9//KX7sLtxH/geHYY9Se9ZTLpCy/0mD+/NxHFxu5LSjrz6afuer358OFchcKFiCwWa1SU33PPRRNRQUFNt26qrl09iOjgwZzu3b2kUjHd8xdhcnLR5cvl/CaOHMkLDHTnVyKXS9tokP5AAb1z5865c+d269btk08+SU9P79u3r40vfOih7Vevqomo/o9isFNDhgRMm9adiBYsSKkvBga686cZJ08+9vnnlw8ezCGi3Nyqb78dGRPjl5xc+swzx/mTPJcuVbi7y4uLdURkMFgeeqhTaOjdkbWXl7ydE7LFxzOwTCaT8B+prNOZevRQhIZ6FhRoxWLav380ESUlXUhPV+/de5WIjh4t8PNz8/GREVF6uvqxxzrLZJLbt7WFhfpDh3KI6LffivV6Y48eKiI6cOD2hAkhgYFuZWV1e/fmu7ndzdL6AYdWazKZLPVttOx4fqCAzs/PDwkJIaLQ0NDc3Fz+gC4qKvrtt980Go1EItm8eXOjL8zLOxMSIlepXKurjenp6vj4AiIS3fK6KfHXfFdReVJdXZ2/bduVX35REhFRzqJFRa6ukmvXKouK9PHxmURUUFB48KAiN/coEVksuWvX5np7y8rK9DpdWXx8PhGlpZVKpaL4+N+ISKfL/+qrzGPHlFYrEeUkJBS7uIgzMyvLyvTx8RlEVFxcuG+f4vp1LyKyWnNXrcpTqWTFxbXV1eXx8XlElJl55+ZNSXz8GSKqrc3//POrP/2kMJksRPnz55eIxaLLl9WVlcb4+HQiKi0t+PHH1CtXVETEcbnvvpvv4eF6+7a2oqIyPv4mEV27VlJY6BIff4qI6uryNm26vnu3e22tmej2vHklRKK0tAqt1hwff5GIKioKdu68eOmSJ/+tWL68UKGQ5uZqy8oq4+NvENGNG8Xl5a4VFT8TkdGY9+mnN3bscKupMVkshfHxxUR04UK5wWCJj79ARJWVt7/9Nj0lxUNEFFIcTkS33865lq8uKqqJj79GRLm5xTU1spKSE0RkMuWtX39z+3Y3jcZgNJbExxcSUWpqmcXCxcenElFVVf6//pVx+rQH39vbbxfJZJLs7KqCAl18fBYRZWUVubqKTSY3IjIYNB98kOPjIysv1xOV1dYWE5FGU5qRoUtJ2dXwIBk5MkgqlRCRj4+8qT9L+/fvP3DgQFuOVVu0+HjWaM589NF5X195RUVdebVoX3FeXVXdBfUFjqP4+BQiwvH8gMeztEjRtaKTvqa2JL6w0eOZiDguZ9myQjc36c2b1U0dz1lZN8vK3DQag8lUsmZNERFlZZWJxZyrK38dSH5AgHdIiAcRpafnuLoGubpKLJaqysrKnTuTiaiqyhgW5unq6k1ERLkVFXkcJysv1+v1JQsXnuIPBoXCxc/PjYhKS2v1enP9QdK3bwA/Hu/QwX38+LBGD6SGx7OI41o+hp0zZ86WLVuI6OjRoxUVFVOnTiWizMzMf/7znxzH6XS60tLGP0xarzfzY2eO4/R6i7u7lIhMes5q5lzdxCIpGQwWiUTE709trdnNTSoSkcXCmUxWuVxCREajRSQSubiI+bXJZBKxWGQ2mzWaMn//jkRkMlmtVk4mkxBRXZ1FKhVLpaJmrc1q5QwGC/8fo8lk5TjO1fW/1sZxpNeb3dykFoupulrj7u7bsrW5uIglkrtr478VZjNnsVj55g0Gi1h8d221tWa5/D9rk0qtWm2VUun7R2uzWixc/drufmM5qq2yyOQSiYwsVs5ksvCnv41GKxHxtxmrqzO7uEgkEtG9Pyaz2Wow1JnNepXK15YfE5GIX5teb3Z1lUgkIquVq6uzNNHb/T8mIhKLRfz3rba2VqfT+fv71x9IEydOnDZtWosP4PvYfjwbDAaNRhMYGNjI8ayzyF1cSERWsaWRb/sDHIGtdTzr9Vqj0ezm5tHU8SwSkdnMmc0t/O1o8fH8O2uzmqlOa3Fzl9TUqaVSmUQib+IbKxGJRBZL847nZv2YHuR41mhKPT29XV1d64fb92nkeOYewOuvv67VajmO++67744cOfIgq2oVBQUFI0aMaP/tXr16dezYse2/3fPnz0+dOrX9t/vzzz/Pnj27/bd74MCB+Pj4tlu/7cfz2bNnp0+f3nadtKkvvvjinXfeEbqLFpo/f/7u3buF7qKFHn/88Zs3bzbrJQ90qnrw4MFHjhzR6/XJycl/+tOfHmRVrUIqlXbs2LH9t+vi4lI/mGpPMpksIECA91C5ubndO4xtN+7u7n5+fm23ftuPZ7lcLsh3vlV4enp6eXkJ3UUL+fj4uLu7C91FC3Xs2FEqbd6s8gNNcVgslu3btxcWFo4ePXrAgAEtXg8AC3A8A2seKKABAKDt2Nk7CY8ePbpz506TybR161a+olarN2zYwHFceHj49OnTbaw0i9FoXLFihcViMZlMzz77bL9+/YjoypUr27dvF4lE48aNGzRokI2V5nrnnXdqa2v1ev3YsWMfffTRdtsuv9dz58596qmnxo0b127bfeONN+RyORHFxMRMnDixPffXdu28ueaqq6t77rnnQkNDiejpp5/mD9c9e/acP3+e47iXX36Znwa0pdJubt269cEHHxQXF3/66af83JHFYvnkk08qKyvd3NxeffVVmUzWsoogzZ87d27Tpk381wsWLFCpVC1vvpWnwdtYZWWl2WyeM2dOfeWTTz45d+4cx3FJSUm5ubk2VprFbDaXlZVxHKdWq1955RW+OG/evOrqar1e/7e//c1sNttYacGmOY6rra194YUX2nO7HMft3LkzKSmp/oRM+2z39ddfv6/S3O22bGebpVW+vW1Hr9e/9dZb91bKy8sTEhKsVmt2dva7775rY6U98dfnvP3223fu3OErv/zyy5dffslx3M6dO/fu3dviiiDNp6SkbN++/d7ntLh5O3s/q0qlkkgk91YyMzP561X79euXkZFhY6VZJBIJf25KLBbzW6+qqnJ1dfXw8JDL5YGBgYWFhbZUWrC//OYMBgN/fW67bVen02VmZtafKGu37YpEosWLFyclJeXl5bVsuy3YaLO0ym62tby8vIULF3744Yf8+zmvXr0aHR0tEonCw8P5b6wtlfbk5uZ236m/K1eu1P/OXrlypcUVQZonohMnTixYsODf//43x3EP0rydTXE0ZDAYXFxciEihUJSXl9tYaZnNmzdPmTKFiLRabf2PRKFQ1NTUSCSSP6y0bKMrVqy4fv36M888057b3blz54QJE3Jzc/nFdttuYmKih4fH9evX165du379+hZstwUbbZaGLbX1FptLJpNt2rRJqVTu2bPn66+/fvnll+/t2Wq10n/vRVMVYdX3o1Qq+W9yyyqCiIqK2rhxIxGtXr369OnTsbGxLW7ezkbQDclkMpPJREQ6nU6pVNpYaYF//vOfAQEBw4YNIyKlUllbW8vXdTqdh4eHLZWWbXfJkiWfffbZrl27ampq2me7Go2msLAwKiqqvtJu+8u/Kjw83GKxGI3GFmy3BRttltb6sbYdkUjEH+QPP/zwzZs36b975m8CZUtFWPX9aLVa/pvcsoog3N3dJRKJRCKJjY29cePGgzQv/E/iAUVFRaWlpRHRpUuXoqOjbaw01/79+9Vq9axZs/hFlUplNBpramrq6upKSko6d+5sS6W5G+U4zmKxEJHr/9c+283JyVGr1YmJiQcOHPjpp5/S09PbZ7tWq9VoNBJReXm52Wxu2f42d6PN9eC72dbq6ur4P6szMjL4c329evXKyMjgOC47O5ufK7OlIqyW/V4/+G96q9Dr9fwXV65c4X8ELW7ezi6zO3fu3J49e7Kysnr16jVr1qywsDC1Wr1+/Xqr1dqzZ8/6azb+sNIsVVVVM2fOjIqKkkgkUql02bJlRJSRkcGfyh8/fjx/Kt+WSrPo9fq///3vYrHYYDA88cQT/FUc7bDdevv377dYLPxVHO2w3dra2rffftvNzc1oNMbFxfHzce25vzZq5801V0ZGxpdffimXy8VicXx8PH8twZ49e86dO0dEr7zySv01G39YaTfl5eUfffTRrVu3unTpMmrUqMcee4y/pEGtVisUinsvcmhuRZDmd+3alZycLJFIOnToEB8fL5FIWty8nQU0AIDzsPspDgAAR4WABgBgFAIaAIBRCGgAAEYhoAEAGIWABgBg1P8Dc1cf54gTvukAAAAASUVORK5CYII=",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "plot_covariates(sceptre_object_highmoi, p_mito_threshold = 0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a7e35402",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178488.23-P1P2 (1 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178620.23-P1P2 (5 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACTL7B_-_111617995.23-P1P2 (10 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR1B_+_52345762.23-P1P2 (15 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2A_+_148602343.23-P1P2 (20 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_+_38495339.23-P1P2 (25 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_-_38495819.23-P1P2 (30 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ADCY2_-_7396429.23-P1P2 (35 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA AMOTL2_+_134093279.23-P1P2 (40 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ANP32E_+_150208342.23-P1P2 (45 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA APAF1_-_99039071.23-P1P2 (50 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_+_72766725.23-P1P2 (55 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_-_72766755.23-P1P2 (60 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARTN_-_44399429.23-P1 (65 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASCC3_-_101329193.23-P1P2 (70 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASH2L_-_37963059.23-P1P2 (75 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATMIN_+_81069661.23-P1P2 (80 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_+_111992112.23-P1P2 (85 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_-_111992129.23-P1P2 (90 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP6AP1_-_153657075.23-P1P2 (95 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATXN2L_-_28834668.23-P1P2 (100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAD_+_64052102.23-P1P2 (105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAG6_+_31620398.23-P1P2 (110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_+_49458268.23-P1P2 (115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_-_49458323.23-P1P2 (120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BCAS2_-_115124045.23-P1P2 (125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BMPR1A_-_88516386.23-P1P2 (130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BRD3_-_136932911.23-P1P2 (135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C17orf49_+_6918114.23-P2 (140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_+_31532436.23-P1P2 (145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_-_31532650.23-P1P2 (150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACFD1_-_136325160.23-P1P2 (155 of 1926)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing gRNA-to-cell assignments for gRNA CACNG1_+_65040756.23-P1P2 (160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CAPN5_+_76778225.23-P1P2 (165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CASP3_+_185570560.23-P1P2 (170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_+_107384793.23-P1P2 (175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_-_107384635.23-P1P2 (180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCM2_-_45039941.23-P1P2 (185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNC_-_100016267.23-P1P2 (190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNT2_-_135676416.23-P1P2 (195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCT6B_+_33288381.23-P1P2 (200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_+_30366413.23-P1P2 (205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_-_30366630.23-P1P2 (210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD99L2_-_150067120.23-P1P2 (215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDH5_-_66400630.23-P1P2 (220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDKN1A_-_36646538.23-P1P2 (225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CENPC_+_68411190.23-P1P2 (230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_+_30773828.23-P1P2 (235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_-_30773391.23-P1P2 (240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHMP3_-_86790451.23-P1P2 (245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHORDC1_-_89955894.23-P1P2 (250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNEP1R1_+_50059284.23-P1P2 (255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNOT2_+_70637651.23-P1P2 (260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_+_83956339.23-P1P2 (265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_-_83956572.23-P1P2 (270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS6_-_99686676.23-P1P2 (275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_-_31634088.23-P1P2 (280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CTR9_+_10772881.23-P1P2 (285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CUL3_-_225449667.23-P1P2 (290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_+_145150267.23-P1P2 (295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_-_145149996.23-P1P2 (300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYCS_+_25164869.23-P1P2 (305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DACH1_-_72441101.23-P1P2 (310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DBR1_+_137893673.23-P1P2 (315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DCTN2_-_57940910.23-P1P2 (320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_+_61851532.23-P1P2 (325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_-_61851812.23-P1P2 (330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DGCR8_-_20068158.23-P1P2 (335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNAJC17_-_41099215.23-P1P2 (340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNM2_+_10829157.23-P1P2 (345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM1_+_49575018.23-P1P2 (350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_+_130700056.23-P1P2 (355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_-_130700076.23-P1P2 (360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM3_+_155113020.23-P1P2 (365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DROSHA_-_31531982.23-P1P2 (370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_+_3985081.23-P1 (375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_-_3985443.23-P1 (380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EHMT2_-_31865199.23-P1P2 (385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3F_+_8009347.23-P1P2 (390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3M_-_32605405.23-P1P2 (395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4E_+_99849831.23-P1P2 (400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184032364.23-P1P2 (405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184033042.23-P1P2 (410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G2_+_10830447.23-P1 (415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF5A_-_7211217.23-P1P2 (420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TCEB3_-_24069975.23-P1P2 (425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EMC2_+_109456080.23-P1P2 (430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823220.23-P1P2 (435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823252.23-P1P2 (440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EOMES_-_27763342.23-P1P2 (445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERCC2_+_45873802.23-P1P2 (450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERO1L_+_53162425.23-P1P2 (455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETS1_+_128392044.23-P1P2 (460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_+_36134592.23-P1P2 (465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_-_36134789.23-P1P2 (470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EWSR1_-_29664347.23-P1P2 (475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EXOSC8_-_37574979.23-P1P2 (480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FBXW11_-_171433300.23-P1P2 (485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FEN1_+_61560782.23-P1P2 (490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_+_75532080.23-P1P2 (495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_-_75532461.23-P1P2 (500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT7_-_139926495.23-P1P2 (505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA G3BP1_-_151151499.23-P1P2 (510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GCK_+_44229021.23-P1P2 (515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOLT1B_+_21654777.23-P1P2 (520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_+_45000529.23-P1P2 (525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_-_45000948.23-P1P2 (530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPI_+_34856284.23-P1P2 (535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPR107_-_132816044.23-P1P2 (540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GSPT1_-_12009714.23-P1P2 (545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A1_+_81687215.23-P1P2 (550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_+_59949355.23-P1P2 (555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_-_59949657.23-P1P2 (560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2H4_-_30876185.23-P1P2 (565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF3C2_+_27579604.23-P1P2 (570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HAND1_+_153857732.23-P1P2 (575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HDAC2_-_114292253.23-P1P2 (580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_+_81399312.23-P1P2 (585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_-_81399351.23-P1P2 (590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HINT2_+_35814995.23-P1P2 (595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNF4A_+_43029971.23-P1P2 (600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPM_-_8509885.23-P1 (605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPU_+_245027753.23-P1P2 (610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_+_43702292.23-P1P2 (615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_-_43702305.23-P1P2 (620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSPA5_-_128003534.23-P1P2 (625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HUWE1_-_53710540.23-P1P2 (630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_-_30007588.23-P1P2 (635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INTS10_+_19675000.23-P1P2 (640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_+_65903046.23-P1P2 (645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_-_65903523.23-P1P2 (650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C7orf26_+_6630358.23-P1P2 (655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA IPO8_+_30848771.23-P1P2 (660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ITGB1BP1_+_9563543.23-P1P2 (665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSFA2_+_182756743.23-P1P2 (670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_+_109914946.23-P1P2 (675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_-_109915112.23-P1P2 (680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM1A_-_23346409.23-P1P2 (685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM3B_-_137688431.23-P1P2 (690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDR_-_55991659.23-P1P2 (695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA0232_+_6784651.23-P1P2 (700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153706791.23-P1P2 (705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153707248.23-P1P2 (710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAMTOR4_-_99746921.23-P1P2 (715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LARP4_-_50794864.23-P1P2 (720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LATS2_-_21635336.23-P1P2 (725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LEMD3_+_65563441.23-P1P2 (730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_+_54960428.23-P1P2 (735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_-_54960616.23-P1P2 (740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LHFPL2_-_77944519.23-P1P2 (745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LSM12_-_42143603.23-P1P2 (750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MANBAL_+_35918136.23-P1P2 (755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MAU2_+_19431732.23-P1P2 (760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_+_47807678.23-P1P2 (765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_-_47807877.23-P1P2 (770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEAF6_-_37980319.23-P1P2 (775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED1_-_37607140.23-P1P2 (780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED12_+_70338541.23-P1P2 (785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED15_+_20862189.23-P1 (790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_+_57479335.23-P1P2 (795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_-_57479437.23-P1P2 (800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEMO1_-_32264882.23-P1 (805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MIXL1_+_226411425.23-P1P2 (810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL28_+_420487.23-P1P2 (815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL32_+_42972018.23-P1P2 (820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_+_179322181.23-P1P2 (825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_-_179322213.23-P1P2 (830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LUZP6_-_135612750.23-P1P2 (835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA20_-_19998023.23-P1P2 (840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA25_+_112546538.23-P1P2 (845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCBP1_+_100396082.23-P1P2 (850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_+_172428630.23-P1P2 (855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_-_172428841.23-P1P2 (860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NEDD8_-_24701363.23-P1P2 (865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NF2_+_29999635.23-P1P2 (870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NFIB_+_14314559.23-P1P2 (875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_+_139418329.23-P2 (880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139418349.23-P2 (885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139440503.23-P1 (890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUFIP2_+_27621073.23-P1P2 (895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUP93_-_56764090.23-P1P2 (900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OGT_+_70752948.23-P1P2 (905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_+_59383170.23-P1P2 (910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_-_59383348.23-P1P2 (915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAM16_-_4401236.23-P1P2 (920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAXBP1_-_34143614.23-P1P2 (925 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PBLD_-_70092436.23-P1 (930 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PCBP2_+_53835614.23-P2 (935 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_+_94292752.23-P2 (940 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_-_94293027.23-P2 (945 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PGRMC2_+_129208913.23-P1P2 (950 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_-_178866211.23-P1P2 (955 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PITPNC1_-_65373433.23-P1P2 (960 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLCG1_+_39766176.23-P1P2 (965 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_+_151042273.23-P2 (970 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_-_151042422.23-P2 (975 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLRG1_-_155471154.23-P1P2 (980 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLXNA2_-_208417275.23-P1P2 (985 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLE4_-_75185725.23-P1P2 (990 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR2M_+_57998962.23-P1P2 (995 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940045.23-P1P2 (1000 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940414.23-P1P2 (1005 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3K_-_103144.23-P1P2 (1010 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPHLN1_-_42720007.23-P1P2 (1015 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPIL4_+_149867146.23-P1P2 (1020 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPP2R1A_+_52693317.23-P1 (1025 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_+_53195217.23-P1P2 (1030 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_-_53195332.23-P1P2 (1035 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKD1_-_30396996.23-P1P2 (1040 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRMT5_+_23398554.23-P1 (1045 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C2orf53_-_27362230.23-P1P2 (1050 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRSS8_+_31147010.23-P1P2 (1055 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_+_36909062.23-P1P2 (1060 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_-_36909230.23-P1P2 (1065 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMD4_-_151227289.23-P1P2 (1070 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMG4_-_3259227.23-P1P2 (1075 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTAR1_+_72374835.23-P1P2 (1080 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTBP1_+_797463.23-P1P2 (1085 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_+_89623388.23-P1P2 (1090 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_-_89623526.23-P1P2 (1095 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTMS_-_6875591.23-P1P2 (1100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PYGO2_-_154934135.23-P1P2 (1105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB18_-_27793292.23-P1P2 (1110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB6A_+_73471983.23-P1P2 (1115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_+_76251918.23-P1P2 (1120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_-_76252008.23-P1P2 (1125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GNB2L1_-_180670840.23-P1P2 (1130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAD21_-_117886635.23-P1P2 (1135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RANGRF_+_8192257.23-P1P2 (1140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBBP5_+_205091050.23-P1P2 (1145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM14-RBM4_+_66406280.23-P2 (1150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM33_+_155437802.23-P1P2 (1155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM4_+_66406319.23-P2 (1160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_+_114271352.23-P1P2 (1165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_-_114271700.23-P1P2 (1170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBX1_-_41347852.23-P1P2 (1175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1551_+_32112399.23-P1P2 (1180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RFX3_-_3525750.23-P1P2 (1185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RGP1_+_35749328.23-P1P2 (1190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_+_95067102.23-P1P2 (1195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_-_95067164.23-P1P2 (1200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RIBC1_-_53449606.23-P1P2 (1205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1432_+_5629173.23-P1P2 (1210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RLF_+_40627522.23-P1P2 (1215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF115_-_145611148.23-P1P2 (1220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_+_104296142.23-P1P2 (1225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_-_104296250.23-P1P2 (1230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF40_-_30773681.23-P1P2 (1235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPL37A_-_217363600.23-P1P2 (1240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPS15A_-_18801590.23-P1P2 (1245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAV1_+_51134992.23-P1P2 (1250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_+_33104038.23-P1P2 (1255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_-_33104343.23-P1P2 (1260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF8_+_155054976.23-P1P2 (1265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EPT1_+_26569050.23-P1P2 (1270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERF2_+_44069659.23-P2 (1275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERPING1_+_57365137.23-P1P2 (1280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_+_122242041.23-P1P2 (1285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_-_122242294.23-P1P2 (1290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETDB1_-_150898815.23-P1P2 (1295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SF1_-_64545937.23-P1P2 (1300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCDC101_-_28565268.23-P1P2 (1305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SKA3_+_21750519.23-P1P2 (1310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_+_78591497.23-P1P2 (1315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_-_78591737.23-P1P2 (1320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC29A1_-_44187438.23-P2 (1325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC35C1_-_45826711.23-P1P2 (1330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD1_+_146403428.23-P1P2 (1335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD3_+_67358229.23-P1P2 (1340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_+_48556623.23-P1P2 (1345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_-_48556657.23-P1P2 (1350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD5_-_135468694.23-P1P2 (1355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA4_-_11071741.23-P1P2 (1360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA5_+_144435243.23-P1P2 (1365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCB1_+_24129228.23-P1 (1370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_+_47822975.23-P1P2 (1375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_-_47822893.23-P1P2 (1380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCD3_-_150974212.23-P1 (1385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCE1_-_38803669.23-P1P2 (1390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMC2_-_106856631.23-P1P2 (1395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMG5_+_156252537.23-P1P2 (1400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_+_103852.23-P1P2 (1405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_-_104242.23-P1P2 (1410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP27_-_70121181.23-P1P2 (1415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNUPN_-_75917960.23-P1P2 (1420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SOX4_+_21594213.23-P1 (1425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPIN1_+_91003737.23-P1P2 (1430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_+_2328680.23-P1P2 (1435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_-_2328912.23-P1P2 (1440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPTY2D1_-_18655846.23-P1P2 (1445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP19_+_112197188.23-P1P2 (1450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP54_-_35452244.23-P1P2 (1455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP68_+_74068540.23-P1P2 (1460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_+_126138730.23-P1P2 (1465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_-_126138742.23-P1P2 (1470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPRB_-_133524719.23-P1P2 (1475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRRM2_-_2802714.23-P1P2 (1480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRSF6_+_42086993.23-P1 (1485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSB_+_170655450.23-P1P2 (1490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_+_57102902.23-P1P2 (1495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_-_57103355.23-P1P2 (1500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STAG2_-_123095658.23-P1 (1505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STK40_+_36851507.23-P1P2 (1510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUPT20H_+_37633798.23-P1P2 (1515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUZ12_+_30264157.23-P1P2 (1520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_+_64901973.23-P1P2 (1525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_-_64901981.23-P1P2 (1530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TADA2B_-_7045121.23-P1P2 (1535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF10_+_6633406.23-P1P2 (1540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF2_+_120845032.23-P1P2 (1545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF3_-_7860540.23-P1P2 (1550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_+_60640814.23-P1P2 (1555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_-_60640498.23-P1P2 (1560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF6L_-_62538954.23-P1P2 (1565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAOK1_-_27717907.23-P1P2 (1570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FAM208A_+_56717273.23-P1P2 (1575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBC1D9_+_141677481.23-P1P2 (1580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_+_170863524.23-P1P2 (1585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_-_170863481.23-P1P2 (1590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBX3_-_115121998.23-P1P2 (1595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA T_-_166581333.23-P1P2 (1600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TGFBR2_+_30648526.23-P1P2 (1605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA THBD_+_23029196.23-P2 (1610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_+_2977592.23-P1P2 (1615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_-_2977643.23-P1P2 (1620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MESDC1_-_81293340.23-P1P2 (1625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMEM41B_+_9336156.23-P1P2 (1630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMTC3_-_88536127.23-P1P2 (1635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TOP1_+_39657538.23-P1P2 (1640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_+_7590531.23-P1P2 (1645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_-_7590704.23-P1P2 (1650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRIM24_-_138145308.23-P1P2 (1655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRRAP_-_98476465.23-P1P2 (1660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TSTD2_+_100395594.23-P1P2 (1665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UBE2E1_+_23847541.23-P1P2 (1670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_+_38924011.23-P1P2 (1675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_-_38924421.23-P1P2 (1680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UPF2_+_12084799.23-P1P2 (1685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UXT_-_47518505.23-P1P2 (1690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VHL_-_10183525.23-P1P2 (1695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VPS54_+_64246104.23-P1P2 (1700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_+_28821698.23-P1P2 (1705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_-_28822074.23-P1P2 (1710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAPAL_+_88281533.23-P1P2 (1715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR12_-_203776383.23-P1P2 (1720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR77_+_111991860.23-P1P2 (1725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WIZ_-_15543016.23-P1P2 (1730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_+_9983328.23-P1P2 (1735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_-_9983510.23-P1P2 (1740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XPO5_-_43543670.23-P1P2 (1745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XRN1_-_142166648.23-P1P2 (1750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA YARS2_+_32908787.23-P1P2 (1755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB22_+_33285454.23-P1P2 (1760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_+_6640093.23-P1P2 (1765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_-_6640138.23-P1P2 (1770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZC3H8_-_113012435.23-P1P2 (1775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZNF281_+_200379151.23-P1P2 (1780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00117 (1785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00233 (1790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00518 (1795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00656 (1800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01226 (1805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01730 (1810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02009 (1815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02288 (1820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02777 (1825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03026 (1830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03252 (1835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03620 (1840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_SGTN_04 (1845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_SGTN_25 (1850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_SGTN_26 (1855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1044 (1860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1414 (1865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1554 (1870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1785 (1875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2009 (1880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2271 (1885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2425 (1890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2614 (1895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_293 (1900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_541 (1905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_725 (1910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_860 (1915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_SGTN_14 (1920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_SGTN_01 (1925 of 1926)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- assign_grnas(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  method = \"mixture\",\n",
    "  parallel = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3a4a320d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     cell_id               grna_target value\n",
      "1 CCTAATCCAACAACCATCTATTAC_2    HGF_+_81399312.23-P1P2     1\n",
      "2 CCTAATCCAACAACCATCTATTAC_2  SMG5_+_156252585.23-P1P2     1\n",
      "3 CCTAATCCAACAACCATCTATTAC_2 WDR77_+_111991472.23-P1P2     1\n",
      "4 CCTAATCCAACAACCATCTATTAC_2 ZBTB22_-_33285488.23-P1P2     1\n",
      "5 CCTAATCCAACAACCATCTATTAC_2             WDR61_SGTN_13     1\n",
      "6 GGTGCGAACCTCCTGAATCCGCGA_2   EIF5A_-_7210911.23-P1P2     1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "⚠️ Matrix missing colnames. Fetching from sceptre_object...\n",
       "✅ Saved long CSV to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/grna_assignments_long.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(Matrix)\n",
    "\n",
    "# 1. Get the assignment matrix\n",
    "A <- get_grna_assignments(sceptre_object_highmoi)\n",
    "\n",
    "# 2. Ensure it is a numeric sparse matrix (1/0)\n",
    "if (inherits(A, \"Matrix\") || is.matrix(A)) {\n",
    "    A_num <- if (inherits(A, \"Matrix\")) drop0(A * 1) else (A * 1)\n",
    "} else {\n",
    "    stop(\"Output is not a matrix.\")\n",
    "}\n",
    "\n",
    "# --- FIX: ROBUST NAME RECOVERY ---\n",
    "# We retrieve names using Sceptre's standard metadata accessors\n",
    "# instead of touching internal slots like @grna_matrix.\n",
    "\n",
    "# Try to get cell names from the matrix first\n",
    "current_cell_names <- colnames(A_num)\n",
    "current_guide_names <- rownames(A_num)\n",
    "\n",
    "# If missing, grab them from the main object's covariate data frame\n",
    "if (is.null(current_cell_names)) {\n",
    "    message(\"⚠️ Matrix missing colnames. Fetching from sceptre_object...\")\n",
    "    # The covariate data frame ALWAYS has the cell IDs as rownames\n",
    "    current_cell_names <- rownames(sceptre_object_highmoi@covariate_data_frame)\n",
    "    \n",
    "    # Safety check: lengths must match\n",
    "    if (length(current_cell_names) != ncol(A_num)) {\n",
    "        stop(paste(\"Mismatch! Matrix has\", ncol(A_num), \"cols but object has\", length(current_cell_names), \"cells.\"))\n",
    "    }\n",
    "    colnames(A_num) <- current_cell_names\n",
    "}\n",
    "\n",
    "# If guide names are missing, they should match the unique values in your guide mapping\n",
    "if (is.null(current_guide_names)) {\n",
    "    message(\"⚠️ Matrix missing rownames. Fetching from guide mapping...\")\n",
    "    # Get unique targets from the mapping table you loaded earlier\n",
    "    # Note: This assumes the matrix rows are sorted/aligned to these unique targets.\n",
    "    # In 'mixture' method, rows usually correspond to the 'grna_target' labels.\n",
    "    \n",
    "    # SAFE FALLBACK: If we can't be 100% sure of order, we warn the user.\n",
    "    # But usually, they match the 'grna_target' column from your input.\n",
    "    unique_targets <- sort(unique(guide_mapping$grna_target))\n",
    "    \n",
    "    if (length(unique_targets) == nrow(A_num)) {\n",
    "         rownames(A_num) <- unique_targets\n",
    "    } else {\n",
    "         warning(\"Could not automatically align target names. Using generic IDs.\")\n",
    "         rownames(A_num) <- paste0(\"target_\", seq_len(nrow(A_num)))\n",
    "    }\n",
    "}\n",
    "# ---------------------------------\n",
    "\n",
    "# 3. Convert to Long Format\n",
    "s <- summary(A_num)\n",
    "assign_long <- data.frame(\n",
    "    cell_id = colnames(A_num)[s$j],\n",
    "    grna_target = rownames(A_num)[s$i],\n",
    "    value = s$x\n",
    ")\n",
    "\n",
    "# 4. Save\n",
    "out_path <- file.path(data_dir, \"grna_assignments_long.csv\")\n",
    "write.csv(assign_long, out_path, row.names = FALSE)\n",
    "\n",
    "message(\"✅ Saved long CSV to: \", out_path)\n",
    "print(head(assign_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bf6f0a89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/MOI.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "png(file.path(data_dir, \"MOI.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"MOI.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "014235ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/QC.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_qc(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  p_mito_threshold = 0.25\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"QC.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"QC.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "233e48fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Constructing negative control pairs. ✓\n",
      "Running precomputation on response ENSG00000160741.17 (1 of 259)\n",
      "Running precomputation on response ENSG00000286221.1 (5 of 259)\n",
      "Running precomputation on response ENSG00000162971.11 (10 of 259)\n",
      "Running precomputation on response ENSG00000183087.15 (15 of 259)\n",
      "Running precomputation on response ENSG00000228566.2 (20 of 259)\n",
      "Running precomputation on response ENSG00000213235.3 (25 of 259)\n",
      "Running precomputation on response ENSG00000120645.12 (30 of 259)\n",
      "Running precomputation on response ENSG00000149100.13 (35 of 259)\n",
      "Running precomputation on response ENSG00000142046.15 (40 of 259)\n",
      "Running precomputation on response ENSG00000143748.18 (45 of 259)\n",
      "Running precomputation on response ENSG00000149311.20 (50 of 259)\n",
      "Running precomputation on response ENSG00000133980.5 (55 of 259)\n",
      "Running precomputation on response ENSG00000196312.16 (60 of 259)\n",
      "Running precomputation on response ENSG00000113522.14 (65 of 259)\n",
      "Running precomputation on response ENSG00000259673.6 (70 of 259)\n",
      "Running precomputation on response ENSG00000100580.8 (75 of 259)\n",
      "Running precomputation on response ENSG00000235655.3 (80 of 259)\n",
      "Running precomputation on response ENSG00000196323.14 (85 of 259)\n",
      "Running precomputation on response ENSG00000204520.14 (90 of 259)\n",
      "Running precomputation on response ENSG00000146834.15 (95 of 259)\n",
      "Running precomputation on response ENSG00000102978.13 (100 of 259)\n",
      "Running precomputation on response ENSG00000164758.7 (105 of 259)\n",
      "Running precomputation on response ENSG00000171097.14 (110 of 259)\n",
      "Running precomputation on response ENSG00000111704.11 (115 of 259)\n",
      "Running precomputation on response ENSG00000008086.13 (120 of 259)\n",
      "Running precomputation on response ENSG00000175390.15 (125 of 259)\n",
      "Running precomputation on response ENSG00000100300.18 (130 of 259)\n",
      "Running precomputation on response ENSG00000196678.14 (135 of 259)\n",
      "Running precomputation on response ENSG00000119638.14 (140 of 259)\n",
      "Running precomputation on response ENSG00000116120.11 (145 of 259)\n",
      "Running precomputation on response ENSG00000112667.13 (150 of 259)\n",
      "Running precomputation on response ENSG00000111203.13 (155 of 259)\n",
      "Running precomputation on response ENSG00000160606.11 (160 of 259)\n",
      "Running precomputation on response ENSG00000242612.7 (165 of 259)\n",
      "Running precomputation on response ENSG00000103404.15 (170 of 259)\n",
      "Running precomputation on response ENSG00000097046.13 (175 of 259)\n",
      "Running precomputation on response ENSG00000213722.10 (180 of 259)\n",
      "Running precomputation on response ENSG00000133104.14 (185 of 259)\n",
      "Running precomputation on response ENSG00000107771.17 (190 of 259)\n",
      "Running precomputation on response ENSG00000187720.15 (195 of 259)\n",
      "Running precomputation on response ENSG00000108344.15 (200 of 259)\n",
      "Running precomputation on response ENSG00000135424.19 (205 of 259)\n",
      "Running precomputation on response ENSG00000198408.14 (210 of 259)\n",
      "Running precomputation on response ENSG00000162374.18 (215 of 259)\n",
      "Running precomputation on response ENSG00000187987.10 (220 of 259)\n",
      "Running precomputation on response ENSG00000126464.15 (225 of 259)\n",
      "Running precomputation on response ENSG00000125149.12 (230 of 259)\n",
      "Running precomputation on response ENSG00000154079.6 (235 of 259)\n",
      "Running precomputation on response ENSG00000162769.13 (240 of 259)\n",
      "Running precomputation on response ENSG00000135763.10 (245 of 259)\n",
      "Running precomputation on response ENSG00000290018.1 (250 of 259)\n",
      "Running precomputation on response ENSG00000102241.12 (255 of 259)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00259&non-targeting_02159&non-targeting_02571&non-targeting_03029&non-targeting_03081 (1 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00355&non-targeting_02025&non-targeting_02089&non-targeting_03081&non-targeting_03384 (5 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00518&non-targeting_03081&non-targeting_03342&non-targeting_03427&non-targeting_03635&non-targeting_03664 (10 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_01509&non-targeting_01812&non-targeting_02025&non-targeting_02677&non-targeting_02777&non-targeting_03384 (15 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_00231&non-targeting_01185&non-targeting_02025&non-targeting_03384&non-targeting_03427 (20 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_02193&non-targeting_02543&non-targeting_02837&non-targeting_03026&non-targeting_03342 (25 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_00117&non-targeting_00196&non-targeting_00555&non-targeting_02025&non-targeting_02193 (30 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00355&non-targeting_00518&non-targeting_01601&non-targeting_02677&non-targeting_03427 (35 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00390&non-targeting_01601&non-targeting_02009&non-targeting_02288&non-targeting_03029 (40 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00231&non-targeting_00614&non-targeting_00629&non-targeting_00918&non-targeting_02193&non-targeting_02543 (45 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00390&non-targeting_01732&non-targeting_02193&non-targeting_03081&non-targeting_03384 (50 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_00196&non-targeting_00390&non-targeting_01730&non-targeting_03081&non-targeting_03252 (55 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00043&non-targeting_00355&non-targeting_01601&non-targeting_02814&non-targeting_03081&non-targeting_03620 (60 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_01185&non-targeting_03029&non-targeting_03081&non-targeting_03180&non-targeting_03384 (65 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00259&non-targeting_00972&non-targeting_01601&non-targeting_02089&non-targeting_03620 (70 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_01858&non-targeting_02777&non-targeting_03158&non-targeting_03252&non-targeting_03483 (75 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00614&non-targeting_00625&non-targeting_01601&non-targeting_02837&non-targeting_03180 (80 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00629&non-targeting_00656&non-targeting_02193&non-targeting_02543&non-targeting_03026&non-targeting_03483 (85 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_00213&non-targeting_00656&non-targeting_02911&non-targeting_03158&non-targeting_03483 (90 of 95)\n",
      "Analyzing pairs containing gRNA group non-targeting_00629&non-targeting_00794&non-targeting_02962&non-targeting_03029&non-targeting_03620&non-targeting_03635 (95 of 95)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/calibration_check.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Calibration check\n",
    "\n",
    "\n",
    "sceptre_object_highmoi <- run_calibration_check(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"calibration_check.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"calibration_check.png\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58b65dac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          response_id\n",
      "               <char>\n",
      "1:  ENSG00000288053.1\n",
      "2: ENSG00000102893.18\n",
      "3:  ENSG00000137970.7\n",
      "4:  ENSG00000290682.1\n",
      "5:  ENSG00000248671.9\n",
      "6:  ENSG00000044574.9\n",
      "                                                                                                               grna_target\n",
      "                                                                                                                    <char>\n",
      "1: non-targeting_00231&non-targeting_00518&non-targeting_00629&non-targeting_00794&non-targeting_02777&non-targeting_02837\n",
      "2: non-targeting_00518&non-targeting_00656&non-targeting_01730&non-targeting_02574&non-targeting_03029&non-targeting_03158\n",
      "3: non-targeting_00184&non-targeting_02193&non-targeting_02543&non-targeting_02837&non-targeting_03026&non-targeting_03342\n",
      "4: non-targeting_00196&non-targeting_00918&non-targeting_01730&non-targeting_02025&non-targeting_02911&non-targeting_03384\n",
      "5: non-targeting_00231&non-targeting_00518&non-targeting_00629&non-targeting_00794&non-targeting_02777&non-targeting_02837\n",
      "6: non-targeting_00213&non-targeting_02543&non-targeting_02777&non-targeting_02814&non-targeting_03252&non-targeting_03483\n",
      "   n_nonzero_trt n_nonzero_cntrl pass_qc     p_value log_2_fold_change\n",
      "           <int>           <int>  <lgcl>       <num>             <num>\n",
      "1:            25            1171    TRUE 0.001359857         0.7726258\n",
      "2:            65            5944    TRUE 0.003678742        -0.4971951\n",
      "3:           137            8700    TRUE 0.004943804        -0.3413673\n",
      "4:            62            3031    TRUE 0.007133584         0.4393944\n",
      "5:            16             845    TRUE 0.007279064         0.7678848\n",
      "6:           343           24721    TRUE 0.009819351         0.1739685\n",
      "   significant\n",
      "        <lgcl>\n",
      "1:       FALSE\n",
      "2:       FALSE\n",
      "3:       FALSE\n",
      "4:       FALSE\n",
      "5:       FALSE\n",
      "6:       FALSE\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "calibration_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_calibration_check\"\n",
    ")\n",
    "head(calibration_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "55e8c8f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Running precomputation on response ENSG00000004487.18 (1 of 261)\n",
      "Running precomputation on response ENSG00000117000.9 (5 of 261)\n",
      "Running precomputation on response ENSG00000265491.5 (10 of 261)\n",
      "Running precomputation on response ENSG00000179085.8 (15 of 261)\n",
      "Running precomputation on response ENSG00000119185.13 (20 of 261)\n",
      "Running precomputation on response ENSG00000143952.20 (25 of 261)\n",
      "Running precomputation on response ENSG00000138385.16 (30 of 261)\n",
      "Running precomputation on response ENSG00000134086.9 (35 of 261)\n",
      "Running precomputation on response ENSG00000173473.11 (40 of 261)\n",
      "Running precomputation on response ENSG00000144959.11 (45 of 261)\n",
      "Running precomputation on response ENSG00000128052.10 (50 of 261)\n",
      "Running precomputation on response ENSG00000153147.6 (55 of 261)\n",
      "Running precomputation on response ENSG00000113360.17 (60 of 261)\n",
      "Running precomputation on response ENSG00000113658.18 (65 of 261)\n",
      "Running precomputation on response ENSG00000180822.12 (70 of 261)\n",
      "Running precomputation on response ENSG00000204371.12 (75 of 261)\n",
      "Running precomputation on response ENSG00000112249.14 (80 of 261)\n",
      "Running precomputation on response ENSG00000112592.15 (85 of 261)\n",
      "Running precomputation on response ENSG00000242265.6 (90 of 261)\n",
      "Running precomputation on response ENSG00000105887.11 (95 of 261)\n",
      "Running precomputation on response ENSG00000104412.8 (100 of 261)\n",
      "Running precomputation on response ENSG00000107036.12 (105 of 261)\n",
      "Running precomputation on response ENSG00000136925.15 (110 of 261)\n",
      "Running precomputation on response ENSG00000136908.18 (115 of 261)\n",
      "Running precomputation on response ENSG00000151461.20 (120 of 261)\n",
      "Running precomputation on response ENSG00000107779.14 (125 of 261)\n",
      "Running precomputation on response ENSG00000198730.9 (130 of 261)\n",
      "Running precomputation on response ENSG00000149136.9 (135 of 261)\n",
      "Running precomputation on response ENSG00000162227.8 (140 of 261)\n",
      "Running precomputation on response ENSG00000173933.21 (145 of 261)\n",
      "Running precomputation on response ENSG00000182934.12 (150 of 261)\n",
      "Running precomputation on response ENSG00000174718.12 (155 of 261)\n",
      "Running precomputation on response ENSG00000197111.16 (160 of 261)\n",
      "Running precomputation on response ENSG00000120868.14 (165 of 261)\n",
      "Running precomputation on response ENSG00000165480.16 (170 of 261)\n",
      "Running precomputation on response ENSG00000100462.16 (175 of 261)\n",
      "Running precomputation on response ENSG00000151748.15 (180 of 261)\n",
      "Running precomputation on response ENSG00000255529.9 (185 of 261)\n",
      "Running precomputation on response ENSG00000169371.14 (190 of 261)\n",
      "Running precomputation on response ENSG00000167978.18 (195 of 261)\n",
      "Running precomputation on response ENSG00000168488.19 (200 of 261)\n",
      "Running precomputation on response ENSG00000102900.13 (205 of 261)\n",
      "Running precomputation on response ENSG00000108961.14 (210 of 261)\n",
      "Running precomputation on response ENSG00000125686.12 (215 of 261)\n",
      "Running precomputation on response ENSG00000154217.16 (220 of 261)\n",
      "Running precomputation on response ENSG00000005206.17 (225 of 261)\n",
      "Running precomputation on response ENSG00000011451.21 (230 of 261)\n",
      "Running precomputation on response ENSG00000105568.19 (235 of 261)\n",
      "Running precomputation on response ENSG00000124181.15 (240 of 261)\n",
      "Running precomputation on response ENSG00000159086.15 (245 of 261)\n",
      "Running precomputation on response ENSG00000186575.19 (250 of 261)\n",
      "Running precomputation on response ENSG00000086758.17 (255 of 261)\n",
      "Running precomputation on response ENSG00000071553.18 (260 of 261)\n",
      "Analyzing pairs containing gRNA group KDM1A (1 of 261)\n",
      "Analyzing pairs containing gRNA group RLF (5 of 261)\n",
      "Analyzing pairs containing gRNA group RNF115 (10 of 261)\n",
      "Analyzing pairs containing gRNA group DPM3 (15 of 261)\n",
      "Analyzing pairs containing gRNA group ITGB1BP1 (20 of 261)\n",
      "Analyzing pairs containing gRNA group VPS54 (25 of 261)\n",
      "Analyzing pairs containing gRNA group SSB (30 of 261)\n",
      "Analyzing pairs containing gRNA group VHL (35 of 261)\n",
      "Analyzing pairs containing gRNA group SMARCC1 (40 of 261)\n",
      "Analyzing pairs containing gRNA group NCEH1 (45 of 261)\n",
      "Analyzing pairs containing gRNA group KDR (50 of 261)\n",
      "Analyzing pairs containing gRNA group SMARCA5 (55 of 261)\n",
      "Analyzing pairs containing gRNA group DROSHA (60 of 261)\n",
      "Analyzing pairs containing gRNA group SMAD5 (65 of 261)\n",
      "Analyzing pairs containing gRNA group PSMG4 (70 of 261)\n",
      "Analyzing pairs containing gRNA group EHMT2 (75 of 261)\n",
      "Analyzing pairs containing gRNA group ASCC3 (80 of 261)\n",
      "Analyzing pairs containing gRNA group TBP (85 of 261)\n",
      "Analyzing pairs containing gRNA group PEG10 (90 of 261)\n",
      "Analyzing pairs containing gRNA group MTPN (95 of 261)\n",
      "Analyzing pairs containing gRNA group EMC2 (100 of 261)\n",
      "Analyzing pairs containing gRNA group RIC1 (105 of 261)\n",
      "Analyzing pairs containing gRNA group TSTD2 (110 of 261)\n",
      "Analyzing pairs containing gRNA group DPM2 (115 of 261)\n",
      "Analyzing pairs containing gRNA group UPF2 (120 of 261)\n",
      "Analyzing pairs containing gRNA group BMPR1A (125 of 261)\n",
      "Analyzing pairs containing gRNA group CTR9 (130 of 261)\n",
      "Analyzing pairs containing gRNA group SSRP1 (135 of 261)\n",
      "Analyzing pairs containing gRNA group TAF6L (140 of 261)\n",
      "Analyzing pairs containing gRNA group RBM4 (145 of 261)\n",
      "Analyzing pairs containing gRNA group SRPRA (150 of 261)\n",
      "Analyzing pairs containing gRNA group RESF1 (155 of 261)\n",
      "Analyzing pairs containing gRNA group PCBP2 (160 of 261)\n",
      "Analyzing pairs containing gRNA group APAF1 (165 of 261)\n",
      "Analyzing pairs containing gRNA group SKA3 (170 of 261)\n",
      "Analyzing pairs containing gRNA group PRMT5 (175 of 261)\n",
      "Analyzing pairs containing gRNA group SAV1 (180 of 261)\n",
      "Analyzing pairs containing gRNA group POLR2M (185 of 261)\n",
      "Analyzing pairs containing gRNA group SNUPN (190 of 261)\n",
      "Analyzing pairs containing gRNA group SRRM2 (195 of 261)\n",
      "Analyzing pairs containing gRNA group ATXN2L (200 of 261)\n",
      "Analyzing pairs containing gRNA group NUP93 (205 of 261)\n",
      "Analyzing pairs containing gRNA group RANGRF (210 of 261)\n",
      "Analyzing pairs containing gRNA group MED1 (215 of 261)\n",
      "Analyzing pairs containing gRNA group PITPNC1 (220 of 261)\n",
      "Analyzing pairs containing gRNA group SPPL2B (225 of 261)\n",
      "Analyzing pairs containing gRNA group WIZ (230 of 261)\n",
      "Analyzing pairs containing gRNA group PPP2R1A (235 of 261)\n",
      "Analyzing pairs containing gRNA group PLCG1 (240 of 261)\n",
      "Analyzing pairs containing gRNA group PAXBP1 (245 of 261)\n",
      "Analyzing pairs containing gRNA group NF2 (250 of 261)\n",
      "Analyzing pairs containing gRNA group HUWE1 (255 of 261)\n",
      "Analyzing pairs containing gRNA group ATP6AP1 (260 of 261)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/discovery.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_discovery_analysis(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"discovery.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"discovery.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "02e82574",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved discovery to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d0_sceptre_transfer/TargetKD.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "discovery_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_discovery_analysis\"\n",
    ")\n",
    "\n",
    "write.csv(discovery_result, file.path(data_dir, \"TargetKD.csv\"), row.names = FALSE)\n",
    "message(\"✅ Saved discovery to: \", file.path(data_dir, \"TargetKD.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6eeb819",
   "metadata": {},
   "source": [
    "### Day 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bd348cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io as sio\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2844d514",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1598: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1461: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "d1_mdata = mu.read(f\"{data_dir}/d1_normalized_gex_and_guide.h5mu\")\n",
    "mdata = d1_mdata\n",
    "\n",
    "# Filter out cells that have no guides\n",
    "#mdata = mdata[mdata['GEX'].obs['guides_per_cell'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "651115ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing sparse matrices...\n"
     ]
    }
   ],
   "source": [
    "# Define a temporary folder for transfer\n",
    "temp_dir = working_dir + \"/d1_sceptre_transfer\"\n",
    "os.makedirs(temp_dir, exist_ok=True)\n",
    "\n",
    "# 1. EXTRACT MATRICES (Transpose them!)\n",
    "# Scanpy is (Cells × Genes), Sceptre needs (Genes × Cells)\n",
    "print(\"Writing sparse matrices...\")\n",
    "sio.mmwrite(f\"{temp_dir}/response_matrix.mtx\", mdata['GEX'].layers['raw_counts'].T)\n",
    "sio.mmwrite(f\"{temp_dir}/grna_matrix.mtx\", mdata['guide'].X.T)\n",
    "\n",
    "# 2. EXTRACT NAMES (Row/Col names are lost in .mtx)\n",
    "# Save Gene names and Guide names to assign them later in R\n",
    "pd.Series(mdata['GEX'].var_names).to_csv(f\"{temp_dir}/gene_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['guide'].var_names).to_csv(f\"{temp_dir}/guide_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['GEX'].obs_names).to_csv(f\"{temp_dir}/cell_names.csv\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b6b0bb9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16925 × 13383\n",
       "    obs: 'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "    var: 'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "    uns: 'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "    varm: 'PCs'\n",
       "    layers: 'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['GEX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "4690bf5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>grna_n_nonzero</th>\n",
       "      <th>grna_n_umis</th>\n",
       "      <th>response_n_nonzero</th>\n",
       "      <th>response_n_umis</th>\n",
       "      <th>response_p_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACGTGATGTGTTCTATACGGTTA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>3</td>\n",
       "      <td>321.0</td>\n",
       "      <td>1769</td>\n",
       "      <td>2897.497314</td>\n",
       "      <td>12.067884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCATCCTCCAACCACATTGGGAGA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>10</td>\n",
       "      <td>463.0</td>\n",
       "      <td>1605</td>\n",
       "      <td>2809.700684</td>\n",
       "      <td>18.792038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTACAAGGCGAGTAATTATTCTG_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>6</td>\n",
       "      <td>155.0</td>\n",
       "      <td>1636</td>\n",
       "      <td>2930.134033</td>\n",
       "      <td>15.733069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTACGCAAACCACTGTTACGGTTA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>7</td>\n",
       "      <td>102.0</td>\n",
       "      <td>527</td>\n",
       "      <td>720.762817</td>\n",
       "      <td>24.765429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGATTCCCGACAACTACGGTTA_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>6</td>\n",
       "      <td>332.0</td>\n",
       "      <td>854</td>\n",
       "      <td>1462.475586</td>\n",
       "      <td>18.803732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATCGCATACCAATTGTCTTA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>2</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1680</td>\n",
       "      <td>2683.880615</td>\n",
       "      <td>9.948281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATCAAGTACTATGCATTGTCTTA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>12</td>\n",
       "      <td>424.0</td>\n",
       "      <td>3617</td>\n",
       "      <td>8417.620117</td>\n",
       "      <td>8.529727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACCTTACGCTCGGTATTATTCTG_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1884</td>\n",
       "      <td>3545.153809</td>\n",
       "      <td>9.999566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACAGTGCAACAACCACGCGACTA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>2425</td>\n",
       "      <td>3860.632080</td>\n",
       "      <td>12.225977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAAGCTACCATCCTCTACGGTTA_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>3</td>\n",
       "      <td>53.0</td>\n",
       "      <td>577</td>\n",
       "      <td>824.888916</td>\n",
       "      <td>24.609373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16925 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             batch  grna_n_nonzero  \\\n",
       "barcode                                                              \n",
       "AACGTGATGTGTTCTATACGGTTA_2  300genes:eTN5_GEX_sub2               3   \n",
       "CCATCCTCCAACCACATTGGGAGA_3  300genes:eTN5_GEX_sub3              10   \n",
       "AGTACAAGGCGAGTAATTATTCTG_1  300genes:eTN5_GEX_sub1               6   \n",
       "GTACGCAAACCACTGTTACGGTTA_2  300genes:eTN5_GEX_sub2               7   \n",
       "ACAGATTCCCGACAACTACGGTTA_1  300genes:eTN5_GEX_sub1               6   \n",
       "...                                            ...             ...   \n",
       "AAACATCGCATACCAATTGTCTTA_2  300genes:eTN5_GEX_sub2               2   \n",
       "CATCAAGTACTATGCATTGTCTTA_3  300genes:eTN5_GEX_sub3              12   \n",
       "CACCTTACGCTCGGTATTATTCTG_1  300genes:eTN5_GEX_sub1               0   \n",
       "GACAGTGCAACAACCACGCGACTA_3  300genes:eTN5_GEX_sub3               0   \n",
       "ACAAGCTACCATCCTCTACGGTTA_1  300genes:eTN5_GEX_sub1               3   \n",
       "\n",
       "                            grna_n_umis  response_n_nonzero  response_n_umis  \\\n",
       "barcode                                                                        \n",
       "AACGTGATGTGTTCTATACGGTTA_2        321.0                1769      2897.497314   \n",
       "CCATCCTCCAACCACATTGGGAGA_3        463.0                1605      2809.700684   \n",
       "AGTACAAGGCGAGTAATTATTCTG_1        155.0                1636      2930.134033   \n",
       "GTACGCAAACCACTGTTACGGTTA_2        102.0                 527       720.762817   \n",
       "ACAGATTCCCGACAACTACGGTTA_1        332.0                 854      1462.475586   \n",
       "...                                 ...                 ...              ...   \n",
       "AAACATCGCATACCAATTGTCTTA_2         43.0                1680      2683.880615   \n",
       "CATCAAGTACTATGCATTGTCTTA_3        424.0                3617      8417.620117   \n",
       "CACCTTACGCTCGGTATTATTCTG_1         19.0                1884      3545.153809   \n",
       "GACAGTGCAACAACCACGCGACTA_3         42.0                2425      3860.632080   \n",
       "ACAAGCTACCATCCTCTACGGTTA_1         53.0                 577       824.888916   \n",
       "\n",
       "                            response_p_mito  \n",
       "barcode                                      \n",
       "AACGTGATGTGTTCTATACGGTTA_2        12.067884  \n",
       "CCATCCTCCAACCACATTGGGAGA_3        18.792038  \n",
       "AGTACAAGGCGAGTAATTATTCTG_1        15.733069  \n",
       "GTACGCAAACCACTGTTACGGTTA_2        24.765429  \n",
       "ACAGATTCCCGACAACTACGGTTA_1        18.803732  \n",
       "...                                     ...  \n",
       "AAACATCGCATACCAATTGTCTTA_2         9.948281  \n",
       "CATCAAGTACTATGCATTGTCTTA_3         8.529727  \n",
       "CACCTTACGCTCGGTATTATTCTG_1         9.999566  \n",
       "GACAGTGCAACAACCACGCGACTA_3        12.225977  \n",
       "ACAAGCTACCATCCTCTACGGTTA_1        24.609373  \n",
       "\n",
       "[16925 rows x 6 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. EXTRACT COVARIATES \n",
    "rna = mdata['GEX']\n",
    "covariates_df = pd.DataFrame({\n",
    "    \"batch\": rna.obs['sample_id'],\n",
    "    \"grna_n_nonzero\": rna.obs['guides_per_cell'],\n",
    "    \"grna_n_umis\": rna.obs['guide_umi_counts'],\n",
    "    \"response_n_nonzero\": rna.obs['n_genes_by_counts'],\n",
    "    \"response_n_umis\": rna.obs['total_counts'],\n",
    "    \"response_p_mito\": rna.obs['pct_counts_mt'],\n",
    "})\n",
    "covariates_df.to_csv(f\"{temp_dir}/covariates.csv\")\n",
    "covariates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "64f876cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_meta_df = pd.read_csv(f\"{working_dir}/300genes_guide_metadata_v43.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c791f3ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ All guides mapped successfully.\n",
      "\n",
      "First 5 rows of mapping (Must match mdata['guide'].var_names order):\n",
      "                    grna_id grna_target\n",
      "0  ACAA1_+_38178488.23-P1P2       ACAA1\n",
      "1  ACAA1_+_38178517.23-P1P2       ACAA1\n",
      "2  ACAA1_+_38178559.23-P1P2       ACAA1\n",
      "3  ACAA1_+_38178570.23-P1P2       ACAA1\n",
      "4  ACAA1_+_38178620.23-P1P2       ACAA1\n"
     ]
    }
   ],
   "source": [
    "# 4. PREPARE GUIDE MAPPING \n",
    "# Sceptre needs a table mapping 'grna_id' to 'grna_target'.\n",
    "\n",
    "# 1. Create a \"skeleton\" DataFrame from the actual matrix order\n",
    "# This ensures the order matches mdata['guide'].X perfectly\n",
    "# We name the index 'grna_id' to match Sceptre's expectation immediately\n",
    "guide_mapping = pd.DataFrame(index=mdata['guide'].var_names)\n",
    "guide_mapping.index.name = 'grna_id'\n",
    "\n",
    "# 2. Prepare metadata for merging\n",
    "# Ensure 'guide_id' is the index so we can map efficiently\n",
    "lookup_df = guide_meta_df.set_index('guide_id')\n",
    "\n",
    "# 3. Map the targets onto the skeleton\n",
    "# We use .loc[] to pull the specific target for each guide in the exact order required\n",
    "guide_mapping['grna_target'] = lookup_df.loc[guide_mapping.index, 'intended_target_name']\n",
    "\n",
    "# 4. Reset index so 'grna_id' becomes a column (Sceptre requires this)\n",
    "guide_mapping = guide_mapping.reset_index()\n",
    "\n",
    "# ==========================================\n",
    "# VALIDATION CHECKS \n",
    "# ==========================================\n",
    "\n",
    "# Check 1: Are there any NAs? (Guides in mdata that were missing from meta_df)\n",
    "na_count = guide_mapping['grna_target'].isna().sum()\n",
    "if na_count > 0:\n",
    "    print(f\"⚠️ WARNING: {na_count} guides have NaN targets! These will crash Sceptre.\")\n",
    "else:\n",
    "    print(\"✅ All guides mapped successfully.\")\n",
    "\n",
    "# Check 2: Verify the first few rows manually\n",
    "print(\"\\nFirst 5 rows of mapping (Must match mdata['guide'].var_names order):\")\n",
    "print(guide_mapping.head())\n",
    "\n",
    "# 5. Save for Sceptre\n",
    "guide_mapping.to_csv(f\"{temp_dir}/guide_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6982f652",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>grna_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1_+_38178488.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1_+_38178517.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1_+_38178559.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1_+_38178570.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAA1_+_38178620.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>SUZ12_SGTN_23</td>\n",
       "      <td>SUZ12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1922</th>\n",
       "      <td>MESDC1_SGTN_11</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1923</th>\n",
       "      <td>MESDC1_SGTN_12</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>TP53_SGTN_01</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1925</th>\n",
       "      <td>TP53_SGTN_02</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1926 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       grna_id grna_target\n",
       "0     ACAA1_+_38178488.23-P1P2       ACAA1\n",
       "1     ACAA1_+_38178517.23-P1P2       ACAA1\n",
       "2     ACAA1_+_38178559.23-P1P2       ACAA1\n",
       "3     ACAA1_+_38178570.23-P1P2       ACAA1\n",
       "4     ACAA1_+_38178620.23-P1P2       ACAA1\n",
       "...                        ...         ...\n",
       "1921             SUZ12_SGTN_23       SUZ12\n",
       "1922            MESDC1_SGTN_11      TLNRD1\n",
       "1923            MESDC1_SGTN_12      TLNRD1\n",
       "1924              TP53_SGTN_01        TP53\n",
       "1925              TP53_SGTN_02        TP53\n",
       "\n",
       "[1926 rows x 2 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "56a9225a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Gene mapping saved.\n"
     ]
    }
   ],
   "source": [
    "# Create a mapping from the Gene Symbol (which matches your guides)\n",
    "# to the Gene Index (which matches the matrix rownames)\n",
    "\n",
    "# Check which column in mdata['GEX'].var holds the symbol\n",
    "symbol_col = 'symbol' \n",
    "\n",
    "gene_map = mdata['GEX'].var[[symbol_col]].reset_index()\n",
    "gene_map.columns = ['response_id', 'grna_target'] # Rename for easy merging in R\n",
    "\n",
    "# Save it\n",
    "gene_map.to_csv(f\"{temp_dir}/gene_symbol_map.csv\", index=False)\n",
    "print(\"✅ Gene mapping saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "3834c6e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>ENSG00000268903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>WASH9P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13378</th>\n",
       "      <td>ENSG00000275757.1</td>\n",
       "      <td>5_8S_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13379</th>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "      <td>PRODH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13380</th>\n",
       "      <td>ENSG00000276256.1</td>\n",
       "      <td>ENSG00000276256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13381</th>\n",
       "      <td>ENSG00000274917.1</td>\n",
       "      <td>RNA5-8SN5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13382</th>\n",
       "      <td>ENSG00000273730.1</td>\n",
       "      <td>5_8S_rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13383 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             response_id      grna_target\n",
       "0      ENSG00000227232.5           WASH7P\n",
       "1      ENSG00000238009.6  ENSG00000238009\n",
       "2      ENSG00000268903.1  ENSG00000268903\n",
       "3      ENSG00000241860.7  ENSG00000241860\n",
       "4      ENSG00000279457.4           WASH9P\n",
       "...                  ...              ...\n",
       "13378  ENSG00000275757.1        5_8S_rRNA\n",
       "13379  ENSG00000277196.4            PRODH\n",
       "13380  ENSG00000276256.1  ENSG00000276256\n",
       "13381  ENSG00000274917.1        RNA5-8SN5\n",
       "13382  ENSG00000273730.1        5_8S_rRNA\n",
       "\n",
       "[13383 rows x 2 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af083308",
   "metadata": {},
   "source": [
    "#### Import data to SCEPTRE (Rcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f3624030",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5a734b6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading matrices into R...\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(sceptre)\n",
    "library(Matrix)\n",
    "\n",
    "# Define path\n",
    "data_dir <- \"/oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer\"\n",
    "\n",
    "# 1. LOAD MATRICES\n",
    "cat(\"Loading matrices into R...\\n\")\n",
    "response_mat <- readMM(file.path(data_dir, \"response_matrix.mtx\"))\n",
    "grna_mat <- readMM(file.path(data_dir, \"grna_matrix.mtx\"))\n",
    "\n",
    "# 2. ATTACH NAMES\n",
    "gene_names <- read.csv(file.path(data_dir, \"gene_names.csv\"), header=FALSE)$V1\n",
    "guide_names <- read.csv(file.path(data_dir, \"guide_names.csv\"), header=FALSE)$V1\n",
    "cell_names <- read.csv(file.path(data_dir, \"cell_names.csv\"), header=FALSE)$V1\n",
    "\n",
    "rownames(response_mat) <- gene_names\n",
    "colnames(response_mat) <- cell_names\n",
    "\n",
    "rownames(grna_mat) <- guide_names\n",
    "colnames(grna_mat) <- cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "19e8ecd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renaming 'response_p_mito' to 'mt_percent'...\n",
      "[1] \"\\u2705 Sceptre object created!\"\n",
      "                           response_n_nonzero response_n_umis grna_n_nonzero\n",
      "AACGTGATGTGTTCTATACGGTTA_2               1487       2897.4972             19\n",
      "CCATCCTCCAACCACATTGGGAGA_3               1382       2809.7007             48\n",
      "AGTACAAGGCGAGTAATTATTCTG_1               1415       2930.1340             50\n",
      "GTACGCAAACCACTGTTACGGTTA_2                448        720.7629             23\n",
      "ACAGATTCCCGACAACTACGGTTA_1                755       1462.4755             57\n",
      "AGAGTCAAGGAGAACATTGTCTTA_1               1086       2194.7055             42\n",
      "                           grna_n_umis                  batch mt_percent\n",
      "AACGTGATGTGTTCTATACGGTTA_2         321 300genes:eTN5_GEX_sub2   12.06788\n",
      "CCATCCTCCAACCACATTGGGAGA_3         463 300genes:eTN5_GEX_sub3   18.79204\n",
      "AGTACAAGGCGAGTAATTATTCTG_1         155 300genes:eTN5_GEX_sub1   15.73307\n",
      "GTACGCAAACCACTGTTACGGTTA_2         102 300genes:eTN5_GEX_sub2   24.76543\n",
      "ACAGATTCCCGACAACTACGGTTA_1         332 300genes:eTN5_GEX_sub1   18.80373\n",
      "AGAGTCAAGGAGAACATTGTCTTA_1         244 300genes:eTN5_GEX_sub1   10.79872\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# 3. LOAD METADATA\n",
    "covariates <- read.csv(file.path(data_dir, \"covariates.csv\"), row.names=1)\n",
    "guide_mapping <- read.csv(file.path(data_dir, \"guide_mapping.csv\"))\n",
    "\n",
    "# ==============================================================================\n",
    "# FIX: RENAME 'response_p_mito' SO WE KEEP IT\n",
    "# Sceptre blocks the name 'response_p_mito' but doesn't calculate it.\n",
    "# We rename it to 'mt_percent' so we can use it as a covariate later.\n",
    "# ==============================================================================\n",
    "\n",
    "# 1. Rename response_p_mito if it exists\n",
    "if (\"response_p_mito\" %in% colnames(covariates)) {\n",
    "    cat(\"Renaming 'response_p_mito' to 'mt_percent'...\\n\")\n",
    "    names(covariates)[names(covariates) == \"response_p_mito\"] <- \"mt_percent\"\n",
    "}\n",
    "\n",
    "# 2. DROP OTHER RESERVED COLUMNS (which Sceptre DOES calculate)\n",
    "# We still drop n_umis and n_nonzero because Sceptre re-calculates those automatically.\n",
    "reserved_cols <- c(\"response_n_nonzero\", \"response_n_umis\", \"grna_n_nonzero\", \"grna_n_umis\")\n",
    "covariates <- covariates[, !(colnames(covariates) %in% reserved_cols), drop=FALSE]\n",
    "\n",
    "# 4. CREATE SCEPTRE OBJECT\n",
    "sceptre_object <- import_data(\n",
    "  response_matrix = response_mat,\n",
    "  grna_matrix = grna_mat,\n",
    "  grna_target_data_frame = guide_mapping,\n",
    "  extra_covariates = covariates,\n",
    "  moi = \"high\"\n",
    ")\n",
    "\n",
    "print(\"✅ Sceptre object created!\")\n",
    "print(head(sceptre_object@covariate_data_frame))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8881f692",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Unique pairs to test: 281\"\n",
      "   grna_target        response_id\n",
      "1        ACAA1 ENSG00000060971.19\n",
      "7       ACVR1B ENSG00000135503.13\n",
      "13      ACVR2A ENSG00000121989.15\n",
      "19      ACVR2B ENSG00000114739.14\n",
      "25       ADCY2 ENSG00000078295.17\n",
      "31      AMOTL2 ENSG00000114019.15\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 1. Load the translation map\n",
    "gene_map <- read.csv(file.path(data_dir, \"gene_symbol_map.csv\"))\n",
    "\n",
    "# 2. Start with guide mapping (which has 'grna_target' as Symbol)\n",
    "discovery_pairs_highmoi <- guide_mapping\n",
    "\n",
    "# 3. Merge to attach the correct response_id (Gene ID)\n",
    "#    This matches 'grna_target' in guides to 'grna_target' in the gene map\n",
    "discovery_pairs_highmoi <- merge(discovery_pairs_highmoi, gene_map, by=\"grna_target\")\n",
    "\n",
    "# 4. Select ONLY the two columns Sceptre needs\n",
    "#    (We filter out non-targeting guides automatically via the merge)\n",
    "discovery_pairs_highmoi <- discovery_pairs_highmoi[, c(\"grna_target\", \"response_id\")]\n",
    "\n",
    "# 5. DEDUPLICATE\n",
    "#    Sceptre performs analysis at the *Target* level, not the *Guide* level.\n",
    "#    We only need one row per Target-Gene pair.\n",
    "discovery_pairs_highmoi <- unique(discovery_pairs_highmoi)\n",
    "\n",
    "print(paste(\"Unique pairs to test:\", nrow(discovery_pairs_highmoi)))\n",
    "print(head(discovery_pairs_highmoi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f60caed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "sceptre_object_highmoi <- set_analysis_parameters(\n",
    "  sceptre_object = sceptre_object,\n",
    "  discovery_pairs = discovery_pairs_highmoi,\n",
    "  side = \"both\", control_group = \"complement\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f46242bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/grna_count_distributions.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "png(file.path(data_dir, \"grna_count_distributions.png\"), width = 3000, height = 2000, res = 300)\n",
    "\n",
    "print(plot_grna_count_distributions(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  n_grnas_to_plot = 8\n",
    "))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"grna_count_distributions.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "bd43891f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3deVxU9f4/8PfMADMwAwgDyCIMOwoornhJ3E3NTE3xaoZali3XFrtl5rUSrVtiammZN7W6N/2ZuWUumaCkpaCIIgSooGyy78KwDLP9/jh+uVwWQ4Q5H5jX8+Efcz4zc87rwOHlhzNnBoFerycAAGCPkO8AAADQNhQ0AACjUNAAAIxCQQMAMEoUERHBdwZD2LBhQ3R09Pnz57Ozs729vc3MzPhO1K4NGzbk5+cHBARwixs3bvT19bWwsOA3FfQU8fHxO3fuPHv2bGJioomJiYuLC9+J2sVF9ff3l0qlRJSQkBAbG+vv79/0gKioqNu3b3t5eXGLZ8+eTUlJ8fX15Z7o7Owsl8u5u3bs2HHmzJlRo0bFx8dfuHCh6cenpzOWGfSGDRskEkm/fv2OHj0aHBzc2NjId6J2bdiwYcmSJXl5edzixo0by8rK+I0EPUh8fPzPP//s7e2t1+unTZu2Z88evhO1Kz4+/oMPPvjggw+4xYSEhIMHDzZ/QFRUVFRUVNPi2bNnjx8/zj1x8+bNW7du5caLioreeeedDRs2cHft37/fQDvQ/Uz4DmA406dPDwwMfPrpp21sbFJTU4cMGUJEP//8c2pqqpeX16xZs4RCIRElJyefPn1aq9UOHTp04sSJRBQREbFkyZJjx47JZLL58+eLxWIi0mg0hw8fzs3NDQ4OHjNmDLeJiIiIl19++ccffxSJROHh4ebm5m2usM3tNvfkk0+uXbt2586dbe5Im1vpYJ6rV68ePXq0aVULFizw9fVtM09ERMSLL7547Ngxd3f3CRMmtF45MMvNzS08PJyIVCrVgQMHuNt5eXknTpxQqVSPP/44NydtaGg4dOhQTk6Ovb39zJkzHRwczp49S0Q1NTUZGRkTJkwYPHgwt8IbN25ERUVJpdInn3zS1taWiLhHajSapKSkMWPGjBgxos0Vtrnd5h5//PEffvjh73//u4eHxwPt49SpU0+cOLF582YzM7Pdu3fPnj37yJEjzR/QZpgex1hm0E20Wq1er+cOsgULFkRHR7u6up45cyYsLIyIUlNT58yZY2dn169fv7i4OO4pa9euXb169fDhw4uKiqZNm0ZEer3+scceO3nypIODw4oVK9avX9/0yHfffTcgICA3N3fBggXtrbD1dltYsWLF8ePHb9682eYutN5Kx/M4OjqGhoaGhoa6u7tv3rxZq9W2l2ft2rUrV6708PDo169fmysH9jU2NtrZ2RFRQkJCWFiYSCSSSqUzZsyIj48nomeffTYuLq5///51dXXcwXb27NmFCxdWVFSMGDHiueeeO3PmDBGdOXNm2rRpEonkzp07Q4cOLS8v5x65dOnS3NzcIUOGhIeHJyYmtrnCNrfbnK2t7fLly99///0H3TULC4uJEydys409e/YsXLiwxQNah+mJjGgGvWfPHnt7+1OnTn366acKhSIhISE+Pv7dd99taGgYMWLEW2+9VVBQcOfOHQcHhylTpvTt27f5c99//30fH5+RI0cePHjw0qVLGo0mPz8/OjqaiMaNG+fv7//2229zE8+1a9c6OzuHhIT06dNHr9e3XmGb23V2dm6+OSsrqxUrVrz33nvt/bLWYiuxsbEdzOPs7Ozs7KxUKkePHr1z584BAwbcJ09ERISnp+eFCxfaWzmw6datW5999llJScmlS5e4UxwRERGhoaEmJiZEFBISsmvXruDg4KysrBkzZsycOVMkEjU9d8yYMYsXLyaidevWbdy4ceLEievXr//444/nzZtHRLm5ubt27Vq5ciURjR8/fsmSJUT01FNPnTp1asiQIa1X2OZ2W6Rdvny5r69vSkrKg+7ms88++/HHHysUCicnp9an2tvcux7HuH7MdDpdRUUF979uRkaGpaVl010bN26USCRTpkyZMWPGtGnTBg0atG3btqZ73dzcuBsKhSInJyc7O9vPz6/5XU2niblq445IlUrVeoVtbrd11GXLlsXHx1+5cqXNHWmxlY7nISKtVjt//vywsDDuR+4+ebjfOu+zcmBZfX395cuXb9++Tf/7XQ4NDZ0/fz4R7dy58/Dhwz4+PnPmzMnJyeHubXGoE1F2dnb//v25QX9//6ZHNs0qZDJZVVVVmytsc7stWFhYrF69+h//+Efru8RiMXfQclQqVfMfllGjRmVlZUVGRj7zzDOtn9vm3vU4RjSDDg8PDwwMfOGFFwYMGPDTTz95enqWlJTMnz+/RT+uXLly5cqVJSUlAQEB06dPVygURHT9+nXufFxaWppCodBoNDdu3NDr9QKBIDc3l4i4XyRbEwgELVbY3nZbEIvFa9asWbVqVUd2zd3dvYN5iGj58uVyuXz16tXc4n3yCASCB105sMDb23v58uVE5Ozs/Nprr125csXT09PW1rZFkQ0cOPDAgQN6vT4iImLt2rXffPMNEV2/fp27lzvUicjd3f369etBQUHcYNP/1q21XmGb223t+eef37x5c58+fVqM+/r6cqk4ycnJjz/+ePMHLFq0aPPmzXv27Gl6Uf3+e9fjGFFBc6ytrdetW7dmzZrExMQxY8ZMmzZt4cKFSqUyOjr66NGjP/30U0JCQkBAQH5+vkwma3ph4YMPPpg5c+a5c+ecnJxGjhxJRP369Vu8ePH48eO3b9/+7rvvtvcrf+sVKhSK1ttt87mLFi3auHFjaWnpn+7UqFGjOphn9+7dhw8f3rlz5+nTp4lo8ODBI0eOvH+ejq8cWPP666//61//+vHHH9etWzdr1qyKigoPD4+kpKTAwMBnn3126dKlw4cPt7a2/u2332bNmsU9JS0tbfXq1R4eHh999NGuXbuI6J133nn++ecrKysLCwtjYmI++eST9jbXeoVtbrf1E01NTdetWxceHj537tzm4/Pmzdu0adOSJUtCQkKuXr1669atRYsWNX8AN/vpYJieyFiugyaiUaNGcZdbBgUFVVVV+fj4PPPMM87OzllZWaamposXL3Z0dHRwcKiurs7Ozra2tt60aRP3X/ratWtjYmL++OMPLy+v9evXc6cLFixYoNVqi4qKnnnmGe5Vcs64ceOabo8ePdrR0bH1CsPCwlpst3VUiUQiFAqHDh3q4uIybty4FtdBt9iKUCjsYJ7y8nJ7e/v8/Pzc3Nzc3Fxvb2+5XN5enqbntrdyYJOLiwt3IbBIJBo6dGhlZeWUKVOefvrp4uLikpKSgQMHzpgxQywWOzg45OTkVFRUzJs3j3sN+ezZswMHDhwzZsydO3feeeedkJAQIvL09Jw6dWp6erq9vf1nn33WdOmxu7u7u7s7d9vV1dXT07P1Cl1cXFpvt0VU7vxJYGCgRCIJDQ1tfh20iYnJkiVLNBpNUVHRwIEDt2zZwv0IN39iE3Nz81GjRjXtfuswPZEAn2b3pwQCfJXAKHDTNeOZtLEP1QMAwCicTAQAYBQKGgCAUShoAABGoaABABiFggYAYFR3vVGlsrLy/fff//zzzzv4+MITVHOd5KNIHtL5jZaU1L3zzq/W1mZEpFJpR4zo9+yzgZ1fnVFqKKbc3SQ0J+9lfEdhSQeP505/9X744cavv2aZm5sQ0d27qvfeG+3hYd3ptEYl7wDV5ZDDo9QniO8o3aC7ZtC1tbUP9N7KqkQqOEbK9Ifa6N27qiNHMlesCFixIsDRUfzrr7kPtTqjpK6igmNU/AvfORjTweO501+9uLgCqVTIHbpnz+aXlNR1JqVRKr9IBceorqd+2MafwCkOAABGoaABABiFggYAYJTRfZodACNiYnLDwu59cGBdnXrZMv/7Px6MEGbQAPxQq3WOjpITJyafODHZzU3KdxxgEWbQALwRi0UKhYyIxGJMlaANOCwAABiFggYAYFSvPcVRUlKfkVG1Y0cytzhxopuXV8u/eAYAwLJeW9CZmTWZmRV79iQTUUpK5bZtj6KgAaBn6bUFTURjxjht2RJCRE8//SvfWQAAHhjOQQMAMKo3z6Cb1NdrsrLuXrlSzC0OHuwgEgn4jQQA8KeMoqAzMqo//fTyjh2JRJSTo6yqetXaWvynzwIA4JdRFDQRrV07bNYshVard3P7nu8sAAAdgnPQAACMQkEDADAKBQ0AwCgUNAAAo/5b0EVFRTdv3mx+n1KpzM7O1ul0DzQCAABd4t5VHIcPH7569WpxcfHOnTu5kfT09K+//trf3z8rK2vNmjUCgaAjI/ztCABAb3NvBj179uwPP/ywecMeOnTopZdeWrx4sb29fXJycgdHAACgq7R7Djo3N9fd3Z2IPD09s7OzOzhCRJWVlfHx8UlJSaampt0aHQCgd2v3jSoajYabUJuamlZXV3dwhIgyMzM///zzxsZGc3NzA+wAAEBv1W5BW1pa1tbWSqXSyspKuVzewREiGjZs2O7du/Py8vz8/AyzDwAAvVK7pzhGjhwZHR1dX18fFxc3bNiwDo4AwP0dPJh+4MDNAwdu/v57Ht9ZgHX3Cvry5cvvvfeevb39e++9d+vWLSIKCwurqanZsmVLeHi4jY1NB0cA4P7mzz+2efOlzz+//P/+XyrfWYB1905xjBgxYsSIEc3vEIlECxcufNARAPhTO3aEyuWSrVtTT53CJBrupzd8mt3duyqdTs/d4DsLAECX6Q0FPWLEnoyMSu62jQ0+6BkAeole8lkcx49Pyc9fsG/fBL6DAAB0mV5S0AAAvQ8KGgCAUShoAABGoaABABiFggYAYBQKGgCAUb3hOuiO0+uJiI4cybCwMCUiOzvz8ePdeM4EANAO4ypo7g2HX3111cREWFGh6tPH4vx5FDQAMMq4Cprz7bdjLS1No6Pzd+68xXcWMApLl56Kicnlbmu1en7DQA9ijAUNYGDFxXVPPuk2aZILET322C98x4EeAwUNYAiurrJBg2z5TgE9DK7iAABgFAoaAIBROMUBwJbyctXq1edtbSVEZGlp9vXXU/hOBLzBDBqALfX1GldXsY+PuVwu/OknXGhk1DCDBmBOWJjHkCHyrKyaAwey+c4CfMIMGgCAUShoAABG9dRTHDdvVty5U8PdrqtT8xsGAKA79NSC3r792vffpzk6WhBRQYGS7zgAAF2vpxY0EYWH+7z55kAicnff14mnZ2crb96s+Otfj3GL8+f3nz3bpyvzAQA8HOM9B11VpZJIhD4+5j4+5rdvl12/Xs53IgCA/9GDZ9APz8nJYtkyfyLKy6vlOwsAQEvGO4MGAGAcChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGBUT/qLKgkJRatW/c7dTk+vCAtz5zUOAED36kkz6PLyhoyM8rlzXefOddXrdXzHAQDoXj2poInIxkY8fbrb9Olu1tamfGcBAOhePaygAQCMBwoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGNWT3uoNYFTy8mrLy+sFgo3c4sKF/t99N43fSGBgKGgAdvXpYxYbO4OIdu/OyMtT8x0HDA0FDcAugUBgbW1GRBKJCREK2ujgHDQAAKNQ0AAAjMIpDiKihITSAweyNm1K4BbPnp03aJA9v5EAADCDJiLSavWLFnmfODH5xInJYrFAq9XznQgAADPo/9Onj1ihkBGRiQn+0wIAJqCMAAAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYZdLeHVFRUeXl5dztUaNGubm5JSYm3rhxg4hkMtkTTzzB3XXt2rU7d+6EhITY2dkZIK4BKJXqf/87JSpKRkQymemyZUP4TgQARqrdGbSXl1dgYGBgYOCZM2ckEgkRJSYmSqXSwMBAPz8/7jG//fZbTEyMi4vLP//5T5VKZaDI3aymRn3rVmlKSsH589n//OdFvuMAEBE1NGiVSnVm5l3un1Kp5jsRGEK7M2gvLy8iKiwslMvlDg4O3GBdXV1jY+OAAQO4xVOnTr3xxht2dnYpKSkJCQmjRo0yQGIDeP31wMBAm5s3q55++hzfWQCIiK5dKz99Oj8mJoeIamvV3333+FNP9ec7FHS7dguac+7cuTFjxnC3vb29q6urk5KSDhw48NFHHwmFwtLSUu7MRt++fYuKiriHZWZmnjhxoqamxsLColujAxiVmTMVW7aEENHTT//KdxYwkD95kfDChQuhoaHc7TFjxkyfPn3JkiU2NjbcyWgAAOg+9yvozMxMuVxubW3dYlwsFms0GiKyt7cvKysjouLiYkdHR+5eT0/PV199ddGiRXV1dd2TGeBPlJWVnT59+ujRo80Hr127duzYMe6I7fgIAI/uV9Dnzp0bO3Zs0+LRo0evXLnyyy+/ZGRkcK8TTp48+T//+U9CQkJiYuLw4cO7PSxAx6SlpZWXlx8/frxppPUL2h0ZAeDX/QraycnpL3/5S9OiQqHIysoiosjISLFYTERjx46dMGFCYWHh6tWruREAFowZM2bevHnNR06dOrVo0aKhQ4cOGTIkISGhgyMA/Lrfi4RTp05tvhgUFBQUFNTiMUOGDBkyBFcKA+tav6DdkREiKigouHjxYmVlJaYgYHh/chUHgJGrqqqKj4+vr683MXmwH5bExJKoqGzu9q1bVRMnOnR9OOjt8FZvMAqtX9DuyAgR+fv7r1+/fsWKFbW1tQ+0xYsXC7Zvv5qSUpCSUpCVVdW1uwNGAgUNvVBlZWVaWpparU5LS+OKtfUL2h0ZeUhBQfLIyODIyGBLS9OHXxsYIZzigF6ovLw8KSlp8uTJSUlJNjY2Uql07NixVlZWeXl5TS9od2QEgF8oaOiFvL29vb29Wwy2fkG7IyMAPMIpDgAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGBUD7gO+syZXL1eT0SJiSV8ZwEAMJweUNBTphzw87M2MRGWlDQ4OprzHQcAwEB6ximOffsmnDw5dfFiH76DAAAYTs8oaAAAI4SCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRPeDDkviiUmnVat2VK8Xcop2duUJhxW8kADAqKOh2FRbWVVY2zJnzIxHdvds4Z47frl1T+A4FAEYEBX0/crk4NnYGEW3ffr2wUMN3HAAwLjgHDQDAKMygAXqYjIy7q1f/vmnTZW7x7Nn5Mpkpv5Ggm6CgAXqY+nrtnDkejzzSV6fTL1jwq1ar4zsRdBcUNEDP4+fXZ/RoR61Wz3cQ6F44Bw0AwCgUNAAAo3CKo0MKCmr/+KM6MjKeW5w1y9vPz5bfSADQ66GgOyQnp7aioi4lpYCIoqPzvb37oKABoLuhoDsqONg+MjKYiDIzz/CdBQCMAs5BAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoE74D9Dw3blQtXnxy6dIobrGg4CWJBF9GAOh6aJYHptXqV68ePG6ck0qlHT/+hF7PdyAA6KVQ0J1hZydWKGQNDVq+gwBAb4Zz0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjMJVHJ2n0+mJKCYmVywWEZGNjWTYsL58hwKA3gMF3XmNjToieuedX4VCwd27jZ6etjExf+U7FAD0Hijoh/XTT5MlEtGJE7nff5/LdxYA6FVwDhoAgFEoaAAARqGgAQAYhYIGAGAUXiQE6Km4Cz2XL48xMxMRkZdXn7ffDuY7FHQlzKABeiruo27lcqGdnbCysubo0dt8J4Iuhhk0QM+2dGl/S0vTf/3rRnR06qOPHuAGn302cMGCAfwGg4eHGTRAb1BT0yiXi+fOdZ0711WtbszKust3IugCmEED9BK2tuLp092I6MKFYr6zQNdgtKBv3aq6e1fF3cbflAIA48RoQf/977/GxubLZCb0fy9VAwAYG0YLmojee2/I3LkeROTispfvLAAAPGi3oE+ePPn7778TkVwuf/PNN4lIq9Xu3bu3oKDg0UcfHTp0aJsjAADQVdot6MLCwqeeemrgwIFNIwcPHrSyspo7d+4//vEPDw8PGxub1iMGyQwAYBTud5nd8ePH//3vfxcVFXGLly5dmjRpkkQiCQkJuXLlSpsjAADQVdqdQYeGhorF4uLi4oiIiM2bN1tYWNTU1EilUiKysbEpLy8notYjRHTlypXPPvussbHRysrKILsAANA7tVvQvr6+RKRQKOLj49PT0wcPHmxiYqLX6wUCgVqtlkgkRNR6hIg8PT1fffXV0tLSU6dOGWYfAAB6pT9/J2FFRYWFhQURubm5ZWdnE1FmZqa7u3ubI0RkY2MTHBwcFBSkVqu7KTQAgDFodwa9adMmuVxeWlpqYmLi4+NDRHPmzNm+ffuAAQNKS0sHDRrU5ggAAHSVdgv6pZdeysvLk8lkLi4u3Iivr+/7779fVlbm5uYmEAjaHAEAgK7SbkFLpVI/P78WgzKZTCaT3X8EAAC6BD7NDgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUez+yauepaioPje3OjIynlscP941ONiJ30gA0NOhoLtGQUGtUqlKSSkgori4EqFQgIIGgIeEgu4yXl5WkZHBRPTWW5f4zgIAvQHOQQMAMAoFDQDAKBQ0QG9z6VLJu++eFwg2cv8SE0v4TgSdhIIG6IVefz0gLS0sLS3M2dmC7yzQeXiREKAXkkhMrK3NiEgoxJ866sEwgwYAYBQKGgCAUTjFAdBlamvVRKYNDZrIyKsXLxbwHQd6PMygAbpMdXUjEWk0upSUgmvXiviOAz0eChqgi1lYmERGBgcFyfkOAj0eChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARrH1gf0ff3xp/xuXiUipbJw40YHvOAAAfGKroJ2czE98O5mIpk8/xXcWAACesVXQZmYihUJGRCIR/hQxABg7nIMGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFFsXWYH0E2KiopKS0uJSCwW+/r6coNKpbKsrMzNzU0oFLY3AsAjFDQYhZ9//rm6urpv374ymYwr6PT09K+//trf3z8rK2vNmjUCgaD1CN+pwdihoMFYPPLII4MGDZJIJNzioUOHXnrpJQ8Pj23btiUnJwcFBbUe4Tdwlygurh837gfunV99+1pcv76E70TwAFDQYBSkUumZM2d++OGHgICAJUuWEFFubq67uzsReXp6ZmdnBwUFtR4hooKCgosXL1ZWVorFYj53oLN0Ov3nn//Fx8c6K6vmrbfi+Y4DDwYF3fVqazV5ecorV4q5xYAAuUSCrzPP5s2bR0R6vX7VqlU5OTkKhUKj0XAnMUxNTaurq4mo9QgRVVVVxcfH19fXm5j01G+io6OFQiFraNDwHQQeWE895liWklL5++9FP/10k4jy8mqTk5/x95fzHQqIiAQCgUKhKC8vVygUlpaWtbW1Uqm0srJSLpcTUesRIvL391+/fn1eXt6uXbt4zQ7GCC9Vd4tXXvGPjZ0RGztDLpfwnQWIiK5fv97Y2Jifn5+UlOTp6UlEI0eOjI6Orq+vj4uLGzZsWJsjAPxCQYNRSE1N/eijj/bu3fv666/36dOHiMLCwmpqarZs2RIeHm5jY9PmCAC/cIoDjEJYWFhYWFjzEZFItHDhwvuPAPALM2gAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWPG+1eZWUNAQHfcrfd3a2zspbymwcAehDMoLvdsWNT0tLCDh6cyHcQAOhhUNDdTiYzsbY2k8lM+Q4CAD0MChoAgFEoaAAARqGgAQAYhYIGAGAULrMzEJ2OdDp9ZWUDtygWiyws8LIhANwPCtpAMjOrc3OrbW2/4Bb/9rfB27ZN4jcSADAOBW04rq6yixdnENGWLalKJd9pAIB5OAcNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNwmR2AUWho0KpU2tOnc7hFZ2eZv7+c30jwp1DQAEahqKiuulr15ptniKi0tGHaNK9du6bwHQr+BAoawFjY2opPnpxKRNu3Xy8s1PAdB/4czkEDADAKBQ0AwCgUNAAAo1DQAACMQkEDADAKBQ0AwCgUNAAAo1DQAACMaveNKjt27MjMzNRqtSEhIbNnzyaib775JikpSSqVyuXyN998k4gqKiq++OILvV7v4+OzYMECw6UGADAC7Rb0jBkzHB0ddTrd66+/Pn78eBsbGyJ6/vnnBw4c2PSYffv2TZs2bfjw4evWrcvJyVEoFIaIDABgHNo9xeHo6EhEQqFQIBAIhfce9u23365atSouLo5bTEtLCwoKIqLBgwenpKRwg/X19fn5+cXFxU3PAgCATviTz+I4efJk//79ra2tiWju3LlLliyprq5etWqVj4+PnZ2dSqUyNTUlIqlUWlZWxj3l6tWr7733nk6ns8hsuTEAABagSURBVLKy6u70vcDly0V376qaFidOVAgEPMYBAIbcr6ATEhJiY2Pff/99btHS0pKIrKysAgMDc3Jy7OzsxGKxWq02NTWtra2VyWTcw0aNGhUTE5OXl+fn59fd6XuB5ctjCgqqrazMtFp9amqlVvumAA0NAER0n1Mc6enp+/fvX7VqFTdHJqL6+noi0mq1GRkZ3AmQgICApKQkIrp27VpgYKBBAvdCa9cOO3ly6sGDk/gOAgBsaXcGvXXrVpFI9PHHHxPRCy+84Orq+sUXX1RUVKjV6rFjx7q4uBDRvHnztm7devTo0f79++MVQoCeoqCg9o8/qiMj47nFWbO8/fxs+Y0EbWq3oL/44osWIytWrGgxYmtrGxER0eWZAKBb5eTUVlTUpaQUEFF0dL63dx8UNJvwgf0Axig42D4yMpiIMjPP8J0F2oWCBjBqWq2+tlZdWdnALfbpI8Gr1OxAQfNAqdSUlmquXCkmIqVSzXccMGo3blQtXnyyabGubrm5OWqBFfhO8OCPPyquXSuLj88nojt3lHzHAWO3Y0fo44+7NTRovbx+4DsL/A+82Y8fs2e7x8bOiI2dYWGB/yMBoG0oaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUPgiCFTU1aiKys9vGLU6c6HbgwAxeEwEAz1DQbDl+fLJAQFFReRcvVvKdBQB4hoJmi5ubVCgU2NlJ+A4CAPxDQQMAEZFeT0RUVdXQ0GBCRCYmQktLM54zGT0UNAAQEalUWiJydv4Xtzh+vFtMzF95TQQoaABoJjFxtlgsjI7OP3Ikn+8swHdBFxfXhYefIKKZmpGB5MZvGACwsjKVSETm5iK+gwAR79dBNzRozp/PmzvXVaGw4DcJAABr+H+jiomJcPp0NxcXKd9BAADYwn9BAwBAm1DQAACMQkEDADAKBQ0AwCgUNAAAo/BGFQBoKTdXefNmxYsvRnGLc+b4Tp7szmsiI4UZNAC0VFbWYGpKdnZCOzthfHx+UlIp34mMFGbQANAGFxfpsmX+RJSVVcN3FuOFGTQAAKMwg2ZRdXVjeXn96dM53KKfn62rqyW/kQDA8FDQLEpPv3vjRvmbb54hotxc5dq1oa+9NpTvUABgaChoRg0fbr9nzzgieu21OL6zAAA/cA4aAIBRKGgAAEahoAEAGIVz0KxLTi4/e7Zwy5Yr3GJMzDyFworfSABgGCho1qlUurlzPWbOVBDRX/96Rq3W8Z0IAAwEBd0DODtbDBpkS0QikYDvLABgODgHDQDAKBQ0AACjcIoDAO6nokKVnFx64MBNIlq+/Ne6Oo1AQETk5mZ57dpinsP1dihoALifjIzq9PS7OTkVRFRQoNy7d7y7u2V6+t2IiGt8R+v9UNAA8CfCw71femkAEbm6fu/kZKFQyKqrG/kOZRRwDhoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFG4iqMnuXu30cdnF3fb19fm5s3n+M0DAN0KM+ge5pdfpqalhX3//QS+gwBAt0NB9zCWlmbW1mYymSnfQQCg26GgAQAYhXPQPVJFhaqioiEyMp5bHD7cceJEN34jgVGprm6sqWncsSOZW/T3l4eGuvAbqVdCQfdIpaUNKpUmJaWAiJKSKoqKalHQYEglJQ319epjx24Q0a1b1RMmuKOguwMKuqeyt5dERgYT0caNyXo932nA+Mjlkq++CiWiLVtSlUq+0/RSOAcNAMAozKABHlZeXk1jo46IiooaiaR8x4HeAwXd4+Xn1xUUVDa9YPjXv/p5eFjzG8nYTJ16KCfnrkgk8BDbf9r/Kb7jQO+Bgu7xcnNrqqoauRcMT57MGzLEAQVteN99N3bkSIfEE7W0h+8o0IugoHuD0aOdIiKGEtEff1TynQUAugwKulcpKanbsuXKoUPpRCQUCrZvf5TvRADQeSjoXuXuXbW1tdDOTlhd3bh9+/UJE+5dHO3sLBs1CpepAvQwKOje5okn3MaOdUpPv7ttW9rnn18mouLi+v797X766Um+o0HvlJlZnZZW/eKLUdziK68MHTjQjt9IvQaug+61RCLhvn0T9u2b8PLLA/jOAr1ZcXGDREJ2dkI7O+HPP9/OybnLd6LeAzNoAHhYAwfaLlvmT0RRUXl8Z+lVMIMGAGAUZtAA0L1yc6svXSpsWhw82MHHx4bHPD0IChoAutf58/mvvHI6MNCGiG7evLt6dQgKuoNQ0ADQ7QYPlu/ZM46IXnstju8sPQkKuvdLSqo4fTrHy2snt/jBB6ELFuC6DoAeAAXd+6lU2kGDbNesGUpEH398rbq6ke9E0GvV1KjPn8+vr9fU1amXLTtjZiYiosZGbXCwPd/ReiQUtFGwsjIdNMiWiKytzfjOAr1Zfn7t8eMZsbG5DQ3a2lp1VNRjAgFt2pRcUYFpQWegoI3X++9faGzUcrf79pW+8cYwfvNA7/D224MmTXLJz68LDj7i5iYVCgWWlqYo6M7BddDG69NPEwoLK5VKZXp6yTff/MF3HABoCTNo45KSUhkXd/6TT+KJqK5O/cor/i4u0osXS9asucZ3NABoCQVtXFQq7VNPeU6b5kpEjz32S+sHlJbWEVloNLorV0qJyNfXxtISp60B+IGCNjouLlLuBUOBoI179++/+RcaolSqF8/5saCgLipq7rhxroaOCMZhwoT91dUq7rabm9XhwzP5zcMgFDS0wcrKNDZ2xqRJP/MdBHqbxMSykyfzIiJiiaiqquGzz0L69jXPyLi7e3cm39FYhIIGAMPRaPQvvug3d64nEY0adXT4cHt3d5lEYqLV3q6sbOAeIxaLLCxMeY3JChQ0UFlZQ2lpPfd3wf/4o/ovfOeB3s3GRqxQyJqP3LpVfft2la3tF9zi8uXDPv10PBH95z+pRUW13KBQKFixYoSBo/IOBQ1UUtKgUqm5vwt++3YjefAdCIyPp6fl778/QUQbNybr9fcGv/gi0cpK6Oxs0dCgPXIkBwUNRsrR0SIyMpiIlj15le8sAP/1/PN+Y8c6lZU1HDmSw3cWHvBT0IGB/05NLeNuy2Q42cSosrKGTZsuf//9dZ1OX13d+Pbbwdy4jY3E09Oa32xgVFQqrU5HO3Ykc4sKhdWUKe68JjIQ3mbQu3ePGzbMLjGx/MUXz/OVAe6vulrdp4/Qzk5YWFh34ED6pUv5RFRTo54wQXHgwAy+04ERqa3V6PW6Y8duEFFeXq2bWx8UdPeSSk2src1kMpxjYdqsWe4hIQ5JSeX79mXGxs4goj17bl2+XMV3LjA6IpHwq69CiWj//sxffy3lO46BoB8BgCEajV6j0XCX3Gm1utYPuHq1/OTJbO6SD72edj+6WEaWdXXq+fN/0Wrvvf3K1dVy06ZxBs3dPVDQ0HmHD2dkZFQ2Lb7yyhCpFK8owEO5fLnk4sXS7duT/m8goMUDNBrdyJF2GzaMJKJZs6Jv3KgYbmP597+fPVZ4c9WqoD59xFlZNVFR2dyD4+MLr13773R7+nRPZ2cZ9RwoaHgwly6VHj6cJRBs5BZDQvp6eVkS0d69txcvDkBBw8NburR/RMRQIvLx2d/mAywsTLgrqU1MBD4+VlRG69ePOP5M0pNPuru4SM+eLYyOLrxypZiItm9P+v333AED+hDRuXOFvr42KGjo5SZPdvnssxAiGjv2eFiY+/z5XkT0ww/33qobH1/4t7+dbnrwc88NfPnlwbzkBOOUmVmTm1s9Z86PRFRUVP/cc36rVw8mop740QUoaHhgpqZC7i+zCJt9nLhWq/vnPy9KpabZ2dUVFXUffzyCiHbvvlVQoOQeMHLk/2s6H2JtLc7KWmro3GA0vLyszpyZRkQzZkS3vre8vP7ll+/NIWpr1XZ2kunTvbjF4cMdPTwYuoT0YQtaqVSWlZW5ubkJhfjsf6Om01Ftba1eb5KbWymTmY4e7UhEMTEFTQ+orlZFRo4IDLQpKKh9/nlGr63E8dyLFRbWvfXWWRsbiUqluXix8MMPhxPRDz9kxscXZGVVEFFaWtWGDWM9PAYS0caNl7Xae+9o7NvX4plnAnnJ/FAFnZ6e/vXXX/v7+2dlZa1Zs0bQ5udXgtF47bUABwfzL79MO3o0lxspKKi7dauO+5SP8vKGvn3NFQqZUqlWqbTcIBH5+8ufeMKLt9DN4Hju3errtWPHOvj5WWdnKxMSisPDvYno0qVSJyfzHTtGE9HUqScPHkxPT68koo0bE2bNUpibi+7cqS0ubhw48N4fvXVykhryLPZDFfShQ4deeuklDw+Pbdu2JScnBwUFdVUs6B1yc5U6nZ77lI/y8npusLpardHouMHLl0u1WsHx47e5u557bmBwsBNfaXE893qPPNKXu67/s89SWt9bUlJvb2+uVCqJSKvVvfvuYG7C8cknydwZ7cLCOqFQaG5+rzZ/+WVOi8N1587klSt/426rVJpHHnH29OxDRP/5T6q5uSn3P/6IEY6nToV1MLBA3/TBJA/u5Zdf/vLLLwUCwcmTJxsbG2fOnElEBQUFFy9erKysfOONNzZu3NjmE9eujfXxsZTLJeXlDRcuFM+YobC/4yirskxR35EObSCiEydyBw60dXOTEdHhw1nTp7uZmYlu3qwqKKgfP96JiGJi8p2cpNyLs0eOZI8d62RjIy4trb90qXT6dDciunixxMREMHy4Pbe2oCDbfv1kOh0dOZL1xBMKU1NhWlpVaWn92LFORHT6dL6rq9TPrw8R/fhj9oQJTtbW4sLCusTEsmnT3IgoLq5YLBYNHWpHRMeO5Q4dKndxkarV2mPHcmfNchcKBX/8UVFV1cj9Xh8VlefhYenjY82tbdIkZ0tLszt3lKmpVVOn9iOi8+eLZDLTwYPlRHT0aE5wsL2jo0VdneaXX+7Mnu1OJEhKKlcqNaNG9SWiU6fyvL2tvLysiOjHH7MmT3aVSk2ys5UZGVWPPtqPiH77rbBPH7NBg+6t7S9/cXBwMK+pUZ8+nf/kk+5EdPVqmUqlDQnpS0S//HLHz6+Ph4cl94V97DFXc3OT27ers7JqJk1yIaKUC8onrIboTHQ5A24fOZITGupgZ2deWan67beimTMVRJSQUKrV6keOdCCin3/O9fe3cXe/t7bHH3cTi0Xp6Xfz8monTHAmopiYAkdHc39/G+7bNGaMk62tuKysPi6u9Ikn3Ijo0qWSsrIG7u3+lZWNtrZi7jqQwsLavn2l3GkGFxfLxYtbXmvFGTJkyIgRXfYBOg9/PDcUCQaUenNfvcuXS/V6Cg6+dwTieO6m43me23BbrVWZS8l/zl5rfTyfO1col4sDA22JqEuOZ2trUw8PKy6nTtdGefbtaz5ggA0RXbhQJBaLxGIREZWXN4SGOpqYCMvLG2prdStXBrd5ILU+nh+qoJcuXbpz504iOn36dHl5+bx584goLS3tu+++0+v1tbW1JSUleXl5Tk5OIpGo+RPr6zXcvun1+vp6rYWFiV5Nmka9TqAXWwiJSKXSikQCExMhEdXVaczNTQQC0mr1arVOIhERUWOjViAQmJoKubWJxSKhUKDT6VUqbV1dmY2NvVYr0On03FenoUFrYiI0MRF0Ym3c/5ZqtU6v15uZ/c/a9Hqqr9eYm5tUVBRZWtrpdILOrc3UVCgS3VubhYUJEWk0eq1Wx4VXqbRC4b211dVpJBKRUCgoLy8yN7e1sDDr2Np0Wq2+aW3/+4UVCQQCrVavVmslEhMialTpSC0wMRUKxfqGBo2pqUgkEuj1+vLyYltbe6FQdN+1tfGFJRKYmd37UpiZiUQigU6nb2jQtpPtv9+m+voyGxt7oVAkFAqa5iwtzJ49e/78+Z0+gFvoyPHc/PElJSUymczCwqLpeNZp9Q01Oom5iVCsf9Av1J8eM117PCuVFVKpzMxM0vp4Fgi4d4t08qejE8fzA65NV11dbmlpZWoqVqm0Qp1QJBQKTPT1qraO50YdEXFHYPPjmaudto7A7jqeubURUVVVmVxuI5FIWh+BbRzP+ofwxhtvKJVKvV6/f//+6OjoNh8zZsyY/Pz8h9nKg8IWscXO6cjx3NwLL7xw5syZ7s/VLZYsWXLu3Dm+U3TS4sWLL1y4wHeKTgoPD7948WIHH/xQL1WPHDkyOjq6vr4+Li5u2LBhbT7G2dnZxMSgF/Nhi9hi53TkeG7Ozs7O3NzcAMG6g729fY8O3+YMtEewt7cXi8UdfPBDneLQarV79+7Nz8+fPHny0KFDO70eABbgeAbWPFRBAwBA9+nGXyErKiq++OILvV7v4+OzYMGCrlrt6dOnDx06pFard+3a1d6GOjLScTt27MjMzNRqtSEhIbNnzzbAFvft25ecnKxWq319fZcuXWqALXLWr1+v1WpXr15tgC1+8803SUlJUqlULpe/+eabBtvHLqHVardt21ZVVWVubv7aa691/DdWg7l8+fJXX33l4OBARCtXrrS2tm6duSMjhsycmZn56aefFhYWfvnll1zyzmXmZS9ah++Sb8FDvUh4f9u2bbt8+bJer1+7dm12dnZXrbaqqkqj0SxduvQ+G+rISMcVFhbq9XqtVvvKK69UVFQYYIsajYa7sWrVqo6v/yG/4Ddv3ly3bt2HH37ILXb3Fr/++uvk5OTmIwbYx67y22+/ffvtt3q9/tChQ8eOHeMrxn3Ex8fv3bu3+UjrzB0ZMaS6urra2tp33323uLj4YTLzshetw3fJt6Ab38+alpbGXeo/ePDglJQ2LgvvHGtr6xYX7bXeUEdGOs7R0ZGIhEKhQCDg3gHc3VvkdlCr1UokErlcboAtEtHBgwe53w84Btjit99+u2rVqri4OINtsaukpqY2xUhNTeUrxv39+uuvK1eu/P777/V6PbWVuSMjhmRubm5hYdF8pHOZedmL1uGpK74F3VjQKpXK1NSUiKRSKffmHINtqCMjD+rkyZP9+/e3trY2zBb37Nnzt7/9zdLSkvuud/cWr1696unpKZP99z2s3b3FuXPnbt68edWqVXv27CkrKzPAFruQUqnkvi8ymaympoavGPcREBCwffv2jz76KCcn58KFC9RW5o6M8KtzmRnZiy75FnRjQYvFYrVaTUS1tbXNf/INsKGOjDyQhISE2NhY7nSwYbYYHh6+ffv2hoaGhISE7t6iXq8/cuQI98a5Jt29j5aWlkRkZWUVGBiYk5NjgC12IZlMVldXR0RKpZLbEdZYWFiIRCKRSBQaGnrr1i1qK3NHRvjVucyM7EWXfAu6saADAgKSkpKI6Nq1a4GB3fhZUK031JGRjktPT9+/f/+qVau4uZsBtsh1kFAoNDc3NzMz6+4t1tbW3r17NzIy8ssvv7xx48bRo0cNsI/19fVEpNVqMzIyuJNI3b3FLsRIjPvgvrxElJqa6uTkRD3qy9ukc5kZ2Ysu+RZ042V2FRUVW7du1el0/fv378IX3C9fvnz06NHr168PGDBg8eLF3t7erTfUkZGOe+WVV0QikZWVFRG98MILrq6u3b3FrVu3lpaWajQaLy+v5557TiAQdPcWOTk5OXv27Gm6iqNbt/jJJ59UVFSo1erRo0dzk3fD7GOX4F5qr6iokEqlbF7F8eOPP8bFxYlEor59+7766qsikah15o6MGDJzWVnZli1bMjMz+/XrN2nSpEcffbRzmXnZi9bhu+RbgOugAQAYhU8lBwBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBR/x/mHgoUMnXTWQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "plot_covariates(sceptre_object_highmoi, p_mito_threshold = 0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "1d947863",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178488.23-P1P2 (1 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178620.23-P1P2 (5 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACTL7B_-_111617995.23-P1P2 (10 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR1B_+_52345762.23-P1P2 (15 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2A_+_148602343.23-P1P2 (20 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_+_38495339.23-P1P2 (25 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_-_38495819.23-P1P2 (30 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ADCY2_-_7396429.23-P1P2 (35 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA AMOTL2_+_134093279.23-P1P2 (40 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ANP32E_+_150208342.23-P1P2 (45 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA APAF1_-_99039071.23-P1P2 (50 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_+_72766725.23-P1P2 (55 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_-_72766755.23-P1P2 (60 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARTN_-_44399429.23-P1 (65 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASCC3_-_101329193.23-P1P2 (70 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASH2L_-_37963059.23-P1P2 (75 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATMIN_+_81069661.23-P1P2 (80 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_+_111992112.23-P1P2 (85 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_-_111992129.23-P1P2 (90 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP6AP1_-_153657075.23-P1P2 (95 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATXN2L_-_28834668.23-P1P2 (100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAD_+_64052102.23-P1P2 (105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAG6_+_31620398.23-P1P2 (110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_+_49458268.23-P1P2 (115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_-_49458323.23-P1P2 (120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BCAS2_-_115124045.23-P1P2 (125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BMPR1A_-_88516386.23-P1P2 (130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BRD3_-_136932911.23-P1P2 (135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C17orf49_+_6918114.23-P2 (140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_+_31532436.23-P1P2 (145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_-_31532650.23-P1P2 (150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACFD1_-_136325160.23-P1P2 (155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACNG1_+_65040756.23-P1P2 (160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CAPN5_+_76778225.23-P1P2 (165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CASP3_+_185570560.23-P1P2 (170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_+_107384793.23-P1P2 (175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_-_107384635.23-P1P2 (180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCM2_-_45039941.23-P1P2 (185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNC_-_100016267.23-P1P2 (190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNT2_-_135676416.23-P1P2 (195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCT6B_+_33288381.23-P1P2 (200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_+_30366413.23-P1P2 (205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_-_30366630.23-P1P2 (210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD99L2_-_150067120.23-P1P2 (215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDH5_-_66400630.23-P1P2 (220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDKN1A_-_36646538.23-P1P2 (225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CENPC_+_68411190.23-P1P2 (230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_+_30773828.23-P1P2 (235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_-_30773391.23-P1P2 (240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHMP3_-_86790451.23-P1P2 (245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHORDC1_-_89955894.23-P1P2 (250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNEP1R1_+_50059284.23-P1P2 (255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNOT2_+_70637651.23-P1P2 (260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_+_83956339.23-P1P2 (265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_-_83956572.23-P1P2 (270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS6_-_99686676.23-P1P2 (275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_-_31634088.23-P1P2 (280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CTR9_+_10772881.23-P1P2 (285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CUL3_-_225449667.23-P1P2 (290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_+_145150267.23-P1P2 (295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_-_145149996.23-P1P2 (300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYCS_+_25164869.23-P1P2 (305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DACH1_-_72441101.23-P1P2 (310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DBR1_+_137893673.23-P1P2 (315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DCTN2_-_57940910.23-P1P2 (320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_+_61851532.23-P1P2 (325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_-_61851812.23-P1P2 (330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DGCR8_-_20068158.23-P1P2 (335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNAJC17_-_41099215.23-P1P2 (340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNM2_+_10829157.23-P1P2 (345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM1_+_49575018.23-P1P2 (350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_+_130700056.23-P1P2 (355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_-_130700076.23-P1P2 (360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM3_+_155113020.23-P1P2 (365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DROSHA_-_31531982.23-P1P2 (370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_+_3985081.23-P1 (375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_-_3985443.23-P1 (380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EHMT2_-_31865199.23-P1P2 (385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3F_+_8009347.23-P1P2 (390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3M_-_32605405.23-P1P2 (395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4E_+_99849831.23-P1P2 (400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184032364.23-P1P2 (405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184033042.23-P1P2 (410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G2_+_10830447.23-P1 (415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF5A_-_7211217.23-P1P2 (420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TCEB3_-_24069975.23-P1P2 (425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EMC2_+_109456080.23-P1P2 (430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823220.23-P1P2 (435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823252.23-P1P2 (440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EOMES_-_27763342.23-P1P2 (445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERCC2_+_45873802.23-P1P2 (450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERO1L_+_53162425.23-P1P2 (455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETS1_+_128392044.23-P1P2 (460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_+_36134592.23-P1P2 (465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_-_36134789.23-P1P2 (470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EWSR1_-_29664347.23-P1P2 (475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EXOSC8_-_37574979.23-P1P2 (480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FBXW11_-_171433300.23-P1P2 (485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FEN1_+_61560782.23-P1P2 (490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_+_75532080.23-P1P2 (495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_-_75532461.23-P1P2 (500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT7_-_139926495.23-P1P2 (505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA G3BP1_-_151151499.23-P1P2 (510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GCK_+_44229021.23-P1P2 (515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOLT1B_+_21654777.23-P1P2 (520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_+_45000529.23-P1P2 (525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_-_45000948.23-P1P2 (530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPI_+_34856284.23-P1P2 (535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPR107_-_132816044.23-P1P2 (540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GSPT1_-_12009714.23-P1P2 (545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A1_+_81687215.23-P1P2 (550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_+_59949355.23-P1P2 (555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_-_59949657.23-P1P2 (560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2H4_-_30876185.23-P1P2 (565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF3C2_+_27579604.23-P1P2 (570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HAND1_+_153857732.23-P1P2 (575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HDAC2_-_114292253.23-P1P2 (580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_+_81399312.23-P1P2 (585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_-_81399351.23-P1P2 (590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HINT2_+_35814995.23-P1P2 (595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNF4A_+_43029971.23-P1P2 (600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPM_-_8509885.23-P1 (605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPU_+_245027753.23-P1P2 (610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_+_43702292.23-P1P2 (615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_-_43702305.23-P1P2 (620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSPA5_-_128003534.23-P1P2 (625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HUWE1_-_53710540.23-P1P2 (630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_-_30007588.23-P1P2 (635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INTS10_+_19675000.23-P1P2 (640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_+_65903046.23-P1P2 (645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_-_65903523.23-P1P2 (650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C7orf26_+_6630358.23-P1P2 (655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA IPO8_+_30848771.23-P1P2 (660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ITGB1BP1_+_9563543.23-P1P2 (665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSFA2_+_182756743.23-P1P2 (670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_+_109914946.23-P1P2 (675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_-_109915112.23-P1P2 (680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM1A_-_23346409.23-P1P2 (685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM3B_-_137688431.23-P1P2 (690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDR_-_55991659.23-P1P2 (695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA0232_+_6784651.23-P1P2 (700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153706791.23-P1P2 (705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153707248.23-P1P2 (710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAMTOR4_-_99746921.23-P1P2 (715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LARP4_-_50794864.23-P1P2 (720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LATS2_-_21635336.23-P1P2 (725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LEMD3_+_65563441.23-P1P2 (730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_+_54960428.23-P1P2 (735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_-_54960616.23-P1P2 (740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LHFPL2_-_77944519.23-P1P2 (745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LSM12_-_42143603.23-P1P2 (750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MANBAL_+_35918136.23-P1P2 (755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MAU2_+_19431732.23-P1P2 (760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_+_47807678.23-P1P2 (765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_-_47807877.23-P1P2 (770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEAF6_-_37980319.23-P1P2 (775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED1_-_37607140.23-P1P2 (780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED12_+_70338541.23-P1P2 (785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED15_+_20862189.23-P1 (790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_+_57479335.23-P1P2 (795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_-_57479437.23-P1P2 (800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEMO1_-_32264882.23-P1 (805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MIXL1_+_226411425.23-P1P2 (810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL28_+_420487.23-P1P2 (815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL32_+_42972018.23-P1P2 (820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_+_179322181.23-P1P2 (825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_-_179322213.23-P1P2 (830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LUZP6_-_135612750.23-P1P2 (835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA20_-_19998023.23-P1P2 (840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA25_+_112546538.23-P1P2 (845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCBP1_+_100396082.23-P1P2 (850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_+_172428630.23-P1P2 (855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_-_172428841.23-P1P2 (860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NEDD8_-_24701363.23-P1P2 (865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NF2_+_29999635.23-P1P2 (870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NFIB_+_14314559.23-P1P2 (875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_+_139418329.23-P2 (880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139418349.23-P2 (885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139440503.23-P1 (890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUFIP2_+_27621073.23-P1P2 (895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUP93_-_56764090.23-P1P2 (900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OGT_+_70752948.23-P1P2 (905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_+_59383170.23-P1P2 (910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_-_59383348.23-P1P2 (915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAM16_-_4401236.23-P1P2 (920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAXBP1_-_34143614.23-P1P2 (925 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PBLD_-_70092436.23-P1 (930 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PCBP2_+_53835614.23-P2 (935 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_+_94292752.23-P2 (940 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_-_94293027.23-P2 (945 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PGRMC2_+_129208913.23-P1P2 (950 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_-_178866211.23-P1P2 (955 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PITPNC1_-_65373433.23-P1P2 (960 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLCG1_+_39766176.23-P1P2 (965 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_+_151042273.23-P2 (970 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_-_151042422.23-P2 (975 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLRG1_-_155471154.23-P1P2 (980 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLXNA2_-_208417275.23-P1P2 (985 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLE4_-_75185725.23-P1P2 (990 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR2M_+_57998962.23-P1P2 (995 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940045.23-P1P2 (1000 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940414.23-P1P2 (1005 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3K_-_103144.23-P1P2 (1010 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPHLN1_-_42720007.23-P1P2 (1015 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPIL4_+_149867146.23-P1P2 (1020 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPP2R1A_+_52693317.23-P1 (1025 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_+_53195217.23-P1P2 (1030 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_-_53195332.23-P1P2 (1035 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKD1_-_30396996.23-P1P2 (1040 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRMT5_+_23398554.23-P1 (1045 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C2orf53_-_27362230.23-P1P2 (1050 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRSS8_+_31147010.23-P1P2 (1055 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_+_36909062.23-P1P2 (1060 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_-_36909230.23-P1P2 (1065 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMD4_-_151227289.23-P1P2 (1070 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMG4_-_3259227.23-P1P2 (1075 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTAR1_+_72374835.23-P1P2 (1080 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTBP1_+_797463.23-P1P2 (1085 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_+_89623388.23-P1P2 (1090 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_-_89623526.23-P1P2 (1095 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTMS_-_6875591.23-P1P2 (1100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PYGO2_-_154934135.23-P1P2 (1105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB18_-_27793292.23-P1P2 (1110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB6A_+_73471983.23-P1P2 (1115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_+_76251918.23-P1P2 (1120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_-_76252008.23-P1P2 (1125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GNB2L1_-_180670840.23-P1P2 (1130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAD21_-_117886635.23-P1P2 (1135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RANGRF_+_8192257.23-P1P2 (1140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBBP5_+_205091050.23-P1P2 (1145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM14-RBM4_+_66406280.23-P2 (1150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM33_+_155437802.23-P1P2 (1155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM4_+_66406319.23-P2 (1160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_+_114271352.23-P1P2 (1165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_-_114271700.23-P1P2 (1170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBX1_-_41347852.23-P1P2 (1175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1551_+_32112399.23-P1P2 (1180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RFX3_-_3525750.23-P1P2 (1185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RGP1_+_35749328.23-P1P2 (1190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_+_95067102.23-P1P2 (1195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_-_95067164.23-P1P2 (1200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RIBC1_-_53449606.23-P1P2 (1205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1432_+_5629173.23-P1P2 (1210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RLF_+_40627522.23-P1P2 (1215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF115_-_145611148.23-P1P2 (1220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_+_104296142.23-P1P2 (1225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_-_104296250.23-P1P2 (1230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF40_-_30773681.23-P1P2 (1235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPL37A_-_217363600.23-P1P2 (1240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPS15A_-_18801590.23-P1P2 (1245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAV1_+_51134992.23-P1P2 (1250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_+_33104038.23-P1P2 (1255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_-_33104343.23-P1P2 (1260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF8_+_155054976.23-P1P2 (1265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EPT1_+_26569050.23-P1P2 (1270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERF2_+_44069659.23-P2 (1275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERPING1_+_57365137.23-P1P2 (1280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_+_122242041.23-P1P2 (1285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_-_122242294.23-P1P2 (1290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETDB1_-_150898815.23-P1P2 (1295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SF1_-_64545937.23-P1P2 (1300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCDC101_-_28565268.23-P1P2 (1305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SKA3_+_21750519.23-P1P2 (1310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_+_78591497.23-P1P2 (1315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_-_78591737.23-P1P2 (1320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC29A1_-_44187438.23-P2 (1325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC35C1_-_45826711.23-P1P2 (1330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD1_+_146403428.23-P1P2 (1335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD3_+_67358229.23-P1P2 (1340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_+_48556623.23-P1P2 (1345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_-_48556657.23-P1P2 (1350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD5_-_135468694.23-P1P2 (1355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA4_-_11071741.23-P1P2 (1360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA5_+_144435243.23-P1P2 (1365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCB1_+_24129228.23-P1 (1370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_+_47822975.23-P1P2 (1375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_-_47822893.23-P1P2 (1380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCD3_-_150974212.23-P1 (1385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCE1_-_38803669.23-P1P2 (1390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMC2_-_106856631.23-P1P2 (1395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMG5_+_156252537.23-P1P2 (1400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_+_103852.23-P1P2 (1405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_-_104242.23-P1P2 (1410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP27_-_70121181.23-P1P2 (1415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNUPN_-_75917960.23-P1P2 (1420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SOX4_+_21594213.23-P1 (1425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPIN1_+_91003737.23-P1P2 (1430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_+_2328680.23-P1P2 (1435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_-_2328912.23-P1P2 (1440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPTY2D1_-_18655846.23-P1P2 (1445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP19_+_112197188.23-P1P2 (1450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP54_-_35452244.23-P1P2 (1455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP68_+_74068540.23-P1P2 (1460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_+_126138730.23-P1P2 (1465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_-_126138742.23-P1P2 (1470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPRB_-_133524719.23-P1P2 (1475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRRM2_-_2802714.23-P1P2 (1480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRSF6_+_42086993.23-P1 (1485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSB_+_170655450.23-P1P2 (1490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_+_57102902.23-P1P2 (1495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_-_57103355.23-P1P2 (1500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STAG2_-_123095658.23-P1 (1505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STK40_+_36851507.23-P1P2 (1510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUPT20H_+_37633798.23-P1P2 (1515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUZ12_+_30264157.23-P1P2 (1520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_+_64901973.23-P1P2 (1525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_-_64901981.23-P1P2 (1530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TADA2B_-_7045121.23-P1P2 (1535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF10_+_6633406.23-P1P2 (1540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF2_+_120845032.23-P1P2 (1545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF3_-_7860540.23-P1P2 (1550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_+_60640814.23-P1P2 (1555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_-_60640498.23-P1P2 (1560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF6L_-_62538954.23-P1P2 (1565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAOK1_-_27717907.23-P1P2 (1570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FAM208A_+_56717273.23-P1P2 (1575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBC1D9_+_141677481.23-P1P2 (1580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_+_170863524.23-P1P2 (1585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_-_170863481.23-P1P2 (1590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBX3_-_115121998.23-P1P2 (1595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA T_-_166581333.23-P1P2 (1600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TGFBR2_+_30648526.23-P1P2 (1605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA THBD_+_23029196.23-P2 (1610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_+_2977592.23-P1P2 (1615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_-_2977643.23-P1P2 (1620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MESDC1_-_81293340.23-P1P2 (1625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMEM41B_+_9336156.23-P1P2 (1630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMTC3_-_88536127.23-P1P2 (1635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TOP1_+_39657538.23-P1P2 (1640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_+_7590531.23-P1P2 (1645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_-_7590704.23-P1P2 (1650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRIM24_-_138145308.23-P1P2 (1655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRRAP_-_98476465.23-P1P2 (1660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TSTD2_+_100395594.23-P1P2 (1665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UBE2E1_+_23847541.23-P1P2 (1670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_+_38924011.23-P1P2 (1675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_-_38924421.23-P1P2 (1680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UPF2_+_12084799.23-P1P2 (1685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UXT_-_47518505.23-P1P2 (1690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VHL_-_10183525.23-P1P2 (1695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VPS54_+_64246104.23-P1P2 (1700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_+_28821698.23-P1P2 (1705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_-_28822074.23-P1P2 (1710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAPAL_+_88281533.23-P1P2 (1715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR12_-_203776383.23-P1P2 (1720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR77_+_111991860.23-P1P2 (1725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WIZ_-_15543016.23-P1P2 (1730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_+_9983328.23-P1P2 (1735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_-_9983510.23-P1P2 (1740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XPO5_-_43543670.23-P1P2 (1745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XRN1_-_142166648.23-P1P2 (1750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA YARS2_+_32908787.23-P1P2 (1755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB22_+_33285454.23-P1P2 (1760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_+_6640093.23-P1P2 (1765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_-_6640138.23-P1P2 (1770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZC3H8_-_113012435.23-P1P2 (1775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZNF281_+_200379151.23-P1P2 (1780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00117 (1785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00233 (1790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00518 (1795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00656 (1800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01226 (1805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01730 (1810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02009 (1815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02288 (1820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02777 (1825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03026 (1830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03252 (1835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03620 (1840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_SGTN_04 (1845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_SGTN_25 (1850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_SGTN_26 (1855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1044 (1860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1414 (1865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1554 (1870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1785 (1875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2009 (1880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2271 (1885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2425 (1890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2614 (1895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_293 (1900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_541 (1905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_725 (1910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_860 (1915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_SGTN_14 (1920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_SGTN_01 (1925 of 1926)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- assign_grnas(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  method = \"mixture\",\n",
    "  parallel = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2e64f05e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     cell_id                grna_target value\n",
      "1 AACGTGATGTGTTCTATACGGTTA_2 CNEP1R1_+_50059217.23-P1P2     1\n",
      "2 AACGTGATGTGTTCTATACGGTTA_2 SPTY2D1_-_18655660.23-P1P2     1\n",
      "3 AACGTGATGTGTTCTATACGGTTA_2   TRRAP_-_98476562.23-P1P2     1\n",
      "4 CCATCCTCCAACCACATTGGGAGA_3 ATP5F1_+_111992171.23-P1P2     1\n",
      "5 CCATCCTCCAACCACATTGGGAGA_3  C7orf26_+_6629634.23-P1P2     1\n",
      "6 CCATCCTCCAACCACATTGGGAGA_3   PLCG1_+_39766234.23-P1P2     1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "⚠️ Matrix missing colnames. Fetching from sceptre_object...\n",
       "✅ Saved long CSV to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/grna_assignments_long.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(Matrix)\n",
    "\n",
    "# 1. Get the assignment matrix\n",
    "A <- get_grna_assignments(sceptre_object_highmoi)\n",
    "\n",
    "# 2. Ensure it is a numeric sparse matrix (1/0)\n",
    "if (inherits(A, \"Matrix\") || is.matrix(A)) {\n",
    "    A_num <- if (inherits(A, \"Matrix\")) drop0(A * 1) else (A * 1)\n",
    "} else {\n",
    "    stop(\"Output is not a matrix.\")\n",
    "}\n",
    "\n",
    "# --- FIX: ROBUST NAME RECOVERY ---\n",
    "# We retrieve names using Sceptre's standard metadata accessors\n",
    "# instead of touching internal slots like @grna_matrix.\n",
    "\n",
    "# Try to get cell names from the matrix first\n",
    "current_cell_names <- colnames(A_num)\n",
    "current_guide_names <- rownames(A_num)\n",
    "\n",
    "# If missing, grab them from the main object's covariate data frame\n",
    "if (is.null(current_cell_names)) {\n",
    "    message(\"⚠️ Matrix missing colnames. Fetching from sceptre_object...\")\n",
    "    # The covariate data frame ALWAYS has the cell IDs as rownames\n",
    "    current_cell_names <- rownames(sceptre_object_highmoi@covariate_data_frame)\n",
    "    \n",
    "    # Safety check: lengths must match\n",
    "    if (length(current_cell_names) != ncol(A_num)) {\n",
    "        stop(paste(\"Mismatch! Matrix has\", ncol(A_num), \"cols but object has\", length(current_cell_names), \"cells.\"))\n",
    "    }\n",
    "    colnames(A_num) <- current_cell_names\n",
    "}\n",
    "\n",
    "# If guide names are missing, they should match the unique values in your guide mapping\n",
    "if (is.null(current_guide_names)) {\n",
    "    message(\"⚠️ Matrix missing rownames. Fetching from guide mapping...\")\n",
    "    # Get unique targets from the mapping table you loaded earlier\n",
    "    # Note: This assumes the matrix rows are sorted/aligned to these unique targets.\n",
    "    # In 'mixture' method, rows usually correspond to the 'grna_target' labels.\n",
    "    \n",
    "    # SAFE FALLBACK: If we can't be 100% sure of order, we warn the user.\n",
    "    # But usually, they match the 'grna_target' column from your input.\n",
    "    unique_targets <- sort(unique(guide_mapping$grna_target))\n",
    "    \n",
    "    if (length(unique_targets) == nrow(A_num)) {\n",
    "         rownames(A_num) <- unique_targets\n",
    "    } else {\n",
    "         warning(\"Could not automatically align target names. Using generic IDs.\")\n",
    "         rownames(A_num) <- paste0(\"target_\", seq_len(nrow(A_num)))\n",
    "    }\n",
    "}\n",
    "# ---------------------------------\n",
    "\n",
    "# 3. Convert to Long Format\n",
    "s <- summary(A_num)\n",
    "assign_long <- data.frame(\n",
    "    cell_id = colnames(A_num)[s$j],\n",
    "    grna_target = rownames(A_num)[s$i],\n",
    "    value = s$x\n",
    ")\n",
    "\n",
    "# 4. Save\n",
    "out_path <- file.path(data_dir, \"grna_assignments_long.csv\")\n",
    "write.csv(assign_long, out_path, row.names = FALSE)\n",
    "\n",
    "message(\"✅ Saved long CSV to: \", out_path)\n",
    "print(head(assign_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "a09cda94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/MOI.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "png(file.path(data_dir, \"MOI.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"MOI.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "2aef1561",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/QC.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_qc(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  p_mito_threshold = 0.25\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"QC.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"QC.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b5dca792",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Constructing negative control pairs. ✓\n",
      "Running precomputation on response ENSG00000197747.9 (1 of 242)\n",
      "Running precomputation on response ENSG00000103035.11 (5 of 242)\n",
      "Running precomputation on response ENSG00000077150.21 (10 of 242)\n",
      "Running precomputation on response ENSG00000175198.17 (15 of 242)\n",
      "Running precomputation on response ENSG00000236859.8 (20 of 242)\n",
      "Running precomputation on response ENSG00000176171.12 (25 of 242)\n",
      "Running precomputation on response ENSG00000162733.19 (30 of 242)\n",
      "Running precomputation on response ENSG00000269547.1 (35 of 242)\n",
      "Running precomputation on response ENSG00000145996.11 (40 of 242)\n",
      "Running precomputation on response ENSG00000116141.17 (45 of 242)\n",
      "Running precomputation on response ENSG00000166025.18 (50 of 242)\n",
      "Running precomputation on response ENSG00000089775.12 (55 of 242)\n",
      "Running precomputation on response ENSG00000099204.22 (60 of 242)\n",
      "Running precomputation on response ENSG00000157224.16 (65 of 242)\n",
      "Running precomputation on response ENSG00000173207.13 (70 of 242)\n",
      "Running precomputation on response ENSG00000008869.12 (75 of 242)\n",
      "Running precomputation on response ENSG00000111145.8 (80 of 242)\n",
      "Running precomputation on response ENSG00000137815.15 (85 of 242)\n",
      "Running precomputation on response ENSG00000213614.11 (90 of 242)\n",
      "Running precomputation on response ENSG00000048342.18 (95 of 242)\n",
      "Running precomputation on response ENSG00000197302.11 (100 of 242)\n",
      "Running precomputation on response ENSG00000065243.20 (105 of 242)\n",
      "Running precomputation on response ENSG00000080546.14 (110 of 242)\n",
      "Running precomputation on response ENSG00000138835.24 (115 of 242)\n",
      "Running precomputation on response ENSG00000118971.9 (120 of 242)\n",
      "Running precomputation on response ENSG00000182287.15 (125 of 242)\n",
      "Running precomputation on response ENSG00000143786.8 (130 of 242)\n",
      "Running precomputation on response ENSG00000034713.8 (135 of 242)\n",
      "Running precomputation on response ENSG00000133874.2 (140 of 242)\n",
      "Running precomputation on response ENSG00000101413.12 (145 of 242)\n",
      "Running precomputation on response ENSG00000269609.6 (150 of 242)\n",
      "Running precomputation on response ENSG00000067829.20 (155 of 242)\n",
      "Running precomputation on response ENSG00000269955.2 (160 of 242)\n",
      "Running precomputation on response ENSG00000121897.15 (165 of 242)\n",
      "Running precomputation on response ENSG00000110651.13 (170 of 242)\n",
      "Running precomputation on response ENSG00000030419.17 (175 of 242)\n",
      "Running precomputation on response ENSG00000136866.14 (180 of 242)\n",
      "Running precomputation on response ENSG00000181830.9 (185 of 242)\n",
      "Running precomputation on response ENSG00000176681.15 (190 of 242)\n",
      "Running precomputation on response ENSG00000148730.7 (195 of 242)\n",
      "Running precomputation on response ENSG00000069869.17 (200 of 242)\n",
      "Running precomputation on response ENSG00000109670.17 (205 of 242)\n",
      "Running precomputation on response ENSG00000107521.21 (210 of 242)\n",
      "Running precomputation on response ENSG00000086015.23 (215 of 242)\n",
      "Running precomputation on response ENSG00000112343.11 (220 of 242)\n",
      "Running precomputation on response ENSG00000181027.11 (225 of 242)\n",
      "Running precomputation on response ENSG00000171314.9 (230 of 242)\n",
      "Running precomputation on response ENSG00000044090.13 (235 of 242)\n",
      "Running precomputation on response ENSG00000143727.16 (240 of 242)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00259&non-targeting_02159&non-targeting_02571&non-targeting_03029&non-targeting_03081 (1 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00355&non-targeting_02025&non-targeting_02089&non-targeting_03081&non-targeting_03384 (5 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_01730&non-targeting_02193&non-targeting_03180&non-targeting_03620&non-targeting_03664 (10 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00043&non-targeting_00184&non-targeting_02677&non-targeting_02777&non-targeting_02814&non-targeting_03427 (15 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00355&non-targeting_00390&non-targeting_00918&non-targeting_01509&non-targeting_01599 (20 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00972&non-targeting_02193&non-targeting_02574&non-targeting_02777&non-targeting_02962 (25 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00355&non-targeting_01601&non-targeting_03026&non-targeting_03081&non-targeting_03158 (30 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00231&non-targeting_00518&non-targeting_00629&non-targeting_00794&non-targeting_02777&non-targeting_02837 (35 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00390&non-targeting_01601&non-targeting_02009&non-targeting_02288&non-targeting_03029 (40 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00231&non-targeting_00324&non-targeting_02025&non-targeting_03252&non-targeting_03384 (45 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00043&non-targeting_00213&non-targeting_00259&non-targeting_00355&non-targeting_00614&non-targeting_01934 (50 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00629&non-targeting_01185&non-targeting_01732&non-targeting_02089&non-targeting_02677&non-targeting_03483 (55 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00518&non-targeting_00614&non-targeting_02025&non-targeting_03342&non-targeting_03427 (60 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00231&non-targeting_00656&non-targeting_02025&non-targeting_02288&non-targeting_02777 (65 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00614&non-targeting_00972&non-targeting_02911&non-targeting_03081&non-targeting_03180&non-targeting_03384 (70 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_00555&non-targeting_01934&non-targeting_02193&non-targeting_02677&non-targeting_03252 (75 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00656&non-targeting_00918&non-targeting_01934&non-targeting_02574&non-targeting_02814 (80 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_00355&non-targeting_02837&non-targeting_02911&non-targeting_03342&non-targeting_03620 (85 of 92)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00518&non-targeting_00614&non-targeting_01226&non-targeting_02025&non-targeting_02837 (90 of 92)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/calibration_check.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Calibration check\n",
    "\n",
    "\n",
    "sceptre_object_highmoi <- run_calibration_check(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"calibration_check.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"calibration_check.png\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "8966c825",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          response_id\n",
      "               <char>\n",
      "1: ENSG00000121957.15\n",
      "2: ENSG00000115414.21\n",
      "3:  ENSG00000122692.9\n",
      "4: ENSG00000150995.21\n",
      "5: ENSG00000048471.14\n",
      "6: ENSG00000149262.19\n",
      "                                                                                                               grna_target\n",
      "                                                                                                                    <char>\n",
      "1: non-targeting_00259&non-targeting_01934&non-targeting_02677&non-targeting_03081&non-targeting_03252&non-targeting_03664\n",
      "2: non-targeting_00355&non-targeting_00518&non-targeting_00555&non-targeting_00794&non-targeting_01185&non-targeting_02159\n",
      "3: non-targeting_00184&non-targeting_00259&non-targeting_02159&non-targeting_02571&non-targeting_03029&non-targeting_03081\n",
      "4: non-targeting_01504&non-targeting_01601&non-targeting_01732&non-targeting_01812&non-targeting_02962&non-targeting_03483\n",
      "5: non-targeting_00231&non-targeting_01730&non-targeting_02089&non-targeting_03180&non-targeting_03342&non-targeting_03384\n",
      "6: non-targeting_00043&non-targeting_00117&non-targeting_00555&non-targeting_00918&non-targeting_03252&non-targeting_03427\n",
      "   n_nonzero_trt n_nonzero_cntrl pass_qc    p_value log_2_fold_change\n",
      "           <int>           <int>  <lgcl>      <num>             <num>\n",
      "1:            31            1118    TRUE 0.02530799         0.5678968\n",
      "2:            59            4094    TRUE 0.02962683         0.3888507\n",
      "3:            81            3788    TRUE 0.03600000        -0.3017934\n",
      "4:            16            1773    TRUE 0.03600000        -0.6002785\n",
      "5:            41            1598    TRUE 0.04000000         0.3800087\n",
      "6:            38            1655    TRUE 0.04000000         0.4606734\n",
      "   significant\n",
      "        <lgcl>\n",
      "1:       FALSE\n",
      "2:       FALSE\n",
      "3:       FALSE\n",
      "4:       FALSE\n",
      "5:       FALSE\n",
      "6:       FALSE\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "calibration_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_calibration_check\"\n",
    ")\n",
    "head(calibration_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5119bbb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Running precomputation on response ENSG00000004487.18 (1 of 243)\n",
      "Running precomputation on response ENSG00000117000.9 (5 of 243)\n",
      "Running precomputation on response ENSG00000265491.5 (10 of 243)\n",
      "Running precomputation on response ENSG00000198952.8 (15 of 243)\n",
      "Running precomputation on response ENSG00000153187.21 (20 of 243)\n",
      "Running precomputation on response ENSG00000162961.14 (25 of 243)\n",
      "Running precomputation on response ENSG00000121989.15 (30 of 243)\n",
      "Running precomputation on response ENSG00000036257.14 (35 of 243)\n",
      "Running precomputation on response ENSG00000114739.14 (40 of 243)\n",
      "Running precomputation on response ENSG00000114127.11 (45 of 243)\n",
      "Running precomputation on response ENSG00000128052.10 (50 of 243)\n",
      "Running precomputation on response ENSG00000153147.6 (55 of 243)\n",
      "Running precomputation on response ENSG00000113360.17 (60 of 243)\n",
      "Running precomputation on response ENSG00000113658.18 (65 of 243)\n",
      "Running precomputation on response ENSG00000180822.12 (70 of 243)\n",
      "Running precomputation on response ENSG00000124571.18 (75 of 243)\n",
      "Running precomputation on response ENSG00000131013.4 (80 of 243)\n",
      "Running precomputation on response ENSG00000106591.4 (85 of 243)\n",
      "Running precomputation on response ENSG00000188186.11 (90 of 243)\n",
      "Running precomputation on response ENSG00000104613.12 (95 of 243)\n",
      "Running precomputation on response ENSG00000179091.5 (100 of 243)\n",
      "Running precomputation on response ENSG00000106723.17 (105 of 243)\n",
      "Running precomputation on response ENSG00000136908.18 (110 of 243)\n",
      "Running precomputation on response ENSG00000151461.20 (115 of 243)\n",
      "Running precomputation on response ENSG00000107779.14 (120 of 243)\n",
      "Running precomputation on response ENSG00000198730.9 (125 of 243)\n",
      "Running precomputation on response ENSG00000149136.9 (130 of 243)\n",
      "Running precomputation on response ENSG00000162298.19 (135 of 243)\n",
      "Running precomputation on response ENSG00000076053.12 (140 of 243)\n",
      "Running precomputation on response ENSG00000133704.10 (145 of 243)\n",
      "Running precomputation on response ENSG00000197111.16 (150 of 243)\n",
      "Running precomputation on response ENSG00000110906.13 (155 of 243)\n",
      "Running precomputation on response ENSG00000165480.16 (160 of 243)\n",
      "Running precomputation on response ENSG00000100462.16 (165 of 243)\n",
      "Running precomputation on response ENSG00000165417.12 (170 of 243)\n",
      "Running precomputation on response ENSG00000166949.17 (175 of 243)\n",
      "Running precomputation on response ENSG00000161981.11 (180 of 243)\n",
      "Running precomputation on response ENSG00000134419.15 (185 of 243)\n",
      "Running precomputation on response ENSG00000052344.16 (190 of 243)\n",
      "Running precomputation on response ENSG00000141510.18 (195 of 243)\n",
      "Running precomputation on response ENSG00000125686.12 (200 of 243)\n",
      "Running precomputation on response ENSG00000154217.16 (205 of 243)\n",
      "Running precomputation on response ENSG00000005206.17 (210 of 243)\n",
      "Running precomputation on response ENSG00000011451.21 (215 of 243)\n",
      "Running precomputation on response ENSG00000105568.19 (220 of 243)\n",
      "Running precomputation on response ENSG00000124193.16 (225 of 243)\n",
      "Running precomputation on response ENSG00000128191.17 (230 of 243)\n",
      "Running precomputation on response ENSG00000100387.9 (235 of 243)\n",
      "Running precomputation on response ENSG00000147162.15 (240 of 243)\n",
      "Analyzing pairs containing gRNA group KDM1A (1 of 243)\n",
      "Analyzing pairs containing gRNA group RLF (5 of 243)\n",
      "Analyzing pairs containing gRNA group RNF115 (10 of 243)\n",
      "Analyzing pairs containing gRNA group SMG5 (15 of 243)\n",
      "Analyzing pairs containing gRNA group HNRNPU (20 of 243)\n",
      "Analyzing pairs containing gRNA group DPY30 (25 of 243)\n",
      "Analyzing pairs containing gRNA group ACVR2A (30 of 243)\n",
      "Analyzing pairs containing gRNA group CUL3 (35 of 243)\n",
      "Analyzing pairs containing gRNA group ACVR2B (40 of 243)\n",
      "Analyzing pairs containing gRNA group XRN1 (45 of 243)\n",
      "Analyzing pairs containing gRNA group KDR (50 of 243)\n",
      "Analyzing pairs containing gRNA group SMARCA5 (55 of 243)\n",
      "Analyzing pairs containing gRNA group DROSHA (60 of 243)\n",
      "Analyzing pairs containing gRNA group SMAD5 (65 of 243)\n",
      "Analyzing pairs containing gRNA group PSMG4 (70 of 243)\n",
      "Analyzing pairs containing gRNA group XPO5 (75 of 243)\n",
      "Analyzing pairs containing gRNA group PPIL4 (80 of 243)\n",
      "Analyzing pairs containing gRNA group MRPL32 (85 of 243)\n",
      "Analyzing pairs containing gRNA group LAMTOR4 (90 of 243)\n",
      "Analyzing pairs containing gRNA group INTS10 (95 of 243)\n",
      "Analyzing pairs containing gRNA group CYC1 (100 of 243)\n",
      "Analyzing pairs containing gRNA group SPIN1 (105 of 243)\n",
      "Analyzing pairs containing gRNA group DPM2 (110 of 243)\n",
      "Analyzing pairs containing gRNA group UPF2 (115 of 243)\n",
      "Analyzing pairs containing gRNA group BMPR1A (120 of 243)\n",
      "Analyzing pairs containing gRNA group CTR9 (125 of 243)\n",
      "Analyzing pairs containing gRNA group SSRP1 (130 of 243)\n",
      "Analyzing pairs containing gRNA group SYVN1 (135 of 243)\n",
      "Analyzing pairs containing gRNA group RBM7 (140 of 243)\n",
      "Analyzing pairs containing gRNA group IPO8 (145 of 243)\n",
      "Analyzing pairs containing gRNA group PCBP2 (150 of 243)\n",
      "Analyzing pairs containing gRNA group KCTD10 (155 of 243)\n",
      "Analyzing pairs containing gRNA group SKA3 (160 of 243)\n",
      "Analyzing pairs containing gRNA group PRMT5 (165 of 243)\n",
      "Analyzing pairs containing gRNA group GTF2A1 (170 of 243)\n",
      "Analyzing pairs containing gRNA group SMAD3 (175 of 243)\n",
      "Analyzing pairs containing gRNA group SNRNP25 (180 of 243)\n",
      "Analyzing pairs containing gRNA group RPS15A (185 of 243)\n",
      "Analyzing pairs containing gRNA group PRSS8 (190 of 243)\n",
      "Analyzing pairs containing gRNA group TP53 (195 of 243)\n",
      "Analyzing pairs containing gRNA group MED1 (200 of 243)\n",
      "Analyzing pairs containing gRNA group PITPNC1 (205 of 243)\n",
      "Analyzing pairs containing gRNA group SPPL2B (210 of 243)\n",
      "Analyzing pairs containing gRNA group WIZ (215 of 243)\n",
      "Analyzing pairs containing gRNA group PPP2R1A (220 of 243)\n",
      "Analyzing pairs containing gRNA group SRSF6 (225 of 243)\n",
      "Analyzing pairs containing gRNA group DGCR8 (230 of 243)\n",
      "Analyzing pairs containing gRNA group RBX1 (235 of 243)\n",
      "Analyzing pairs containing gRNA group OGT (240 of 243)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/discovery.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_discovery_analysis(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"discovery.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"discovery.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0737a97d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved discovery to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d1_sceptre_transfer/TargetKD.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "discovery_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_discovery_analysis\"\n",
    ")\n",
    "\n",
    "write.csv(discovery_result, file.path(data_dir, \"TargetKD.csv\"), row.names = FALSE)\n",
    "message(\"✅ Saved discovery to: \", file.path(data_dir, \"TargetKD.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07e64002",
   "metadata": {},
   "source": [
    "### Day 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e86afea5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1598: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1461: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "d2_mdata = mu.read(f\"{data_dir}/d2_normalized_gex_and_guide.h5mu\")\n",
    "mdata = d2_mdata\n",
    "\n",
    "# Filter out cells that have no guides\n",
    "#mdata = mdata[mdata['GEX'].obs['guides_per_cell'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a37798be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing sparse matrices...\n"
     ]
    }
   ],
   "source": [
    "# Define a temporary folder for transfer\n",
    "temp_dir = working_dir + \"/d2_sceptre_transfer\"\n",
    "os.makedirs(temp_dir, exist_ok=True)\n",
    "\n",
    "# 1. EXTRACT MATRICES (Transpose them!)\n",
    "# Scanpy is (Cells × Genes), Sceptre needs (Genes × Cells)\n",
    "print(\"Writing sparse matrices...\")\n",
    "sio.mmwrite(f\"{temp_dir}/response_matrix.mtx\", mdata['GEX'].layers['raw_counts'].T)\n",
    "sio.mmwrite(f\"{temp_dir}/grna_matrix.mtx\", mdata['guide'].X.T)\n",
    "\n",
    "# 2. EXTRACT NAMES (Row/Col names are lost in .mtx)\n",
    "# Save Gene names and Guide names to assign them later in R\n",
    "pd.Series(mdata['GEX'].var_names).to_csv(f\"{temp_dir}/gene_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['guide'].var_names).to_csv(f\"{temp_dir}/guide_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['GEX'].obs_names).to_csv(f\"{temp_dir}/cell_names.csv\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a37798be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 21980 × 12732\n",
       "    obs: 'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "    var: 'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "    uns: 'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "    varm: 'PCs'\n",
       "    layers: 'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['GEX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ec11d9e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>grna_n_nonzero</th>\n",
       "      <th>grna_n_umis</th>\n",
       "      <th>response_n_nonzero</th>\n",
       "      <th>response_n_umis</th>\n",
       "      <th>response_p_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGGATGAATTGAGGATGCTTGGG_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>1</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1276</td>\n",
       "      <td>2078.232178</td>\n",
       "      <td>13.304576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGGATGAAGCCATGCCGCCCGGA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>2</td>\n",
       "      <td>142.0</td>\n",
       "      <td>561</td>\n",
       "      <td>849.433411</td>\n",
       "      <td>17.541105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCAGTTCACAATGGAATGCTTGGG_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>5</td>\n",
       "      <td>245.0</td>\n",
       "      <td>1141</td>\n",
       "      <td>1890.696289</td>\n",
       "      <td>15.073812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCATTCCAGTCACTACGCCCGGA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>1</td>\n",
       "      <td>18.0</td>\n",
       "      <td>407</td>\n",
       "      <td>549.833374</td>\n",
       "      <td>16.914215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGAGTAAGCTAACGACACAATTG_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>10</td>\n",
       "      <td>198.0</td>\n",
       "      <td>680</td>\n",
       "      <td>1287.249268</td>\n",
       "      <td>22.101393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTGAGAACGCTCGACTTTGGTC_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>5</td>\n",
       "      <td>165.0</td>\n",
       "      <td>1783</td>\n",
       "      <td>3590.150391</td>\n",
       "      <td>14.275167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACCTTACACAGCAGACACAATTG_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>17</td>\n",
       "      <td>1121.0</td>\n",
       "      <td>1462</td>\n",
       "      <td>2993.500000</td>\n",
       "      <td>23.300484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAGACTAGCCAAGACTAAATATC_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>2</td>\n",
       "      <td>134.0</td>\n",
       "      <td>1345</td>\n",
       "      <td>2546.260254</td>\n",
       "      <td>9.209585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAATGGAAGCTCGGTAGTGCTAGC_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>6</td>\n",
       "      <td>105.0</td>\n",
       "      <td>714</td>\n",
       "      <td>1144.333252</td>\n",
       "      <td>19.181475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCAGTTCACTGAGCCATAAATATC_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>9</td>\n",
       "      <td>139.0</td>\n",
       "      <td>879</td>\n",
       "      <td>1110.250244</td>\n",
       "      <td>14.636339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21980 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             batch  grna_n_nonzero  \\\n",
       "barcode                                                              \n",
       "TAGGATGAATTGAGGATGCTTGGG_2  300genes:eTN5_GEX_sub2               1   \n",
       "TAGGATGAAGCCATGCCGCCCGGA_2  300genes:eTN5_GEX_sub2               2   \n",
       "CCAGTTCACAATGGAATGCTTGGG_3  300genes:eTN5_GEX_sub3               5   \n",
       "ATCATTCCAGTCACTACGCCCGGA_2  300genes:eTN5_GEX_sub2               1   \n",
       "GCGAGTAAGCTAACGACACAATTG_3  300genes:eTN5_GEX_sub3              10   \n",
       "...                                            ...             ...   \n",
       "CCGTGAGAACGCTCGACTTTGGTC_3  300genes:eTN5_GEX_sub3               5   \n",
       "CACCTTACACAGCAGACACAATTG_1  300genes:eTN5_GEX_sub1              17   \n",
       "CAAGACTAGCCAAGACTAAATATC_3  300genes:eTN5_GEX_sub3               2   \n",
       "CAATGGAAGCTCGGTAGTGCTAGC_2  300genes:eTN5_GEX_sub2               6   \n",
       "CCAGTTCACTGAGCCATAAATATC_2  300genes:eTN5_GEX_sub2               9   \n",
       "\n",
       "                            grna_n_umis  response_n_nonzero  response_n_umis  \\\n",
       "barcode                                                                        \n",
       "TAGGATGAATTGAGGATGCTTGGG_2         98.0                1276      2078.232178   \n",
       "TAGGATGAAGCCATGCCGCCCGGA_2        142.0                 561       849.433411   \n",
       "CCAGTTCACAATGGAATGCTTGGG_3        245.0                1141      1890.696289   \n",
       "ATCATTCCAGTCACTACGCCCGGA_2         18.0                 407       549.833374   \n",
       "GCGAGTAAGCTAACGACACAATTG_3        198.0                 680      1287.249268   \n",
       "...                                 ...                 ...              ...   \n",
       "CCGTGAGAACGCTCGACTTTGGTC_3        165.0                1783      3590.150391   \n",
       "CACCTTACACAGCAGACACAATTG_1       1121.0                1462      2993.500000   \n",
       "CAAGACTAGCCAAGACTAAATATC_3        134.0                1345      2546.260254   \n",
       "CAATGGAAGCTCGGTAGTGCTAGC_2        105.0                 714      1144.333252   \n",
       "CCAGTTCACTGAGCCATAAATATC_2        139.0                 879      1110.250244   \n",
       "\n",
       "                            response_p_mito  \n",
       "barcode                                      \n",
       "TAGGATGAATTGAGGATGCTTGGG_2        13.304576  \n",
       "TAGGATGAAGCCATGCCGCCCGGA_2        17.541105  \n",
       "CCAGTTCACAATGGAATGCTTGGG_3        15.073812  \n",
       "ATCATTCCAGTCACTACGCCCGGA_2        16.914215  \n",
       "GCGAGTAAGCTAACGACACAATTG_3        22.101393  \n",
       "...                                     ...  \n",
       "CCGTGAGAACGCTCGACTTTGGTC_3        14.275167  \n",
       "CACCTTACACAGCAGACACAATTG_1        23.300484  \n",
       "CAAGACTAGCCAAGACTAAATATC_3         9.209585  \n",
       "CAATGGAAGCTCGGTAGTGCTAGC_2        19.181475  \n",
       "CCAGTTCACTGAGCCATAAATATC_2        14.636339  \n",
       "\n",
       "[21980 rows x 6 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. EXTRACT COVARIATES \n",
    "rna = mdata['GEX']\n",
    "covariates_df = pd.DataFrame({\n",
    "    \"batch\": rna.obs['sample_id'],\n",
    "    \"grna_n_nonzero\": rna.obs['guides_per_cell'],\n",
    "    \"grna_n_umis\": rna.obs['guide_umi_counts'],\n",
    "    \"response_n_nonzero\": rna.obs['n_genes_by_counts'],\n",
    "    \"response_n_umis\": rna.obs['total_counts'],\n",
    "    \"response_p_mito\": rna.obs['pct_counts_mt'],\n",
    "})\n",
    "covariates_df.to_csv(f\"{temp_dir}/covariates.csv\")\n",
    "covariates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ec11d9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_meta_df = pd.read_csv(f\"{working_dir}/300genes_guide_metadata_v43.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "716ef3c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ All guides mapped successfully.\n",
      "\n",
      "First 5 rows of mapping (Must match mdata['guide'].var_names order):\n",
      "                    grna_id grna_target\n",
      "0  ACAA1_+_38178488.23-P1P2       ACAA1\n",
      "1  ACAA1_+_38178517.23-P1P2       ACAA1\n",
      "2  ACAA1_+_38178559.23-P1P2       ACAA1\n",
      "3  ACAA1_+_38178570.23-P1P2       ACAA1\n",
      "4  ACAA1_+_38178620.23-P1P2       ACAA1\n"
     ]
    }
   ],
   "source": [
    "# 4. PREPARE GUIDE MAPPING \n",
    "# Sceptre needs a table mapping 'grna_id' to 'grna_target'.\n",
    "\n",
    "# 1. Create a \"skeleton\" DataFrame from the actual matrix order\n",
    "# This ensures the order matches mdata['guide'].X perfectly\n",
    "# We name the index 'grna_id' to match Sceptre's expectation immediately\n",
    "guide_mapping = pd.DataFrame(index=mdata['guide'].var_names)\n",
    "guide_mapping.index.name = 'grna_id'\n",
    "\n",
    "# 2. Prepare metadata for merging\n",
    "# Ensure 'guide_id' is the index so we can map efficiently\n",
    "lookup_df = guide_meta_df.set_index('guide_id')\n",
    "\n",
    "# 3. Map the targets onto the skeleton\n",
    "# We use .loc[] to pull the specific target for each guide in the exact order required\n",
    "guide_mapping['grna_target'] = lookup_df.loc[guide_mapping.index, 'intended_target_name']\n",
    "\n",
    "# 4. Reset index so 'grna_id' becomes a column (Sceptre requires this)\n",
    "guide_mapping = guide_mapping.reset_index()\n",
    "\n",
    "# ==========================================\n",
    "# VALIDATION CHECKS \n",
    "# ==========================================\n",
    "\n",
    "# Check 1: Are there any NAs? (Guides in mdata that were missing from meta_df)\n",
    "na_count = guide_mapping['grna_target'].isna().sum()\n",
    "if na_count > 0:\n",
    "    print(f\"⚠️ WARNING: {na_count} guides have NaN targets! These will crash Sceptre.\")\n",
    "else:\n",
    "    print(\"✅ All guides mapped successfully.\")\n",
    "\n",
    "# Check 2: Verify the first few rows manually\n",
    "print(\"\\nFirst 5 rows of mapping (Must match mdata['guide'].var_names order):\")\n",
    "print(guide_mapping.head())\n",
    "\n",
    "# 5. Save for Sceptre\n",
    "guide_mapping.to_csv(f\"{temp_dir}/guide_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "fb2a022a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>grna_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1_+_38178488.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1_+_38178517.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1_+_38178559.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1_+_38178570.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAA1_+_38178620.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>SUZ12_SGTN_23</td>\n",
       "      <td>SUZ12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1922</th>\n",
       "      <td>MESDC1_SGTN_11</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1923</th>\n",
       "      <td>MESDC1_SGTN_12</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>TP53_SGTN_01</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1925</th>\n",
       "      <td>TP53_SGTN_02</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1926 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       grna_id grna_target\n",
       "0     ACAA1_+_38178488.23-P1P2       ACAA1\n",
       "1     ACAA1_+_38178517.23-P1P2       ACAA1\n",
       "2     ACAA1_+_38178559.23-P1P2       ACAA1\n",
       "3     ACAA1_+_38178570.23-P1P2       ACAA1\n",
       "4     ACAA1_+_38178620.23-P1P2       ACAA1\n",
       "...                        ...         ...\n",
       "1921             SUZ12_SGTN_23       SUZ12\n",
       "1922            MESDC1_SGTN_11      TLNRD1\n",
       "1923            MESDC1_SGTN_12      TLNRD1\n",
       "1924              TP53_SGTN_01        TP53\n",
       "1925              TP53_SGTN_02        TP53\n",
       "\n",
       "[1926 rows x 2 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "e67e1ac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Gene mapping saved.\n"
     ]
    }
   ],
   "source": [
    "# Create a mapping from the Gene Symbol (which matches your guides)\n",
    "# to the Gene Index (which matches the matrix rownames)\n",
    "\n",
    "# Check which column in mdata['GEX'].var holds the symbol\n",
    "symbol_col = 'symbol' \n",
    "\n",
    "gene_map = mdata['GEX'].var[[symbol_col]].reset_index()\n",
    "gene_map.columns = ['response_id', 'grna_target'] # Rename for easy merging in R\n",
    "\n",
    "# Save it\n",
    "gene_map.to_csv(f\"{temp_dir}/gene_symbol_map.csv\", index=False)\n",
    "print(\"✅ Gene mapping saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2372ddb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>WASH9P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12727</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12728</th>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12729</th>\n",
       "      <td>ENSG00000271254.7</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12730</th>\n",
       "      <td>ENSG00000276345.1</td>\n",
       "      <td>ENSG00000276345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12731</th>\n",
       "      <td>ENSG00000276256.1</td>\n",
       "      <td>ENSG00000276256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12732 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              response_id      grna_target\n",
       "0       ENSG00000227232.5           WASH7P\n",
       "1       ENSG00000238009.6  ENSG00000238009\n",
       "2       ENSG00000241860.7  ENSG00000241860\n",
       "3       ENSG00000279457.4           WASH9P\n",
       "4      ENSG00000228463.10  ENSG00000228463\n",
       "...                   ...              ...\n",
       "12727   ENSG00000198727.2           MT-CYB\n",
       "12728   ENSG00000210196.2            MT-TP\n",
       "12729   ENSG00000271254.7  ENSG00000271254\n",
       "12730   ENSG00000276345.1  ENSG00000276345\n",
       "12731   ENSG00000276256.1  ENSG00000276256\n",
       "\n",
       "[12732 rows x 2 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cdd51f3",
   "metadata": {},
   "source": [
    "#### Import data to SCEPTRE (Rcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "db796010",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "db796010",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading matrices into R...\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(sceptre)\n",
    "library(Matrix)\n",
    "\n",
    "# Define path\n",
    "data_dir <- \"/oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer\"\n",
    "\n",
    "# 1. LOAD MATRICES\n",
    "cat(\"Loading matrices into R...\\n\")\n",
    "response_mat <- readMM(file.path(data_dir, \"response_matrix.mtx\"))\n",
    "grna_mat <- readMM(file.path(data_dir, \"grna_matrix.mtx\"))\n",
    "\n",
    "# 2. ATTACH NAMES\n",
    "gene_names <- read.csv(file.path(data_dir, \"gene_names.csv\"), header=FALSE)$V1\n",
    "guide_names <- read.csv(file.path(data_dir, \"guide_names.csv\"), header=FALSE)$V1\n",
    "cell_names <- read.csv(file.path(data_dir, \"cell_names.csv\"), header=FALSE)$V1\n",
    "\n",
    "rownames(response_mat) <- gene_names\n",
    "colnames(response_mat) <- cell_names\n",
    "\n",
    "rownames(grna_mat) <- guide_names\n",
    "colnames(grna_mat) <- cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6c57ae30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renaming 'response_p_mito' to 'mt_percent'...\n",
      "[1] \"\\u2705 Sceptre object created!\"\n",
      "                           response_n_nonzero response_n_umis grna_n_nonzero\n",
      "TAGGATGAATTGAGGATGCTTGGG_2               1101       2078.2321              6\n",
      "TAGGATGAAGCCATGCCGCCCGGA_2                494        849.4334             28\n",
      "CCAGTTCACAATGGAATGCTTGGG_3                969       1890.6964             24\n",
      "ATCATTCCAGTCACTACGCCCGGA_2                355        549.8333              9\n",
      "GCGAGTAAGCTAACGACACAATTG_3                613       1287.2493             43\n",
      "TCTTCACACAATGGAACTTTGGTC_1                694       1432.8333             59\n",
      "                           grna_n_umis                  batch mt_percent\n",
      "TAGGATGAATTGAGGATGCTTGGG_2          98 300genes:eTN5_GEX_sub2   13.30458\n",
      "TAGGATGAAGCCATGCCGCCCGGA_2         142 300genes:eTN5_GEX_sub2   17.54111\n",
      "CCAGTTCACAATGGAATGCTTGGG_3         245 300genes:eTN5_GEX_sub3   15.07381\n",
      "ATCATTCCAGTCACTACGCCCGGA_2          18 300genes:eTN5_GEX_sub2   16.91421\n",
      "GCGAGTAAGCTAACGACACAATTG_3         198 300genes:eTN5_GEX_sub3   22.10139\n",
      "TCTTCACACAATGGAACTTTGGTC_1         417 300genes:eTN5_GEX_sub1   22.12400\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# 3. LOAD METADATA\n",
    "covariates <- read.csv(file.path(data_dir, \"covariates.csv\"), row.names=1)\n",
    "guide_mapping <- read.csv(file.path(data_dir, \"guide_mapping.csv\"))\n",
    "\n",
    "# ==============================================================================\n",
    "# FIX: RENAME 'response_p_mito' SO WE KEEP IT\n",
    "# Sceptre blocks the name 'response_p_mito' but doesn't calculate it.\n",
    "# We rename it to 'mt_percent' so we can use it as a covariate later.\n",
    "# ==============================================================================\n",
    "\n",
    "# 1. Rename response_p_mito if it exists\n",
    "if (\"response_p_mito\" %in% colnames(covariates)) {\n",
    "    cat(\"Renaming 'response_p_mito' to 'mt_percent'...\\n\")\n",
    "    names(covariates)[names(covariates) == \"response_p_mito\"] <- \"mt_percent\"\n",
    "}\n",
    "\n",
    "# 2. DROP OTHER RESERVED COLUMNS (which Sceptre DOES calculate)\n",
    "# We still drop n_umis and n_nonzero because Sceptre re-calculates those automatically.\n",
    "reserved_cols <- c(\"response_n_nonzero\", \"response_n_umis\", \"grna_n_nonzero\", \"grna_n_umis\")\n",
    "covariates <- covariates[, !(colnames(covariates) %in% reserved_cols), drop=FALSE]\n",
    "\n",
    "# 4. CREATE SCEPTRE OBJECT\n",
    "sceptre_object <- import_data(\n",
    "  response_matrix = response_mat,\n",
    "  grna_matrix = grna_mat,\n",
    "  grna_target_data_frame = guide_mapping,\n",
    "  extra_covariates = covariates,\n",
    "  moi = \"high\"\n",
    ")\n",
    "\n",
    "print(\"✅ Sceptre object created!\")\n",
    "print(head(sceptre_object@covariate_data_frame))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5ab55bf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Unique pairs to test: 279\"\n",
      "   grna_target        response_id\n",
      "1        ACAA1 ENSG00000060971.19\n",
      "7       ACVR1B ENSG00000135503.13\n",
      "13      ACVR2A ENSG00000121989.15\n",
      "19      ACVR2B ENSG00000114739.14\n",
      "25       ADCY2 ENSG00000078295.17\n",
      "31      AMOTL2 ENSG00000114019.15\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 1. Load the translation map\n",
    "gene_map <- read.csv(file.path(data_dir, \"gene_symbol_map.csv\"))\n",
    "\n",
    "# 2. Start with guide mapping (which has 'grna_target' as Symbol)\n",
    "discovery_pairs_highmoi <- guide_mapping\n",
    "\n",
    "# 3. Merge to attach the correct response_id (Gene ID)\n",
    "#    This matches 'grna_target' in guides to 'grna_target' in the gene map\n",
    "discovery_pairs_highmoi <- merge(discovery_pairs_highmoi, gene_map, by=\"grna_target\")\n",
    "\n",
    "# 4. Select ONLY the two columns Sceptre needs\n",
    "#    (We filter out non-targeting guides automatically via the merge)\n",
    "discovery_pairs_highmoi <- discovery_pairs_highmoi[, c(\"grna_target\", \"response_id\")]\n",
    "\n",
    "# 5. DEDUPLICATE\n",
    "#    Sceptre performs analysis at the *Target* level, not the *Guide* level.\n",
    "#    We only need one row per Target-Gene pair.\n",
    "discovery_pairs_highmoi <- unique(discovery_pairs_highmoi)\n",
    "\n",
    "print(paste(\"Unique pairs to test:\", nrow(discovery_pairs_highmoi)))\n",
    "print(head(discovery_pairs_highmoi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "82ce6bf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "sceptre_object_highmoi <- set_analysis_parameters(\n",
    "  sceptre_object = sceptre_object,\n",
    "  discovery_pairs = discovery_pairs_highmoi,\n",
    "  side = \"both\", control_group = \"complement\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "1f19e55f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/grna_count_distributions.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "png(file.path(data_dir, \"grna_count_distributions.png\"), width = 3000, height = 2000, res = 300)\n",
    "\n",
    "print(plot_grna_count_distributions(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  n_grnas_to_plot = 8\n",
    "))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"grna_count_distributions.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "d9608e07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3deVwTd/4G8E8uEkggEA4RlEMBlUMRUZd61fssWo9qFW1ra2u3rdptXa1WhdpDXbXVam3V3V7+XFulWvGo91GPingjKir3KUc4EiCQZH5/jMu6EC0qMF/geb/8I/NNMvMEJo/DzGQi4jiOAACAPWKhAwAAgGUoaAAARqGgAQAYhYIGAGCUJDIyUugMjWHFihWHDh06depUSkqKj4+PlZWV0IkeasWKFZmZmQEBAfzkypUr/fz8bGxshE0FTUVsbOymTZuOHz9+6dIlqVTq7u4udKKH4qP6+/srlUoiiouLO3PmjL+/f/UDDh48ePfu3fbt2/OTx48fj4+P9/Pz45/o5ubm6OjI37Vx48YjR4706tUrNjb29OnT1W+fpq6lbEGvWLFCoVC0adNm9+7dPXr0qKysFDrRQ61YsWL69OkZGRn85MqVK/Pz84WNBE1IbGzsvn37fHx8OI4bMWLEli1bhE70ULGxsUuXLl26dCk/GRcXt2PHjgcfcPDgwYMHD1ZPHj9+fM+ePfwTV69evXbtWn48Jydn/vz5K1as4O/6+eefG+kFNDyp0AEaz6hRowIDA6dMmeLg4HD9+vWuXbsS0b59+65fv96+ffsxY8aIxWIiunr16uHDh00mU0hIyMCBA4koMjJy+vTpMTExKpVq0qRJcrmciIxG4y+//JKWltajR4++ffvyi4iMjHzzzTd37twpkUgiIiKsra0tztDich/0/PPPR0VFbdq0yeILsbiUOua5ePHi7t27q2c1efJkPz8/i3kiIyPfeOONmJgYLy+vAQMG1J45MMvDwyMiIoKIDAbD9u3b+dsZGRl79+41GAwjR47kt0krKiqio6NTU1OdnZ1Hjx7t4uJy/PhxIiotLb19+/aAAQOCg4P5Gd68efPgwYNKpfL555/XaDRExD/SaDReuXKlb9++3bt3tzhDi8t90MiRI3/66ae//e1v3t7ej/Uahw0btnfv3tWrV1tZWf34449jx47dtWvXgw+wGKbJaSlb0NVMJhPHcfxKNnny5EOHDrVt2/bIkSPjx48nouvXr48bN87JyalNmzZnz57lnxIVFbVw4cLQ0NCcnJwRI0YQEcdxw4cP379/v4uLy9y5c5ctW1b9yA8//DAgICAtLW3y5MkPm2Ht5dYwd+7cPXv23Lp1y+JLqL2UuudxdXXt3bt37969vby8Vq9ebTKZHpYnKipq3rx53t7ebdq0sThzYF9lZaWTkxMRxcXFjR8/XiKRKJXK8PDw2NhYInrllVfOnj3bsWPHsrIyfmU7fvz41KlTCwsLu3fv/uqrrx45coSIjhw5MmLECIVCkZ6eHhISUlBQwD9yxowZaWlpXbt2jYiIuHTpksUZWlzugzQazZw5cxYvXvy4L83GxmbgwIH81saWLVumTp1a4wG1wzRFLWgLesuWLc7OzgcOHPj88889PT3j4uJiY2M//PDDioqK7t27v//++1lZWenp6S4uLkOHDm3VqtWDz128eLGvr2/Pnj137Nhx7tw5o9GYmZl56NAhInr22Wf9/f3//ve/8xueUVFRbm5uYWFh9vb2HMfVnqHF5bq5uT24ODs7u7lz5y5atOhhf6zVWMqZM2fqmMfNzc3NzU2n0/Xp02fTpk2dOnV6RJ7IyMh27dqdPn36YTMHNt25c+eLL764d+/euXPn+F0ckZGRvXv3lkqlRBQWFrZ58+YePXokJyeHh4ePHj1aIpFUP7dv374vvfQSEX300UcrV64cOHDgsmXLPvvss4kTJxJRWlra5s2b582bR0T9+/efPn06Eb344osHDhzo2rVr7RlaXG6NtHPmzPHz84uPj3/cl/nKK6989tlnnp6erVu3rr2r3eKra3Ja1tvMbDYXFhby/+vevn3b1ta2+q6VK1cqFIqhQ4eGh4ePGDGic+fO69evr77Xw8ODv+Hp6ZmampqSktKhQ4cH76reTcxXG79GGgyG2jO0uNzaUd96663Y2NgLFy5YfCE1llL3PERkMpkmTZo0fvx4/i33iDz8X52PmDmwrLy8/Pz583fv3qX//S337t170qRJRLRp06ZffvnF19d33Lhxqamp/L01VnUiSklJ6dixIz/o7+9f/cjqrQqVSlVUVGRxhhaXW4ONjc3ChQsXLFhQ+y65XM6vtDyDwfDgm6VXr17JycnLly9/+eWXaz/X4qtrclrQFnRERERgYODrr7/eqVOnX3/9tV27dvfu3Zs0aVKNfpw3b968efPu3bsXEBAwatQoT09PIrpx4wa/Py4hIcHT09NoNN68eZPjOJFIlJaWRkT8H5K1iUSiGjN82HJrkMvlS5Ys+eCDD+ry0ry8vOqYh4jmzJnj6Oi4cOFCfvIReUQi0ePOHFjg4+MzZ84cInJzc5s1a9aFCxfatWun0WhqFFlQUND27ds5jouMjIyKivrXv/5FRDdu3ODv5Vd1IvLy8rpx40aXLl34wer/rWurPUOLy63ttddeW716tb29fY1xPz8/PhXv6tWrI0eOfPAB06ZNW7169ZYtW6oPqj/61TU5LaigeWq1+qOPPlqyZMmlS5f69u07YsSIqVOn6nS6Q4cO7d69+9dff42LiwsICMjMzFSpVNUHFpYuXTp69OgTJ060bt26Z8+eRNSmTZuXXnqpf//+GzZs+PDDDx/2J3/tGXp6etZersXnTps2beXKlXl5eX/6onr16lXHPD/++OMvv/yyadOmw4cPE1FwcHDPnj0fnafuMwfWzJ49++uvv965c+dHH300ZsyYwsJCb2/vK1euBAYGvvLKKzNmzAgNDVWr1SdPnhwzZgz/lISEhIULF3p7e3/66aebN28movnz57/22mtarTY7O/vo0aP/+Mc/Hra42jO0uNzaT5TJZB999FFERMSECRMeHJ84ceKqVaumT58eFhZ28eLFO3fuTJs27cEH8Fs/dQzTFLWU86CJqFevXvzpll26dCkqKvL19X355Zfd3NySk5NlMtlLL73k6urq4uJSUlKSkpKiVqtXrVrF/5ceFRV19OjRa9eutW/fftmyZfzugsmTJ5tMppycnJdffpk/Ss579tlnq2/36dPH1dW19gzHjx9fY7m1oyoUCrFYHBIS4u7u/uyzz9Y4D7rGUsRicR3zFBQUODs7Z2ZmpqWlpaWl+fj4ODo6PixP9XMfNnNgk7u7O38isEQiCQkJ0Wq1Q4cOnTJlSm5u7r1794KCgsLDw+VyuYuLS2pqamFh4cSJE/ljyMePHw8KCurbt296evr8+fPDwsKIqF27dsOGDUtMTHR2dv7iiy+qTz328vLy8vLib7dt27Zdu3a1Z+ju7l57uTWi8vtPAgMDFQpF7969HzwPWiqVTp8+3Wg05uTkBAUFrVmzhn8LP/jEatbW1r169ap++bXDNEUiXM3uT4lE+ClBi8BvrrWcjTb2oXoAABiFnYkAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxqqA+qaLXaxYsXf/nll4/xlAuUd4xUvuQ2+vGWtWjRqaIivVQqNps5iUS2enX/x3s+1EHa/1FFNrmOJLtOQkcRQt3X58oCSvmWRFLynfMnj0xIKFi9+g9bWysiKiszjhrl99xzNS/2Bg0h5VuqLCD3saRsJ3SUP9NQW9B6vf5xP1upu0tZMVR4/rGX9dNPt4YNc5s7N+CFFzx/+CHhsZ8PdZB/irJiqDxT6BwCqfv6XFVKWTGUvffPH5meXnrhwr25cwPmzg0wm40XLuQ+bUqom3vHKCuGKprCz/v+FnR8fPyhQ4eMRuPcuXOr7zt9+vTJkyetrKxmzpypVCpNJtPWrVuzsrIGDx4cEhJCRLVHAACgvtzfgrayshoyZMjt27er7+Avpfbuu+9Onz5dJpMR0Y4dO+zs7GbPnr1lyxatVmtxBAAA6sv9gvbz86vxLV779u3r3Lnztm3bUlNT+W/wO3fu3KBBgxQKRVhYGH8ZzNojAABQXx56kDAnJ+fGjRvh4eHr1q2ztrbu0KFDaWkpf6USBwcH/lsVao8Q0dWrV7/++muDwfDgVYYBAOBxPfQgoZWV1fDhw93d3QcPHsx/n41UKuUv3FFVVcVfO7j2CBG5urqOHTt2yJAhD15pGwAAHtdDC9rb27u0tJSISkpK+Gtdenh4pKSkEFFSUhJ/mcHaI0Tk4uIyaNCgXr16sfzN2QAA7Lu/i+Pu3btxcXF6vf6nn37q168fvxW8YcOGoKCgS5cuLVmyhIjGjRu3YcOGTp065eXlde7c2eIIADyBa9cK9+7N2LPnLj/53XfDAwPxtTXwn4J2cHDw9/fnL5XN71b29PRcsmRJZmbmhAkT+IOEfn5+ixcvzs/P9/Dw4L8MqfYIADwBvb6qVy+X8eO9iWjOnLN6fZXQiYAJ9wtao9FoNJoa91lbW/v4+Dw4olKpVCrVo0cA4Al4eKj69HElImvrFvdFdPAwuBYHAACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKFyWBaBRSaWrTCaOvx0c7ChsGGActqABGtupU88lJIyfPTtQ6CDAuma1Ba3TVVVUmDZuvMpP+vra9+/vIWwkgNrs7GRqtZVCIRE6CLCuWRV0UVGlwVC1ZctVIsrKKnvmmTYoaABouppVQRORra3Vtm0DiGjz5lt37pQLHQcA4MlhHzQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMCopnrB/u++i9+3L5m/nZ2tFzYMAEBDaKpb0Jcv38vOLvL1tfb1tS4vrxI6DgBA/WuqBU1EPXq4vPWW/1tv+YvFIqGzAADUvyZc0AAAzRsKGgCAUShoAABGoaABABiFggYAYFRTPQ8aoLkyGrnMTF1SUjERSSQiT087oROBYFDQAGzJySl76aV9EonIbOasrKT5+W8JnQgEg10cAMzZtm1AQsL4/fuHCR0EBIaCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRzfY86GvXCo8dyw4N/ZGf/OSTPkOHegmaCADg8TTbgtbrjf7+9m+95U9En312pbCwQuhEAACPpznv4mjVyrpPH9c+fVwdHKyEzgIA8Niac0EDADRpKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEY12+tBQ0sWHx9/6NAho9E4d+5cfmT//v2///47ETk6Or733ntEZDKZtm7dmpWVNXjw4JCQEIsjAMJCQUMzZGVlNWTIkDVr1lSPZGdnv/jii0FBQdUjO3bssLOzmzBhwoIFC7y9vR0cHGqPCJEd4L+wiwOaIT8/v4CAgBqDe/bs+e6773JycvjJc+fODRo0SKFQhIWFXbhwweIIgLCwBQ0tQu/eveVyeW5ubmRk5OrVq21sbEpLS5VKJRE5ODgUFBQQUe0RIkpISPjhhx/Ky8v5uxpTVZXZaDRfuJDLTzo6Kry81I2cAYTVIgo6I6Ns8+arx4+n8ZNffDHA2rpFvHCo5ufnR0Senp6xsbGJiYnBwcFSqZTjOJFIVFVVpVAoiKj2CBHZ29v36NFDq9X++OOPjZw5N7dcp6scN24nEZWUVI0a1f6HH0Y0cgYQVovYxVFYWGFnJ3JyEms0oo0br1ZWmoROBIIpLCy0sbEhIg8Pj5SUFCJKSkry8vKyOEJEbm5uY8eOHTp0qMFgaPy0dnZWZ86EnzkTPmdOYOMvHQTXUjYkBw9uM2aMp8nELVt2Regs0ODu3r0bFxen1+t/+umnfv36ubq6rlq1ytHRMS8vTyqV+vr6EtG4ceM2bNjQqVOnvLy8zp07WxwBEFZLKWhoURwcHPz9/f39/YmI33c8c+bMjIwMlUrl7u7OP8bPz2/x4sX5+fkeHh4ikcjiCICwUNDQDGk0Go1G8+CIUqns0KFDjYepVCqVSvXoEQABtYh90AAATREKGgCAUShoAABGoaABABiFggYAYFTLOovDbOaIaM6co1ZWEiJq397+73/vIXQoAADLWtYWNMcRETk6ip2cxFpt6e7dd4VOBADwUC1rC5o3Y0ZHW1vZoUOZmzbdEToLAMBDtawtaACAJqQpbUEnJmr/9a9r/O3TpzP79nURNg8AQINqSlvQycnF3357VafT6XS6lJRioeMAADSsprQFTURubsoFC4KJ6MSJLKGzAAA0rKa0BQ0A0KKgoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGNXETrMDaIrWr7909Woef5u/YhdAXaCgARrcgQMpRFVBQRr6zxW7AOoCBQ3QGAYOdJ8wwZuIPv30stBZoMnAPmgAAEahoAEAGIWCBgBg1P190IcPH46Ojq6qqtq8eTM/snHjxqSkJJPJFBYWNnbsWCIqLCxct24dx3G+vr6TJ0+2OAIAAPXl/hZ09+7d161bJxb/d4M6PDx82bJly5cvP3bsmFarJaJt27aNGDFiyZIld+7cSU1NtTgCAAD15X4jq9VqiUTy4B2urq5EJBaLRSIRX9wJCQldunQhouDg4Pj4eIsjRFReXp6ZmZmbm/tg3QMAwOP6k9Ps9u/f37FjR7VaTUQGg0EmkxGRUqnMz8+3OEJEFy9eXLRokdlstrOza9jsAADN2qMKOi4u7syZM4sXL+Yn5XJ5VVWVTCbT6/UqlcriCBH16tXr6NGjGRkZHTp0aOj0AADN2EP3QiQmJv78888ffPABv41MRAEBAVeuXCGiy5cvBwYGWhwBAID6cr+gz58/v2jRosLCwkWLFt25c4eI1q5dazAYPvvss0WLFqWnpxPRxIkT9+zZs3jxYh8fH09PT4sjAABQX+7v4ujevXv37t0fvGPdunU1HqrRaCIjIx89AgAA9QUnWgAAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACj8JVXAE1AQUFFamrJ9u23+MmePVt7eOBaN80fChqgCbh9u+TmzfwvvzxPRPHx2vXrB6OgWwIUNEDT0Ldv6zVrwohoypRjQmeBRoJ90AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjGoCV7PbtOkqxxERXb9eIHQWAIDG0wQK+s03D3Xv7iyTidPT9fb2VkLHAQBoJE1jF8fGjb23bRswcWI7oYMAADSeplHQAAAtEAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRTeA86AaSkqK7davwhRdi+MlJkzqOHesrbCQAgAe13C3ooiKDUikJDVWHhqozMopu3MDHFAGALS13C5qIXFysIyJ8iOj6da3QWQAAamq5W9AAAIxDQQMAMAoFDQDAKBQ0AACjWvRBQoCmqLi48ty5LKlUxE+OGeMrk2FLq3lCQQM0MampugMHki5ezOI47tSp3KKid9RqudChoEGgoAGannfeCRgzxtNk4jw8/i10FmhA+MsIAIBRKGgAAEZhFwcR0Y0bRadP5x0/ns5Pbtgw2MfHXthIAAAoaCKikpLKwED7ESPaEtGCBXGlpZVCJwIAwC6O/+jY0X7UKI9RozysrSVCZwEAIEJBAwAwCwUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjJIKHQCgeTIYTGVlVfztqiqzsGGgiUJBAzSIf/3r2l//erh6ctgwVwHDQBOFXRwADSU83DMzc3Jm5mRnZ4XQWaBJQkEDADAKBQ0AwCjsgwZoqkwmjojs7b/kJ3v1cj916kVBE0E9Q0EDNG3nzo22tZUdO5a9dWuK0FmgnmEXBzRPOTk5t27denBEp9OlpKSYzebHGmGfWm2lVlspldjYaobwS4Vm6Jdffrl48WJubu6mTZv4kcTExH/+85/+/v7JyclLliwRiUR1GRH2VQBgC7qm4uLKjz8++8YbB9944+Df/35C6DjwJMaOHfvxxx8/2LDR0dEzZ8586aWXnJ2dr169WscRAGGhoGsqKzM6OIidnMQcZ9iyJUHoOFA/0tLSvLy8iKhdu3YpKSl1HCEirVYbGxt75coVmUwmQG5o2bCLw4IXXmgXGOhw61bRnj0ZQmeB+mE0GvkNaplMVlJSUscRIkpKSvryyy8rKyutra0FSw8tFQoaWgRbW1u9Xq9UKrVaraOjYx1HiKhbt24//vhjRkZGhw4dhHwB0CJhFwe0CD179jx06FB5efnZs2e7detWxxEAYaGgoRk6f/78okWLnJ2dFy1adOfOHSIaP358aWnpmjVrIiIiHBwc6jgCICzs4oBmqHv37t27d39wRCKRTJ069XFHAISFLWgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEY9tKBNJtPJkyd37tyZmppaPXj58uWYmJj8/PxHjAAAQL14aEFv2bLlzp07vr6+n3/+eW5uLhGdPHny6NGj7u7un3zyicFgsDjSnBgMpqoq84ULufy/1NQSoRMBQMvy0IJOT08fOnRoYGBgUFAQvxF94MCBadOmhYSEdO3aNS4uzuJIc5KdXabVVowbt3PcuJ2DBv20dOlZoRMBQMvy0Av2T5o0adeuXT4+PtnZ2VOmTCGivLw8JycnImrVqlVOTo7FESJKSkrau3dvaWmpjY1NY7yChuToKD9zJpyINmy4kZ1tFDoOALQsD92CPnz4sMlkEolElZWVxcXFjZkJAADoEQV98eLFt956a8iQIT179jxz5gwROTs78wcDc3NzXV1dLY4QUbt27d55551p06aVlZU1xisAAGimHlrQnTp1io6Ovnbt2okTJ/z8/IhoyJAh33//fVxc3KVLl0JDQy2OAABAfXnoPui333771KlTt2/ffvXVVzt06EBE/fr1s7Ozy8jIWLhwoVwutzgCAAD15aEFLZPJ+vfvX2Owa9euXbt2ffQIAADUi4cWNAA0IVlZZampxfPnn+Qnhw716t/fQ9hI8PRQ0ADNwb175UajSafTEdGJE9m2tlYo6GYABQ3QTLRtq1qwIJiISkurhM4C9QMXSwIAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUYx+1Pvdd4/9/nsGf9ts5oQNAwAgCEYL+u7dop49Hfv2dSWiSZNyhY4DACAARguaiPz81H36uAqdAgBAMNgHDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACj2P2oN1OuXSs8ffre5cv3+MkVK/oNGOAhbCQAaPZQ0HWi0xmDghzeeKMjEX300UWttkLoRADQ/KGg68rV1Zq/eJO9vVzoLADQImAfNAAAo1DQAACMwi4OgOYmK6ssJydDKo3lJ19+ObBVKxthI8GTQUEDNDfp6Tpra2l8fBYR7dyZMmSIFwq6iUJBAzRDQ4e2mTUrgIiOH88WOgs8OeyDBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEbhPOjHlpVV9s03Vw4eTOEn160bJJPh/zkAqH9olsem1Rrs7MROTmJ7e9HGjVeNRrPQiQCgecIW9JMYOtR95EiPigrTihVXhM4CAM0WtqABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABG4WJJT66qykxEERH7JBIREQUEOC5Z8ozQoQD+R1GRYd68k/b2ciJycJB/880QoRPBY8AW9JMzmTgi6tbNLjRULZebTpzIEDoRQE3l5SZ/f2VoqLptW6uYmLtCx4HHg4J+WpMmtY+I8Onb11XoIACWjRjhERHhM2pUW6GDwGNDQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQPUG52uavv2RCIymbhTpzKFjgNNHgoaoN4UFJQvXXqGiEwm86FDyULHgSYPBQ1Qn+RyCRHJZOJnnmkldBZo8lDQAACMQkEDADAKBQ0AwCh8aSxAi5CRob93r6x9+0385IQJHZYt6ytsJPhTDBX0rVuFUtKcP5/z0QvnL1zIHTjQRehEAM2H0WhWq63Wrw8joujo5Pz8cqETwZ9jaBdHQUE5Ednby0JD1SUlBqHjADQ3Uqmoc2dN584aV1cbobNAnTBU0Ly2bZURET7W1hKhgwAACIy5ggYAAB4KGgCAUQwdJGzSbt4siovLCQ39kZ+cPbvb1Kn+wkYCgKYOBV0/ysqMbdvazJ0bQETffHMzJ0cvdCIAaPJQ0PXG3l7ep48rEf36a6rQWQCgOcA+aAAARmELGqDFSUoquXhR+8YbB/nJN97oEhKCa++xCFvQAC1Obm6FQiFychI7OYkPHEhKTi4WOhFYhi3o+ldcXHn7tvbw4ft7osPC3JRKmbCRYP/+/b///jsROTo6vvfee0RkMpm2bt2alZU1ePDgkJAQiyPNWKdO9m+95U9Ex49nC50FHgoFXf9u3iy+dKng3LkMIkpI0F658rK/v6PQoVq67OzsF198MSgoqHpkx44ddnZ2EyZMWLBggbe3t4ODQ+0RAQMDEHZxNJDXXuuwf/+w/fuHaTQKobPAfXv27Pnuu+9ycnL4yXPnzg0aNEihUISFhWsQEQ4AABmGSURBVF24cMHiCICwsAUNLULv3r3lcnlubm5kZOTq1attbGxKS0uVSiUROTg4FBQUEFHtESK6cOHCF198UVlZaWdnJ2B+aJlQ0NAi+Pn5EZGnp2dsbGxiYmJwcLBUKuU4TiQSVVVVKRQKIqo9QkTt2rV755138vLyDhw4IOQLgBYJuzigZSksLLSxsSEiDw+PlJQUIkpKSvLy8rI4QkQODg49evTo0qVLVVWVMImhBcMWNLQIq1atcnR0zMvLk0qlvr6+RDRu3LgNGzZ06tQpLy+vc+fOFkcAhIWChhZh5syZGRkZKpXK3d2dH/Hz81u8eHF+fr6Hh4dIJLI4AiAsFDS0CEqlskOHDjUGVSqVSqV69AiAgLAPGgCAUShoAABG/ckujmXLlplMpoULFxJRYWHhunXrOI7z9fWdPHmyxREAAKgvj9qCTkxMfPDUom3bto0YMWLJkiV37txJTU21OAIAAPXlUQW9Y8eOsWPHVk8mJCR06dKFiIKDg+Pj4y2OAABAfXloQV+8eLFdu3YPHtE2GAwymYyIlEqlTqezOEJEx48fDw0NHTNmjEajadjsAADNmuWC5jhu165do0ePfnBQLpfzezz0ej1f3LVHiKh3795Hjx7dsmWLVqtt2OxNAcdxJSUGrbZCq60oLjYIHQcAmhLLBa3X64uLi5cvX/7VV1/dvHlz9+7dRBQQEHDlyhUiunz5cmBgoMURIpJKpXZ2diqViuO4RnoRDCsoMISFbdVo1mk064KDfxA6DkBN6em6VaviXnghhv9XWWkSOhH8l+WzOFQq1Zo1a4goNTV1y5Yt4eHhRDRx4sS1a9fu3r27Y8eOnp6eFkegtmPHRvr5qa9dK5w586zQWQBqKimpat/eJihIbTSaFy6M+/774UIngv/6k9PsPD09+XPsiEij0URGRj54b+0RAGhy+vZ1HTnSo6LCtHBhnNBZ4H/ggyoAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKHyrdyPJySnPyyt74YUYfnLIEK/XXgsSNhIAMA5b0I1Ep6uyshKHhqpDQ9UlJfpLl3KFTgQArMMWdONRqWQRET5EVFBg+M/3zwCwwmAwEZGb29ciERFR797uu3c/L3CmFg8FDQBERPwXbOzcOVAulxw/nn3oEP7IEx4KGgD+y8NDpVBInJzkQgcBIuyDBgBgFgoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFK5mJ4DMzLKkJP3y5bH85Pjxfu3b2wsbCQAYhIIWQFqarqCgLD4+i4gOHMgICHBEQQNAbShoYYSFuXz6aXciunmzSOgsADVlZpbdvVs0f/5JIsrPLw8IcAwKcubvCgtzUyplgqZrQVDQAFBTbm6ZyWTS6XREFB19d98+ibOzgogSErRXrrzs7+8odMCWAgUtsMxM/fLlsT/8kMBP/vTTc/w3wgEIy9NTtWBBMBHt358xZUr7mTM7EVHXrjuFztWy4CwOgZWWVvn6KkND1QEByu3bb3H8F8MBAKCgWfDss60jInwmTmwvdBAAYAsKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUbgWBytKS6uISCJZxU8OG+a9f/84QRMB1FRebvznP6+5uNgQkVotnzmzi9CJmjkUNFvi48eJxaJ16xJ++SX1hRdi+MFXXgkcPtxb2GAARKTXGy9ezFSpZMXFlenp5SjohoZdHGxRq63Uaquysiq1WhYaqg4NVd+5k3/7tlboXAD3LVkS8s03vRcuDBY6SIuALWhGubnZRET4EFFsbJ7QWQBAGChoAHhsqamlGRml7dtv4icjIvyjonoJG6lZQkEDwGOrquI0Gqv168OI6N//vpufXy50ouYJ+6AB4EnI5dLOnTWdO2tcXW2EztJsoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYJTA50FXVpquXcvnb2u1hlbCpmHSrVtF8fGXY2Lu8pPffz/czU0lbCQAaBwCF3R2tr5Hjy1t2yqJqLfEv2ObdsLmYVBJSVVYmMuAAW5E9Le//VFWZhQ6EQA0EuF3cdjYSM+cCT9zJtzPTy10FkYFBDiMGuUxapSHTCb87wsAGg0+6g0ATyUxsfjataLqq+POn98jJAR7K+sHCropMZu59PRSsVhERDKZuG1bW6ETAVBBgcHFRR4aqiair75KyMrSoaDrCwq6KSktrRo9+hexWGQ0muVy6YwZnfnxXr3cn3uuvbDZoCXz87Pjr467fXuS0FmaFezTbGJ++214QsL4qKhQs9ms0+l0Ot3Ro8lHj6YJnQsA6h+2oJsqjUa+YEEwEa1ceZXjhE4DAA0AW9AAAIxCQQMAMAq7OACg/hUVGRwcvrR415o1A2bNCmnkPE0UChoA6k1KSum77x6LjDxjMnFEFB8/TiwWLV16KT1dt3FjHyKaNy9W6IxNCQq6WYmNzS4pqeRvi0SigQM9hM0DLU1FhWnsWM+QEMe8vIp33rmnVluJxSK5XCyTSdRqKyKyspIInbEpQUE3K2+9dSQtrcjaWmoycTk55VVVfxM6EbQ4AQH2ffq4ZmaWCR2kOUBBNzdffBHWr1/r/PyKbt12CZ0FAJ4KCrrJu3q18Nq1ot9/zyCiGzcKhY4DAPUGBd3k6fVVoaGO06b5EtFLL50QOg4A1BsUdHPg7q7s08eViCQSkdBZAKDe4IMqAACMwhY0ADSeS5fy9+/PiIw8Q0SVlaYFC/7i5GRNRGq11cSJHYVOxxwUNAA0HqORmzbNZ/RoTyIaMeK3I0fu2thItVpDXl4lCro2FDQANCo3N5vOnTX87aVLQ728VBcvFrz3Hj5haAH2QQMAMApb0M1TUVGl0WjWaNYRkcnEdeyoefXVIP6uPn3cO3VyFDQdANQJCrp5Mps5iUS0d+8QIvrss8uxsXkxMTeJ6OLF/EWLeqGgAZoEFHSzJRKJPD1VRKRSybp00XzzTW8imjnzlNC5AB7l1KnM06czqydfeKGDt7dawDzCQkEDgMD0emNZWdXhw6lE9P331+PiMkNDnYlo//6Mrl1dUNAAAILJyNDn5ZW/994RIrp7t2TyZJ/IyBAiunZNK3Q0gaGgW5acnPKDB1OKiw1ElJWlGzfOT6mUEZG1tdTfHzumQTDu7jb79w8jonHjDgmdhSEo6JYlM1MvFovi4zki2rLlTnT0LalUVFFhcnZWXrv2stDpAOB/oKBbnOef95o61YeI/u//7uzcOcjdXfnHH/eWLLksdC6AmrKzyz755I/Nm68RkUQi+ve/RwmdqLGhoOHPFRcb0tIM9mS3bdvN7MP3goKcp0zpJHQoaP50uio/P6Wvr7qoqHL58ivBwS5EdPu29vz5HBcXG/4xs2eHjBrVXtCYDQgFDXTrVvGNG4X8p1oqK02enuqAAEciysgoHTLEy81NpdVW2GW3tbe1q6ioiI3NuHu3CAX9oNdeO5CaWkJEFRVGobM0NwMGuPXr1zoxsXj58ivx8VlE9McfuUql1YQJbYlozZrrhw+nKRT3e6x3b/fq281Ds3ox8GRMJs7TU7llS38iev31352cZKGhaiLaufM2kdHRUVFcXDlZ1JaIRo/2FCWXnD9fJHBixvzxR/bw4W6+vnZ375bG3MwROk7zJBaLli/vQUTvv3+uoKBi1CgPIoqKurhjx81jx5KJKD5em5w8w8PDTuCg9QoFDUREVlYS/lMtcrnU398hIsKHiBYvvvDee53DwlyuXClI/fD+I+PjtUeOpIWG/shPLl78THh4s/0Ds+6eecalZ0+XuLi8mH+hoBvVO+8E8MdU/Px+FjpL/UNBw+PR641+fup33w0gopUrr+bn48ubgS3Z2fqYmLv87djY7JSUEo1GwU8uWRIWEOAkXLTHhoKGx+bkJOe/YWvu3HPz55/85JM/+PE//pji7GwjaDQAun1b+/77x/r1a01Ep0/ndumiCQ1tRUSff34tLy9Y6HSPBwUNT66y0hQR4TNkSBsieu65AyYTJ3QiaLkqKkxTpuxVKKRarcHdXclffCY8/FD1LrsNG25cuJBrNJpFpa0lZPXdd/F5vxYSEctnJaGg4am0aaPkL75uNHKtW2/gB0NCWl24MJWIvv76SnR0YvWD16wZgM8rQgMxm7mRI900GvmxY9mXL1fUfkB2dtlXX11QqWTzVc+3llpVVFTodLoTJ3I2brzy+edx/Bzef7/78OHe/OPVarlYLPC3MKOgod6cPDnKyUlx6lTu+vU3tdoKIrp2Lc/Kysx/v9GSJReKigxCZ4TmbOBAN3d35b17FZcvF1h8wIoVPcPCXG6+bUMF9MIL7ZRd3bKzzyqVUv6YysyZp6ZM2Vv94IgIfxsbKRF5eNgtXPiXxnkJNaCgod7Y2srUaqvU1NIrV/L4s6qJaPbsAP6MqH/846qg6QAsqz6mIpeLV63qOWlSeyJyd9+q0YhUKnFCQtFvv6XwB1eSkorEYrGX1/0z+UaP9mnVqmEPuqCgof4FBWl++20YEQ0f/pvQWQCe0Cuv+Lm4WH/1VcKxY1n89138/nuOi4t1hw5qIjp+PDsw0In1gtbpdPn5+R4eHmLxY3y94fbtt7RaAxEVFFjYVQTNUkaGftiwHVKpmIjUanly8gyhE1nwZOszNG8eHir+kOPw4b/179/673/vQkTe3tt69drKP0AqFT/7bFv+9oABHh980JOIRoyI3r8/uXomhYVvOzgoHnfRT1XQiYmJ//znP/39/ZOTk5csWSIS1XWHelTUWbVaotHI+ZqGlsBs5lat6hkY6JCVpX/tNRa/2OWJ12domTZu7BMY6HDpUsHcuef4j55///3tlSvP8wfG79zRRkZ2GzLEnYieeWb3ky3iqQo6Ojp65syZ3t7e69evv3r1apcuXer+3PffD+I/eTVlyvGnyQBNSKtW1p6eqoICg05XNXjwdiJKTy+5deu/F2VfurR39TH04GAXiaRRK/Jp1mdogfj1OSurTCoV8QdaYmJSO3VSv/NOABG9/vrvjo5yT08VxxER/frrHf7a646O1gMGeNRxESKOe/JzV998882vvvpKJBLt37+/srJy9OjRRJSVlfXHH39otdp333135cqVFp8YFXXG19fW0VFRUFBx+nRueLgnERVfkgVLvfRq/T2PrL1704KCNB4eKiL65ZfkUaM8rKwkt24VZWWV9+/fmoiOHs1s3VrZqZM9Ee3aldKvX2sHB3leXvm5c3n8T+qPP+5JpSL+i3P27k3r0kXTpo3KbKZdu5Kfe85TJhMnJBTl5ZXzZ7MfPpzZtq2yQwd7Itq5M2XAgNZqtTw7u+zSpfwRIzyI6OzZXLlcEhLiREQxMWkhIY7u7sqqKlNMTNqYMV5isejatcKiokr+UMPBgxne3ra+vvzlLFIGDXKztbVKT9ddv140bFgbIjp1KkelkgUHOxLR7t2pPXo4u7ralJUZf/stfexYLyLRlSsFOp2xV69WRHTgQIaPj1379nZEtHNn8pAhbZVKaUqK7vbtosGD2xDRyZPZ9vZWnTvfn9tf/uLi4mJdWlp1+HDm8897EdHFi/kGgyksrBUR/fZbeocO9t7etvwPdvjwttbW0rt3S5KTSwcNcieiEyeyHR3lgYEaItq1K7V3bxcnJ2ut1tD6TtvWCnWeR87x20kmE9ezpwsR7duX5u/v4OV1f24jR3rI5ZLExOKMDP2AAW5EdPRolqurtb+/A/9r6tu3tUYjz88vP3XqXvv2tkSUnV1mZ2fVpYuGiI4cyaqsNFWvJF26uPD7Q1q1shk92sfiitS1a9fu3bs/ybprSb2szzfO66e6/8Us4naUn+U46tHj/hqI9Zmd9bn1TQ9FlTzXK6vMVh8Xl/f06/PZs3nPPedBROfO3ROLRd273/81de6sadtWxXHczp0p1WtLu3b28+b1sLgi1V6fn6qgZ8yYsWnTJiI6fPhwQUHBxIkTiSghIeGHH37gOE6v1+fm5qanp7dp06bGHr3ycqPZzBERx3Hl5Sb+XJYqA2eu5KwUYpGMDAaTRCLi359lZUZra6lIRCYTV1VlVigkRFRZaRKJRDKZmJ+bXC4Ri0VmM2cwmKytpURUVWU2mzm5XEJEFRUmqVQslYr+dG7FxQVisUKptKk9N47jrKz+Z24cR+Xl9+dmNHJG42Nke3BuMplYIrk/N/5HYTRyJpOZD28wmMTi+3MrKzMqFP+dG5HBZKqSy+3+bG5mk4mrntv//mAlIpHIZOKqqkz8ZcAqK81EZGUlJqKKCqNMJpFIRBzHlZeaFXKpSMqZuEfMzcIPlkjEz02vNxgMRRpNK7OZq6gwPSRbzV8TEYnFIv7nVtvYsWMnTZr0xCtwDX+6Pt+7d6/6wZmZma1atZJKpVRjfS4zKaQyIjJLTI/1g3qy9VmvrywvL3Byav24c6v39ZmoqqxMp1A4PPH6XOd3x5+szzpdqclkUqvt6SHrs9lAFeUma1uJSPLouT1qfS4vN1pZSSQS0SPW54qKErlcrlDYPPn6zD2Fd999V6fTcRz3888/Hzp0qPYDTCZTt27dSkpKnmYpjemvf/3rgQMHhE5RVz/88ENUVJTQKerqzp07w4YNEzrFo/zp+vyggQMHpqamNkquR8nNze3Vq5fQKTiO406ePPnKK68InYLjOG7z5s3Lli0TOgXHcdy8efOio6OfZg5Pdai6Z8+ehw4dKi8vP3v2bLdu3Sw+xt3dvQkdEHdycrKxaTJXk7C1tXVwcBA6RV3JZDJXV1ehUzxKXdbnaq1bt+Y3n4UlkUjc3NyETkFEpFAonJ2dhU5BRGRra2tvby90CiIijUajVCqfZg5PtYvDZDJt3bo1MzNzyJAhISEhT5MDQHBYn4E1T1XQAADQcBrwb7TCwsJ169ZxHOfr6zt58uSGW1DdJSUlff7559nZ2V999ZWLiwsRmUym9evXFxUVWVtbz5o1Sy6X12WkcdLevXt306ZNEomEiGbPns0H3r1794ULFziOe/PNN1u3bl3HkcbxySeflJWVlZeXjxo1asCAAUR0/fr1rVu3ikSi8PDwHj161HGEcY28Ym/cuDEpKclkMoWFhY0dO9ZigLqM1IvKysqZM2eOGTMmPDychPv93rp1a8uWLUTUpUuX8ePHk0Dvi5KSklWrVslkssrKylmzZjk5OdV/nzzNDuxHW79+/fnz5zmOi4qKSklJabgF1V1ZWZler//www9zc3P5kZMnT3777bccx0VHR8fExNRxpHEUFRWVl5dzHPf777+vX7+e47j8/Py5c+eazebExMRPP/20jiONxmg0chxXVlb26quv8iOzZ88uKSkpLy9/++23+XvrMsK4Rl6xs7OzOY4zmUxvv/12YWGhxQB1GakX0dHRUVFRv/76Kz8pyO/XYDDMnTu3rKysekSo98WuXbv4Y4BHjhz5/vvvuQbokwY8fJeQkMCf6h8cHBwfH99wC6o7a2vrGscAr1+/Xh3y+vXrdRxpHGq1WqFQEJFIJOK3o2/evBkYGCgSiXx9fVNTU+s40mj4kAaDwcvLi4iKi4utrKxsbW0VCoWrq2tmZmZdRhoz8JNp5BWbP7IqFotFIhF/vL12gLqMPD29Xp+QkFB9+FSo3++NGzdUKtWXX3758ccfJycnk3DvC09PT51OR0Q6nY4/LFnvfdKABW0wGGQyGREplUr+ZTBIp9Pxla1SqUpLS+s40pj0ev327dv5D01UJyEis9lcx5HGtHTp0lmzZvHv4QeTKJXK0tLSuow0cuAnIMiKvX///o4dO6rVaosB6jLy9KKjo59//vnqSaF+v4WFhWlpabNmzXr11VfXrl1Lwr0vvL29L1++PG/evAMHDgwaNIgaoE8asKDlcnlVVRUR6fV6lUrVcAt6GiqVqqysjIh0Op2trW0dRxpNVVXVsmXLpkyZwm9GVSchIn5jqi4jjWnRokVff/31zp07S0tLH0yi1+ttbW3rMtLIgZ9A46/YcXFxZ86cmTFjxsMC1GXkKWm12szMzICAgOoRoX6/tra2vr6+CoWidevW5eXlZrNZqPfFzz//PHjw4OXLl0dERHz77bfUAH3SgG/ggICAK1euENHly5cDAwMbbkFPo3bIuow0Do7jPv/88379+lV/+rNTp07x8fEcxyUmJvK7Eeoy0mhpTSYTEVn9h1qtrqysLC0traioyMnJcXd3r8tIowV+Yo28PiQmJv78888ffPABv0VsMUAjrLTJycmFhYWRkZH79u377bffrl69KtTvt0OHDvwxpNLSUqlUKhaLBXxf2NnZEZGtrS3/Z0q9/2oa8DS7wsLCtWvXms3mjh07MnIWR35+/po1a5KSktq0aTNo0KDBgwfzR1QLCwuVSuWDx1gfPdI4aU+fPr127Vo/Pz8i8vX1nTZtGhHt3r37/PnzRPTXv/61+tj0n440gvLy8o8//lgsFhsMhmHDhvFnccTHx/NH8EePHs0fwa/LCOMaecV+++23JRIJXwSvv/5627Ztaweoy0h92bt3r8lk4s/iEOr3e/jw4RMnThiNxkmTJvH7cwV5X9y7d2/t2rVWVlbl5eVvvPGGl5dXvfcJzoMGAGBUk/kQNgBAS4OCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBg1P8DaA3gY6omBDQAAAAASUVORK5CYII=",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "plot_covariates(sceptre_object_highmoi, p_mito_threshold = 0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "cc34da4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178488.23-P1P2 (1 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178620.23-P1P2 (5 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACTL7B_-_111617995.23-P1P2 (10 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR1B_+_52345762.23-P1P2 (15 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2A_+_148602343.23-P1P2 (20 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_+_38495339.23-P1P2 (25 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_-_38495819.23-P1P2 (30 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ADCY2_-_7396429.23-P1P2 (35 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA AMOTL2_+_134093279.23-P1P2 (40 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ANP32E_+_150208342.23-P1P2 (45 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA APAF1_-_99039071.23-P1P2 (50 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_+_72766725.23-P1P2 (55 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_-_72766755.23-P1P2 (60 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARTN_-_44399429.23-P1 (65 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASCC3_-_101329193.23-P1P2 (70 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASH2L_-_37963059.23-P1P2 (75 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATMIN_+_81069661.23-P1P2 (80 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_+_111992112.23-P1P2 (85 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_-_111992129.23-P1P2 (90 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP6AP1_-_153657075.23-P1P2 (95 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATXN2L_-_28834668.23-P1P2 (100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAD_+_64052102.23-P1P2 (105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAG6_+_31620398.23-P1P2 (110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_+_49458268.23-P1P2 (115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_-_49458323.23-P1P2 (120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BCAS2_-_115124045.23-P1P2 (125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BMPR1A_-_88516386.23-P1P2 (130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BRD3_-_136932911.23-P1P2 (135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C17orf49_+_6918114.23-P2 (140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_+_31532436.23-P1P2 (145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_-_31532650.23-P1P2 (150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACFD1_-_136325160.23-P1P2 (155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACNG1_+_65040756.23-P1P2 (160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CAPN5_+_76778225.23-P1P2 (165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CASP3_+_185570560.23-P1P2 (170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_+_107384793.23-P1P2 (175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_-_107384635.23-P1P2 (180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCM2_-_45039941.23-P1P2 (185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNC_-_100016267.23-P1P2 (190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNT2_-_135676416.23-P1P2 (195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCT6B_+_33288381.23-P1P2 (200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_+_30366413.23-P1P2 (205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_-_30366630.23-P1P2 (210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD99L2_-_150067120.23-P1P2 (215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDH5_-_66400630.23-P1P2 (220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDKN1A_-_36646538.23-P1P2 (225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CENPC_+_68411190.23-P1P2 (230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_+_30773828.23-P1P2 (235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_-_30773391.23-P1P2 (240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHMP3_-_86790451.23-P1P2 (245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHORDC1_-_89955894.23-P1P2 (250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNEP1R1_+_50059284.23-P1P2 (255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNOT2_+_70637651.23-P1P2 (260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_+_83956339.23-P1P2 (265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_-_83956572.23-P1P2 (270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS6_-_99686676.23-P1P2 (275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_-_31634088.23-P1P2 (280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CTR9_+_10772881.23-P1P2 (285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CUL3_-_225449667.23-P1P2 (290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_+_145150267.23-P1P2 (295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_-_145149996.23-P1P2 (300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYCS_+_25164869.23-P1P2 (305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DACH1_-_72441101.23-P1P2 (310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DBR1_+_137893673.23-P1P2 (315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DCTN2_-_57940910.23-P1P2 (320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_+_61851532.23-P1P2 (325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_-_61851812.23-P1P2 (330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DGCR8_-_20068158.23-P1P2 (335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNAJC17_-_41099215.23-P1P2 (340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNM2_+_10829157.23-P1P2 (345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM1_+_49575018.23-P1P2 (350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_+_130700056.23-P1P2 (355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_-_130700076.23-P1P2 (360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM3_+_155113020.23-P1P2 (365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DROSHA_-_31531982.23-P1P2 (370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_+_3985081.23-P1 (375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_-_3985443.23-P1 (380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EHMT2_-_31865199.23-P1P2 (385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3F_+_8009347.23-P1P2 (390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3M_-_32605405.23-P1P2 (395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4E_+_99849831.23-P1P2 (400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184032364.23-P1P2 (405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184033042.23-P1P2 (410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G2_+_10830447.23-P1 (415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF5A_-_7211217.23-P1P2 (420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TCEB3_-_24069975.23-P1P2 (425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EMC2_+_109456080.23-P1P2 (430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823220.23-P1P2 (435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823252.23-P1P2 (440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EOMES_-_27763342.23-P1P2 (445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERCC2_+_45873802.23-P1P2 (450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERO1L_+_53162425.23-P1P2 (455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETS1_+_128392044.23-P1P2 (460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_+_36134592.23-P1P2 (465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_-_36134789.23-P1P2 (470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EWSR1_-_29664347.23-P1P2 (475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EXOSC8_-_37574979.23-P1P2 (480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FBXW11_-_171433300.23-P1P2 (485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FEN1_+_61560782.23-P1P2 (490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_+_75532080.23-P1P2 (495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_-_75532461.23-P1P2 (500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT7_-_139926495.23-P1P2 (505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA G3BP1_-_151151499.23-P1P2 (510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GCK_+_44229021.23-P1P2 (515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOLT1B_+_21654777.23-P1P2 (520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_+_45000529.23-P1P2 (525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_-_45000948.23-P1P2 (530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPI_+_34856284.23-P1P2 (535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPR107_-_132816044.23-P1P2 (540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GSPT1_-_12009714.23-P1P2 (545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A1_+_81687215.23-P1P2 (550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_+_59949355.23-P1P2 (555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_-_59949657.23-P1P2 (560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2H4_-_30876185.23-P1P2 (565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF3C2_+_27579604.23-P1P2 (570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HAND1_+_153857732.23-P1P2 (575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HDAC2_-_114292253.23-P1P2 (580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_+_81399312.23-P1P2 (585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_-_81399351.23-P1P2 (590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HINT2_+_35814995.23-P1P2 (595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNF4A_+_43029971.23-P1P2 (600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPM_-_8509885.23-P1 (605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPU_+_245027753.23-P1P2 (610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_+_43702292.23-P1P2 (615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_-_43702305.23-P1P2 (620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSPA5_-_128003534.23-P1P2 (625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HUWE1_-_53710540.23-P1P2 (630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_-_30007588.23-P1P2 (635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INTS10_+_19675000.23-P1P2 (640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_+_65903046.23-P1P2 (645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_-_65903523.23-P1P2 (650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C7orf26_+_6630358.23-P1P2 (655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA IPO8_+_30848771.23-P1P2 (660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ITGB1BP1_+_9563543.23-P1P2 (665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSFA2_+_182756743.23-P1P2 (670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_+_109914946.23-P1P2 (675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_-_109915112.23-P1P2 (680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM1A_-_23346409.23-P1P2 (685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM3B_-_137688431.23-P1P2 (690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDR_-_55991659.23-P1P2 (695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA0232_+_6784651.23-P1P2 (700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153706791.23-P1P2 (705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153707248.23-P1P2 (710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAMTOR4_-_99746921.23-P1P2 (715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LARP4_-_50794864.23-P1P2 (720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LATS2_-_21635336.23-P1P2 (725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LEMD3_+_65563441.23-P1P2 (730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_+_54960428.23-P1P2 (735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_-_54960616.23-P1P2 (740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LHFPL2_-_77944519.23-P1P2 (745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LSM12_-_42143603.23-P1P2 (750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MANBAL_+_35918136.23-P1P2 (755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MAU2_+_19431732.23-P1P2 (760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_+_47807678.23-P1P2 (765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_-_47807877.23-P1P2 (770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEAF6_-_37980319.23-P1P2 (775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED1_-_37607140.23-P1P2 (780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED12_+_70338541.23-P1P2 (785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED15_+_20862189.23-P1 (790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_+_57479335.23-P1P2 (795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_-_57479437.23-P1P2 (800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEMO1_-_32264882.23-P1 (805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MIXL1_+_226411425.23-P1P2 (810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL28_+_420487.23-P1P2 (815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL32_+_42972018.23-P1P2 (820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_+_179322181.23-P1P2 (825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_-_179322213.23-P1P2 (830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LUZP6_-_135612750.23-P1P2 (835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA20_-_19998023.23-P1P2 (840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA25_+_112546538.23-P1P2 (845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCBP1_+_100396082.23-P1P2 (850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_+_172428630.23-P1P2 (855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_-_172428841.23-P1P2 (860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NEDD8_-_24701363.23-P1P2 (865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NF2_+_29999635.23-P1P2 (870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NFIB_+_14314559.23-P1P2 (875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_+_139418329.23-P2 (880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139418349.23-P2 (885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139440503.23-P1 (890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUFIP2_+_27621073.23-P1P2 (895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUP93_-_56764090.23-P1P2 (900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OGT_+_70752948.23-P1P2 (905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_+_59383170.23-P1P2 (910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_-_59383348.23-P1P2 (915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAM16_-_4401236.23-P1P2 (920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAXBP1_-_34143614.23-P1P2 (925 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PBLD_-_70092436.23-P1 (930 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PCBP2_+_53835614.23-P2 (935 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_+_94292752.23-P2 (940 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_-_94293027.23-P2 (945 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PGRMC2_+_129208913.23-P1P2 (950 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_-_178866211.23-P1P2 (955 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PITPNC1_-_65373433.23-P1P2 (960 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLCG1_+_39766176.23-P1P2 (965 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_+_151042273.23-P2 (970 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_-_151042422.23-P2 (975 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLRG1_-_155471154.23-P1P2 (980 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLXNA2_-_208417275.23-P1P2 (985 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLE4_-_75185725.23-P1P2 (990 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR2M_+_57998962.23-P1P2 (995 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940045.23-P1P2 (1000 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940414.23-P1P2 (1005 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3K_-_103144.23-P1P2 (1010 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPHLN1_-_42720007.23-P1P2 (1015 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPIL4_+_149867146.23-P1P2 (1020 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPP2R1A_+_52693317.23-P1 (1025 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_+_53195217.23-P1P2 (1030 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_-_53195332.23-P1P2 (1035 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKD1_-_30396996.23-P1P2 (1040 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRMT5_+_23398554.23-P1 (1045 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C2orf53_-_27362230.23-P1P2 (1050 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRSS8_+_31147010.23-P1P2 (1055 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_+_36909062.23-P1P2 (1060 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_-_36909230.23-P1P2 (1065 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMD4_-_151227289.23-P1P2 (1070 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMG4_-_3259227.23-P1P2 (1075 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTAR1_+_72374835.23-P1P2 (1080 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTBP1_+_797463.23-P1P2 (1085 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_+_89623388.23-P1P2 (1090 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_-_89623526.23-P1P2 (1095 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTMS_-_6875591.23-P1P2 (1100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PYGO2_-_154934135.23-P1P2 (1105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB18_-_27793292.23-P1P2 (1110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB6A_+_73471983.23-P1P2 (1115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_+_76251918.23-P1P2 (1120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_-_76252008.23-P1P2 (1125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GNB2L1_-_180670840.23-P1P2 (1130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAD21_-_117886635.23-P1P2 (1135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RANGRF_+_8192257.23-P1P2 (1140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBBP5_+_205091050.23-P1P2 (1145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM14-RBM4_+_66406280.23-P2 (1150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM33_+_155437802.23-P1P2 (1155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM4_+_66406319.23-P2 (1160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_+_114271352.23-P1P2 (1165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_-_114271700.23-P1P2 (1170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBX1_-_41347852.23-P1P2 (1175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1551_+_32112399.23-P1P2 (1180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RFX3_-_3525750.23-P1P2 (1185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RGP1_+_35749328.23-P1P2 (1190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_+_95067102.23-P1P2 (1195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_-_95067164.23-P1P2 (1200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RIBC1_-_53449606.23-P1P2 (1205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1432_+_5629173.23-P1P2 (1210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RLF_+_40627522.23-P1P2 (1215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF115_-_145611148.23-P1P2 (1220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_+_104296142.23-P1P2 (1225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_-_104296250.23-P1P2 (1230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF40_-_30773681.23-P1P2 (1235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPL37A_-_217363600.23-P1P2 (1240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPS15A_-_18801590.23-P1P2 (1245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAV1_+_51134992.23-P1P2 (1250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_+_33104038.23-P1P2 (1255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_-_33104343.23-P1P2 (1260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF8_+_155054976.23-P1P2 (1265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EPT1_+_26569050.23-P1P2 (1270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERF2_+_44069659.23-P2 (1275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERPING1_+_57365137.23-P1P2 (1280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_+_122242041.23-P1P2 (1285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_-_122242294.23-P1P2 (1290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETDB1_-_150898815.23-P1P2 (1295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SF1_-_64545937.23-P1P2 (1300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCDC101_-_28565268.23-P1P2 (1305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SKA3_+_21750519.23-P1P2 (1310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_+_78591497.23-P1P2 (1315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_-_78591737.23-P1P2 (1320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC29A1_-_44187438.23-P2 (1325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC35C1_-_45826711.23-P1P2 (1330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD1_+_146403428.23-P1P2 (1335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD3_+_67358229.23-P1P2 (1340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_+_48556623.23-P1P2 (1345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_-_48556657.23-P1P2 (1350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD5_-_135468694.23-P1P2 (1355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA4_-_11071741.23-P1P2 (1360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA5_+_144435243.23-P1P2 (1365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCB1_+_24129228.23-P1 (1370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_+_47822975.23-P1P2 (1375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_-_47822893.23-P1P2 (1380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCD3_-_150974212.23-P1 (1385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCE1_-_38803669.23-P1P2 (1390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMC2_-_106856631.23-P1P2 (1395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMG5_+_156252537.23-P1P2 (1400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_+_103852.23-P1P2 (1405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_-_104242.23-P1P2 (1410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP27_-_70121181.23-P1P2 (1415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNUPN_-_75917960.23-P1P2 (1420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SOX4_+_21594213.23-P1 (1425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPIN1_+_91003737.23-P1P2 (1430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_+_2328680.23-P1P2 (1435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_-_2328912.23-P1P2 (1440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPTY2D1_-_18655846.23-P1P2 (1445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP19_+_112197188.23-P1P2 (1450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP54_-_35452244.23-P1P2 (1455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP68_+_74068540.23-P1P2 (1460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_+_126138730.23-P1P2 (1465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_-_126138742.23-P1P2 (1470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPRB_-_133524719.23-P1P2 (1475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRRM2_-_2802714.23-P1P2 (1480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRSF6_+_42086993.23-P1 (1485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSB_+_170655450.23-P1P2 (1490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_+_57102902.23-P1P2 (1495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_-_57103355.23-P1P2 (1500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STAG2_-_123095658.23-P1 (1505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STK40_+_36851507.23-P1P2 (1510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUPT20H_+_37633798.23-P1P2 (1515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUZ12_+_30264157.23-P1P2 (1520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_+_64901973.23-P1P2 (1525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_-_64901981.23-P1P2 (1530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TADA2B_-_7045121.23-P1P2 (1535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF10_+_6633406.23-P1P2 (1540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF2_+_120845032.23-P1P2 (1545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF3_-_7860540.23-P1P2 (1550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_+_60640814.23-P1P2 (1555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_-_60640498.23-P1P2 (1560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF6L_-_62538954.23-P1P2 (1565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAOK1_-_27717907.23-P1P2 (1570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FAM208A_+_56717273.23-P1P2 (1575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBC1D9_+_141677481.23-P1P2 (1580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_+_170863524.23-P1P2 (1585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_-_170863481.23-P1P2 (1590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBX3_-_115121998.23-P1P2 (1595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA T_-_166581333.23-P1P2 (1600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TGFBR2_+_30648526.23-P1P2 (1605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA THBD_+_23029196.23-P2 (1610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_+_2977592.23-P1P2 (1615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_-_2977643.23-P1P2 (1620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MESDC1_-_81293340.23-P1P2 (1625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMEM41B_+_9336156.23-P1P2 (1630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMTC3_-_88536127.23-P1P2 (1635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TOP1_+_39657538.23-P1P2 (1640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_+_7590531.23-P1P2 (1645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_-_7590704.23-P1P2 (1650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRIM24_-_138145308.23-P1P2 (1655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRRAP_-_98476465.23-P1P2 (1660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TSTD2_+_100395594.23-P1P2 (1665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UBE2E1_+_23847541.23-P1P2 (1670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_+_38924011.23-P1P2 (1675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_-_38924421.23-P1P2 (1680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UPF2_+_12084799.23-P1P2 (1685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UXT_-_47518505.23-P1P2 (1690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VHL_-_10183525.23-P1P2 (1695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VPS54_+_64246104.23-P1P2 (1700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_+_28821698.23-P1P2 (1705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_-_28822074.23-P1P2 (1710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAPAL_+_88281533.23-P1P2 (1715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR12_-_203776383.23-P1P2 (1720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR77_+_111991860.23-P1P2 (1725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WIZ_-_15543016.23-P1P2 (1730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_+_9983328.23-P1P2 (1735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_-_9983510.23-P1P2 (1740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XPO5_-_43543670.23-P1P2 (1745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XRN1_-_142166648.23-P1P2 (1750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA YARS2_+_32908787.23-P1P2 (1755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB22_+_33285454.23-P1P2 (1760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_+_6640093.23-P1P2 (1765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_-_6640138.23-P1P2 (1770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZC3H8_-_113012435.23-P1P2 (1775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZNF281_+_200379151.23-P1P2 (1780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00117 (1785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00233 (1790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00518 (1795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00656 (1800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01226 (1805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01730 (1810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02009 (1815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02288 (1820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02777 (1825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03026 (1830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03252 (1835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03620 (1840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_SGTN_04 (1845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_SGTN_25 (1850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_SGTN_26 (1855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1044 (1860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1414 (1865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1554 (1870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1785 (1875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2009 (1880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2271 (1885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2425 (1890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2614 (1895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_293 (1900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_541 (1905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_725 (1910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_860 (1915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_SGTN_14 (1920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_SGTN_01 (1925 of 1926)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- assign_grnas(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  method = \"mixture\",\n",
    "  parallel = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "479b6e65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     cell_id                grna_target value\n",
      "1 TAGGATGAATTGAGGATGCTTGGG_2    PTEN_+_89623388.23-P1P2     1\n",
      "2 TAGGATGAAGCCATGCCGCCCGGA_2  CCNT2_-_135676427.23-P1P2     1\n",
      "3 TAGGATGAAGCCATGCCGCCCGGA_2      PBLD_-_70092630.23-P1     1\n",
      "4 TAGGATGAAGCCATGCCGCCCGGA_2 PIK3CA_-_178866211.23-P1P2     1\n",
      "5 TAGGATGAAGCCATGCCGCCCGGA_2 SMARCC1_+_47822975.23-P1P2     1\n",
      "6 CCAGTTCACAATGGAATGCTTGGG_3   COPS4_-_83956484.23-P1P2     1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "⚠️ Matrix missing colnames. Fetching from sceptre_object...\n",
       "✅ Saved long CSV to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/grna_assignments_long.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(Matrix)\n",
    "\n",
    "# 1. Get the assignment matrix\n",
    "A <- get_grna_assignments(sceptre_object_highmoi)\n",
    "\n",
    "# 2. Ensure it is a numeric sparse matrix (1/0)\n",
    "if (inherits(A, \"Matrix\") || is.matrix(A)) {\n",
    "    A_num <- if (inherits(A, \"Matrix\")) drop0(A * 1) else (A * 1)\n",
    "} else {\n",
    "    stop(\"Output is not a matrix.\")\n",
    "}\n",
    "\n",
    "# --- FIX: ROBUST NAME RECOVERY ---\n",
    "# We retrieve names using Sceptre's standard metadata accessors\n",
    "# instead of touching internal slots like @grna_matrix.\n",
    "\n",
    "# Try to get cell names from the matrix first\n",
    "current_cell_names <- colnames(A_num)\n",
    "current_guide_names <- rownames(A_num)\n",
    "\n",
    "# If missing, grab them from the main object's covariate data frame\n",
    "if (is.null(current_cell_names)) {\n",
    "    message(\"⚠️ Matrix missing colnames. Fetching from sceptre_object...\")\n",
    "    # The covariate data frame ALWAYS has the cell IDs as rownames\n",
    "    current_cell_names <- rownames(sceptre_object_highmoi@covariate_data_frame)\n",
    "    \n",
    "    # Safety check: lengths must match\n",
    "    if (length(current_cell_names) != ncol(A_num)) {\n",
    "        stop(paste(\"Mismatch! Matrix has\", ncol(A_num), \"cols but object has\", length(current_cell_names), \"cells.\"))\n",
    "    }\n",
    "    colnames(A_num) <- current_cell_names\n",
    "}\n",
    "\n",
    "# If guide names are missing, they should match the unique values in your guide mapping\n",
    "if (is.null(current_guide_names)) {\n",
    "    message(\"⚠️ Matrix missing rownames. Fetching from guide mapping...\")\n",
    "    # Get unique targets from the mapping table you loaded earlier\n",
    "    # Note: This assumes the matrix rows are sorted/aligned to these unique targets.\n",
    "    # In 'mixture' method, rows usually correspond to the 'grna_target' labels.\n",
    "    \n",
    "    # SAFE FALLBACK: If we can't be 100% sure of order, we warn the user.\n",
    "    # But usually, they match the 'grna_target' column from your input.\n",
    "    unique_targets <- sort(unique(guide_mapping$grna_target))\n",
    "    \n",
    "    if (length(unique_targets) == nrow(A_num)) {\n",
    "         rownames(A_num) <- unique_targets\n",
    "    } else {\n",
    "         warning(\"Could not automatically align target names. Using generic IDs.\")\n",
    "         rownames(A_num) <- paste0(\"target_\", seq_len(nrow(A_num)))\n",
    "    }\n",
    "}\n",
    "# ---------------------------------\n",
    "\n",
    "# 3. Convert to Long Format\n",
    "s <- summary(A_num)\n",
    "assign_long <- data.frame(\n",
    "    cell_id = colnames(A_num)[s$j],\n",
    "    grna_target = rownames(A_num)[s$i],\n",
    "    value = s$x\n",
    ")\n",
    "\n",
    "# 4. Save\n",
    "out_path <- file.path(data_dir, \"grna_assignments_long.csv\")\n",
    "write.csv(assign_long, out_path, row.names = FALSE)\n",
    "\n",
    "message(\"✅ Saved long CSV to: \", out_path)\n",
    "print(head(assign_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "f1923840",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/MOI.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "png(file.path(data_dir, \"MOI.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"MOI.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "70ebd23d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/QC.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_qc(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  p_mito_threshold = 0.25\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"QC.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"QC.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "8824bbda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Constructing negative control pairs. ✓\n",
      "Running precomputation on response ENSG00000143545.10 (1 of 226)\n",
      "Running precomputation on response ENSG00000132603.15 (5 of 226)\n",
      "Running precomputation on response ENSG00000155252.14 (10 of 226)\n",
      "Running precomputation on response ENSG00000154175.18 (15 of 226)\n",
      "Running precomputation on response ENSG00000230590.13 (20 of 226)\n",
      "Running precomputation on response ENSG00000290146.1 (25 of 226)\n",
      "Running precomputation on response ENSG00000035115.22 (30 of 226)\n",
      "Running precomputation on response ENSG00000115207.15 (35 of 226)\n",
      "Running precomputation on response ENSG00000140259.7 (40 of 226)\n",
      "Running precomputation on response ENSG00000169180.12 (45 of 226)\n",
      "Running precomputation on response ENSG00000110330.10 (50 of 226)\n",
      "Running precomputation on response ENSG00000226361.5 (55 of 226)\n",
      "Running precomputation on response ENSG00000100150.20 (60 of 226)\n",
      "Running precomputation on response ENSG00000129696.13 (65 of 226)\n",
      "Running precomputation on response ENSG00000133983.16 (70 of 226)\n",
      "Running precomputation on response ENSG00000170949.18 (75 of 226)\n",
      "Running precomputation on response ENSG00000167930.17 (80 of 226)\n",
      "Running precomputation on response ENSG00000145901.16 (85 of 226)\n",
      "Running precomputation on response ENSG00000163214.21 (90 of 226)\n",
      "Running precomputation on response ENSG00000249803.6 (95 of 226)\n",
      "Running precomputation on response ENSG00000237350.1 (100 of 226)\n",
      "Running precomputation on response ENSG00000185627.19 (105 of 226)\n",
      "Running precomputation on response ENSG00000290263.1 (110 of 226)\n",
      "Running precomputation on response ENSG00000100441.10 (115 of 226)\n",
      "Running precomputation on response ENSG00000147799.12 (120 of 226)\n",
      "Running precomputation on response ENSG00000182378.15 (125 of 226)\n",
      "Running precomputation on response ENSG00000108424.11 (130 of 226)\n",
      "Running precomputation on response ENSG00000196850.6 (135 of 226)\n",
      "Running precomputation on response ENSG00000127948.17 (140 of 226)\n",
      "Running precomputation on response ENSG00000159899.16 (145 of 226)\n",
      "Running precomputation on response ENSG00000172530.22 (150 of 226)\n",
      "Running precomputation on response ENSG00000065029.15 (155 of 226)\n",
      "Running precomputation on response ENSG00000158623.14 (160 of 226)\n",
      "Running precomputation on response ENSG00000177731.16 (165 of 226)\n",
      "Running precomputation on response ENSG00000127947.16 (170 of 226)\n",
      "Running precomputation on response ENSG00000166037.11 (175 of 226)\n",
      "Running precomputation on response ENSG00000250802.7 (180 of 226)\n",
      "Running precomputation on response ENSG00000038358.15 (185 of 226)\n",
      "Running precomputation on response ENSG00000214425.11 (190 of 226)\n",
      "Running precomputation on response ENSG00000150275.20 (195 of 226)\n",
      "Running precomputation on response ENSG00000030582.18 (200 of 226)\n",
      "Running precomputation on response ENSG00000109118.14 (205 of 226)\n",
      "Running precomputation on response ENSG00000187778.14 (210 of 226)\n",
      "Running precomputation on response ENSG00000182919.15 (215 of 226)\n",
      "Running precomputation on response ENSG00000254726.3 (220 of 226)\n",
      "Running precomputation on response ENSG00000285816.1 (225 of 226)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00259&non-targeting_02159&non-targeting_02571&non-targeting_03029&non-targeting_03081 (1 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00355&non-targeting_02025&non-targeting_02089&non-targeting_03081&non-targeting_03384 (5 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_01730&non-targeting_02193&non-targeting_03180&non-targeting_03620&non-targeting_03664 (10 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00043&non-targeting_00117&non-targeting_01509&non-targeting_02543&non-targeting_03026&non-targeting_03252 (15 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00259&non-targeting_01732&non-targeting_01934&non-targeting_02193&non-targeting_02837&non-targeting_03081 (20 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00259&non-targeting_01504&non-targeting_02009&non-targeting_02574&non-targeting_03081&non-targeting_03252 (25 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_02543&non-targeting_02777&non-targeting_02814&non-targeting_03252&non-targeting_03483 (30 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00918&non-targeting_01730&non-targeting_02025&non-targeting_02911&non-targeting_03384 (35 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00390&non-targeting_00518&non-targeting_00625&non-targeting_03081&non-targeting_03252 (40 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00231&non-targeting_00614&non-targeting_00629&non-targeting_00918&non-targeting_02193&non-targeting_02543 (45 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00390&non-targeting_01732&non-targeting_02193&non-targeting_03081&non-targeting_03384 (50 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00518&non-targeting_00614&non-targeting_01226&non-targeting_02025&non-targeting_02837 (55 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_02193&non-targeting_02571&non-targeting_02911&non-targeting_03081&non-targeting_03180 (60 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00324&non-targeting_00629&non-targeting_00972&non-targeting_02677&non-targeting_03252&non-targeting_03664 (65 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00656&non-targeting_02543&non-targeting_02574&non-targeting_03081&non-targeting_03384 (70 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_00259&non-targeting_00629&non-targeting_01601&non-targeting_01732&non-targeting_02837 (75 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_00355&non-targeting_02837&non-targeting_02911&non-targeting_03342&non-targeting_03620 (80 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00614&non-targeting_02009&non-targeting_02574&non-targeting_03026&non-targeting_03180 (85 of 90)\n",
      "Analyzing pairs containing gRNA group non-targeting_00625&non-targeting_01226&non-targeting_01858&non-targeting_02009&non-targeting_02962&non-targeting_03158 (90 of 90)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/calibration_check.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Calibration check\n",
    "\n",
    "\n",
    "sceptre_object_highmoi <- run_calibration_check(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"calibration_check.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"calibration_check.png\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "f6e44c8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          response_id\n",
      "               <char>\n",
      "1:  ENSG00000290263.1\n",
      "2: ENSG00000134371.14\n",
      "3:  ENSG00000100591.8\n",
      "4: ENSG00000143742.14\n",
      "5: ENSG00000163930.10\n",
      "6:  ENSG00000237350.1\n",
      "                                                                                                               grna_target\n",
      "                                                                                                                    <char>\n",
      "1: non-targeting_00233&non-targeting_00390&non-targeting_01601&non-targeting_02009&non-targeting_02288&non-targeting_03029\n",
      "2:"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " non-targeting_00014&non-targeting_00259&non-targeting_00629&non-targeting_01601&non-targeting_01732&non-targeting_02837\n",
      "3: non-targeting_00614&non-targeting_00972&non-targeting_02911&non-targeting_03081&non-targeting_03180&non-targeting_03384\n",
      "4: non-targeting_00324&non-targeting_00629&non-targeting_00972&non-targeting_02677&non-targeting_03252&non-targeting_03664\n",
      "5: non-targeting_00629&non-targeting_01185&non-targeting_01732&non-targeting_02089&non-targeting_02677&non-targeting_03483\n",
      "6: non-targeting_00014&non-targeting_00117&non-targeting_00196&non-targeting_00555&non-targeting_02025&non-targeting_02193\n",
      "   n_nonzero_trt n_nonzero_cntrl pass_qc     p_value log_2_fold_change\n",
      "           <int>           <int>  <lgcl>       <num>             <num>\n",
      "1:            25             843    TRUE 0.003853224         0.8886069\n",
      "2:            44            3161    TRUE 0.016145901        -0.4608229\n",
      "3:            13             997    TRUE 0.018890360        -0.8134084\n",
      "4:            53            3475    TRUE 0.024000000        -0.2138143\n",
      "5:            24             776    TRUE 0.024000000         0.6536926\n",
      "6:            33             964    TRUE 0.032000000         0.6869690\n",
      "   significant\n",
      "        <lgcl>\n",
      "1:       FALSE\n",
      "2:       FALSE\n",
      "3:       FALSE\n",
      "4:       FALSE\n",
      "5:       FALSE\n",
      "6:       FALSE\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "calibration_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_calibration_check\"\n",
    ")\n",
    "head(calibration_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ddfa4fbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Running precomputation on response ENSG00000004487.18 (1 of 227)\n",
      "Running precomputation on response ENSG00000116459.11 (5 of 227)\n",
      "Running precomputation on response ENSG00000163348.4 (10 of 227)\n",
      "Running precomputation on response ENSG00000138018.19 (15 of 227)\n",
      "Running precomputation on response ENSG00000124380.11 (20 of 227)\n",
      "Running precomputation on response ENSG00000138385.16 (25 of 227)\n",
      "Running precomputation on response ENSG00000170142.12 (30 of 227)\n",
      "Running precomputation on response ENSG00000114019.15 (35 of 227)\n",
      "Running precomputation on response ENSG00000128052.10 (40 of 227)\n",
      "Running precomputation on response ENSG00000109436.8 (45 of 227)\n",
      "Running precomputation on response ENSG00000113360.17 (50 of 227)\n",
      "Running precomputation on response ENSG00000120733.14 (55 of 227)\n",
      "Running precomputation on response ENSG00000124766.7 (60 of 227)\n",
      "Running precomputation on response ENSG00000112237.13 (65 of 227)\n",
      "Running precomputation on response ENSG00000213079.10 (70 of 227)\n",
      "Running precomputation on response ENSG00000136280.17 (75 of 227)\n",
      "Running precomputation on response ENSG00000188186.11 (80 of 227)\n",
      "Running precomputation on response ENSG00000104613.12 (85 of 227)\n",
      "Running precomputation on response ENSG00000179091.5 (90 of 227)\n",
      "Running precomputation on response ENSG00000106723.17 (95 of 227)\n",
      "Running precomputation on response ENSG00000136908.18 (100 of 227)\n",
      "Running precomputation on response ENSG00000151461.20 (105 of 227)\n",
      "Running precomputation on response ENSG00000171862.14 (110 of 227)\n",
      "Running precomputation on response ENSG00000179119.15 (115 of 227)\n",
      "Running precomputation on response ENSG00000156603.20 (120 of 227)\n",
      "Running precomputation on response ENSG00000248643.5 (125 of 227)\n",
      "Running precomputation on response ENSG00000182934.12 (130 of 227)\n",
      "Running precomputation on response ENSG00000134283.18 (135 of 227)\n",
      "Running precomputation on response ENSG00000174106.3 (140 of 227)\n",
      "Running precomputation on response ENSG00000111300.10 (145 of 227)\n",
      "Running precomputation on response ENSG00000120699.14 (150 of 227)\n",
      "Running precomputation on response ENSG00000129559.13 (155 of 227)\n",
      "Running precomputation on response ENSG00000165417.12 (160 of 227)\n",
      "Running precomputation on response ENSG00000166949.17 (165 of 227)\n",
      "Running precomputation on response ENSG00000167978.18 (170 of 227)\n",
      "Running precomputation on response ENSG00000103549.22 (175 of 227)\n",
      "Running precomputation on response ENSG00000141510.18 (180 of 227)\n",
      "Running precomputation on response ENSG00000125686.12 (185 of 227)\n",
      "Running precomputation on response ENSG00000154217.16 (190 of 227)\n",
      "Running precomputation on response ENSG00000005206.17 (195 of 227)\n",
      "Running precomputation on response ENSG00000011451.21 (200 of 227)\n",
      "Running precomputation on response ENSG00000105568.19 (205 of 227)\n",
      "Running precomputation on response ENSG00000124193.16 (210 of 227)\n",
      "Running precomputation on response ENSG00000128191.17 (215 of 227)\n",
      "Running precomputation on response ENSG00000100387.9 (220 of 227)\n",
      "Running precomputation on response ENSG00000147162.15 (225 of 227)\n",
      "Analyzing pairs containing gRNA group KDM1A (1 of 227)\n",
      "Analyzing pairs containing gRNA group ATP5PB (5 of 227)\n",
      "Analyzing pairs containing gRNA group PYGO2 (10 of 227)\n",
      "Analyzing pairs containing gRNA group SELENOI (15 of 227)\n",
      "Analyzing pairs containing gRNA group SNRNP27 (20 of 227)\n",
      "Analyzing pairs containing gRNA group SSB (25 of 227)\n",
      "Analyzing pairs containing gRNA group UBE2E1 (30 of 227)\n",
      "Analyzing pairs containing gRNA group AMOTL2 (35 of 227)\n",
      "Analyzing pairs containing gRNA group KDR (40 of 227)\n",
      "Analyzing pairs containing gRNA group TBC1D9 (45 of 227)\n",
      "Analyzing pairs containing gRNA group DROSHA (50 of 227)\n",
      "Analyzing pairs containing gRNA group KDM3B (55 of 227)\n",
      "Analyzing pairs containing gRNA group SOX4 (60 of 227)\n",
      "Analyzing pairs containing gRNA group CCNC (65 of 227)\n",
      "Analyzing pairs containing gRNA group SCAF8 (70 of 227)\n",
      "Analyzing pairs containing gRNA group CCM2 (75 of 227)\n",
      "Analyzing pairs containing gRNA group LAMTOR4 (80 of 227)\n",
      "Analyzing pairs containing gRNA group INTS10 (85 of 227)\n",
      "Analyzing pairs containing gRNA group CYC1 (90 of 227)\n",
      "Analyzing pairs containing gRNA group SPIN1 (95 of 227)\n",
      "Analyzing pairs containing gRNA group DPM2 (100 of 227)\n",
      "Analyzing pairs containing gRNA group UPF2 (105 of 227)\n",
      "Analyzing pairs containing gRNA group PTEN (110 of 227)\n",
      "Analyzing pairs containing gRNA group SPTY2D1 (115 of 227)\n",
      "Analyzing pairs containing gRNA group MED19 (120 of 227)\n",
      "Analyzing pairs containing gRNA group RBM14-RBM4 (125 of 227)\n",
      "Analyzing pairs containing gRNA group SRPRA (130 of 227)\n",
      "Analyzing pairs containing gRNA group PPHLN1 (135 of 227)\n",
      "Analyzing pairs containing gRNA group LEMD3 (140 of 227)\n",
      "Analyzing pairs containing gRNA group NAA25 (145 of 227)\n",
      "Analyzing pairs containing gRNA group EXOSC8 (150 of 227)\n",
      "Analyzing pairs containing gRNA group NEDD8 (155 of 227)\n",
      "Analyzing pairs containing gRNA group GTF2A1 (160 of 227)\n",
      "Analyzing pairs containing gRNA group SMAD3 (165 of 227)\n",
      "Analyzing pairs containing gRNA group SRRM2 (170 of 227)\n",
      "Analyzing pairs containing gRNA group RNF40 (175 of 227)\n",
      "Analyzing pairs containing gRNA group TP53 (180 of 227)\n",
      "Analyzing pairs containing gRNA group MED1 (185 of 227)\n",
      "Analyzing pairs containing gRNA group PITPNC1 (190 of 227)\n",
      "Analyzing pairs containing gRNA group SPPL2B (195 of 227)\n",
      "Analyzing pairs containing gRNA group WIZ (200 of 227)\n",
      "Analyzing pairs containing gRNA group PPP2R1A (205 of 227)\n",
      "Analyzing pairs containing gRNA group SRSF6 (210 of 227)\n",
      "Analyzing pairs containing gRNA group DGCR8 (215 of 227)\n",
      "Analyzing pairs containing gRNA group RBX1 (220 of 227)\n",
      "Analyzing pairs containing gRNA group OGT (225 of 227)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/discovery.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_discovery_analysis(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"discovery.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"discovery.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "f9327040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved discovery to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d2_sceptre_transfer/TargetKD.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "discovery_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_discovery_analysis\"\n",
    ")\n",
    "\n",
    "write.csv(discovery_result, file.path(data_dir, \"TargetKD.csv\"), row.names = FALSE)\n",
    "message(\"✅ Saved discovery to: \", file.path(data_dir, \"TargetKD.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cddc34a9",
   "metadata": {},
   "source": [
    "### Day 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6d1a288f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io as sio\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "76562a92",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1598: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/oak/stanford/groups/engreitz/Users/tri/envs/sceptre_env/lib/python3.10/site-packages/mudata/_core/mudata.py:1461: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "d3_mdata = mu.read(f\"{data_dir}/d3_normalized_gex_and_guide.h5mu\")\n",
    "mdata = d3_mdata\n",
    "\n",
    "# Filter out cells that have no guides\n",
    "#mdata = mdata[mdata['GEX'].obs['guides_per_cell'] > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "68d7ecb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing sparse matrices...\n"
     ]
    }
   ],
   "source": [
    "# Define a temporary folder for transfer\n",
    "temp_dir = working_dir + \"/d3_sceptre_transfer\"\n",
    "os.makedirs(temp_dir, exist_ok=True)\n",
    "\n",
    "# 1. EXTRACT MATRICES (Transpose them!)\n",
    "# Scanpy is (Cells × Genes), Sceptre needs (Genes × Cells)\n",
    "print(\"Writing sparse matrices...\")\n",
    "sio.mmwrite(f\"{temp_dir}/response_matrix.mtx\", mdata['GEX'].layers['raw_counts'].T)\n",
    "sio.mmwrite(f\"{temp_dir}/grna_matrix.mtx\", mdata['guide'].X.T)\n",
    "\n",
    "# 2. EXTRACT NAMES (Row/Col names are lost in .mtx)\n",
    "# Save Gene names and Guide names to assign them later in R\n",
    "pd.Series(mdata['GEX'].var_names).to_csv(f\"{temp_dir}/gene_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['guide'].var_names).to_csv(f\"{temp_dir}/guide_names.csv\", index=False, header=False)\n",
    "pd.Series(mdata['GEX'].obs_names).to_csv(f\"{temp_dir}/cell_names.csv\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2ea00cf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 44102 × 14712\n",
       "    obs: 'guides_per_cell', 'guide_umi_counts', 'pct_counts_ribo', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'pool', 'sample_id', 'round1', 'well', 'biological_sample', 'plate', 'bci', 'stype', 'day', 'cell_pool', 'log1p_n_genes_by_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'log1p_total_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'S_score', 'G2M_score', 'phase', 'outlier', 'size_factors', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0'\n",
       "    var: 'gene_id', 'symbol', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells', 'n_counts', 'highly_deviant', 'highly_deviant_2k', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'orig_highly_variable', 'highly_variable_2k'\n",
       "    uns: 'hvg', 'leiden_res_0.1', 'leiden_res_0.2', 'leiden_res_0.3', 'leiden_res_0.4', 'leiden_res_0.5', 'leiden_res_0.6', 'leiden_res_0.7', 'leiden_res_0.8', 'leiden_res_0.9', 'leiden_res_1.0', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'guide_assignment', 'guide_counts'\n",
       "    varm: 'PCs'\n",
       "    layers: 'ambiguous', 'analytic_pearson_residuals', 'log1p_norm', 'mature', 'nascent', 'raw_counts', 'scran_normalization'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['GEX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e1d4d6e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>grna_n_nonzero</th>\n",
       "      <th>grna_n_umis</th>\n",
       "      <th>response_n_nonzero</th>\n",
       "      <th>response_n_umis</th>\n",
       "      <th>response_p_mito</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GACAGTGCCCTCTATCTCACTTTA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>4</td>\n",
       "      <td>70.0</td>\n",
       "      <td>786</td>\n",
       "      <td>1109.333252</td>\n",
       "      <td>9.375001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGCAGACTGAGCCACTATTTCA_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>1</td>\n",
       "      <td>908.0</td>\n",
       "      <td>1086</td>\n",
       "      <td>1671.621948</td>\n",
       "      <td>9.092965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCTGAATCCTGTACGCTTAAA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>1</td>\n",
       "      <td>121.0</td>\n",
       "      <td>1493</td>\n",
       "      <td>2472.592041</td>\n",
       "      <td>15.483077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTACGCAAATCATTCCCATTTGGA_3</th>\n",
       "      <td>300genes:eTN5_GEX_sub3</td>\n",
       "      <td>2</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1792</td>\n",
       "      <td>3231.275391</td>\n",
       "      <td>12.533751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAACCACAAAACATCGGACCTTTC_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>3</td>\n",
       "      <td>380.0</td>\n",
       "      <td>847</td>\n",
       "      <td>1556.620483</td>\n",
       "      <td>18.148289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATCCGTCTCTTCACAGTTACGTA_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>870</td>\n",
       "      <td>1732.666626</td>\n",
       "      <td>9.320893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCATTCCCGACACACTCACTTTA_2</th>\n",
       "      <td>300genes:eTN5_GEX_sub2</td>\n",
       "      <td>1</td>\n",
       "      <td>33.0</td>\n",
       "      <td>607</td>\n",
       "      <td>912.434875</td>\n",
       "      <td>18.521870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAACCAACACAGAAGACCTTTC_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>9</td>\n",
       "      <td>249.0</td>\n",
       "      <td>2098</td>\n",
       "      <td>4470.594727</td>\n",
       "      <td>14.371690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGTTGCATCCTGTAGGTGGAGC_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>3</td>\n",
       "      <td>223.0</td>\n",
       "      <td>1213</td>\n",
       "      <td>2456.680908</td>\n",
       "      <td>11.784192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACTGTGGTGCGAACGTCTAGG_1</th>\n",
       "      <td>300genes:eTN5_GEX_sub1</td>\n",
       "      <td>4</td>\n",
       "      <td>75.0</td>\n",
       "      <td>612</td>\n",
       "      <td>1291.583984</td>\n",
       "      <td>14.439633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44102 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             batch  grna_n_nonzero  \\\n",
       "barcode                                                              \n",
       "GACAGTGCCCTCTATCTCACTTTA_2  300genes:eTN5_GEX_sub2               4   \n",
       "ACAGCAGACTGAGCCACTATTTCA_1  300genes:eTN5_GEX_sub1               1   \n",
       "CCTCCTGAATCCTGTACGCTTAAA_3  300genes:eTN5_GEX_sub3               1   \n",
       "GTACGCAAATCATTCCCATTTGGA_3  300genes:eTN5_GEX_sub3               2   \n",
       "CAACCACAAAACATCGGACCTTTC_1  300genes:eTN5_GEX_sub1               3   \n",
       "...                                            ...             ...   \n",
       "AATCCGTCTCTTCACAGTTACGTA_1  300genes:eTN5_GEX_sub1               1   \n",
       "ATCATTCCCGACACACTCACTTTA_2  300genes:eTN5_GEX_sub2               1   \n",
       "AACAACCAACACAGAAGACCTTTC_1  300genes:eTN5_GEX_sub1               9   \n",
       "AATGTTGCATCCTGTAGGTGGAGC_1  300genes:eTN5_GEX_sub1               3   \n",
       "ACCACTGTGGTGCGAACGTCTAGG_1  300genes:eTN5_GEX_sub1               4   \n",
       "\n",
       "                            grna_n_umis  response_n_nonzero  response_n_umis  \\\n",
       "barcode                                                                        \n",
       "GACAGTGCCCTCTATCTCACTTTA_2         70.0                 786      1109.333252   \n",
       "ACAGCAGACTGAGCCACTATTTCA_1        908.0                1086      1671.621948   \n",
       "CCTCCTGAATCCTGTACGCTTAAA_3        121.0                1493      2472.592041   \n",
       "GTACGCAAATCATTCCCATTTGGA_3         59.0                1792      3231.275391   \n",
       "CAACCACAAAACATCGGACCTTTC_1        380.0                 847      1556.620483   \n",
       "...                                 ...                 ...              ...   \n",
       "AATCCGTCTCTTCACAGTTACGTA_1         16.0                 870      1732.666626   \n",
       "ATCATTCCCGACACACTCACTTTA_2         33.0                 607       912.434875   \n",
       "AACAACCAACACAGAAGACCTTTC_1        249.0                2098      4470.594727   \n",
       "AATGTTGCATCCTGTAGGTGGAGC_1        223.0                1213      2456.680908   \n",
       "ACCACTGTGGTGCGAACGTCTAGG_1         75.0                 612      1291.583984   \n",
       "\n",
       "                            response_p_mito  \n",
       "barcode                                      \n",
       "GACAGTGCCCTCTATCTCACTTTA_2         9.375001  \n",
       "ACAGCAGACTGAGCCACTATTTCA_1         9.092965  \n",
       "CCTCCTGAATCCTGTACGCTTAAA_3        15.483077  \n",
       "GTACGCAAATCATTCCCATTTGGA_3        12.533751  \n",
       "CAACCACAAAACATCGGACCTTTC_1        18.148289  \n",
       "...                                     ...  \n",
       "AATCCGTCTCTTCACAGTTACGTA_1         9.320893  \n",
       "ATCATTCCCGACACACTCACTTTA_2        18.521870  \n",
       "AACAACCAACACAGAAGACCTTTC_1        14.371690  \n",
       "AATGTTGCATCCTGTAGGTGGAGC_1        11.784192  \n",
       "ACCACTGTGGTGCGAACGTCTAGG_1        14.439633  \n",
       "\n",
       "[44102 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 3. EXTRACT COVARIATES \n",
    "rna = mdata['GEX']\n",
    "covariates_df = pd.DataFrame({\n",
    "    \"batch\": rna.obs['sample_id'],\n",
    "    \"grna_n_nonzero\": rna.obs['guides_per_cell'],\n",
    "    \"grna_n_umis\": rna.obs['guide_umi_counts'],\n",
    "    \"response_n_nonzero\": rna.obs['n_genes_by_counts'],\n",
    "    \"response_n_umis\": rna.obs['total_counts'],\n",
    "    \"response_p_mito\": rna.obs['pct_counts_mt'],\n",
    "})\n",
    "covariates_df.to_csv(f\"{temp_dir}/covariates.csv\")\n",
    "covariates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a71b3f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_meta_df = pd.read_csv(f\"{working_dir}/300genes_guide_metadata_v43.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "36793d55",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ All guides mapped successfully.\n",
      "\n",
      "First 5 rows of mapping (Must match mdata['guide'].var_names order):\n",
      "                    grna_id grna_target\n",
      "0  ACAA1_+_38178488.23-P1P2       ACAA1\n",
      "1  ACAA1_+_38178517.23-P1P2       ACAA1\n",
      "2  ACAA1_+_38178559.23-P1P2       ACAA1\n",
      "3  ACAA1_+_38178570.23-P1P2       ACAA1\n",
      "4  ACAA1_+_38178620.23-P1P2       ACAA1\n"
     ]
    }
   ],
   "source": [
    "# 4. PREPARE GUIDE MAPPING \n",
    "# Sceptre needs a table mapping 'grna_id' to 'grna_target'.\n",
    "\n",
    "# 1. Create a \"skeleton\" DataFrame from the actual matrix order\n",
    "# This ensures the order matches mdata['guide'].X perfectly\n",
    "# We name the index 'grna_id' to match Sceptre's expectation immediately\n",
    "guide_mapping = pd.DataFrame(index=mdata['guide'].var_names)\n",
    "guide_mapping.index.name = 'grna_id'\n",
    "\n",
    "# 2. Prepare metadata for merging\n",
    "# Ensure 'guide_id' is the index so we can map efficiently\n",
    "lookup_df = guide_meta_df.set_index('guide_id')\n",
    "\n",
    "# 3. Map the targets onto the skeleton\n",
    "# We use .loc[] to pull the specific target for each guide in the exact order required\n",
    "guide_mapping['grna_target'] = lookup_df.loc[guide_mapping.index, 'intended_target_name']\n",
    "\n",
    "# 4. Reset index so 'grna_id' becomes a column (Sceptre requires this)\n",
    "guide_mapping = guide_mapping.reset_index()\n",
    "\n",
    "# ==========================================\n",
    "# VALIDATION CHECKS \n",
    "# ==========================================\n",
    "\n",
    "# Check 1: Are there any NAs? (Guides in mdata that were missing from meta_df)\n",
    "na_count = guide_mapping['grna_target'].isna().sum()\n",
    "if na_count > 0:\n",
    "    print(f\"⚠️ WARNING: {na_count} guides have NaN targets! These will crash Sceptre.\")\n",
    "else:\n",
    "    print(\"✅ All guides mapped successfully.\")\n",
    "\n",
    "# Check 2: Verify the first few rows manually\n",
    "print(\"\\nFirst 5 rows of mapping (Must match mdata['guide'].var_names order):\")\n",
    "print(guide_mapping.head())\n",
    "\n",
    "# 5. Save for Sceptre\n",
    "guide_mapping.to_csv(f\"{temp_dir}/guide_mapping.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6bfc3653",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>grna_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1_+_38178488.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1_+_38178517.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1_+_38178559.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1_+_38178570.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAA1_+_38178620.23-P1P2</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>SUZ12_SGTN_23</td>\n",
       "      <td>SUZ12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1922</th>\n",
       "      <td>MESDC1_SGTN_11</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1923</th>\n",
       "      <td>MESDC1_SGTN_12</td>\n",
       "      <td>TLNRD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1924</th>\n",
       "      <td>TP53_SGTN_01</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1925</th>\n",
       "      <td>TP53_SGTN_02</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1926 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       grna_id grna_target\n",
       "0     ACAA1_+_38178488.23-P1P2       ACAA1\n",
       "1     ACAA1_+_38178517.23-P1P2       ACAA1\n",
       "2     ACAA1_+_38178559.23-P1P2       ACAA1\n",
       "3     ACAA1_+_38178570.23-P1P2       ACAA1\n",
       "4     ACAA1_+_38178620.23-P1P2       ACAA1\n",
       "...                        ...         ...\n",
       "1921             SUZ12_SGTN_23       SUZ12\n",
       "1922            MESDC1_SGTN_11      TLNRD1\n",
       "1923            MESDC1_SGTN_12      TLNRD1\n",
       "1924              TP53_SGTN_01        TP53\n",
       "1925              TP53_SGTN_02        TP53\n",
       "\n",
       "[1926 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guide_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "894a7ca1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Gene mapping saved.\n"
     ]
    }
   ],
   "source": [
    "# Create a mapping from the Gene Symbol (which matches your guides)\n",
    "# to the Gene Index (which matches the matrix rownames)\n",
    "\n",
    "# Check which column in mdata['GEX'].var holds the symbol\n",
    "symbol_col = 'symbol' \n",
    "\n",
    "gene_map = mdata['GEX'].var[[symbol_col]].reset_index()\n",
    "gene_map.columns = ['response_id', 'grna_target'] # Rename for easy merging in R\n",
    "\n",
    "# Save it\n",
    "gene_map.to_csv(f\"{temp_dir}/gene_symbol_map.csv\", index=False)\n",
    "print(\"✅ Gene mapping saved.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7d3aac11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response_id</th>\n",
       "      <th>grna_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>ENSG00000241860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>WASH9P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000228463.10</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14707</th>\n",
       "      <td>ENSG00000275757.1</td>\n",
       "      <td>5_8S_rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14708</th>\n",
       "      <td>ENSG00000276256.1</td>\n",
       "      <td>ENSG00000276256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14709</th>\n",
       "      <td>ENSG00000273748.1</td>\n",
       "      <td>ENSG00000273748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14710</th>\n",
       "      <td>ENSG00000274917.1</td>\n",
       "      <td>RNA5-8SN5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14711</th>\n",
       "      <td>ENSG00000273730.1</td>\n",
       "      <td>5_8S_rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14712 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              response_id      grna_target\n",
       "0       ENSG00000227232.5           WASH7P\n",
       "1       ENSG00000238009.6  ENSG00000238009\n",
       "2       ENSG00000241860.7  ENSG00000241860\n",
       "3       ENSG00000279457.4           WASH9P\n",
       "4      ENSG00000228463.10  ENSG00000228463\n",
       "...                   ...              ...\n",
       "14707   ENSG00000275757.1        5_8S_rRNA\n",
       "14708   ENSG00000276256.1  ENSG00000276256\n",
       "14709   ENSG00000273748.1  ENSG00000273748\n",
       "14710   ENSG00000274917.1        RNA5-8SN5\n",
       "14711   ENSG00000273730.1        5_8S_rRNA\n",
       "\n",
       "[14712 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a8121d3",
   "metadata": {},
   "source": [
    "#### Import data to SCEPTRE (Rcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "96f67cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "56df1153",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading matrices into R...\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(sceptre)\n",
    "library(Matrix)\n",
    "\n",
    "# Define path\n",
    "data_dir <- \"/oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer\"\n",
    "\n",
    "# 1. LOAD MATRICES\n",
    "cat(\"Loading matrices into R...\\n\")\n",
    "response_mat <- readMM(file.path(data_dir, \"response_matrix.mtx\"))\n",
    "grna_mat <- readMM(file.path(data_dir, \"grna_matrix.mtx\"))\n",
    "\n",
    "# 2. ATTACH NAMES\n",
    "gene_names <- read.csv(file.path(data_dir, \"gene_names.csv\"), header=FALSE)$V1\n",
    "guide_names <- read.csv(file.path(data_dir, \"guide_names.csv\"), header=FALSE)$V1\n",
    "cell_names <- read.csv(file.path(data_dir, \"cell_names.csv\"), header=FALSE)$V1\n",
    "\n",
    "rownames(response_mat) <- gene_names\n",
    "colnames(response_mat) <- cell_names\n",
    "\n",
    "rownames(grna_mat) <- guide_names\n",
    "colnames(grna_mat) <- cell_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7683bd61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Renaming 'response_p_mito' to 'mt_percent'...\n",
      "[1] \"\\u2705 Sceptre object created!\"\n",
      "                           response_n_nonzero response_n_umis grna_n_nonzero\n",
      "GACAGTGCCCTCTATCTCACTTTA_2                693       1109.3333             24\n",
      "ACAGCAGACTGAGCCACTATTTCA_1                920       1671.6219             97\n",
      "CCTCCTGAATCCTGTACGCTTAAA_3               1281       2472.5923             26\n",
      "GTACGCAAATCATTCCCATTTGGA_3               1581       3231.2755             16\n",
      "CAACCACAAAACATCGGACCTTTC_1                759       1556.6205             96\n",
      "TGGTGGTAAATGTTGCACTATATA_3                515        858.7499             25\n",
      "                           grna_n_umis                  batch mt_percent\n",
      "GACAGTGCCCTCTATCTCACTTTA_2          70 300genes:eTN5_GEX_sub2   9.375001\n",
      "ACAGCAGACTGAGCCACTATTTCA_1         908 300genes:eTN5_GEX_sub1   9.092965\n",
      "CCTCCTGAATCCTGTACGCTTAAA_3         121 300genes:eTN5_GEX_sub3  15.483077\n",
      "GTACGCAAATCATTCCCATTTGGA_3          59 300genes:eTN5_GEX_sub3  12.533751\n",
      "CAACCACAAAACATCGGACCTTTC_1         380 300genes:eTN5_GEX_sub1  18.148289\n",
      "TGGTGGTAAATGTTGCACTATATA_3         142 300genes:eTN5_GEX_sub3   9.024746\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# 3. LOAD METADATA\n",
    "covariates <- read.csv(file.path(data_dir, \"covariates.csv\"), row.names=1)\n",
    "guide_mapping <- read.csv(file.path(data_dir, \"guide_mapping.csv\"))\n",
    "\n",
    "# ==============================================================================\n",
    "# FIX: RENAME 'response_p_mito' SO WE KEEP IT\n",
    "# Sceptre blocks the name 'response_p_mito' but doesn't calculate it.\n",
    "# We rename it to 'mt_percent' so we can use it as a covariate later.\n",
    "# ==============================================================================\n",
    "\n",
    "# 1. Rename response_p_mito if it exists\n",
    "if (\"response_p_mito\" %in% colnames(covariates)) {\n",
    "    cat(\"Renaming 'response_p_mito' to 'mt_percent'...\\n\")\n",
    "    names(covariates)[names(covariates) == \"response_p_mito\"] <- \"mt_percent\"\n",
    "}\n",
    "\n",
    "# 2. DROP OTHER RESERVED COLUMNS (which Sceptre DOES calculate)\n",
    "# We still drop n_umis and n_nonzero because Sceptre re-calculates those automatically.\n",
    "reserved_cols <- c(\"response_n_nonzero\", \"response_n_umis\", \"grna_n_nonzero\", \"grna_n_umis\")\n",
    "covariates <- covariates[, !(colnames(covariates) %in% reserved_cols), drop=FALSE]\n",
    "\n",
    "# 4. CREATE SCEPTRE OBJECT\n",
    "sceptre_object <- import_data(\n",
    "  response_matrix = response_mat,\n",
    "  grna_matrix = grna_mat,\n",
    "  grna_target_data_frame = guide_mapping,\n",
    "  extra_covariates = covariates,\n",
    "  moi = \"high\"\n",
    ")\n",
    "\n",
    "print(\"✅ Sceptre object created!\")\n",
    "print(head(sceptre_object@covariate_data_frame))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d7717ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Unique pairs to test: 287\"\n",
      "   grna_target        response_id\n",
      "1        ACAA1 ENSG00000060971.19\n",
      "7       ACVR1B ENSG00000135503.13\n",
      "13      ACVR2A ENSG00000121989.15\n",
      "19      ACVR2B ENSG00000114739.14\n",
      "25       ADCY2 ENSG00000078295.17\n",
      "31      AMOTL2 ENSG00000114019.15\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 1. Load the translation map\n",
    "gene_map <- read.csv(file.path(data_dir, \"gene_symbol_map.csv\"))\n",
    "\n",
    "# 2. Start with guide mapping (which has 'grna_target' as Symbol)\n",
    "discovery_pairs_highmoi <- guide_mapping\n",
    "\n",
    "# 3. Merge to attach the correct response_id (Gene ID)\n",
    "#    This matches 'grna_target' in guides to 'grna_target' in the gene map\n",
    "discovery_pairs_highmoi <- merge(discovery_pairs_highmoi, gene_map, by=\"grna_target\")\n",
    "\n",
    "# 4. Select ONLY the two columns Sceptre needs\n",
    "#    (We filter out non-targeting guides automatically via the merge)\n",
    "discovery_pairs_highmoi <- discovery_pairs_highmoi[, c(\"grna_target\", \"response_id\")]\n",
    "\n",
    "# 5. DEDUPLICATE\n",
    "#    Sceptre performs analysis at the *Target* level, not the *Guide* level.\n",
    "#    We only need one row per Target-Gene pair.\n",
    "discovery_pairs_highmoi <- unique(discovery_pairs_highmoi)\n",
    "\n",
    "print(paste(\"Unique pairs to test:\", nrow(discovery_pairs_highmoi)))\n",
    "print(head(discovery_pairs_highmoi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fc15db26",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "sceptre_object_highmoi <- set_analysis_parameters(\n",
    "  sceptre_object = sceptre_object,\n",
    "  discovery_pairs = discovery_pairs_highmoi,\n",
    "  side = \"both\", control_group = \"complement\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea8d519",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2c86a51e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/grna_count_distributions.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "png(file.path(data_dir, \"grna_count_distributions.png\"), width = 3000, height = 2000, res = 300)\n",
    "\n",
    "print(plot_grna_count_distributions(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  n_grnas_to_plot = 8\n",
    "))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"grna_count_distributions.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "13b84608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAgAElEQVR4nO3deVxU5f4H8O9szCDLyCqLsimmgIoIcr2uaaXXBZUozVDLbqnVbfdqkQrZzRa0rNTc6ld6y5uJpZW5XjXFVBRBQGXfBWSfgWHW8/vjeIkAc1DgPMDn/fIPzjMz53wOHD4enjkzI+I4jgAAgD1ioQMAAEDrUNAAAIxCQQMAMAoFDQDAKEl0dLTQGTrD+++/f+TIkdOnT+fm5g4YMMDCwkLoRLf1/vvvFxUV+fv784uxsbEDBw7s1auXsKmgqzh//vy2bdtOnDiRmJgolUrd3d2FTnRbfFQ/Pz8rKysiSkhIiI+P9/Pza7zD4cOHs7Ky+vfvzy+eOHEiJSVl4MCB/APd3NwcHBz4m7Zu3Xrs2LHRo0efP3/+zJkzjb8+XV1POYN+//33FQpF37599+/fP3LkSJ1OJ3Si23r//fcXLVpUWFjIL8bGxpaXlwsbCbqQ8+fP//zzzwMGDOA4burUqbt27RI60W2dP39+zZo1a9as4RcTEhK+++67pnc4fPjw4cOHGxdPnDjx448/8g9cv379xx9/zI+XlJSsWLHi/fff52/69ttvO2kHOp5U6ACdZ/r06QEBAY8//ridnV1qaurw4cOJ6Oeff05NTe3fv/+sWbPEYjERJScnHz161Gg0BgUFTZo0iYiio6MXLVp04MABa2vruXPnyuVyIjIYDHFxcfn5+SNHjhw3bhy/iejo6KVLl+7bt08ikURGRlpaWra6wla329Ts2bNjYmK2bdvW6o60uhUz81y6dGn//v2Nq5o3b97AgQNbzRMdHb148eIDBw54eXlNnDix5cqBWR4eHpGRkUSk1Wr37NnDf11YWPjTTz9ptdpp06bx56QNDQ179+7Ny8tzcnKaOXOms7PziRMniEilUmVkZEycODEwMJBf4bVr1w4fPmxlZTV79mx7e3si4u9pMBiSkpLGjRsXEhLS6gpb3W5T06ZN+89//vPKK694e3u3aR+nTJny008/rV+/3sLCYufOneHh4d9//33TO7QapsvpKWfQjYxGI8dx/EE2b968I0eO9OvX79ixYxEREUSUmpr68MMPOzo69u3b9+zZs/xDYmJioqKigoODS0pKpk6dSkQcx/3tb387ePCgs7PzsmXL3n333cZ7vvnmm/7+/vn5+fPmzbvdCltut5lly5b9+OOP169fb3UXWm7F/DwuLi5jxowZM2aMl5fX+vXrjUbj7fLExMQsX77c29u7b9++ra4c2KfT6RwdHYkoISEhIiJCIpFYWVmFhYWdP3+eiJ588smzZ88OGjSovr6eP9hOnDgxf/78ysrKkJCQp5566tixY0R07NixqVOnKhSKgoKCoKCgiooK/p5PP/10fn7+8OHDIyMjExMTW11hq9ttyt7e/qWXXlq1alVbd61Xr16TJk3izzZ27do1f/78ZndoGaYr6kFn0Lt27XJycjp06NCHH37o6emZkJBw/vz5N998s6GhISQk5LXXXisuLi4oKHB2dp48eXKfPn2aPnbVqlW+vr6hoaHffffduXPnDAZDUVHRkSNHiGjChAl+fn7//Oc/+RPPmJgYNze3UaNG9e7dm+O4litsdbtubm5NN2dra7ts2bKVK1fe7o+1ZluJj483M4+bm5ubm5tarR47duy2bdsGDx78J3mio6N9fHzOnDlzu5UDmzIzMz/66KOysrJz587xUxzR0dFjxoyRSqVENGrUqO3bt48cOTInJycsLGzmzJkSiaTxsePGjVu4cCERvfXWW7GxsZMmTXr33XfXrl07Z84cIsrPz9++ffvy5cuJ6P7771+0aBERPfbYY4cOHRo+fHjLFba63WZpX3rppYEDB6akpLR1N5988sm1a9d6enq6urq2nGpvde+6nJ71a2YymSorK/n/dTMyMmxsbBpvio2NVSgUkydPDgsLmzp16tChQzdu3Nh4q4eHB/+Fp6dnXl5ebm7ufffd1/Smxmlivtr4I1Kr1bZcYavbbRn1ueeeO3/+/MWLF1vdkWZbMT8PERmNxrlz50ZERPC/cn+Sh/+r809WDizTaDQXLlzIysqiP/6Ux4wZM3fuXCLatm1bXFycr6/vww8/nJeXx9/a7FAnotzc3EGDBvGDfn5+jfdsPKuwtraurq5udYWtbreZXr16RUVFvfHGGy1vksvl/EHL02q1TX9ZRo8enZOT89577z3xxBMtH9vq3nU5PegMOjIyMiAg4Jlnnhk8ePAPP/zg4+NTVlY2d+7cZv24fPny5cuXl5WV+fv7T58+3dPTk4iuXr3Kz8elpaV5enoaDIZr165xHCcSifLz84mI/0OyJZFI1GyFt9tuM3K5fPXq1a+//ro5u+bl5WVmHiJ66aWXHBwcoqKi+MU/ySMSidq6cmDBgAEDXnrpJSJyc3N74YUXLl686OPjY29v36zIhgwZsmfPHo7joqOjY2JiPv/8cyK6evUqfyt/qBORl5fX1atXhw0bxg82/m/dUssVtrrdlv7+97+vX7++d+/ezcYHDhzIp+IlJydPmzat6R0WLFiwfv36Xbt2NT6p/ud71+X0oILmKZXKt956a/Xq1YmJiePGjZs6der8+fPVavWRI0f279//ww8/JCQk+Pv7FxUVWVtbNz6xsGbNmpkzZ548edLV1TU0NJSI+vbtu3Dhwvvvv3/z5s1vvvnm7f7kb7lCT0/Plttt9bELFiyIjY29efPmHXdq9OjRZubZuXNnXFzctm3bjh49SkSBgYGhoaF/nsf8lQNrXnzxxc8++2zfvn1vvfXWrFmzKisrvb29k5KSAgICnnzyyaeffjo4OFipVJ46dWrWrFn8Q9LS0qKiory9vd95553t27cT0YoVK/7+979XVVXduHHj+PHjH3zwwe0213KFrW635QNlMtlbb70VGRn5yCOPNB2fM2fOunXrFi1aNGrUqEuXLmVmZi5YsKDpHfizHzPDdEU95TpoIho9ejR/ueWwYcOqq6t9fX2feOIJNze3nJwcmUy2cOFCFxcXZ2fn2tra3NxcpVK5bt06/r/0mJiY48ePX7lypX///u+++y4/XTBv3jyj0VhSUvLEE0/wz5LzJkyY0Pj12LFjXVxcWq4wIiKi2XZbRlUoFGKxOCgoyN3dfcKECc2ug262FbFYbGaeiooKJyenoqKi/Pz8/Pz8AQMGODg43C5P42Nvt3Jgk7u7O38hsEQiCQoKqqqqmjx58uOPP15aWlpWVjZkyJCwsDC5XO7s7JyXl1dZWTlnzhz+OeQTJ04MGTJk3LhxBQUFK1asGDVqFBH5+PhMmTIlPT3dycnpo48+arz02MvLy8vLi/+6X79+Pj4+LVfo7u7ecrvNovLzJwEBAQqFYsyYMU2vg5ZKpYsWLTIYDCUlJUOGDNmwYQP/K9z0gY0sLS1Hjx7duPstw3RFIryb3R2JRPguQY/An671nJM29qF6AAAYhclEAABGoaABABiFggYAYBQKGgCAUShoAABGddQLVaqqqlatWvXJJ5+Yef/iH0idQU4TyC64zdtaufJ0dXWdVCo2mTiJRLZ+/f1tXgUQEVFDKeXvJLGCBjwvdBTGmH88l/2Xqi9S7yBynniHe6alVaxf/5uNjQUR1dcbpk8fOGNG8zd7g3Z040dSXSOHMeTwF6GjmK2jzqDr6ura9NrKygtUfIDUWXezrf/85/qUKW7Llvk/+qjnV1+l3c0qgIiI9DVUfIBKDgqdgz3mH8+1qVR8gGrNeOefggLVxYtly5b5L1vmbzIZLl4svdeU8KeqLlHxAVJnCJ2jLTDFAQDAKBQ0AACjUNAAAIxCQQMAMKpbvd2oRmPU6Yx79tz6eBtPT9uRI12FjQQAcNe6VUFXVDRoNPpPPrlARMXF9X/9a9+vvkJBA1suXSrl36AsI6NK6CzAum5V0ERkY2Oxe/dEItq+/XpmpkboOADNjRy5y9W1l0QiqqnReXnZ3PkB0INhDhqgs/388+T4+LDFiwcLHQRYh4IGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGBUd3upN0BXpNUaq6q02dk1/KK7u7VcLhE2ErAABQ0gvMTE8oyMnC+/vEJEKpU+Pn5eaCje5wswxQHAhuee80tLi0hLi/D0tBY6C7ACBQ0AwCgUNAAAo1DQAACMQkEDADAKBQ0AwCgUNAAAo1DQAACMQkEDADCqq76SMDb2wu7d1/iv8/NVLe9QU6O7caPu6NE8fjEgwNHFxarz8gEA3LOuWtCFhSpv714REd5E9Pjj/215h6tXqy9eLHv11WNElJOj2rJl8mOPDerslAAA96CrFjQReXhYjx3rQkRisajVO0ya5LZhwyi6TYMDADAOc9AAAIxCQQMAMOpWQev1+qysrKysLJPJ1HibWq3Ozc1t6wgAALSLW3PQcXFxarVar9dnZmauXbtWJpOlp6fv2LHDz88vJydn9erVIpHInBFhdwYAoDu5dQY9Z86cp556asmSJfb29gUFBUS0d+/eJUuWLFy40MnJKTk52cwRAABoL7/PQdfW1mZlZVlbW3t4eBBRfn6+l5cXEfn4+OTm5po5QkRVVVXnz59PSkqSyWSduCMAAN3N75fZnThx4sqVK5WVlTqdTiqVGgwGfspCJpPV1tYSkTkjRJSdnf3JJ5/odDpLS8vO3x8AgG7j9zPosLCwqKioQYMGXbx4kYhsbGzq6uqIqKqqysHBwcwRIhoxYsTOnTvXrVvXWNkAAHAXbhV0dna2wWCoqanJzc11d3cnotDQ0CNHjmg0mrNnz44YMcLMEQAAaC+3CjopKemdd97ZvHnzjBkzfHx8iCgiIkKlUm3YsCEyMtLOzs7MEQAAaC+35qBnz549e/bspjdIJJL58+e3dQQAANoLXkkIAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMCorvShsYcO5U6Z8l3j4iuvDBEwDABAR+tiZ9ABAXZpaRFpaRGDBimFzgIA0LG60hk0EUmlYqXSgoikUny8FgB0c12soAG6opdf/u+vvxbyX5tMnLBhoAtBQQN0uKys6tBQh3HjXIho7txSoeNAl4GCBugMAwcqx451EToFdDFd7ElCAICeAwUNAMAoFDQAAKNQ0AAAjEJBAwAwCldxALClrk7//feZSUk3iUihkCxY4C90IhBMjyhotVqfklJ+9GgevzhhQj+pFH86AKOqqnSnT+empMjr6w1JSVUo6J6sRxR0drZq166UH39M5zi6cqWyuvofSqVc6FAAt7ViReDw4Q45OaqZM48KnQWE1CMKmoiWLx82a5an0ch5eHwjdBYAALPgL30AAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGIWCBgBgFAoauiG9Xp+VlZWVlWUymRoH1Wp1bm5uW0cABCQVOgBA+4uLi1Or1Xq9PjMzc+3atTKZLD09fceOHX5+fjk5OatXrxaJROaMCL0f0NPhDBq6oTlz5jz11FNLliyxt7cvKCggor179y5ZsmThwoVOTk7JyclmjgAICwUN3VNtbW1WVpa1tbWHhwcR5efne3l5EZGPj09ubq6ZI0RUVlZ29OjRM2fOWFhYdP5eQA+HKQ7onk6cOHHlypXKykqdTieVSg0GAz9lIZPJamtricicESIqKSmJi4vTarVyuVywnYGeCmfQ0D2FhYVFRUUNGjTo4sWLRGRjY1NXV0dEVVVVDg4OZo4Q0dChQzdt2hQTE6NSqYTaF+ixUNDQDWVnZxsMhpqamtzcXHd3dyIKDQ09cuSIRqM5e/bsiBEjzBwBEBYKGrqhpKSkd955Z/PmzTNmzPDx8SGiiIgIlUq1YcOGyMhIOzs7M0cAhIU5aOiGZs+ePXv27KYjEolk/vz5bR0BEBbOoAEAGIWCBgBgFAoaAIBRKGgAAEahoAEAGNWzruIwGjkiCgz8UiwWEdGIES7ffjtD6FAAAK3rWQXNi40NsbKSnTtXduRIqdBZAABuqycWdECAvY2NrLRUQ4SCBnYVFtZVVGhEolh+cf58v6++mipsJOhkXaCgt21L5jgiotTUCqGzAHSq3r0t4uPDiGjnzozCQr3QcaCzdYGCXrr0yKRJbhYWkszMWoVCInQcgM4jEomUSgsiUiikRCjoHqdrXMURGxu6ZcuYmTM9hQ4CANB5ukZBAwD0QChoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABglFToAANxZRkbNr7+WPfroAX7xxReDRo92FzYSdAKcQQN0ARUVWgcHWXCwMjhYmZRUkp+vEjoRdAacQQN0DT4+tpGRA4jo4MECobNAJ8EZNAAAo1DQAACMQkEDADAKBQ0AwCgUNAAAo1DQAACMQkEDADAKBQ0AwCgUNAAAo1DQAACMQkEDADCq574XR3p6TVJSWXDwTn7x2WcDFy0aImwkAICmem5B19cb3N2tli3zJ6IdO67fuFEndCIAgD/ouQVNRLa2srFjXYjo55/x9mAAwJweXdAAHefmzfrGd22uqdEKGwa6KBQ0QIf47rv0ZctOODoqiKiwsC48vJ/QiaDrwVUcAB1l0iT3+Piw+Pgwe3u50FmgS0JBAwAwCgUNAMAoFDQAAKNuPUmYn5+fnJwsEonGjBmjVCr5wcuXLxcUFIwaNcrR0dH8EQAAaBe3zqCPHz9uZ2dnaWkZFRVlMBiI6NSpU8ePH3d3d//Xv/6l1WrNHAEAgPZyq6CfeOKJ0aNHT5w4UalUlpeXE9GhQ4cWLFgQFBQ0fPjwhIQEM0cAAKC9/OE66IqKitraWicnJyK6efMmP2vRp0+fkpISM0eIKC0t7auvvtJoNFZWVp27LwAA3crvTxI2NDR88MEHixcvlkgk97LG3r17jxw5cujQofxUCQAA3J1bBW00Gj/44IPp06cHBATwI05OTvxcR2lpqYuLi5kjROTm5hYeHj558mTMSgMA3ItbBf3ZZ59ZWVnZ29unpaVpNBoieuihh7788suEhITExMTg4GAzRwAAoL3cmoP28PBQq9VJSUlE5OzsbGlpOX78eFtb28LCwqioKLlcTkTmjAAAQHu5VdAzZsxoedvw4cOHDx/e1hEAAGgXeCUhAACj8HajAF2MRmPIyam5eLGUXwwMdJZIRMJGgg6CggboYjIyaj/88MLWrYlElJenrq7+h1KJZ4C6JxQ0QNcTEzNi1ixPo5Hz8PhG6CzQgTAHDQDAKBQ0AACjMMVBRHT5csWPPxZ8/vkVfvHAgXA/PwdhIwEAoKCJiLRa44wZ/ebO7U9ECxee0GqNQicCAEBB/4+LS6+hQ+2JyMLint4rCgCgvWAOGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBSjL1T58svUq1cr+K9NJk7YMAAAgmC0oPfuTa+v1wwa1JuIOPQzAPRIjBY0Ec2Y4fnII95EtHFjmtBZAAAEgDloAABGoaABABiFggYAYBQKGgCAUShoAABGoaABABiFggYAYBQKGgCAUSho6IZMJlNeXl5eXl7TQbVanZubazKZ2jQCICB2X0kIcNfee+89Ozs7nU5XW1sbFRUlEonS09N37Njh5+eXk5OzevVqM0eE3g/o6XAGDd3Q8uXLlyxZ8sILL9TV1ZWVlRHR3r17lyxZsnDhQicnp+TkZDNHAISFgoZuSCwWE5Fer6+urlYqlUSUn5/v5eVFRD4+Prm5uWaOEFFZWdnRo0fPnDljYWHR6fsBPR2mOKDb2rx58/Tp0xUKBREZDAZ+ykImk9XW1po5QkQlJSVxcXFarVYulwu1I9Bj4QwauqdvvvnGyspq2rRp/KKNjU1dXR0RVVVVOTg4mDlCREOHDt20aVNMTIxKpRJkR6AnQ0FDN3To0KHi4uJFixY1joSGhh45ckSj0Zw9e3bEiBFmjgAICwUN3VBqamp1dfWqVatWrlxZUFBARBERESqVasOGDZGRkXZ2dmaOAAgLc9DQDb3yyivNRiQSyfz589s6AiAsnEEDADAKBQ0AwCgUNAAAo1DQAACMwpOEzen13NWrFSYTR0QymXjoUCehEwH8merqBv5wlUrFNjZ4uWO3goJurrxc8+qrx+VyiU5nIhIVFy8VOhFA64xGjoi8vLbxi6NHu58+/ZigiaCdYYqjFV9+OSE+Puzf/54gdBCAO7t27ZGionn/93/jhQ4C7Q8FDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKBQ0AACjUNAAAIxCQQMAMAoFDQDAKKnQAdhlMJiMRi47u4ZftLGROTn1EjYSwO1cvlxx5kyRSBTLL7799pioqL8IGwnuHQr6tgoL68rLNUFBXxKRVmt8/HG/7dsnCx0K4LaGD3f497/vJ6K33koUOgu0DxT0n3F0lCcmhhPR5s1Xb9wwCB0H4M9IpWKl0oKILCwwddlN4AcJAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwCgUNAMAoFDQAAKNQ0AAAjEJBAwAwSip0AABoZ2VlDTU1JVu3JvOL4eG+jo6WwkaCu4OCBuhucnJqRSKR0agjomPHikNCXFDQXRRbBZ2RUXX9aC0R3bypEToLQBc2c6bnCy/4E1Fo6A9CZ4G7x1ZBf/ll6uU96USUllY1d66n0HEAAITEVkH372+z9vMpRBQYGCd0FgAAgeEqDgAARqGgAQAYxdYUB0CXZjJxWq2RSKrVGuv1eqHjQJeHM2iAdlNQoNq8OYmItmxJeu21k0LHgS4PZ9Bmyc6uvXSpavHiw/zi4sXDgoL6CBsJ2CSTiYnoqafuyz+XxXFCp4EuDmfQZiktbVAoRI6OYkdH8aFD2Tk5NUInAoDuD2fQ5ho8uPdzz/kR0YkTN4TOAgA9As6gAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhU9UAejOysu1jz/+U69eUiJydLT85ZcIoRNBG6CgAbozvd4YHt7Pw8O6uLh+69brQseBtsEUR5tdu1a9cOFBe/tP+X8NDQahEwH8mXHjXKdP9xg/3kXoINBmOINuM6ORi4oKnDDBVas13n//TxwndCAA6KZuFfTRo0f37t2r1+u3b9/Oj1RWVn766accx/n6+s6bN8/MkR7C0VHu6Wnd0GAUOggAdGe3pjhCQkI+/fRTsfj3GY/du3dPnTp19erVmZmZeXl5Zo4AAEB7udXISqVSIpE0vSEtLW3YsGFEFBgYmJKSYuYIAAC0l9s+SajVamUyGRFZWVmp1WozR4joxIkTwcHBs2bNsre374QdAADorm77JKFcLtfr9TKZrK6uztra2swRIhozZszx48eLi4uDg4M7Zx8AALql255B+/v7JyUlEdHly5cDAgLMHCEiqVRqa2trbW3N4foGAIB7cKugL1y4sHLlysrKypUrV2ZmZhLRnDlzfvzxx1WrVg0YMMDT09PMEQAAaC+3pjhCQkJCQkKa3mBvbx8dHd3WEQAAaC94JSEAAKPwSsK7p9OZiGj69DixWEREw4Y5xcZOEDgTAHQjKOi7ZzJxRDR7dl8LC/GlS+WXLpUJnQgAuhUU9L2aOrWfQiERiSg9PV/oLADQrWAOGgCAUShoAABGoaABABiFggYAYBQKGgCAUShoAABGoaCheyopKbl+/Q+fkapWq3Nzc00mU5tGAASE66ChG4qLi7t06VJpaem2bdv4kfT09B07dvj5+eXk5KxevVokEpkzIuxeAOAMun2UlzcUF6u3bk3m/yUl3RQ6UY8WHh7+9ttvN23YvXv3LlmyZOHChU5OTsnJyWaOdCcNDUat1nj0aB7/Ly2tQuhEcGc4g24f+fnqior6AweuEVFycuXzz48YNsxJ6FDwu/z8fC8vLyLy8fHJzc0dNmyYOSNEVFZWlpycXFFRYWFhIWD+e1dSUl9bq3311WNEdPNmw9Sp/bdvnyx0KLgDFHS7GThQuWXLGCJ67bVzQmeB5gwGA39CLZPJamtrzRwhopKSkri4OK1WK5fLBUvfTuzt5QcPTiGizZuv3rhhEDoO3BmmOKBHsLGxqaurI6KqqioHBwczR4ho6NChmzZtiomJUalUgqWHngoFDT1CaGjokSNHNBrN2bNnR4wYYeYIgLBQ0NAN8R/h5uTk1PgRbhERESqVasOGDZGRkXZ2dmaOAAgLc9DQDbX8CDeJRDJ//vy2jgAIC2fQAACMwhk0QI9TV2coL9dcvFjKL/r4KO3sFMJGglahoAF6nOTkynPnyi5fLiGi0lLNrl3THn54oNChoBUo6PZXV2coLFQ3np74+zsoFPg+A1tmzPCIjQ0lokceOSZ0FrgtFEf7S0mp+vXXkh9+uE5EhYV1yclP+Pk5CB0KALoeFHSHeP55vyVLBhPR8JPnnDkAABbuSURBVOH7hM4CAF0VruIAAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFEoaAAARqGgAQAYhYIGAGAUChoAgFH4yKuOpVLpX3nlhK2tBRE5OVlu3PiA0IkA/qC8vOH77zMzM6v5xVdfDZZKcd7GChR0x9JqjQEB1s7OlkVFdfv35wgdB6C50lJNfn4lkc5g4HbvznrhhSAUNDtQ0B1u+nSPgQOVV65U7t9fKHQWgFZERg6YNs2jocG4e3eW0FngD/BfJQAAo1DQAACMQkEDADBK4Dlok4mrqdESkV4vI5IIGwYAgCkCF3RBgcrLaysRveU3c6yjr7BhAHoyjiMiysmpVSgkRGRpKXV1tRI4U48n/FUc1tay69cfKV5vr74gdBSAHkyrNRLRqFE7RSKRXm8KDXU7fvxRoUP1dMIXNACwIzExXKGQ/PRT/jff5AudBfAkIQAAq1DQAACMQkEDADAKc9CdpKJCW1XVsGLFKX7xL39xmzVrgLCRAIBxKOhOUlWlNZlMarWaiBISylUqHQoaAP4cCrrz9O4tf+ONQCJ6442EtLSKrVuT+fEpU7w8PGwFjQYALEJBCyA7W5Wfrzpw4BoRnTlT6uZmhYIGgJZQ0MKYMMH1nXdCiGjmzMNCZwEARuEqDgAARqGgAQAYhYIGAGAUChoAgFF4khAAmjMYTAaDqaqqgV+0tJQqFOgKAeAMGgCaS06u/PXXQnv7T/l/n3ySKHSiHgoFDQCt+Otf+xQVzSsqmvfYY/2FztJzoaABABiFeSWBNTQYMzOrL14s5RdHjOgjbB4AYAcKWmC5uaq1a3/76COJyUQFBWqj8VWxWCR0KIDfVVZqk5Nv7tlznV+cPNnb1tZC2Eg9BwpaeB98MPKBB9xra/WDB+8ROgtAcxkZtenpNXl5lUT0229lly8v9PNzEDpUT4GCBoA7iIwcsGTJYCIaPnyf0Fl6FjxJCADAKBQ0AACjMMUBAOaqqtJOnrzHwkJCRH372pw8OVfoRN0cChoAzGU0cqtWDe/Xzyozs3bdulSh43R/KGhWqFR6IpJI1vGLU6Z4Hzz4sKCJAFrh62s7cKBShGtBOwUKmi1nzoSJRHT4cOFvv1UJnQUABIaCZouHh5VYLHJ0VAgdBACEh6s4AAAYhYIGAGAUChoAgFGYg2ZRWlpVfHxRcPBOfvG110Lmzh0kbCQA6HwoaBbV1xt8fGyWLfMnok8/TS0rqxc6EcAfaLVGnc7Y+Da5Tk6WHh62wkbqllDQjLKzk48d60JEe/bkCJ0FoLmCgrqysvqHH95HRDU1unnz/DZufEDoUN0QChoA7oara6/4+DAi+uc/LyQmlr333nl+PCJiYP/+vQWN1n2goAHgnuTnqysq6lNSiono0KFCf38HFHR7QUEDwL0aNcr5nXdCiOjatWqhs3QruMwOAIBROIPuSpKSbhoMJv5rhULi7+8obB4A6FAoaNZlZdVkZaWcPl1ERD/8kOnoKJfJxFqt0dZWcf36U0KnA4AOhIJmXWWlLjDQJjhYSUTffWfcs+cBLy/rS5cqXn31vNDRAJpLT6957LGfZLJbc6fl5c/hU+rvBQq6CwgOdoyMHEBEK1aglIFpJhO3enXQ6NF91Gr9Qw8dFDpOl4eCBoD25Oys8PS0rq3VCx2kO8BVHAAAjEJBAwAwClMcAND+OI6IqKqqgX+SUCaTWFvLBM7UBaGguySVSqdS6fbsuc4v+vraBQY6CxsJoCm1Wk9Ejo4b+UV8CPLdQUF3ScXFmupq7SefXCCivDx1ePh9KGhgUEHBY2KxaPfu7B9+KMrOruEHHRwUSqVc2GBdBQq6q3J1tdy9eyIRxcYm839OArDpypWK+PjCoKAviai+3hAbO+GFF4KEDtU14ElCAOhwY8e6pqVFpKVFhIV5Cp2lK0FBd3lnz5Z+9NFFkSiW/3f4cK7QiQCgfaCgu4P58wfwpyf+/nZCZwGAdoM56O5AoZAqlRZElJOjmjz5O35QKhXr9a8ImgsA7gnOoLubbdvGpqVFnD49Q+ggAHCvcAbd3VhZSZVKC73eJHQQgFYkJpYfPFgYHR3PL168ON/bWylsJJahoAGg8xgM3OLF9z3yiA8RTZly0GjEJaJ/BgXdPdXXG00mbsWKU/zi4MEOCxf6CxsJgGdnJ/f0tCaiujrDxIn/4d882ttbefToo0JHYw4KuntqaDAQkVqtJqLExPJjx/J69br1sw4K6oMPXQYWmExcdPRwN7deJ0/e+PjjtAcf3MOPh4f7Ll0aKGw2RqCguy2xWPTGG4FE9Npr5w4fLuJfF56YWDFwoH1jQW/a9ICjo6WQKaFnGzSot5eXdWpqtZWV5MEH+xDRgQN5p08XjRzpyt/B19fO1tZC0IxCQkH3CCNGOHzxxXgiCgn53svLkv8ArZUrE2JjxxOhoO9VTEx8cbGaiNRqvTO5CR2nS7KxkfEfG7RvX86BA5lnzhQQ0Y0b9QcOhD/0kJfA4YSDy+x6nJEjnSMjB0RGDmj84Di4R3v2pGs09Y6OYpnMKHSW7mDu3P7x8WHx8WGDB/f0F17hDBqgHcye7RUa6pyQcPO/53CBY7vJyqp9+OH9/JmEVCouK3tW6ESdDQXdc2m1xmefPdqrl4yI+vWzWbdugtCJAP7AZOLWrg0OCXGqqtLOnHlE6DgCQEH3XEYjN3SoTe/e8pwcFd5iCdjk5KTw9LS2suqhTdVDdxt4s2d7ubtbffNN1pdfZgQH73QX279sOU2rNRJJhI4GcGdbtyZfvFjSuLh27Th7e4WAedodChpIozE6OiqWLfOX3uxFe0ijMYhEH/I3/fe/cyZM6CdsPACdzsRxHP8Zb+fO3fjmm6sKhZSISkvrx41zGT7cgYg++CA5KuovKGjohqyspGPHujTkyPL3kIWFOC0tgohmzz4qdC4AIiK1Wm8ycfy1/BkZNV5eNqtXBxHRggUnxo93nT9/ABF98EHyu++es7WVE1GfPlYvvzyCiHbsuLJ79zV+JbMMof7kIdg+3BUUNDQnEhH/5qXi/12GZzJxNTXaxjtYWEisrPAJzdCpJBIx/xlvr712rqKiYehQeyJqeqmowWCqq6sXi/VZWarExEqdzkhEv/ySa2FhfPhhbyKq2KIla4HS3y0UNNxWZaV269akX37JUal0mzZdbhyPiBi4Z0+YgMEAWvX8837u7laff57+22+lKSnFRJSSUvbooz7Tp3sQ0b93aIQO2GYoaLit6mqdRqNRq8U3b2osLaWZmY8S0ZtvJvz0U2Hj2ya8+mrwlCnegsYEaM7Fpdd7740koqtXaxoHdToTKejw4dys1HyZTLxmzRjhApoLBQ1/JjLSd9Qo56Skih9/LOBHamr0fftaPvJIPyLavPkq/xJnAPbpdCYi0mp1eXnle/bk/PbbDSK6caMuO7uaf8rRaOSefTZw0qRb89TjxvW1sBD4ciZhCjog4P9SU8v5r62tMZvZxbi69uL/Zty3L1foLABtM2GCa39vbVxcLn+S8fnn6aNGOb/9djARzZhxaNeulJ9/ziCiK1cqi4uXurhYCZtWsDPonTsnjBjhmJhYsXjxaaEywD1KTq48der4a6+dJKKamgZbW7lIJOI48vC49brE2lqtvb2lh4ctf383Nyv+VAVAWBKJmD/JOHSoSKs18O9PLZWKXn45YO7c/kTUr9/XzzxzWKGQ1tTo8vJqhg51IqKrVyvKyur5c8qKioaQEBc7OwURWVnJvvhiSkfkvNffFrVaXV5e7uHhIRa37Z13+E9msrbGr2sXZjSaXnrJnz/Q//rX/bt2TbC3l+/alfnFF+mvvnqMiDIyak0mTqGQEJFKpdu+fQp/oEskosBAZ2HDt+quj2foZkwmCgy0tbaWnTx5Q63W8m8AGR9fOGmS64IFvkQUFnZ44EArT0/r/Hz1li1phYUq/oETJ3q8/nooEe3Zcz07+/cZ8BdfDLqLs5N76sf09PQdO3b4+fnl5OSsXr1aJBLdy9qgK3JwuPXpGETk7t7L2dnSzs5iwADbgwenENHf/vbL/fe7/vOfw4jI23v366+fVCgkWq2xqkr30ksj+EcFBjrPnTtIqPxN4XiGpsLDvZydLWtrdQUFdfxbof7735nOzgr+Cj+RiCZP7hsc7Hj2bNkXX6TzEyYHDuT//HP2gAG9iWjt2nO9eom8vW2IaNeuzGeeGdrZBb13794lS5Z4e3tv3LgxOTl52LBh97I26Pa2bh3LH9Dz5h3nP+3l0qXypKSbvr633lXS21sp4CvBcDzD3ZFKRfzfkQcO5KWnV/IvqElJKV+//i/h4V4cR7t2ZVZXa/93Z7GNjbkfQSDiuLv/0MalS5du2rRJJBIdPHhQp9PNnDmTiIqLi3/77beqqqqXX345Nja21QfGxMT7+to4OCgqKhrOnCkNC/N0zne1qrG+rM9VBumJ6Kef8ocMsffwsCaiuLic6dM9LCwk169XFxdr7r/flYiOHy9ydbUaPLg3EX3/fe748a52dvKbNzXnzt3kv1O//VYmlYqCg534tQ0bZt+3r7XJRN9/nzNjhqdMJk5Lq755UzN+vCsRHT1a1K+f1X339SaifftyJ050VSrlN27UJyaWT53qQURnz5bK5ZKgIEciOnAgPyjIwd3dSq83HjiQP2uWl1gsunKlsrpaN3asCxEdPlzo7W3j66vk1/bAA242NhYFBerU1OopU/oS0enTJdbWssBAByLavz9v5EgnF5de9fWGX34pCA/3IhIlJVWo1YbRo/sQ0aFDhQMG2Pbvb0tE+/blPPRQPysraW6uOiOj+sEH+xLRqVM3eve2GDr01tr+8hdnZ2dLlUp/9GjR7NleRHTpUrlWaxw1qg8R/fJLwX339eb/V4+Ly/nb3/pZWkqzsmpzclQPPOBuoZG7Z3royVA4JIeIvv8+b8wYZ0dHy6oq7alTJTNnehJRQsJNo5ELDXUmop9/zvfzs/PyurW2adM85HJJenpNYWHdxIluRHT8eLGLi6Wfnx3/Yxo3ztXeXl5erjl79uaMGR5EdO5cWVFRXeOB4ehoyb8EpqpKGxLiIpWKicjeXnG7l5sPHz48JCTkbo7d1rTL8Sy5phxu61HrUP1LURrH0ciRt45AHM+dfzwT0cmTNxwc5AEB9kSkPWszyNqlyqUiR3qj445nsVgUEnLrxzR0qH2/ftYcxzV9Ot3Hp/fy5SNbPZBaHs/3VNBPP/30tm3biOjo0aMVFRVz5swhorS0tK+++orjuLq6urS0NIVCYW3d/OU7Go3BZOKIiOM4jcbYq5eU05NBx5nEJrmlhIi0WqNEIuJ/P+vrDZaWUpGIjEZOrzfxE5o6nVEkEvGvI9JoDHK5RCwWmUycVmu0tJQSkV5vMpk4uVxCRA0NRqlULJWK7ri2qqoyCwtbS0t5y7VxHMdfc9O4No4jjebW2gwGzmBoQ7ama5PJxBLJrbXxnxxoMHBGo4kPr9UaxeJba6uvNygUv69NoRBXVJQqlX3utDaT0cg1ru2P31iJSCQyGjm93qhQSDkT6es5IrKwFhFRQ4NBJpNIJKLGH9Od1tbKN5ZIZGFx61shFhs1GpWtrUNDw+3W1vzHRERisYj/vrUUHh4+d+7cuz6Am7nj8VxWVtZ456qqKrFYrFQqqdnxrDYqLGQiCWektn2jOuJ4bmioa2jQ2dr2Zvl41usbNJp6uVxp9trMPZ7pf5fW3ToC1UaZWCKxIJK0+XiWSk21tVXW1g5Nj2cLC4lEIjKZuA48nrl78PLLL6vVao7jvv322yNHjrS8w7Jly/bt23cvm+hkM2bMSEtLEzqFuWpqakaMGGEymYQOYq5z587NmzdP6BS3dcfjuanY2NgtW7Z0Sq67t2vXrujoaKFT3MGxY8cWL14sdIo7SE1NDQsL6/zt3tNT1aGhoUeOHNFoNGfPnh0xYkTLO9jb21tZCXwhYZu4uLhYWHSZT6gUi8Xu7u5Cp2gDhULh7MzixRu8Ox7PTSmVShsbm84JdtdsbGzs7Fj/1CiFQuHo6Ch0ijuwsLDo06dP52/3nqY4jEbj119/XVRU9NBDDwUFBbVjLIDOh+MZWHNPBQ0AAB2no14nYjQaN27cWF1dbWlp+cILL8jl8g7aUJtkZ2d/+OGHN27c2LRpE/+3dsuc5ox0WuCtW7dmZ2cbjcZRo0aFh4cTUWVl5aeffspxnK+v77x588wc6TT/+te/6uvrNRrN9OnTJ06cSESpqalff/21SCQKCwsbOXKkmSOM279//8WLFzmOW7p0qaurq1AxsrKytm3bJpFIiOjFF190dnZuaGh44oknfHx8iOjRRx8NDAxsNW0n53/llVcUCgURBQcH84cxa0eFwWBYtWoV//W1a9e+/vprhUJxd7HbWQfNbZ86deqLL77gOG7v3r0HDhzooK20VX19fV1d3ZtvvllaWsqPtMxpzkinuXHjBsdxRqPx+eefr6ys5Dhu48aNFy5c4DguJiYmNzfXzJFOYzAYOI6rr69/6qmn+JEXX3yxtrZWo9E8//zz/K3mjLCsvLx82bJlJpMpPT39nXfeETBJdXW1RqPhOO7XX3/duHEjx3Eajeb1119vep+WaTs//8svv9xshNmj4tq1a2vWrLmX2O2ro17Pmpqayl/nHxgYmJqa2kFbaStLS8tevXo1HWmZ05yRTuPi4kJEYrFYJBLxLz5OS0trDJOSkmLmSKfhz+a0Wq2XlxcR1dTUWFhY2NjYKBQKFxeXoqIic0Y6M/BduHbtWkBAgEgk8vX1zcvLEzCJUqnkT/FEIhH/nSeivLy8FStWfPTRR/xLgVqm7fz8IpEoKioqJiaG3xzLR8WpU6fGjx9/17HbPU9HTXGo1Wq+Cq2trVUqVQdt5d61zGnOSCc7ePDgoEGD+EtutVqtTCYjIisrq/LycjNHOtOaNWsyMjIee+wxavKt48OoVCqJRHLHkU4O3FZNd8pkMgkbhojq6ur27NmzYsUKIpLL5Vu2bLG2tt6/f//OnTuXLl3aMm3n54+OjraxscnIyFi3bt3HH3/M7FFhMpkSEhIWLFhw17HbPVJHnUFbW1vX19cTkVqtZvlqpJY5zRnpTAkJCfHx8U8//TS/KJfL9Xo9EdXV1fGvADJnpDOtXLnys88+27dvn0qlavzW8WFsbGzMGenkwG3VNLDg76mk1+vffffdxx9/nP9jSyQS8T/x8ePHZ2VlUWtpOz8//zP19fU1Go06nY7Zo+LKlSv33Xdf45NMdxG73SN11I/H398/KSmJiC5fvhwQENBBW7l3LXOaM9Jp0tPTv/3229dff50/I2Y8MMdxRqORiCz+R6lU6nQ6lUrV0NBQUlLi7u5uzkinBb47gwcPTklJ4TguPT2dn8kRCsdxH3744fjx4xtfH9zQ0MBxHBGlpKTwz/61TNvJ+U0mk06nI6Ly8nKDwcDyUXHy5Mlx48bdS+x2j9RRl9nxVz5UVlZaWVmxcxVHeXn5hg0bsrOz+/bt+8ADDzz44IMtc5oz0mmBn3/+eYlEYmtrS0TPPPNMv379KisrP/74Y5PJNGjQoMZrNu440jk0Gs3bb78tFou1Wu2UKVP4qzhSUlL4p7lnzpzJP81tzgjj9u/ff+HCBSJ69tlnBbyK48yZMx9//PHAgQOJyNfXd8GCBSkpKV988YVCoRCLxf/4xz/4S5Vapu3M/PX19W+++aalpaVOp4uMjOSfHWHwqNDr9c8999zmzZv52fy7jt2+cB00AACj8K7kAACMQkEDADAKBQ0AwCgUNAAAo1DQAACMQkEDADDq/wFpRtID/yQzcwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "plot_covariates(sceptre_object_highmoi, p_mito_threshold = 0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "195fcd72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178488.23-P1P2 (1 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACAA1_+_38178620.23-P1P2 (5 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACTL7B_-_111617995.23-P1P2 (10 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR1B_+_52345762.23-P1P2 (15 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2A_+_148602343.23-P1P2 (20 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_+_38495339.23-P1P2 (25 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ACVR2B_-_38495819.23-P1P2 (30 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ADCY2_-_7396429.23-P1P2 (35 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA AMOTL2_+_134093279.23-P1P2 (40 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ANP32E_+_150208342.23-P1P2 (45 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA APAF1_-_99039071.23-P1P2 (50 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_+_72766725.23-P1P2 (55 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARIH1_-_72766755.23-P1P2 (60 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ARTN_-_44399429.23-P1 (65 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASCC3_-_101329193.23-P1P2 (70 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ASH2L_-_37963059.23-P1P2 (75 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATMIN_+_81069661.23-P1P2 (80 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_+_111992112.23-P1P2 (85 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP5F1_-_111992129.23-P1P2 (90 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATP6AP1_-_153657075.23-P1P2 (95 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ATXN2L_-_28834668.23-P1P2 (100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAD_+_64052102.23-P1P2 (105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAG6_+_31620398.23-P1P2 (110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_+_49458268.23-P1P2 (115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BAX_-_49458323.23-P1P2 (120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BCAS2_-_115124045.23-P1P2 (125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BMPR1A_-_88516386.23-P1P2 (130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA BRD3_-_136932911.23-P1P2 (135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C17orf49_+_6918114.23-P2 (140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_+_31532436.23-P1P2 (145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C5orf22_-_31532650.23-P1P2 (150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACFD1_-_136325160.23-P1P2 (155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CACNG1_+_65040756.23-P1P2 (160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CAPN5_+_76778225.23-P1P2 (165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CASP3_+_185570560.23-P1P2 (170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_+_107384793.23-P1P2 (175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CBLL1_-_107384635.23-P1P2 (180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCM2_-_45039941.23-P1P2 (185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNC_-_100016267.23-P1P2 (190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCNT2_-_135676416.23-P1P2 (195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCT6B_+_33288381.23-P1P2 (200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_+_30366413.23-P1P2 (205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD2BP2_-_30366630.23-P1P2 (210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CD99L2_-_150067120.23-P1P2 (215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDH5_-_66400630.23-P1P2 (220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CDKN1A_-_36646538.23-P1P2 (225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CENPC_+_68411190.23-P1P2 (230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_+_30773828.23-P1P2 (235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C16orf93_-_30773391.23-P1P2 (240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHMP3_-_86790451.23-P1P2 (245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CHORDC1_-_89955894.23-P1P2 (250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNEP1R1_+_50059284.23-P1P2 (255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CNOT2_+_70637651.23-P1P2 (260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_+_83956339.23-P1P2 (265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS4_-_83956572.23-P1P2 (270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA COPS6_-_99686676.23-P1P2 (275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_-_31634088.23-P1P2 (280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CTR9_+_10772881.23-P1P2 (285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CUL3_-_225449667.23-P1P2 (290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_+_145150267.23-P1P2 (295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYC1_-_145149996.23-P1P2 (300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CYCS_+_25164869.23-P1P2 (305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DACH1_-_72441101.23-P1P2 (310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DBR1_+_137893673.23-P1P2 (315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DCTN2_-_57940910.23-P1P2 (320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_+_61851532.23-P1P2 (325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DDX42_-_61851812.23-P1P2 (330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DGCR8_-_20068158.23-P1P2 (335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNAJC17_-_41099215.23-P1P2 (340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DNM2_+_10829157.23-P1P2 (345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM1_+_49575018.23-P1P2 (350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_+_130700056.23-P1P2 (355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM2_-_130700076.23-P1P2 (360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DPM3_+_155113020.23-P1P2 (365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA DROSHA_-_31531982.23-P1P2 (370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_+_3985081.23-P1 (375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EEF2_-_3985443.23-P1 (380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EHMT2_-_31865199.23-P1P2 (385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3F_+_8009347.23-P1P2 (390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF3M_-_32605405.23-P1P2 (395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4E_+_99849831.23-P1P2 (400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184032364.23-P1P2 (405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G1_-_184033042.23-P1P2 (410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF4G2_+_10830447.23-P1 (415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EIF5A_-_7211217.23-P1P2 (420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TCEB3_-_24069975.23-P1P2 (425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EMC2_+_109456080.23-P1P2 (430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823220.23-P1P2 (435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EME2_-_1823252.23-P1P2 (440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EOMES_-_27763342.23-P1P2 (445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERCC2_+_45873802.23-P1P2 (450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ERO1L_+_53162425.23-P1P2 (455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETS1_+_128392044.23-P1P2 (460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_+_36134592.23-P1P2 (465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ETV2_-_36134789.23-P1P2 (470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EWSR1_-_29664347.23-P1P2 (475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EXOSC8_-_37574979.23-P1P2 (480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FBXW11_-_171433300.23-P1P2 (485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FEN1_+_61560782.23-P1P2 (490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_+_75532080.23-P1P2 (495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT11_-_75532461.23-P1P2 (500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FUT7_-_139926495.23-P1P2 (505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA G3BP1_-_151151499.23-P1P2 (510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GCK_+_44229021.23-P1P2 (515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOLT1B_+_21654777.23-P1P2 (520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_+_45000529.23-P1P2 (525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GOSR2_-_45000948.23-P1P2 (530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPI_+_34856284.23-P1P2 (535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GPR107_-_132816044.23-P1P2 (540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GSPT1_-_12009714.23-P1P2 (545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A1_+_81687215.23-P1P2 (550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_+_59949355.23-P1P2 (555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2A2_-_59949657.23-P1P2 (560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF2H4_-_30876185.23-P1P2 (565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GTF3C2_+_27579604.23-P1P2 (570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HAND1_+_153857732.23-P1P2 (575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HDAC2_-_114292253.23-P1P2 (580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_+_81399312.23-P1P2 (585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HGF_-_81399351.23-P1P2 (590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HINT2_+_35814995.23-P1P2 (595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNF4A_+_43029971.23-P1P2 (600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPM_-_8509885.23-P1 (605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HNRNPU_+_245027753.23-P1P2 (610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_+_43702292.23-P1P2 (615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSD17B12_-_43702305.23-P1P2 (620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HSPA5_-_128003534.23-P1P2 (625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA HUWE1_-_53710540.23-P1P2 (630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_-_30007588.23-P1P2 (635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INTS10_+_19675000.23-P1P2 (640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_+_65903046.23-P1P2 (645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VWA9_-_65903523.23-P1P2 (650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C7orf26_+_6630358.23-P1P2 (655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA IPO8_+_30848771.23-P1P2 (660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ITGB1BP1_+_9563543.23-P1P2 (665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSFA2_+_182756743.23-P1P2 (670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_+_109914946.23-P1P2 (675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KCTD10_-_109915112.23-P1P2 (680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM1A_-_23346409.23-P1P2 (685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDM3B_-_137688431.23-P1P2 (690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KDR_-_55991659.23-P1P2 (695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA0232_+_6784651.23-P1P2 (700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153706791.23-P1P2 (705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAGE3_+_153707248.23-P1P2 (710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LAMTOR4_-_99746921.23-P1P2 (715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LARP4_-_50794864.23-P1P2 (720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LATS2_-_21635336.23-P1P2 (725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LEMD3_+_65563441.23-P1P2 (730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_+_54960428.23-P1P2 (735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LENG8_-_54960616.23-P1P2 (740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LHFPL2_-_77944519.23-P1P2 (745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LSM12_-_42143603.23-P1P2 (750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MANBAL_+_35918136.23-P1P2 (755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MAU2_+_19431732.23-P1P2 (760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_+_47807678.23-P1P2 (765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MBD1_-_47807877.23-P1P2 (770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEAF6_-_37980319.23-P1P2 (775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED1_-_37607140.23-P1P2 (780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED12_+_70338541.23-P1P2 (785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED15_+_20862189.23-P1 (790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_+_57479335.23-P1P2 (795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MED19_-_57479437.23-P1P2 (800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MEMO1_-_32264882.23-P1 (805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MIXL1_+_226411425.23-P1P2 (810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL28_+_420487.23-P1P2 (815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL32_+_42972018.23-P1P2 (820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_+_179322181.23-P1P2 (825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MRPL47_-_179322213.23-P1P2 (830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA LUZP6_-_135612750.23-P1P2 (835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA20_-_19998023.23-P1P2 (840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NAA25_+_112546538.23-P1P2 (845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCBP1_+_100396082.23-P1P2 (850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_+_172428630.23-P1P2 (855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NCEH1_-_172428841.23-P1P2 (860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NEDD8_-_24701363.23-P1P2 (865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NF2_+_29999635.23-P1P2 (870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NFIB_+_14314559.23-P1P2 (875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_+_139418329.23-P2 (880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139418349.23-P2 (885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NOTCH1_-_139440503.23-P1 (890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUFIP2_+_27621073.23-P1P2 (895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA NUP93_-_56764090.23-P1P2 (900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OGT_+_70752948.23-P1P2 (905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_+_59383170.23-P1P2 (910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA OSBP_-_59383348.23-P1P2 (915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAM16_-_4401236.23-P1P2 (920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PAXBP1_-_34143614.23-P1P2 (925 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PBLD_-_70092436.23-P1 (930 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PCBP2_+_53835614.23-P2 (935 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_+_94292752.23-P2 (940 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PEG10_-_94293027.23-P2 (945 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PGRMC2_+_129208913.23-P1P2 (950 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_-_178866211.23-P1P2 (955 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PITPNC1_-_65373433.23-P1P2 (960 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLCG1_+_39766176.23-P1P2 (965 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_+_151042273.23-P2 (970 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLEKHG1_-_151042422.23-P2 (975 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLRG1_-_155471154.23-P1P2 (980 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PLXNA2_-_208417275.23-P1P2 (985 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLE4_-_75185725.23-P1P2 (990 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR2M_+_57998962.23-P1P2 (995 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940045.23-P1P2 (1000 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3H_+_41940414.23-P1P2 (1005 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA POLR3K_-_103144.23-P1P2 (1010 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPHLN1_-_42720007.23-P1P2 (1015 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPIL4_+_149867146.23-P1P2 (1020 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PPP2R1A_+_52693317.23-P1 (1025 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_+_53195217.23-P1P2 (1030 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKCD_-_53195332.23-P1P2 (1035 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRKD1_-_30396996.23-P1P2 (1040 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRMT5_+_23398554.23-P1 (1045 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA C2orf53_-_27362230.23-P1P2 (1050 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PRSS8_+_31147010.23-P1P2 (1055 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_+_36909062.23-P1P2 (1060 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMB3_-_36909230.23-P1P2 (1065 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMD4_-_151227289.23-P1P2 (1070 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PSMG4_-_3259227.23-P1P2 (1075 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTAR1_+_72374835.23-P1P2 (1080 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTBP1_+_797463.23-P1P2 (1085 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_+_89623388.23-P1P2 (1090 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTEN_-_89623526.23-P1P2 (1095 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PTMS_-_6875591.23-P1P2 (1100 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PYGO2_-_154934135.23-P1P2 (1105 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB18_-_27793292.23-P1P2 (1110 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAB6A_+_73471983.23-P1P2 (1115 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_+_76251918.23-P1P2 (1120 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RABGGTB_-_76252008.23-P1P2 (1125 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA GNB2L1_-_180670840.23-P1P2 (1130 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RAD21_-_117886635.23-P1P2 (1135 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RANGRF_+_8192257.23-P1P2 (1140 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBBP5_+_205091050.23-P1P2 (1145 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM14-RBM4_+_66406280.23-P2 (1150 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM33_+_155437802.23-P1P2 (1155 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM4_+_66406319.23-P2 (1160 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_+_114271352.23-P1P2 (1165 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBM7_-_114271700.23-P1P2 (1170 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RBX1_-_41347852.23-P1P2 (1175 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1551_+_32112399.23-P1P2 (1180 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RFX3_-_3525750.23-P1P2 (1185 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RGP1_+_35749328.23-P1P2 (1190 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_+_95067102.23-P1P2 (1195 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RHOBTB3_-_95067164.23-P1P2 (1200 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RIBC1_-_53449606.23-P1P2 (1205 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA KIAA1432_+_5629173.23-P1P2 (1210 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RLF_+_40627522.23-P1P2 (1215 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF115_-_145611148.23-P1P2 (1220 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_+_104296142.23-P1P2 (1225 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF20_-_104296250.23-P1P2 (1230 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RNF40_-_30773681.23-P1P2 (1235 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPL37A_-_217363600.23-P1P2 (1240 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA RPS15A_-_18801590.23-P1P2 (1245 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAV1_+_51134992.23-P1P2 (1250 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_+_33104038.23-P1P2 (1255 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF4_-_33104343.23-P1P2 (1260 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SCAF8_+_155054976.23-P1P2 (1265 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA EPT1_+_26569050.23-P1P2 (1270 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERF2_+_44069659.23-P2 (1275 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SERPING1_+_57365137.23-P1P2 (1280 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_+_122242041.23-P1P2 (1285 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETD1B_-_122242294.23-P1P2 (1290 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SETDB1_-_150898815.23-P1P2 (1295 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SF1_-_64545937.23-P1P2 (1300 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CCDC101_-_28565268.23-P1P2 (1305 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SKA3_+_21750519.23-P1P2 (1310 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_+_78591497.23-P1P2 (1315 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_-_78591737.23-P1P2 (1320 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC29A1_-_44187438.23-P2 (1325 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SLC35C1_-_45826711.23-P1P2 (1330 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD1_+_146403428.23-P1P2 (1335 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD3_+_67358229.23-P1P2 (1340 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_+_48556623.23-P1P2 (1345 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD4_-_48556657.23-P1P2 (1350 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMAD5_-_135468694.23-P1P2 (1355 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA4_-_11071741.23-P1P2 (1360 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCA5_+_144435243.23-P1P2 (1365 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCB1_+_24129228.23-P1 (1370 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_+_47822975.23-P1P2 (1375 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCC1_-_47822893.23-P1P2 (1380 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCD3_-_150974212.23-P1 (1385 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMARCE1_-_38803669.23-P1P2 (1390 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMC2_-_106856631.23-P1P2 (1395 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SMG5_+_156252537.23-P1P2 (1400 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_+_103852.23-P1P2 (1405 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP25_-_104242.23-P1P2 (1410 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNRNP27_-_70121181.23-P1P2 (1415 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SNUPN_-_75917960.23-P1P2 (1420 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SOX4_+_21594213.23-P1 (1425 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPIN1_+_91003737.23-P1P2 (1430 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_+_2328680.23-P1P2 (1435 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPPL2B_-_2328912.23-P1P2 (1440 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SPTY2D1_-_18655846.23-P1P2 (1445 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP19_+_112197188.23-P1P2 (1450 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP54_-_35452244.23-P1P2 (1455 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRP68_+_74068540.23-P1P2 (1460 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_+_126138730.23-P1P2 (1465 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPR_-_126138742.23-P1P2 (1470 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRPRB_-_133524719.23-P1P2 (1475 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRRM2_-_2802714.23-P1P2 (1480 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SRSF6_+_42086993.23-P1 (1485 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSB_+_170655450.23-P1P2 (1490 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_+_57102902.23-P1P2 (1495 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SSRP1_-_57103355.23-P1P2 (1500 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STAG2_-_123095658.23-P1 (1505 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA STK40_+_36851507.23-P1P2 (1510 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUPT20H_+_37633798.23-P1P2 (1515 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SUZ12_+_30264157.23-P1P2 (1520 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_+_64901973.23-P1P2 (1525 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SYVN1_-_64901981.23-P1P2 (1530 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TADA2B_-_7045121.23-P1P2 (1535 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF10_+_6633406.23-P1P2 (1540 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF2_+_120845032.23-P1P2 (1545 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF3_-_7860540.23-P1P2 (1550 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_+_60640814.23-P1P2 (1555 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF4_-_60640498.23-P1P2 (1560 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAF6L_-_62538954.23-P1P2 (1565 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TAOK1_-_27717907.23-P1P2 (1570 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA FAM208A_+_56717273.23-P1P2 (1575 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBC1D9_+_141677481.23-P1P2 (1580 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_+_170863524.23-P1P2 (1585 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBP_-_170863481.23-P1P2 (1590 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TBX3_-_115121998.23-P1P2 (1595 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA T_-_166581333.23-P1P2 (1600 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TGFBR2_+_30648526.23-P1P2 (1605 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA THBD_+_23029196.23-P2 (1610 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_+_2977592.23-P1P2 (1615 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TLE6_-_2977643.23-P1P2 (1620 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA MESDC1_-_81293340.23-P1P2 (1625 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMEM41B_+_9336156.23-P1P2 (1630 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TMTC3_-_88536127.23-P1P2 (1635 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TOP1_+_39657538.23-P1P2 (1640 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_+_7590531.23-P1P2 (1645 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_-_7590704.23-P1P2 (1650 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRIM24_-_138145308.23-P1P2 (1655 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TRRAP_-_98476465.23-P1P2 (1660 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TSTD2_+_100395594.23-P1P2 (1665 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UBE2E1_+_23847541.23-P1P2 (1670 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_+_38924011.23-P1P2 (1675 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UFM1_-_38924421.23-P1P2 (1680 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UPF2_+_12084799.23-P1P2 (1685 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA UXT_-_47518505.23-P1P2 (1690 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VHL_-_10183525.23-P1P2 (1695 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA VPS54_+_64246104.23-P1P2 (1700 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_+_28821698.23-P1P2 (1705 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAC_-_28822074.23-P1P2 (1710 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WAPAL_+_88281533.23-P1P2 (1715 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR12_-_203776383.23-P1P2 (1720 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR77_+_111991860.23-P1P2 (1725 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WIZ_-_15543016.23-P1P2 (1730 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_+_9983328.23-P1P2 (1735 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WWC3_-_9983510.23-P1P2 (1740 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XPO5_-_43543670.23-P1P2 (1745 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA XRN1_-_142166648.23-P1P2 (1750 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA YARS2_+_32908787.23-P1P2 (1755 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB22_+_33285454.23-P1P2 (1760 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_+_6640093.23-P1P2 (1765 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZBTB48_-_6640138.23-P1P2 (1770 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZC3H8_-_113012435.23-P1P2 (1775 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA ZNF281_+_200379151.23-P1P2 (1780 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00117 (1785 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00233 (1790 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00518 (1795 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_00656 (1800 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01226 (1805 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_01730 (1810 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02009 (1815 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02288 (1820 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_02777 (1825 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03026 (1830 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03252 (1835 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA non-targeting_03620 (1840 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA CSNK2B_SGTN_04 (1845 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA INO80E_SGTN_25 (1850 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA PIK3CA_SGTN_26 (1855 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1044 (1860 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1414 (1865 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1554 (1870 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_1785 (1875 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2009 (1880 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2271 (1885 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2425 (1890 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_2614 (1895 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_293 (1900 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_541 (1905 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_725 (1910 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA SAFE_TARGETING_860 (1915 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA WDR61_SGTN_14 (1920 of 1926)\n",
      "Performing gRNA-to-cell assignments for gRNA TP53_SGTN_01 (1925 of 1926)\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- assign_grnas(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  method = \"mixture\",\n",
    "  parallel = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e113b305",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     cell_id               grna_target value\n",
      "1 GACAGTGCCCTCTATCTCACTTTA_2    BAD_+_64052144.23-P1P2     1\n",
      "2 GACAGTGCCCTCTATCTCACTTTA_2   DNM2_-_10828821.23-P1P2     1\n",
      "3 GACAGTGCCCTCTATCTCACTTTA_2  TCEB3_-_24070127.23-P1P2     1\n",
      "4 GACAGTGCCCTCTATCTCACTTTA_2    GPI_+_34856284.23-P1P2     1\n",
      "5 GACAGTGCCCTCTATCTCACTTTA_2            CSNK2B_SGTN_05     1\n",
      "6 ACAGCAGACTGAGCCACTATTTCA_1 SMAD1_+_146403220.23-P1P2     1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "⚠️ Matrix missing colnames. Fetching from sceptre_object...\n",
       "✅ Saved long CSV to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/grna_assignments_long.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(Matrix)\n",
    "\n",
    "# 1. Get the assignment matrix\n",
    "A <- get_grna_assignments(sceptre_object_highmoi)\n",
    "\n",
    "# 2. Ensure it is a numeric sparse matrix (1/0)\n",
    "if (inherits(A, \"Matrix\") || is.matrix(A)) {\n",
    "    A_num <- if (inherits(A, \"Matrix\")) drop0(A * 1) else (A * 1)\n",
    "} else {\n",
    "    stop(\"Output is not a matrix.\")\n",
    "}\n",
    "\n",
    "# --- FIX: ROBUST NAME RECOVERY ---\n",
    "# We retrieve names using Sceptre's standard metadata accessors\n",
    "# instead of touching internal slots like @grna_matrix.\n",
    "\n",
    "# Try to get cell names from the matrix first\n",
    "current_cell_names <- colnames(A_num)\n",
    "current_guide_names <- rownames(A_num)\n",
    "\n",
    "# If missing, grab them from the main object's covariate data frame\n",
    "if (is.null(current_cell_names)) {\n",
    "    message(\"⚠️ Matrix missing colnames. Fetching from sceptre_object...\")\n",
    "    # The covariate data frame ALWAYS has the cell IDs as rownames\n",
    "    current_cell_names <- rownames(sceptre_object_highmoi@covariate_data_frame)\n",
    "    \n",
    "    # Safety check: lengths must match\n",
    "    if (length(current_cell_names) != ncol(A_num)) {\n",
    "        stop(paste(\"Mismatch! Matrix has\", ncol(A_num), \"cols but object has\", length(current_cell_names), \"cells.\"))\n",
    "    }\n",
    "    colnames(A_num) <- current_cell_names\n",
    "}\n",
    "\n",
    "# If guide names are missing, they should match the unique values in your guide mapping\n",
    "if (is.null(current_guide_names)) {\n",
    "    message(\"⚠️ Matrix missing rownames. Fetching from guide mapping...\")\n",
    "    # Get unique targets from the mapping table you loaded earlier\n",
    "    # Note: This assumes the matrix rows are sorted/aligned to these unique targets.\n",
    "    # In 'mixture' method, rows usually correspond to the 'grna_target' labels.\n",
    "    \n",
    "    # SAFE FALLBACK: If we can't be 100% sure of order, we warn the user.\n",
    "    # But usually, they match the 'grna_target' column from your input.\n",
    "    unique_targets <- sort(unique(guide_mapping$grna_target))\n",
    "    \n",
    "    if (length(unique_targets) == nrow(A_num)) {\n",
    "         rownames(A_num) <- unique_targets\n",
    "    } else {\n",
    "         warning(\"Could not automatically align target names. Using generic IDs.\")\n",
    "         rownames(A_num) <- paste0(\"target_\", seq_len(nrow(A_num)))\n",
    "    }\n",
    "}\n",
    "# ---------------------------------\n",
    "\n",
    "# 3. Convert to Long Format\n",
    "s <- summary(A_num)\n",
    "assign_long <- data.frame(\n",
    "    cell_id = colnames(A_num)[s$j],\n",
    "    grna_target = rownames(A_num)[s$i],\n",
    "    value = s$x\n",
    ")\n",
    "\n",
    "# 4. Save\n",
    "out_path <- file.path(data_dir, \"grna_assignments_long.csv\")\n",
    "write.csv(assign_long, out_path, row.names = FALSE)\n",
    "\n",
    "message(\"✅ Saved long CSV to: \", out_path)\n",
    "print(head(assign_long))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f9f6b6d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/MOI.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "png(file.path(data_dir, \"MOI.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"MOI.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c65cd694",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/QC.png\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_qc(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  p_mito_threshold = 0.25\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"QC.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"QC.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e2408f35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Constructing negative control pairs. ✓\n",
      "Running precomputation on response ENSG00000163346.17 (1 of 233)\n",
      "Running precomputation on response ENSG00000253767.3 (5 of 233)\n",
      "Running precomputation on response ENSG00000234287.1 (10 of 233)\n",
      "Running precomputation on response ENSG00000115808.12 (15 of 233)\n",
      "Running precomputation on response ENSG00000136699.20 (20 of 233)\n",
      "Running precomputation on response ENSG00000117448.14 (25 of 233)\n",
      "Running precomputation on response ENSG00000264324.1 (30 of 233)\n",
      "Running precomputation on response ENSG00000113456.20 (35 of 233)\n",
      "Running precomputation on response ENSG00000178971.17 (40 of 233)\n",
      "Running precomputation on response ENSG00000177663.15 (45 of 233)\n",
      "Running precomputation on response ENSG00000171204.13 (50 of 233)\n",
      "Running precomputation on response ENSG00000131002.15 (55 of 233)\n",
      "Running precomputation on response ENSG00000165061.15 (60 of 233)\n",
      "Running precomputation on response ENSG00000181827.16 (65 of 233)\n",
      "Running precomputation on response ENSG00000291064.1 (70 of 233)\n",
      "Running precomputation on response ENSG00000290192.1 (75 of 233)\n",
      "Running precomputation on response ENSG00000171132.14 (80 of 233)\n",
      "Running precomputation on response ENSG00000067225.21 (85 of 233)\n",
      "Running precomputation on response ENSG00000007129.18 (90 of 233)\n",
      "Running precomputation on response ENSG00000177156.11 (95 of 233)\n",
      "Running precomputation on response ENSG00000156482.11 (100 of 233)\n",
      "Running precomputation on response ENSG00000258526.8 (105 of 233)\n",
      "Running precomputation on response ENSG00000230409.3 (110 of 233)\n",
      "Running precomputation on response ENSG00000047410.14 (115 of 233)\n",
      "Running precomputation on response ENSG00000196839.14 (120 of 233)\n",
      "Running precomputation on response ENSG00000106624.11 (125 of 233)\n",
      "Running precomputation on response ENSG00000163382.13 (130 of 233)\n",
      "Running precomputation on response ENSG00000126821.8 (135 of 233)\n",
      "Running precomputation on response ENSG00000164823.11 (140 of 233)\n",
      "Running precomputation on response ENSG00000273748.1 (145 of 233)\n",
      "Running precomputation on response ENSG00000181264.9 (150 of 233)\n",
      "Running precomputation on response ENSG00000128487.19 (155 of 233)\n",
      "Running precomputation on response ENSG00000172183.16 (160 of 233)\n",
      "Running precomputation on response ENSG00000186260.17 (165 of 233)\n",
      "Running precomputation on response ENSG00000137942.18 (170 of 233)\n",
      "Running precomputation on response ENSG00000175155.11 (175 of 233)\n",
      "Running precomputation on response ENSG00000136997.22 (180 of 233)\n",
      "Running precomputation on response ENSG00000137992.15 (185 of 233)\n",
      "Running precomputation on response ENSG00000189120.5 (190 of 233)\n",
      "Running precomputation on response ENSG00000255245.4 (195 of 233)\n",
      "Running precomputation on response ENSG00000086289.13 (200 of 233)\n",
      "Running precomputation on response ENSG00000182359.15 (205 of 233)\n",
      "Running precomputation on response ENSG00000189058.9 (210 of 233)\n",
      "Running precomputation on response ENSG00000186834.4 (215 of 233)\n",
      "Running precomputation on response ENSG00000116138.13 (220 of 233)\n",
      "Running precomputation on response ENSG00000101230.6 (225 of 233)\n",
      "Running precomputation on response ENSG00000171724.3 (230 of 233)\n",
      "Analyzing pairs containing gRNA group non-targeting_00184&non-targeting_00259&non-targeting_02159&non-targeting_02571&non-targeting_03029&non-targeting_03081 (1 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_01226&non-targeting_01509&non-targeting_01732&non-targeting_02543&non-targeting_02837&non-targeting_03026 (5 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00518&non-targeting_03081&non-targeting_03342&non-targeting_03427&non-targeting_03635&non-targeting_03664 (10 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_01509&non-targeting_01812&non-targeting_02025&non-targeting_02677&non-targeting_02777&non-targeting_03384 (15 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_00324&non-targeting_00390&non-targeting_00629&non-targeting_01934&non-targeting_03180 (20 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00518&non-targeting_00656&non-targeting_01730&non-targeting_02574&non-targeting_03029&non-targeting_03158 (25 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00213&non-targeting_02543&non-targeting_02777&non-targeting_02814&non-targeting_03252&non-targeting_03483 (30 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00196&non-targeting_00918&non-targeting_01730&non-targeting_02025&non-targeting_02911&non-targeting_03384 (35 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00233&non-targeting_00390&non-targeting_01601&non-targeting_02009&non-targeting_02288&non-targeting_03029 (40 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00014&non-targeting_02159&non-targeting_02193&non-targeting_02677&non-targeting_02962&non-targeting_03029 (45 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00390&non-targeting_01732&non-targeting_02193&non-targeting_03081&non-targeting_03384 (50 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00629&non-targeting_01185&non-targeting_01732&non-targeting_02089&non-targeting_02677&non-targeting_03483 (55 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_01185&non-targeting_02543&non-targeting_02677&non-targeting_03029&non-targeting_03342&non-targeting_03664 (60 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00614&non-targeting_01185&non-targeting_01226&non-targeting_02089&non-targeting_02159&non-targeting_02962 (65 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00355&non-targeting_00656&non-targeting_01730&non-targeting_02574&non-targeting_02911&non-targeting_03664 (70 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00355&non-targeting_00518&non-targeting_00555&non-targeting_00794&non-targeting_01185&non-targeting_02159 (75 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00196&non-targeting_00213&non-targeting_00518&non-targeting_02574&non-targeting_02837 (80 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00614&non-targeting_01599&non-targeting_02009&non-targeting_02543&non-targeting_02837&non-targeting_03081 (85 of 91)\n",
      "Analyzing pairs containing gRNA group non-targeting_00117&non-targeting_00555&non-targeting_01226&non-targeting_01812&non-targeting_02911&non-targeting_03180 (90 of 91)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/calibration_check.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "# Calibration check\n",
    "\n",
    "\n",
    "sceptre_object_highmoi <- run_calibration_check(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"calibration_check.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"calibration_check.png\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e21113bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          response_id\n",
      "               <char>\n",
      "1: ENSG00000175984.15\n",
      "2: ENSG00000174695.10\n",
      "3: ENSG00000171570.11\n",
      "4:  ENSG00000278390.7\n",
      "5: ENSG00000137992.15\n",
      "6:  ENSG00000160213.9\n",
      "                                                                                                               grna_target\n",
      "                                                                                                                    <char>\n",
      "1: non-targeting_00233&non-targeting_02193&non-targeting_02571&non-targeting_02911&non-targeting_03081&non-targeting_03180\n",
      "2: non-targeting_00117&non-targeting_01601&non-targeting_01732&non-targeting_02962&non-targeting_03483&non-targeting_03664\n",
      "3: non-targeting_00117&non-targeting_00656&non-targeting_01601&non-targeting_03158&non-targeting_03620&non-targeting_03635\n",
      "4: non-targeting_00233&non-targeting_00972&non-targeting_02193&non-targeting_02574&non-targeting_02777&non-targeting_02962\n",
      "5: non-targeting_00184&non-targeting_02193&non-targeting_02543&non-targeting_02837&non-targeting_03026&non-targeting_03342\n",
      "6: non-targeting_00014&non-targeting_00518&non-targeting_00614&non-targeting_03026&non-targeting_03029&non-targeting_03384\n",
      "   n_nonzero_trt n_nonzero_cntrl pass_qc      p_value log_2_fold_change\n",
      "           <int>           <int>  <lgcl>        <num>             <num>\n",
      "1:            20             371    TRUE 0.0005481779         1.2263258\n",
      "2:            54            3076    TRUE 0.0076809527         0.5482496\n",
      "3:            32            2086    TRUE 0.0089332634         0.7373303\n",
      "4:            21             420    TRUE 0.0102692754         0.8778109\n",
      "5:            17            1575    TRUE 0.0320000000        -0.6913136\n",
      "6:            64            5269    TRUE 0.0320000000        -0.3982145\n",
      "   significant\n",
      "        <lgcl>\n",
      "1:       FALSE\n",
      "2:       FALSE\n",
      "3:       FALSE\n",
      "4:       FALSE\n",
      "5:       FALSE\n",
      "6:       FALSE\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "calibration_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_calibration_check\"\n",
    ")\n",
    "head(calibration_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f075b076",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: If you are on a Mac laptop or desktop, consider setting `parallel = TRUE` to improve speed. Otherwise, keep `parallel = FALSE`.\n",
      "\n",
      "Running precomputation on response ENSG00000004487.18 (1 of 238)\n",
      "Running precomputation on response ENSG00000117000.9 (5 of 238)\n",
      "Running precomputation on response ENSG00000265491.5 (10 of 238)\n",
      "Running precomputation on response ENSG00000198952.8 (15 of 238)\n",
      "Running precomputation on response ENSG00000119185.13 (20 of 238)\n",
      "Running precomputation on response ENSG00000124380.11 (25 of 238)\n",
      "Running precomputation on response ENSG00000138434.17 (30 of 238)\n",
      "Running precomputation on response ENSG00000163513.19 (35 of 238)\n",
      "Running precomputation on response ENSG00000138231.14 (40 of 238)\n",
      "Running precomputation on response ENSG00000170871.12 (45 of 238)\n",
      "Running precomputation on response ENSG00000164040.18 (50 of 238)\n",
      "Running precomputation on response ENSG00000164305.20 (55 of 238)\n",
      "Running precomputation on response ENSG00000164292.13 (60 of 238)\n",
      "Running precomputation on response ENSG00000113196.3 (65 of 238)\n",
      "Running precomputation on response ENSG00000204435.15 (70 of 238)\n",
      "Running precomputation on response ENSG00000112237.13 (75 of 238)\n",
      "Running precomputation on response ENSG00000213079.10 (80 of 238)\n",
      "Running precomputation on response ENSG00000196367.15 (85 of 238)\n",
      "Running precomputation on response ENSG00000082014.17 (90 of 238)\n",
      "Running precomputation on response ENSG00000064313.13 (95 of 238)\n",
      "Running precomputation on response ENSG00000107185.10 (100 of 238)\n",
      "Running precomputation on response ENSG00000155827.12 (105 of 238)\n",
      "Running precomputation on response ENSG00000169925.17 (110 of 238)\n",
      "Running precomputation on response ENSG00000095787.26 (115 of 238)\n",
      "Running precomputation on response ENSG00000175390.15 (120 of 238)\n",
      "Running precomputation on response ENSG00000149100.13 (125 of 238)\n",
      "Running precomputation on response ENSG00000110048.12 (130 of 238)\n",
      "Running precomputation on response ENSG00000173933.21 (135 of 238)\n",
      "Running precomputation on response ENSG00000134954.14 (140 of 238)\n",
      "Running precomputation on response ENSG00000134283.18 (145 of 238)\n",
      "Running precomputation on response ENSG00000174106.3 (150 of 238)\n",
      "Running precomputation on response ENSG00000111300.10 (155 of 238)\n",
      "Running precomputation on response ENSG00000102710.21 (160 of 238)\n",
      "Running precomputation on response ENSG00000184304.17 (165 of 238)\n",
      "Running precomputation on response ENSG00000140264.21 (170 of 238)\n",
      "Running precomputation on response ENSG00000140395.9 (175 of 238)\n",
      "Running precomputation on response ENSG00000134419.15 (180 of 238)\n",
      "Running precomputation on response ENSG00000102900.13 (185 of 238)\n",
      "Running precomputation on response ENSG00000141510.18 (190 of 238)\n",
      "Running precomputation on response ENSG00000125686.12 (195 of 238)\n",
      "Running precomputation on response ENSG00000154217.16 (200 of 238)\n",
      "Running precomputation on response ENSG00000005206.17 (205 of 238)\n",
      "Running precomputation on response ENSG00000011451.21 (210 of 238)\n",
      "Running precomputation on response ENSG00000167615.17 (215 of 238)\n",
      "Running precomputation on response ENSG00000124193.16 (220 of 238)\n",
      "Running precomputation on response ENSG00000128191.17 (225 of 238)\n",
      "Running precomputation on response ENSG00000100387.9 (230 of 238)\n",
      "Running precomputation on response ENSG00000147162.15 (235 of 238)\n",
      "Analyzing pairs containing gRNA group KDM1A (1 of 238)\n",
      "Analyzing pairs containing gRNA group RLF (5 of 238)\n",
      "Analyzing pairs containing gRNA group RNF115 (10 of 238)\n",
      "Analyzing pairs containing gRNA group SMG5 (15 of 238)\n",
      "Analyzing pairs containing gRNA group ITGB1BP1 (20 of 238)\n",
      "Analyzing pairs containing gRNA group SNRNP27 (25 of 238)\n",
      "Analyzing pairs containing gRNA group ITPRID2 (30 of 238)\n",
      "Analyzing pairs containing gRNA group TGFBR2 (35 of 238)\n",
      "Analyzing pairs containing gRNA group DBR1 (40 of 238)\n",
      "Analyzing pairs containing gRNA group KIAA0232 (45 of 238)\n",
      "Analyzing pairs containing gRNA group PGRMC2 (50 of 238)\n",
      "Analyzing pairs containing gRNA group CASP3 (55 of 238)\n",
      "Analyzing pairs containing gRNA group RHOBTB3 (60 of 238)\n",
      "Analyzing pairs containing gRNA group HAND1 (65 of 238)\n",
      "Analyzing pairs containing gRNA group CSNK2B (70 of 238)\n",
      "Analyzing pairs containing gRNA group CCNC (75 of 238)\n",
      "Analyzing pairs containing gRNA group SCAF8 (80 of 238)\n",
      "Analyzing pairs containing gRNA group TRRAP (85 of 238)\n",
      "Analyzing pairs containing gRNA group SMARCD3 (90 of 238)\n",
      "Analyzing pairs containing gRNA group TAF2 (95 of 238)\n",
      "Analyzing pairs containing gRNA group RGP1 (100 of 238)\n",
      "Analyzing pairs containing gRNA group RNF20 (105 of 238)\n",
      "Analyzing pairs containing gRNA group BRD3 (110 of 238)\n",
      "Analyzing pairs containing gRNA group WAC (115 of 238)\n",
      "Analyzing pairs containing gRNA group EIF3F (120 of 238)\n",
      "Analyzing pairs containing gRNA group EIF3M (125 of 238)\n",
      "Analyzing pairs containing gRNA group OSBP (130 of 238)\n",
      "Analyzing pairs containing gRNA group RBM4 (135 of 238)\n",
      "Analyzing pairs containing gRNA group ETS1 (140 of 238)\n",
      "Analyzing pairs containing gRNA group PPHLN1 (145 of 238)\n",
      "Analyzing pairs containing gRNA group LEMD3 (150 of 238)\n",
      "Analyzing pairs containing gRNA group NAA25 (155 of 238)\n",
      "Analyzing pairs containing gRNA group SUPT20H (160 of 238)\n",
      "Analyzing pairs containing gRNA group PRKD1 (165 of 238)\n",
      "Analyzing pairs containing gRNA group SERF2 (170 of 238)\n",
      "Analyzing pairs containing gRNA group SKIC8 (175 of 238)\n",
      "Analyzing pairs containing gRNA group RPS15A (180 of 238)\n",
      "Analyzing pairs containing gRNA group NUP93 (185 of 238)\n",
      "Analyzing pairs containing gRNA group TP53 (190 of 238)\n",
      "Analyzing pairs containing gRNA group MED1 (195 of 238)\n",
      "Analyzing pairs containing gRNA group PITPNC1 (200 of 238)\n",
      "Analyzing pairs containing gRNA group SPPL2B (205 of 238)\n",
      "Analyzing pairs containing gRNA group WIZ (210 of 238)\n",
      "Analyzing pairs containing gRNA group LENG8 (215 of 238)\n",
      "Analyzing pairs containing gRNA group SRSF6 (220 of 238)\n",
      "Analyzing pairs containing gRNA group DGCR8 (225 of 238)\n",
      "Analyzing pairs containing gRNA group RBX1 (230 of 238)\n",
      "Analyzing pairs containing gRNA group OGT (235 of 238)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "✅ Saved PNG plot to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/discovery.png\n",
       "In addition: There were 50 or more warnings (use warnings() to see the first 50)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sceptre_object_highmoi <- run_discovery_analysis(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  parallel = FALSE\n",
    ")\n",
    "\n",
    "png(file.path(data_dir, \"discovery.png\"), width = 2400, height = 2000, res = 300)\n",
    "\n",
    "print(plot(sceptre_object_highmoi))\n",
    "\n",
    "dev.off()\n",
    "\n",
    "message(\"✅ Saved PNG plot to: \", file.path(data_dir, \"discovery.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "07f63f1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "✅ Saved discovery to: /oak/stanford/groups/engreitz/Users/tri/CodeBase/programDE/notebooks/d3_sceptre_transfer/TargetKD.csv\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "discovery_result <- get_result(\n",
    "  sceptre_object = sceptre_object_highmoi,\n",
    "  analysis = \"run_discovery_analysis\"\n",
    ")\n",
    "\n",
    "write.csv(discovery_result, file.path(data_dir, \"TargetKD.csv\"), row.names = FALSE)\n",
    "message(\"✅ Saved discovery to: \", file.path(data_dir, \"TargetKD.csv\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e681854b",
   "metadata": {},
   "source": [
    "### Plotting"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rpy2 (sceptre/scanpy)",
   "language": "python",
   "name": "sceptre_env"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
